
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Building CGAT pipelines &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/sorttable.js"></script>
    <script type="text/javascript" src="_static/jquery.min.js"></script>
    <script type="text/javascript" src="_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="index.html" />
    <link rel="up" title="CGAT Pipelines" href="Pipelines.html" />
    <link rel="next" title="Writing pipeline reports" href="PipelineReports.html" />
    <link rel="prev" title="Using CGAT pipelines" href="UsingPipelines.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PipelineReports.html" title="Writing pipeline reports"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="UsingPipelines.html" title="Using CGAT pipelines"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="Pipelines.html" accesskey="U">CGAT Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="building-cgat-pipelines">
<h1>Building CGAT pipelines<a class="headerlink" href="#building-cgat-pipelines" title="Permalink to this headline">¶</a></h1>
<p>The best way to build a pipeline is to start from an example. There are several
pipelines available, see <a class="reference internal" href="index.html#cgatpipelines"><em>CGAT Pipelines</em></a>. To start a new project, use
<tt class="file docutils literal"><span class="pre">pipeline_quickstart.py</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">python</span> <span class="o">&lt;</span><span class="n">srcdir</span><span class="o">&gt;</span><span class="n">pipeline_quickstart</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">name</span><span class="o">=</span><span class="n">test</span>
</pre></div>
</div>
<p>This will create a new directory called <tt class="docutils literal"><span class="pre">test</span></tt> in the current directory.</p>
<p>Another source of information is the script <tt class="file docutils literal"><span class="pre">pipeline_template.py</span></tt> in
the <a class="reference internal" href="make_pipelines/Gpipe.html#term-source-directory"><em class="xref std std-term">source directory</em></a>.</p>
<p>This section describes how CGAT pipelines can be constructed using the
<a class="reference internal" href="modules/Pipeline.html#module-Pipeline" title="Pipeline"><tt class="xref py py-mod docutils literal"><span class="pre">Pipeline</span></tt></a> module. The Pipeline.py module contains a
variety of useful functions for pipeline construction.</p>
<div class="section" id="overview">
<span id="pipelineorganization"></span><h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Pipelines generally have a similar structure. Pipelines are implemented
as a pipeline script in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-source-directory"><em class="xref std std-term">source directory</em></a> called <tt class="file docutils literal"><span class="pre">pipeline_&lt;somename&gt;.py</span></tt>
and a file <tt class="file docutils literal"><span class="pre">pipeline_&lt;somename&gt;.ini</span></tt> with default configuration values.</p>
<div class="section" id="pipeline-input">
<h3>Pipeline input<a class="headerlink" href="#pipeline-input" title="Permalink to this headline">¶</a></h3>
<p>Pipelines are executed within a dedicated <a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>. They usually require
the following files within this directory:</p>
<blockquote>
<div><ul class="simple">
<li>a pipeline configuration file <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt></li>
<li>input data files, usually linked in from a data repository</li>
</ul>
</div></blockquote>
<p>Other files that might be used in a pipeline are:</p>
<blockquote>
<div><ul class="simple">
<li>external data files such as genomes that a referred to by they their full path name.</li>
<li>conf.py for automated reports.</li>
</ul>
</div></blockquote>
<p>The pipelines will work from the input files in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>, usually identified by their
suffix. For example, a ChIP-Seq pipeline might look for any <tt class="docutils literal"><span class="pre">*.fastq.gz</span></tt> files in the directory,
run QC on these, map the reads to a genome sequence, call peaks, do motif analyses, etc.</p>
</div>
<div class="section" id="pipeline-output">
<h3>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this headline">¶</a></h3>
<p>The pipeline will create files and database tables in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.
When building a pipeline, you can choose any
file/directory layout that suits your needs. Some prefer flat hierarchies with many files, while
others prefer deep directories.</p>
<p>Two directories have a special function and can be used for exporting pipeline results (see <a class="reference internal" href="#pipelinepublishing">PipelinePublishing</a>):</p>
<p>The <tt class="file docutils literal"><span class="pre">export</span></tt> directory contains all files that will be referred to directly in the report
or that later should be published by the pipeline. For example, pdf documents created by the peak caller
or logo images created by a motif tool should go there.</p>
<p>The directory <tt class="file docutils literal"><span class="pre">report</span></tt> will contain the automatically generated report.</p>
</div>
</div>
<div class="section" id="guidelines">
<h2>Guidelines<a class="headerlink" href="#guidelines" title="Permalink to this headline">¶</a></h2>
<p>To preserve disk space, please always work use compressed files as much as possible.
Most data files compress very well, for example fastq files often compress by a factor of 80% or more:
a 10Gb file will use just 2Gb.</p>
<p>Working with compressed files is straight-forward using unix pipes and the commands <tt class="docutils literal"><span class="pre">gzip</span></tt>, <tt class="docutils literal"><span class="pre">gunzip</span></tt> or <tt class="docutils literal"><span class="pre">zcat</span></tt>.</p>
<p>If you require random access to a file, load the file into the database and index it
appropriately. Genomic interval files can be indexed with tabix to allow random access.</p>
</div>
<div class="section" id="running-commands-within-tasks">
<span id="pipelinecommands"></span><h2>Running commands within tasks<a class="headerlink" href="#running-commands-within-tasks" title="Permalink to this headline">¶</a></h2>
<p>To run a command line program within a pipeline task, build a statement and call the
<a class="reference internal" href="modules/Pipeline.html#Pipeline.run" title="Pipeline.run"><tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.run()</span></tt></a> method:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="s">&#39;*.unsorted&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.unsorted&#39;</span><span class="p">),</span> <span class="s">&#39;.sorted&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">sortFile</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>

    <span class="n">statement</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;sort </span><span class="si">%(infile)s</span><span class="s"> &gt; </span><span class="si">%(outfile)s</span><span class="s">&#39;&#39;&#39;</span>
    <span class="n">P</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>On calling the <a class="reference internal" href="modules/Pipeline.html#Pipeline.run" title="Pipeline.run"><tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.run()</span></tt></a> method, the environment of the caller
is examined for a variable called <tt class="docutils literal"><span class="pre">statement</span></tt>. The variable is subjected
to string substitution from other variables in the local namespace. In the
example above, <tt class="docutils literal"><span class="pre">%(infile)s</span></tt> and <tt class="docutils literal"><span class="pre">%(outfile)s</span></tt> are substituted with the
values of the variables <tt class="docutils literal"><span class="pre">infile</span></tt> and <tt class="docutils literal"><span class="pre">outfile</span></tt>, respectively.</p>
<p>The same mechanism also permits setting configuration parameters, for example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="s">&#39;*.unsorted&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.unsorted&#39;</span><span class="p">),</span> <span class="s">&#39;.sorted&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">sortFile</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>

    <span class="n">statement</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;sort -t </span><span class="si">%(tmpdir)s</span><span class="s"> </span><span class="si">%(infile)s</span><span class="s"> &gt; </span><span class="si">%(outfile)s</span><span class="s">&#39;&#39;&#39;</span>
    <span class="n">P</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>will automatically substitute the configuration parameter <tt class="docutils literal"><span class="pre">tmpdir</span></tt>
into the command. See <a class="reference internal" href="#configurationvalues">ConfigurationValues</a> for more on using configuration
parameters.</p>
<p>The pipeline will stop and return an error if the command exits with an error code.</p>
<p>If you chain multiple commands, only the return value of the last command
is used to check for an error. Thus, if an upstream command fails, it will go unnoticed.
To detect these errors, insert the <tt class="docutils literal"><span class="pre">checkpoint</span></tt> statement between commands. For example:</p>
<div class="highlight-python"><div class="highlight"><pre>@files( &#39;*.unsorted.gz&#39;, suffix(&#39;.unsorted.gz&#39;), &#39;.sorted)
def sortFile( infile, outfile ):

    statement = &#39;&#39;&#39;gunzip %(infile)s %(infile)s.tmp;
                   checkpoint;
                   sort -t %(tmpdir)s %(infile)s.tmp &gt; %(outfile)s;
                   checkpoint;
                   rm -f %(infile)s.tmp
    P.run()
</pre></div>
</div>
<p>Of course, the statement aboved could be executed more efficiently using pipes:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="s">&#39;*.unsorted.gz&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.unsorted.gz&#39;</span><span class="p">),</span> <span class="s">&#39;.sorted.gz&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">sortFile</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>

    <span class="n">statement</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;gunzip &lt; </span><span class="si">%(infile)s</span><span class="s"></span>
<span class="s">                   | sort -t </span><span class="si">%(tmpdir)s</span><span class="s"></span>
<span class="s">                   | gzip &gt; </span><span class="si">%(outfile)s</span><span class="s">&#39;&#39;&#39;</span>
    <span class="n">P</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>The pipeline inserts code automatically to check for error return codes if multiple commands are
combined in a pipe.</p>
<div class="section" id="running-commands-on-the-cluster">
<h3>Running commands on the cluster<a class="headerlink" href="#running-commands-on-the-cluster" title="Permalink to this headline">¶</a></h3>
<p>In order to run commands on cluster, use <tt class="docutils literal"><span class="pre">to_cluster=True</span></tt>.</p>
<p>To run the command from the previous section on the cluster:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="s">&#39;*.unsorted.gz&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.unsorted.gz&#39;</span><span class="p">),</span> <span class="s">&#39;.sorted.gz&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">sortFile</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>

    <span class="n">to_cluster</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">statement</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;gunzip &lt; </span><span class="si">%(infile)s</span><span class="s"></span>
<span class="s">                   | sort -t </span><span class="si">%(tmpdir)s</span><span class="s"></span>
<span class="s">                   | gzip &gt; </span><span class="si">%(outfile)s</span><span class="s">&#39;&#39;&#39;</span>
    <span class="n">P</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>The pipeline will automatically create the job submission files, submit
the job to the cluster and wait for its return.</p>
<p>Pipelines will use the command line options <tt class="docutils literal"><span class="pre">--cluster-queue</span></tt>,
<tt class="docutils literal"><span class="pre">--cluster-priority</span></tt>, etc. for global job control. For example, to
change the priority when starting the pipeline, use:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">python</span> <span class="o">&lt;</span><span class="n">pipeline_script</span><span class="o">.</span><span class="n">py</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">cluster</span><span class="o">-</span><span class="n">priority</span><span class="o">=-</span><span class="mi">20</span>
</pre></div>
</div>
<p>To set job options specific to a task, you can define additional variables:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="s">&#39;*.unsorted.gz&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.unsorted.gz&#39;</span><span class="p">),</span> <span class="s">&#39;.sorted.gz&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">sortFile</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>

    <span class="n">to_cluster</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">job_queue</span> <span class="o">=</span> <span class="s">&#39;longjobs.q&#39;</span>
    <span class="n">job_priority</span> <span class="o">=</span> <span class="o">-</span><span class="mi">10</span>
    <span class="n">job_options</span><span class="o">=</span> <span class="s">&quot;-pe dedicated 4 -R y&quot;</span>

    <span class="n">statement</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;gunzip &lt; </span><span class="si">%(infile)s</span><span class="s"></span>
<span class="s">                   | sort -t </span><span class="si">%(tmpdir)s</span><span class="s"></span>
<span class="s">                   | gzip &gt; </span><span class="si">%(outfile)s</span><span class="s">&#39;&#39;&#39;</span>
    <span class="n">P</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>The above statement will be run in the queue <tt class="docutils literal"><span class="pre">longjobs.q</span></tt> at a priority of <tt class="docutils literal"><span class="pre">-10</span></tt>.
Additionally, it will be executed in the parallel environment <tt class="docutils literal"><span class="pre">dedicated</span></tt> with at
least 4 cores.</p>
<p>Array jobs can be controlled through the <tt class="docutils literal"><span class="pre">job_array</span></tt> variable:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="s">&#39;*.in&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.in&#39;</span><span class="p">),</span> <span class="s">&#39;.out&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">myGridTask</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>

    <span class="n">job_array</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">nsnps</span><span class="p">,</span> <span class="n">stepsize</span><span class="p">)</span>

    <span class="n">statement</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;grid_task.bash </span><span class="si">%(infile)s</span><span class="s"> </span><span class="si">%(outfile)s</span><span class="s"></span>
<span class="s">       &gt; </span><span class="si">%(outfile)s</span><span class="s">.$SGE_TASK_ID 2&gt; </span><span class="si">%(outfile)s</span><span class="s">.err.$SGE_TASK_ID</span>
<span class="s">    &#39;&#39;&#39;</span>
    <span class="n">P</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>Note that the <tt class="file docutils literal"><span class="pre">grid_task.bash</span></tt> file must be grid engine aware. This means
it makes use of the <span class="target" id="index-0"></span><tt class="xref std std-envvar docutils literal"><span class="pre">SGE_TASK_ID</span></tt>, <span class="target" id="index-1"></span><tt class="xref std std-envvar docutils literal"><span class="pre">SGE_TASK_FIRST</span></tt>, <span class="target" id="index-2"></span><tt class="xref std std-envvar docutils literal"><span class="pre">SGE_TASK_LAST</span></tt> and
<span class="target" id="index-3"></span><tt class="xref std std-envvar docutils literal"><span class="pre">SGE_TASK_STEPSIZE</span></tt> environment variables to select the chunk of data it wants
to work on.</p>
<p>The job submission files are files called <cite>tmp*</cite> in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.
These files will be deleted automatically. However, the files will remain after
aborted runs to be cleaned up manually.</p>
</div>
</div>
<div class="section" id="tracks">
<span id="pipelinetracks"></span><h2>Tracks<a class="headerlink" href="#tracks" title="Permalink to this headline">¶</a></h2>
<p>A pipeline typically processes the data streams from several experimental
data sources. These data streams are usually processed separately (processing,
quality control) and as aggregates. The module <a class="reference internal" href="pipelinemodules/PipelineTracks.html#module-PipelineTracks" title="PipelineTracks"><tt class="xref py py-mod docutils literal"><span class="pre">PipelineTracks</span></tt></a> helps
implementing this.</p>
</div>
<div class="section" id="databases">
<span id="pipelinedatabases"></span><h2>Databases<a class="headerlink" href="#databases" title="Permalink to this headline">¶</a></h2>
<div class="section" id="loading-data-into-the-database">
<h3>Loading data into the database<a class="headerlink" href="#loading-data-into-the-database" title="Permalink to this headline">¶</a></h3>
<p><tt class="xref py py-mod docutils literal"><span class="pre">Pipeline.py</span></tt> offers various tools for working with databases. By default,
it is configured to use an sqlite3 database in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a> called <tt class="file docutils literal"><span class="pre">csvdb</span></tt>.</p>
<p>Tab-separated output files can be loaded into a table using the <a class="reference internal" href="modules/Pipeline.html#Pipeline.load" title="Pipeline.load"><tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.load()</span></tt></a>
function. For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@transform</span><span class="p">(</span> <span class="s">&#39;data_*.tsv.gz&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.tsv.gz&#39;</span><span class="p">),</span> <span class="s">&#39;.load&#39;</span> <span class="p">)</span>
<span class="k">def</span> <span class="nf">loadTables</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>
   <span class="n">P</span><span class="o">.</span><span class="n">load</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">)</span>
</pre></div>
</div>
<p>The task above will load all tables ending with <tt class="docutils literal"><span class="pre">tsv.gz</span></tt> into the database
Table names are given by the filenames, i.e, the data in <tt class="file docutils literal"><span class="pre">data_1.tsv.gz</span></tt> will
be loaded into the table <tt class="file docutils literal"><span class="pre">data_1</span></tt>.</p>
<p>The load mechanism uses the script <tt class="file docutils literal"><span class="pre">csv2db.py</span></tt> and can be configured using
the configuration options <tt class="docutils literal"><span class="pre">database</span></tt> and <tt class="docutils literal"><span class="pre">csv2db_options</span></tt>. Additional options
can be given via the optional <em>options</em> argument:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@transform</span><span class="p">(</span> <span class="s">&#39;data_*.tsv.gz&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.tsv.gz&#39;</span><span class="p">),</span> <span class="s">&#39;.load&#39;</span> <span class="p">)</span>
<span class="k">def</span> <span class="nf">loadTables</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>
   <span class="n">P</span><span class="o">.</span><span class="n">load</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="s">&quot;--index=gene_id&quot;</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="connecting-to-a-database">
<h3>Connecting to a database<a class="headerlink" href="#connecting-to-a-database" title="Permalink to this headline">¶</a></h3>
<p>To use data in the database in your tasks, you need to first connect to the
database. It helps to encapsulate the connection in a separate function. For
example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">connect</span><span class="p">():</span>
    <span class="n">dbh</span> <span class="o">=</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span> <span class="n">PARAMS</span><span class="p">[</span><span class="s">&quot;database&quot;</span><span class="p">]</span> <span class="p">)</span>
    <span class="n">statement</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;ATTACH DATABASE &#39;</span><span class="si">%s</span><span class="s">&#39; as annotations&#39;&#39;&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">PARAMS</span><span class="p">[</span><span class="s">&quot;annotations_database&quot;</span><span class="p">])</span>
    <span class="n">cc</span> <span class="o">=</span> <span class="n">dbh</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
    <span class="n">cc</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span> <span class="n">statement</span> <span class="p">)</span>
    <span class="n">cc</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">dbh</span>
</pre></div>
</div>
<p>The above function will connect to the database. It will also attach a secondary database
<tt class="docutils literal"><span class="pre">annotations</span></tt>.</p>
<p>The following example illustrates how to use the connection:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@transform</span><span class="p">(</span> <span class="o">...</span> <span class="p">)</span>
<span class="k">def</span> <span class="nf">buildCodingTranscriptSet</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>

    <span class="n">dbh</span> <span class="o">=</span> <span class="n">connect</span><span class="p">()</span>

    <span class="n">statement</span> <span class="o">=</span> <span class="s">&#39;&#39;&#39;SELECT DISTINCT transcript_id FROM transcript_info WHERE transcript_biotype = &#39;protein_coding&#39; &#39;&#39;&#39;</span>
    <span class="n">cc</span> <span class="o">=</span> <span class="n">dbh</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
    <span class="n">transcript_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cc</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">statement</span><span class="p">)]</span> <span class="p">)</span>
    <span class="o">...</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="reports">
<span id="pipelinereports"></span><h2>Reports<a class="headerlink" href="#reports" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="modules/Pipeline.html#Pipeline.run_report" title="Pipeline.run_report"><tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.run_report()</span></tt></a> method builds or updates reports using <a class="reference external" href="http://code.google.com/p/sphinx-report/">SphinxReport</a>. Usually, a pipeline
will simply contain the following:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@follows</span><span class="p">(</span> <span class="n">mkdir</span><span class="p">(</span> <span class="s">&quot;report&quot;</span> <span class="p">)</span> <span class="p">)</span>
<span class="k">def</span> <span class="nf">build_report</span><span class="p">():</span>
    <span class="sd">&#39;&#39;&#39;build report from scratch.&#39;&#39;&#39;</span>

    <span class="n">E</span><span class="o">.</span><span class="n">info</span><span class="p">(</span> <span class="s">&quot;starting report build process from scratch&quot;</span> <span class="p">)</span>
    <span class="n">P</span><span class="o">.</span><span class="n">run_report</span><span class="p">(</span> <span class="n">clean</span> <span class="o">=</span> <span class="bp">True</span> <span class="p">)</span>

<span class="nd">@follows</span><span class="p">(</span> <span class="n">mkdir</span><span class="p">(</span> <span class="s">&quot;report&quot;</span> <span class="p">)</span> <span class="p">)</span>
<span class="k">def</span> <span class="nf">update_report</span><span class="p">():</span>
    <span class="sd">&#39;&#39;&#39;update report.&#39;&#39;&#39;</span>

    <span class="n">E</span><span class="o">.</span><span class="n">info</span><span class="p">(</span> <span class="s">&quot;updating report&quot;</span> <span class="p">)</span>
    <span class="n">P</span><span class="o">.</span><span class="n">run_report</span><span class="p">(</span> <span class="n">clean</span> <span class="o">=</span> <span class="bp">False</span> <span class="p">)</span>
</pre></div>
</div>
<p>This will add the two tasks <tt class="docutils literal"><span class="pre">build_report</span></tt> and <tt class="docutils literal"><span class="pre">update_report</span></tt> to
the pipeline. The former completely rebuilds a report, while the
latter only updates changed pages. The report will be in the directory
<tt class="file docutils literal"><span class="pre">report</span></tt>.</p>
<p>Note that report building requires the file <tt class="file docutils literal"><span class="pre">conf.py</span></tt> in the
<a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>. This file is read by <a class="reference external" href="http://sphinx-doc.org/">sphinx</a> and can be
used to report building options. By default, the file is a stub
reading in common options from the CGAT code base.</p>
<p>The section <a class="reference internal" href="PipelineReports.html#writingreports"><em>Writing pipeline reports</em></a> contains more information.</p>
</div>
<div class="section" id="configuration-values">
<span id="configurationvalues"></span><h2>Configuration values<a class="headerlink" href="#configuration-values" title="Permalink to this headline">¶</a></h2>
<div class="section" id="setting-up-configuration-values">
<h3>Setting up configuration values<a class="headerlink" href="#setting-up-configuration-values" title="Permalink to this headline">¶</a></h3>
<p>Pipelines are configured via a configuration script. The
following snippet can be included at the beginning of a pipeline
to set it all up:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># load options from the config file</span>
<span class="kn">import</span> <span class="nn">Pipeline</span> <span class="kn">as</span> <span class="nn">P</span>
<span class="n">P</span><span class="o">.</span><span class="n">getParameters</span><span class="p">(</span>
       <span class="p">[</span><span class="s">&quot;</span><span class="si">%s</span><span class="s">.ini&quot;</span> <span class="o">%</span> <span class="n">__file__</span><span class="p">[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="s">&quot;.py&quot;</span><span class="p">)],</span>
       <span class="s">&quot;../pipeline.ini&quot;</span><span class="p">,</span>
       <span class="s">&quot;pipeline.ini&quot;</span> <span class="p">]</span> <span class="p">)</span>
<span class="n">PARAMS</span> <span class="o">=</span> <span class="n">P</span><span class="o">.</span><span class="n">PARAMS</span>
</pre></div>
</div>
<p>Default values of configuration parameters will be read from a global
configuration file that is part of the CGAT code base
in <tt class="file docutils literal"><span class="pre">CGATPipelines/configuration/pipeline.ini</span></tt>.</p>
<p>These values will be updated with the file named
<tt class="file docutils literal"><span class="pre">pipeline_&lt;pipeline_name&gt;.ini</span></tt> in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-source-directory"><em class="xref std std-term">source directory</em></a>.</p>
<p>Next, the file <tt class="file docutils literal"><span class="pre">../pipeline.ini</span></tt> will be read (if it exists) and configuration
values that are specific to a certain project will overwrite default values.</p>
<p>Finally, run specific configuration will be read from the file <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt>
in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
<p>The method <a class="reference internal" href="modules/Pipeline.html#Pipeline.getParameters" title="Pipeline.getParameters"><tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.getParameters()</span></tt></a> reads parameters and updates a global
dictionary of parameter values. It automatically guesses the type of parameters
in the order of <tt class="docutils literal"><span class="pre">int()</span></tt>, <tt class="docutils literal"><span class="pre">float()</span></tt> or <tt class="docutils literal"><span class="pre">str()</span></tt>.</p>
<p>If a configuration variable is empty (<tt class="docutils literal"><span class="pre">var=</span></tt>), it will be set to <tt class="docutils literal"><span class="pre">None</span></tt>.</p>
<p>Configuration values from another pipeline can be added in a separate namespace:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">PARAMS_ANNOTATIONS</span> <span class="o">=</span> <span class="n">P</span><span class="o">.</span><span class="n">peekParameters</span><span class="p">(</span> <span class="n">PARAMS</span><span class="p">[</span><span class="s">&quot;annotations_dir&quot;</span><span class="p">],</span>
                                              <span class="s">&quot;pipeline_annotations.py&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<p>The statement above will load the parameters from a <a class="reference internal" href="pipelines/pipeline_annotations.html#module-pipeline_annotations" title="pipeline_annotations"><tt class="xref py py-mod docutils literal"><span class="pre">pipeline_annotations</span></tt></a> pipeline with
<a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a> <tt class="docutils literal"><span class="pre">annotations_dir</span></tt>.</p>
</div>
<div class="section" id="using-configuration-values">
<h3>Using configuration values<a class="headerlink" href="#using-configuration-values" title="Permalink to this headline">¶</a></h3>
<p>Configuration values are accessible via the <tt class="xref py py-data docutils literal"><span class="pre">PARAMS</span></tt> variable. The <tt class="xref py py-data docutils literal"><span class="pre">PARAMS</span></tt>
variable is a dictionary mapping configuration parameters to values. Keys are in the
format <tt class="docutils literal"><span class="pre">section_parameter</span></tt>. For example, the key <tt class="docutils literal"><span class="pre">bowtie_threads</span></tt> will provide the
configuration value of:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="n">bowtie</span><span class="p">]</span>
<span class="n">threads</span><span class="o">=</span><span class="mi">4</span>
</pre></div>
</div>
<p>In a script, the value can be accessed via <tt class="docutils literal"><span class="pre">PARAMS[&quot;bowtie_threads&quot;]</span></tt>.</p>
<p>Undefined configuration values will throw a <tt class="xref py py-class docutils literal"><span class="pre">ValueError</span></tt>. To test if
a configuration variable exists, use:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">if</span> <span class="s">&#39;bowtie_threads&#39;</span> <span class="ow">in</span> <span class="n">PARAMS</span><span class="p">:</span> <span class="k">pass</span>
</pre></div>
</div>
<p>To test, if it is unset, use:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">if</span> <span class="s">&#39;bowie_threads&#39;</span> <span class="ow">in</span> <span class="n">PARAMS</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">PARAMS</span><span class="p">[</span><span class="s">&#39;botwie_threads&#39;</span><span class="p">]:</span> <span class="k">pass</span>
</pre></div>
</div>
</div>
<div class="section" id="task-specific-parameters">
<h3>Task specific parameters<a class="headerlink" href="#task-specific-parameters" title="Permalink to this headline">¶</a></h3>
<p>Task specific parameters can be set by creating a task specific section in
the <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt>. The task is identified by the output filename.
For example, given the following task:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="s">&#39;*.fastq&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.fastq&#39;</span><span class="p">),</span> <span class="s">&#39;.bam&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">mapWithBowtie</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>
   <span class="o">...</span>
</pre></div>
</div>
<p>and the files <tt class="file docutils literal"><span class="pre">data1.fastq</span></tt> and <tt class="file docutils literal"><span class="pre">data2.fastq</span></tt> in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>,
two output files <tt class="file docutils literal"><span class="pre">data.bam</span></tt> and <tt class="file docutils literal"><span class="pre">data2.bam</span></tt> will be created on executing
<tt class="docutils literal"><span class="pre">mapWithBowtie</span></tt>. Both will use the same parameters. To set parameters specific to the
execution of <tt class="file docutils literal"><span class="pre">data1.fastq</span></tt>, add the following to <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="n">data1</span><span class="o">.</span><span class="n">fastq</span><span class="p">]</span>
<span class="n">bowtie_threads</span><span class="o">=</span><span class="mi">16</span>
</pre></div>
</div>
<p>This will set the configuration value <tt class="docutils literal"><span class="pre">bowtie_threads</span></tt> to 16 when using the command
line substitution method in <a class="reference internal" href="modules/Pipeline.html#Pipeline.run" title="Pipeline.run"><tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.run()</span></tt></a>. To get an task-specific parameter values in
a python task, use:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="s">&#39;*.fastq&#39;</span><span class="p">,</span> <span class="n">suffix</span><span class="p">(</span><span class="s">&#39;.fastq&#39;</span><span class="p">),</span> <span class="s">&#39;.bam&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">mytask</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>
    <span class="n">MY_PARAMS</span> <span class="o">=</span> <span class="n">P</span><span class="o">.</span><span class="n">substituteParameters</span><span class="p">(</span> <span class="nb">locals</span><span class="p">()</span> <span class="p">)</span>
</pre></div>
</div>
<p>Thus, task specific are implemented generically using the <a class="reference internal" href="modules/Pipeline.html#Pipeline.run" title="Pipeline.run"><tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.run()</span></tt></a> mechanism,
but pipeline authors need to explicitely code for track specific parameters.</p>
</div>
</div>
<div class="section" id="documentation">
<span id="pipelinedocumentation"></span><h2>Documentation<a class="headerlink" href="#documentation" title="Permalink to this headline">¶</a></h2>
<p>Up-to-date and accurate documentation is crucial for writing portable and maintainable
pipelines. To document your pipelines write documentation as you would for a module.
See <tt class="file docutils literal"><span class="pre">pipeline_template.py</span></tt> and other pipelines for an example.</p>
<p>To rebuild all documentation, enter the <tt class="file docutils literal"><span class="pre">doc</span></tt> directory in the <a class="reference internal" href="make_pipelines/Gpipe.html#term-source-directory"><em class="xref std std-term">source directory</em></a> and
type:</p>
<div class="highlight-python"><div class="highlight"><pre>cd doc
python collect.py
</pre></div>
</div>
<p>This will collect all new scripts to the documentation.</p>
<p>Next, edit the file <tt class="file docutils literal"><span class="pre">contents.rst</span></tt> and add your pipeline to the table of pipelines. Finally, type:</p>
<div class="highlight-python"><div class="highlight"><pre>make html
</pre></div>
</div>
<p>to rebuild the documentation.</p>
</div>
<div class="section" id="using-other-pipelines">
<h2>Using other pipelines<a class="headerlink" href="#using-other-pipelines" title="Permalink to this headline">¶</a></h2>
<p>You can use the output of other pipelines within your own pipelines. <a class="reference internal" href="pipelines/pipeline_annotations.html#module-pipeline_annotations" title="pipeline_annotations"><tt class="xref py py-mod docutils literal"><span class="pre">pipeline_annotations</span></tt></a>
is an example - it provides often used annotation data sets for an analysis. How to load another
pipelines parameters, connect to its database and write a modular report have been discussed above.</p>
<p>If you write a pipeline that is likely to be used by others, it is best to provide an interface.
For example, the <a class="reference internal" href="pipelines/pipeline_annotations.html#module-pipeline_annotations" title="pipeline_annotations"><tt class="xref py py-mod docutils literal"><span class="pre">pipeline_annotations</span></tt></a> pipeline has an interface section that list all the
files that are produced by the pipeline. Other pipelines can refer to the interface section without
having to be aware of the actual file names:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">filename_cds</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">PARAMS</span><span class="p">[</span><span class="s">&quot;annotations_dir&quot;</span><span class="p">],</span>
                        <span class="n">PARAMS_ANNOTATIONS</span><span class="p">[</span><span class="s">&quot;interface_geneset_cds_gtf&quot;</span><span class="p">]</span> <span class="p">)</span>
</pre></div>
</div>
<p>Running other pipelines within your pipeline <em>should</em> be possible as well - provided they are within
their own separate <a class="reference internal" href="make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
</div>
<div class="section" id="publishing-data">
<span id="pipelinepublishing"></span><h2>Publishing data<a class="headerlink" href="#publishing-data" title="Permalink to this headline">¶</a></h2>
<p>To publish data and a report, use the <tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.publish_report()</span></tt> method, such as in the
following task:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@follows</span><span class="p">(</span> <span class="n">update_report</span> <span class="p">)</span>
<span class="k">def</span> <span class="nf">publish_report</span><span class="p">():</span>
    <span class="sd">&#39;&#39;&#39;publish report.&#39;&#39;&#39;</span>

    <span class="n">E</span><span class="o">.</span><span class="n">info</span><span class="p">(</span> <span class="s">&quot;publishing report&quot;</span> <span class="p">)</span>
    <span class="n">P</span><span class="o">.</span><span class="n">publish_report</span><span class="p">()</span>
</pre></div>
</div>
<p>On publishing a report, the report (in the directory <tt class="file docutils literal"><span class="pre">report</span></tt>, specified by <tt class="docutils literal"><span class="pre">report_dir</span></tt>)
will get copied to the directory specified in the configuration value <tt class="docutils literal"><span class="pre">web_dir</span></tt>. Also, all files
in the <tt class="file docutils literal"><span class="pre">export</span></tt> directory will get copied over and links pointing to such files will be
automatically corrected.</p>
<p>The report will then be available at <tt class="docutils literal"><span class="pre">http://www.cgat.org/downloads/%(project_id)s/report</span></tt> where
<tt class="docutils literal"><span class="pre">project_id</span></tt> is the unique identifier given to each project. It is looked up automatically, but the
automatic look-up requires that the pipeline is executed within the <tt class="file docutils literal"><span class="pre">/ifs/proj</span></tt> directory.</p>
<p>If the option <em>prefix</em> is given to publish_report, all output directories will be output
prefixed by <em>prefix</em>. This is very useful if there is more than one report per project.</p>
<p>See <tt class="xref py py-meth docutils literal"><span class="pre">Pipeline.publish_report()</span></tt> for more options.</p>
</div>
<div class="section" id="checking-requisites">
<h2>Checking requisites<a class="headerlink" href="#checking-requisites" title="Permalink to this headline">¶</a></h2>
<p>TODO</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Building CGAT pipelines</a><ul>
<li><a class="reference internal" href="#overview">Overview</a><ul>
<li><a class="reference internal" href="#pipeline-input">Pipeline input</a></li>
<li><a class="reference internal" href="#pipeline-output">Pipeline output</a></li>
</ul>
</li>
<li><a class="reference internal" href="#guidelines">Guidelines</a></li>
<li><a class="reference internal" href="#running-commands-within-tasks">Running commands within tasks</a><ul>
<li><a class="reference internal" href="#running-commands-on-the-cluster">Running commands on the cluster</a></li>
</ul>
</li>
<li><a class="reference internal" href="#tracks">Tracks</a></li>
<li><a class="reference internal" href="#databases">Databases</a><ul>
<li><a class="reference internal" href="#loading-data-into-the-database">Loading data into the database</a></li>
<li><a class="reference internal" href="#connecting-to-a-database">Connecting to a database</a></li>
</ul>
</li>
<li><a class="reference internal" href="#reports">Reports</a></li>
<li><a class="reference internal" href="#configuration-values">Configuration values</a><ul>
<li><a class="reference internal" href="#setting-up-configuration-values">Setting up configuration values</a></li>
<li><a class="reference internal" href="#using-configuration-values">Using configuration values</a></li>
<li><a class="reference internal" href="#task-specific-parameters">Task specific parameters</a></li>
</ul>
</li>
<li><a class="reference internal" href="#documentation">Documentation</a></li>
<li><a class="reference internal" href="#using-other-pipelines">Using other pipelines</a></li>
<li><a class="reference internal" href="#publishing-data">Publishing data</a></li>
<li><a class="reference internal" href="#checking-requisites">Checking requisites</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="UsingPipelines.html"
                        title="previous chapter">Using CGAT pipelines</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="PipelineReports.html"
                        title="next chapter">Writing pipeline reports</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/BuildingPipelines.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PipelineReports.html" title="Writing pipeline reports"
             >next</a> |</li>
        <li class="right" >
          <a href="UsingPipelines.html" title="Using CGAT pipelines"
             >previous</a> |</li>
        <li><a href="index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="Pipelines.html" >CGAT Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>