
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>CGAT - Computational Genomics Analysis Tools &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/sorttable.js"></script>
    <script type="text/javascript" src="_static/jquery.min.js"></script>
    <script type="text/javascript" src="_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="index.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="cgat-computational-genomics-analysis-tools">
<span id="cgat"></span><h1>CGAT - Computational Genomics Analysis Tools<a class="headerlink" href="#cgat-computational-genomics-analysis-tools" title="Permalink to this headline">¶</a></h1>
<p>CGAT is a collection of tools for the computational genomicist written
in the python language. The tools have been developed and accumulated in various
genome projects (<a class="reference external" href="http://www.cgat.org">Heger &amp; Ponting, 2007</a>, <a class="reference external" href="http://www.cgat.org">Warren et al., 2008</a>) and NGS projects
(<a class="reference external" href="http://www.cgat.org">Ramagopalan et al., 2010</a>). The tools are continuously being developed
as part of the <a class="reference external" href="http://www.cgat.org">CGAT Training programme</a>.</p>
<p>The tools work from the command line, but can readily be installed
within frameworks such as <a class="reference external" href="https://main.g2.bx.psu.edu/">Galaxy</a>.</p>
<p>Please note that the tools are part of a larger code base also
including genomics and NGS pipelines. More information about those
is <a class="reference internal" href="index.html#contents"><em>here</em></a>.</p>
<p>Detailed instructions on installation, on usage and a tool reference
are below, followed by a <a class="reference internal" href="#quickstart"><em>Quickstart</em></a> guide.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="CGATInstallation.html">Installation instructions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="CGATInstallation.html#quick-installation">Quick installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATInstallation.html#manual-installation">Manual installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATInstallation.html#cgat-code-clean-installation">CGAT Code Clean Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATInstallation.html#installing-in-galaxy">Installing in Galaxy</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CGATUsage.html">Using CGAT Tools</a><ul>
<li class="toctree-l2"><a class="reference internal" href="CGATUsage.html#command-line-usage">Command line usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATUsage.html#indexing-genomes">Indexing genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATUsage.html#pipeline-usage">Pipeline usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CGATRecipes.html">Using CGAT tools - Recipes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="recipes/gat.html">Testing for Functional enrichment</a></li>
<li class="toctree-l2"><a class="reference internal" href="recipes/nfkb_profile.html">What is the binding profile of NFKB across gene models?</a></li>
<li class="toctree-l2"><a class="reference internal" href="recipes/lncrna_cpg.html">Assessing CpG content in long non-coding RNA promoters</a></li>
<li class="toctree-l2"><a class="reference internal" href="recipes/metagenome_contigs_kmers.html">Clustering metagenomic contigs on tetranucleotide frequency</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CGATReference.html">Tool reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="CGATReference.html#genomic-intervals-features">Genomic intervals/features</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATReference.html#gene-sets">Gene sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATReference.html#sequence-data">Sequence data</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATReference.html#ngs-data">NGS data</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATReference.html#variants">Variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATReference.html#genomics">Genomics</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CGATMissionStatement.html">Background</a><ul>
<li class="toctree-l2"><a class="reference internal" href="CGATMissionStatement.html#mission-statement">Mission statement</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATMissionStatement.html#other-toolkits-with-similar-functionality">Other toolkits with similar functionality</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="developing.html">Contributing to CGAT code</a><ul>
<li class="toctree-l2"><a class="reference internal" href="developing.html#checklist-for-new-scripts-modules">Checklist for new scripts/modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="developing.html#building-extensions">Building extensions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CGATRelease.html">Release Notes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="CGATRelease.html#release-0-2">Release 0.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATRelease.html#release-0-1-9">Release 0.1.9</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATRelease.html#release-0-1-8">Release 0.1.8</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATRelease.html#contributors">Contributors</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATRelease.html#rd-party-code">3rd party code</a></li>
<li class="toctree-l2"><a class="reference internal" href="CGATRelease.html#licence">Licence</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="quickstart">
<span id="id1"></span><h2>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h2>
<p>To install the CGAT tools, type:</p>
<div class="highlight-python"><div class="highlight"><pre>pip install cgat
</pre></div>
</div>
<p>This will install the CGAT scripts and libraries together with the
required dependencies. See <a class="reference internal" href="CGATInstallation.html#cgatinstallation"><em>Installation instructions</em></a> for
dependencies and troubleshooting.</p>
<p>CGAT tools are run from the unix command line. Lets assume we have
the results of the binding locations of a ChIP-Seq experiment
(chipseq.hg19.bed) in bed format and we want to know, how many
binding locations are intronic, intergenic and within exons.</p>
<p>Thus, we need to create a set of genomic annotations denoting
intronic, intergenic regions, etc. with respect to a reference gene set.
Here, we download the GENCODE geneset (Harrow et al., 2012) in GTF
format from ENSEMBL (Flicek et al., 2013).</p>
<p>The following unix statement downloads the ENSEMBL gene set containing
over-lapping transcripts, and outputs a set of non-overlapping genomic
annotations in gff format (<tt class="file docutils literal"><span class="pre">annotations.gff</span></tt>) by piping the data
through various GAT tools:</p>
<div class="highlight-python"><div class="highlight"><pre>wget -qO- ftp://ftp.ensembl.org/pub/release-72/gtf/homo_sapiens/Homo_sapiens.GRCh37.72.gtf.gz
| gunzip
| awk &#39;$2 == &quot;protein_coding&quot;&#39;
| cgat gff2ff --genome-file=hg19 --sanitize=ucsc --skip-missing
| cgat gtf2gtf --sort=gene
| cgat gtf2gtf --merge-exons --with-utr
| cgat gtf2gtf --filter=longest-gene
| cgat gtf2gtf --sort=position
| cgat gtf2gff --genome-file=hg19 --flank=5000 --method=genome
| gzip
&gt; annotations.gff.gz
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>The statements above need an indexed genome. To create such an
indexed genome for hg19, type the following:</p>
<p class="last">wget <a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz</a>
| index_fasta.py hg19 - &gt; hg19.log</p>
</div>
<p>CGAT tools can be chained into a single work flow using unix
pipes. The above sequence of commands in turn (1) reconciles UCSC and
ENSEMBL naming schemes for chromosome names, (2) merges all exons of
alternative transcripts per gene, (3) keeps the longest gene in case
of overlapping genes and (4) annotates exonic, intronic, intergenic
and flanking region (size=5kb) within and between genes.</p>
<p>Note that the creation of <tt class="file docutils literal"><span class="pre">annotations.gff.gz</span></tt> goes beyond
simple interval intersection, as gene structures have to be normalized
from multiple possible alternative transcripts to a single transcript
that is chosen by the user depending on what is most relevant for the
analysis.</p>
<p>Choosing different options can provide different sets of
answers. Instead of merging all exons per gene, the longest transcript
might be selected by replacing (2) with <tt class="docutils literal"><span class="pre">gtf2gtf</span>
<span class="pre">--filter=longest-transcript</span></tt>.
Or, instead of genomic annotations, regulatory domains such as defined by GREAT might be obtained by
removing (3) and replacing (4) with <tt class="docutils literal"><span class="pre">gtf2gff</span> <span class="pre">--method=great-domains</span></tt>.</p>
<p>The generated annotations in annotations.gff can then be used to count
the number of transcription factor binding sites using bed-tools or
other interval intersections. Here, we will use another CGAT tool,
<tt class="docutils literal"><span class="pre">gtf2table</span></tt>, to do the counting and classification:</p>
<div class="highlight-python"><div class="highlight"><pre>zcat /ifs/devel/gat/tutorial/data/srf.hg19.bed
| cgat bed2gff --as-gtf
| cgat gtf2table --counter=classifier-chipseq --filename-gff=annotations.gff.gz
</pre></div>
</div>
<p>The scripts follow a consistent naming scheme centered around common
genomic formats. Because of the common genomic formats, the tools can
be easily combined with other tools such as <a class="reference external" href="http://bedtools.readthedocs.org/en/latest/">bedtools</a> (Quinlan and
Hall, 2010) or <a class="reference external" href="http://genome.ucsc.edu/admin/git.html">UCSC Tools</a> (Kuhn et al. 2013).</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">CGAT - Computational Genomics Analysis Tools</a><ul>
<li><a class="reference internal" href="#quickstart">Quickstart</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/cgat.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>