
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tool reference &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/sorttable.js"></script>
    <script type="text/javascript" src="_static/jquery.min.js"></script>
    <script type="text/javascript" src="_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="index.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tool-reference">
<h1>Tool reference<a class="headerlink" href="#tool-reference" title="Permalink to this headline">¶</a></h1>
<p>This page summarizes prominent tools within the CGAT Code
collection. The tools are grouped losely by functionality.</p>
<div class="section" id="genomic-intervals-features">
<h2>Genomic intervals/features<a class="headerlink" href="#genomic-intervals-features" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><a class="reference internal" href="scripts/beds2counts.html"><em>beds2counts.py - compute overlap stats between multiple bed files</em></a></dt>
<dd>Compute overlap statistics of multiple <a class="reference internal" href="glossary.html#term-bed"><em class="xref std std-term">bed</em></a> files.</dd>
<dt><a class="reference internal" href="scripts/bed2fasta.html"><em>bed2fasta.py - get sequences from bed file</em></a></dt>
<dd>Transform interval data in a <a class="reference internal" href="glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file into a
<a class="reference internal" href="glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> formatted file of sequence data.</dd>
<dt><a class="reference internal" href="scripts/bed2gff.html"><em>bed2gff.py - convert bed to gff/gtf</em></a></dt>
<dd>Convert between interval data. Convert a <a class="reference internal" href="glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted
file to a <a class="reference internal" href="glossary.html#term-gff"><em class="xref std std-term">gff</em></a> or <a class="reference internal" href="glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file.</dd>
<dt><a class="reference internal" href="scripts/gff2gff.html"><em>gff2gff.py - manipulate gff files</em></a></dt>
<dd>Work on <a class="reference internal" href="glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted files with genomic features. This
tools sorts/renames feature files, reconciles chromosome names,
and more.</dd>
<dt><a class="reference internal" href="scripts/bed2bed.html"><em>bed2bed.py - manipulate bed files</em></a></dt>
<dd>Filter or merge interval data in a <a class="reference internal" href="glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file.</dd>
<dt><a class="reference internal" href="scripts/bed2graph.html"><em>bed2graph.py - compute the overlap graph between two bed files</em></a></dt>
<dd>Compare two sets of genomic intervals and output a list of
overlapping features.</dd>
<dt><a class="reference internal" href="scripts/bed2stats.html"><em>bed2stats.py - summary of bed file contents</em></a></dt>
<dd>Compute summary statistics of genomic intervals.</dd>
<dt><a class="reference internal" href="scripts/bed2table.html"><em>bed2table.py - annotate intervals</em></a></dt>
<dd>Annotate genomic intervals (composition, peak location, overlap, ...)</dd>
<dt><a class="reference internal" href="scripts/beds2beds.html"><em>beds2beds.py - decompose bed files</em></a></dt>
<dd>Decompose multiple sets of genomic intervals into various
intersections and unions.</dd>
<dt><a class="reference internal" href="scripts/diff_bed.html"><em>diff_bed.py - count differences between several bed files</em></a></dt>
<dd>Compare multiple sets of interval data sets. The tools computes
all-vs-all pairwise overlap summaries. Permits incremental updates
of similarity table.</dd>
<dt><a class="reference internal" href="scripts/gff2bed.html"><em>gff2bed.py - convert from gff to bed</em></a></dt>
<dd>Convert between formats</dd>
<dt><a class="reference internal" href="scripts/split_gff.html"><em>gff2chunks.py - split a gff file into chunks</em></a></dt>
<dd>Split a file in gff format into smaller files. The script ensures
that overlapping intervals remain in the same file.</dd>
<dt><a class="reference internal" href="scripts/gff2coverage.html"><em>gff2coverage.py - compute genomic coverage of gff intervals</em></a></dt>
<dd>This script computes the genomic coverage of intervals
in a <a class="reference internal" href="glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file. The coverage is computed
per feature.</dd>
<dt><a class="reference internal" href="scripts/gff2fasta.html"><em>gff2fasta.py - output sequences from genomic features</em></a></dt>
<dd>Output genomic sequences from intervals.</dd>
<dt><a class="reference internal" href="scripts/gff2histogram.html"><em>gff2histogram.py - compute histograms from intervals in gff or bed format</em></a></dt>
<dd>Compute distributions of interval sizes, intersegmental distances
and interval ovelap from list of intervals.</dd>
<dt><a class="reference internal" href="scripts/gff2stats.html"><em>gff2stats.py - count features, etc. in gff file</em></a></dt>
<dd>Summarize features within a <a class="reference internal" href="glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file.</dd>
<dt><a class="reference internal" href="scripts/gff2psl.html"><em>gff2psl.py - convert from gff to psl</em></a></dt>
<dd>Convert between formats.</dd>
</dl>
</div>
<div class="section" id="gene-sets">
<h2>Gene sets<a class="headerlink" href="#gene-sets" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><a class="reference internal" href="scripts/gtf2gff.html"><em>gtf2gff.py - convert a transcript set to genomic features</em></a></dt>
<dd>Translate a gene set into genomic annotations such as introns,
intergenic regions, regulatory domains, etc.</dd>
<dt><a class="reference internal" href="scripts/gtf2table.html"><em>gtf2table.py - annotate genes/transcripts</em></a></dt>
<dd>Annotate transcripts in a <a class="reference internal" href="glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file. Annotations
can be in reference to a second gene set (fragments, extensions),
aligned reads (coverage, intron overrun, ...) or densities.</dd>
<dt><a class="reference internal" href="scripts/gtf2fasta.html"><em>gtf2fasta.py - annotate genomic bases from a gene set</em></a></dt>
<dd>Annotate each base in the genome according to its use within
a transcript. Outputs lists of junctions.</dd>
<dt><a class="reference internal" href="scripts/gtf2gtf.html"><em>gtf2gtf.py - manipulate transcript models</em></a></dt>
<dd>merge exons/transcripts/genes, filter transcripts/genes, rename
transcripts/genes, ...</dd>
<dt><a class="reference internal" href="scripts/gtf2tsv.html"><em>gtf2tsv.py - convert gtf file to a tab-separated table</em></a></dt>
<dd>convert gene set in <a class="reference internal" href="glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> format to tabular format.</dd>
<dt><a class="reference internal" href="scripts/gtfs2tsv.html"><em>gtfs2tsv.py - compare two genesets</em></a></dt>
<dd>Compare two gene sets - output common and unique lists of genes.</dd>
<dt><a class="reference internal" href="scripts/diff_gtf.html"><em>diff_gtf.py - compute overlap between multiple gtf files</em></a></dt>
<dd>Compare multiple gene sets. The tools computes all-vs-all pairwise
overlap of exons, bases and genes. Permits incremental updates of
similarity table.</dd>
</dl>
</div>
<div class="section" id="sequence-data">
<h2>Sequence data<a class="headerlink" href="#sequence-data" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><a class="reference internal" href="scripts/fastqs2fasta.html"><em>fastqs2fasta.py - interleave two fastq files</em></a></dt>
<dd>Interleave paired reads from two fastq files into a single fasta file.</dd>
<dt><a class="reference internal" href="scripts/index_fasta.html"><em>index_fasta.py - Index fasta formatted files</em></a></dt>
<dd>Build an index for a fasta file. Pre-requisite for many CGAT tools.</dd>
<dt><a class="reference internal" href="scripts/fasta2kmercontent.html"><em>fasta2kmercontent.py</em></a></dt>
<dd>Count kmer content in a set of <a class="reference internal" href="glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> sequences.</dd>
<dt><a class="reference internal" href="scripts/fasta2table.html"><em>fasta2table.py - analyze sequences for codon bias and other sequence properties</em></a></dt>
<dd>Compute features of sequences in <a class="reference internal" href="glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> formatted files</dd>
<dt><a class="reference internal" href="scripts/diff_fasta.html"><em>diff_fasta.py - compare contents of two fasta files</em></a></dt>
<dd>Compare two sets of sequences. Outputs missing, identical
and fragmented sequences.</dd>
<dt><a class="reference internal" href="scripts/fasta2bed.html"><em>&lt;no title&gt;</em></a></dt>
<dd>Segment sequences based on G+C content, gaps, ...</dd>
<dt><a class="reference internal" href="scripts/fastas2fasta.html"><em>fastas2fasta.py - concatenate sequences from multiple fasta files</em></a></dt>
<dd>Concatentate sequences from multiple files.</dd>
<dt><a class="reference internal" href="scripts/fasta2variants.html"><em>fasta2variants.py - create sequence variants from a set of sequences</em></a></dt>
<dd>In-silico creation of variants of protein coding
sequences.</dd>
</dl>
</div>
<div class="section" id="ngs-data">
<h2>NGS data<a class="headerlink" href="#ngs-data" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><a class="reference internal" href="scripts/bam2geneprofile.html"><em>bam2geneprofile.py - build meta-gene profile for a set of transcripts/genes</em></a></dt>
<dd>Compute meta-gene profiles from aligned reads in a <a class="reference internal" href="glossary.html#term-bam"><em class="xref std std-term">bam</em></a>
formatted file. Also accepts <a class="reference internal" href="glossary.html#term-bed"><em class="xref std std-term">bed</em></a> or <a class="reference internal" href="glossary.html#term-bigwig"><em class="xref std std-term">bigwig</em></a>
formatted files.</dd>
<dt><a class="reference internal" href="scripts/bam2bam.html"><em>bam2bam.py - modify bam files</em></a></dt>
<dd>Operate on <a class="reference internal" href="glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted files - filtering, stripping,
setting flags.</dd>
<dt><a class="reference internal" href="scripts/bam2bed.html"><em>cgat_script_template.py - template for CGAT scripts</em></a></dt>
<dd>Convert <a class="reference internal" href="glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file of genomic alignments
into genomic intervals. Permits merging of paired read data
and filtering by insert-size.</dd>
<dt><a class="reference internal" href="scripts/bam2fastq.html"><em>bam2fastq.py - output fastq files from a bam-file</em></a></dt>
<dd>Save sequence and quality information from a <a class="reference internal" href="glossary.html#term-bam"><em class="xref std std-term">bam</em></a>
formatted file.</dd>
<dt><a class="reference internal" href="scripts/bam2peakshape.html"><em>bam2peakshape.py - compute peak shape features from a bam-file</em></a></dt>
<dd>Compute read densities over a collection of intervals. Also
accepts <a class="reference internal" href="glossary.html#term-bed"><em class="xref std std-term">bed</em></a> or <a class="reference internal" href="glossary.html#term-bigwig"><em class="xref std std-term">bigwig</em></a> formatted files.</dd>
<dt><a class="reference internal" href="scripts/bam2stats.html"><em>bam2stats.py - compute stats from a bam-file</em></a></dt>
<dd>Compute summary statistics of a <a class="reference internal" href="glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file.</dd>
<dt><a class="reference internal" href="scripts/bam2wiggle.html"><em>bam2wiggle.py - convert bam to wig/bigwig file</em></a></dt>
<dd>Convert read coverage in a <a class="reference internal" href="glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file into
a <a class="reference internal" href="glossary.html#term-wiggle"><em class="xref std std-term">wiggle</em></a> or <a class="reference internal" href="glossary.html#term-bigwig"><em class="xref std std-term">bigwig</em></a> formatted file.</dd>
<dt><a class="reference internal" href="scripts/bam_vs_gtf.html"><em>bam_vs_gtf.py - compare bam file against gene set</em></a></dt>
<dd>Compute stats on exon over-/underrun and spliced reads.</dd>
<dt><a class="reference internal" href="scripts/bam_vs_bed.html"><em>bam_vs_bed.py - count context that reads map to</em></a></dt>
<dd>Compute coverage of reads within multiple interval types.</dd>
<dt><a class="reference internal" href="scripts/bam_vs_bam.html"><em>bam_vs_bam.py - compute coverage correlation between bam files</em></a></dt>
<dd>Outputs side-by-side comparison of residue level counts
between multiple <a class="reference internal" href="glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted files.</dd>
<dt><a class="reference internal" href="scripts/fastq2fastq.html"><em>fastq2fastq.py - manipulate fastq files</em></a></dt>
<dd>Perform quality score conversion between <a class="reference internal" href="glossary.html#term-36"><em class="xref std std-term">fastq</em></a>
formatted files.</dd>
<dt><a class="reference internal" href="scripts/fastqs2fasta.html"><em>fastqs2fasta.py - interleave two fastq files</em></a></dt>
<dd>Interleave paired end data.</dd>
<dt><a class="reference internal" href="scripts/fastq2table.html"><em>fastq2table.py - compute stats on reads in fastq files</em></a></dt>
<dd>Output bases below quality threshold, number of N&#8217;s, quality score distribution.</dd>
<dt><a class="reference internal" href="scripts/fastqs2fastqs.html"><em>fastqs2fastqs.py - manipulate (merge/reconcile) fastq files</em></a></dt>
<dd>Ensure that paired read <a class="reference internal" href="glossary.html#term-36"><em class="xref std std-term">fastq</em></a> formatted files are consistent
after filtering on the individual files.</dd>
<dt><a class="reference internal" href="scripts/diff_bam.html"><em>diff_bam.py - compare multiple bam files against each other</em></a></dt>
<dd>Perform read-by-read comparison of two bam-files.</dd>
</dl>
</div>
<div class="section" id="variants">
<h2>Variants<a class="headerlink" href="#variants" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><a class="reference internal" href="scripts/vcf2vcf.html"><em>vcf2vcf.py - manipulate vcf files</em></a></dt>
<dd>Sort a vcf file.</dd>
</dl>
</div>
<div class="section" id="genomics">
<h2>Genomics<a class="headerlink" href="#genomics" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt><a class="reference internal" href="scripts/diff_chains.html"><em>diff_chains.py - compare to chain formatted files</em></a></dt>
<dd>How many residues to the same locations, do different locations,
etc.</dd>
<dt><a class="reference internal" href="scripts/chain2stats.html"><em>chain2stats.py</em></a></dt>
<dd>Output coverage statistics for a UCSC liftover chain file.</dd>
</dl>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Tool reference</a><ul>
<li><a class="reference internal" href="#genomic-intervals-features">Genomic intervals/features</a></li>
<li><a class="reference internal" href="#gene-sets">Gene sets</a></li>
<li><a class="reference internal" href="#sequence-data">Sequence data</a></li>
<li><a class="reference internal" href="#ngs-data">NGS data</a></li>
<li><a class="reference internal" href="#variants">Variants</a></li>
<li><a class="reference internal" href="#genomics">Genomics</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/CGATReference.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>