
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Style Guide &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/sorttable.js"></script>
    <script type="text/javascript" src="_static/jquery.min.js"></script>
    <script type="text/javascript" src="_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="index.html" />
    <link rel="next" title="Documentation" href="documenting.html" />
    <link rel="prev" title="Testing" href="testing.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="documenting.html" title="Documentation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="testing.html" title="Testing"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="style-guide">
<span id="styleguide"></span><h1>Style Guide<a class="headerlink" href="#style-guide" title="Permalink to this headline">¶</a></h1>
<div class="section" id="coding-style">
<h2>Coding style<a class="headerlink" href="#coding-style" title="Permalink to this headline">¶</a></h2>
<p>This style guide lays down coding conventions in the CGAT repository.
For new scripts, follow the guidelines below.</p>
<p>As the repository has grown over years and several people contributed,
the style between scripts can vary. For older scripts, follow the style within a
script/module. If you want to apply the newer style, make consistent
changes across the script.</p>
<p>In general, we want to adhere to the following conventions:</p>
<blockquote>
<div><ul>
<li><p class="first"><em>Variable names</em> are lower case throughout with underscores to
separate words, such as <tt class="docutils literal"><span class="pre">peaks_in_interval</span> <span class="pre">=</span> <span class="pre">0</span></tt></p>
</li>
<li><dl class="first docutils">
<dt><em>Function names</em> start with a lower case character and a</dt>
<dd><p class="first last">verb. Additional words start in upper case, such as
<tt class="docutils literal"><span class="pre">doSomethingWithData()</span></tt></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><em>Class names</em> start with an upper case character, additional words</dt>
<dd><p class="first last">start again in upper case, such as <tt class="docutils literal"><span class="pre">class</span> <span class="pre">AFancyClass():</span></tt></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><em>Class methods</em> follow the same convention as functions, such as</dt>
<dd><p class="first last"><tt class="docutils literal"><span class="pre">self.calculateFactor()</span></tt></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><em>Class attributes</em> follow the same convention as variables, such</dt>
<dd><p class="first last">as <tt class="docutils literal"><span class="pre">self.factor</span></tt></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><em>Global variables</em> - in the rare cases they are used, are upper case</dt>
<dd><p class="first last">throughout such as <tt class="docutils literal"><span class="pre">DEBUG=False</span></tt></p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><em>Module names</em> should start with an uppercase letter, for example,</dt>
<dd><p class="first last"><tt class="docutils literal"><span class="pre">TreeTools.py</span></tt> in order to distinguish them from built-in
and third-party python modules.</p>
</dd>
</dl>
</li>
<li><p class="first"><em>Script names</em> are lower-case throughout with underscores to
separate words, for example, <tt class="docutils literal"><span class="pre">bam2geneprofile.py</span></tt> or
<tt class="docutils literal"><span class="pre">join_table.py</span></tt>.</p>
</li>
<li><p class="first"><em>Cython extensions</em> to scripts (via pyximport) should be put
into the script name starting with an underscore. For example,
The extensions to <tt class="docutils literal"><span class="pre">bam2geneprofile.py</span></tt> are in
<tt class="docutils literal"><span class="pre">_bam2geneprofile.pyx</span></tt>.</p>
</li>
</ul>
</div></blockquote>
<p>For new scripts, use the template <tt class="file docutils literal"><span class="pre">script_template.py</span></tt>.</p>
<p>The general rule is to write easily readable and maintainable
code. Thus, please</p>
<blockquote>
<div><ul>
<li><p class="first">document code liberally and accurately</p>
</li>
<li><dl class="first docutils">
<dt>make use of whitespaces and line-breaks to break long statements</dt>
<dd><p class="first last">into easily readable statements.</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<p>In case of uncertainty, follow the python style guides as much as
possible. The relevant documents are:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference external" href="http://www.python.org/dev/peps/pep-0008/">PEP0008 - Style Guide for Python Code</a></li>
<li><a class="reference external" href="http://www.python.org/dev/peps/pep-0257/">PEP0257 - Docstring Conventions</a></li>
</ul>
</div></blockquote>
<p>For documenting CGAT code, we follow the conventions for documenting
python code:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference external" href="http://docs.python.org/devguide/documenting.html">Python Developer&#8217;s guide</a></li>
</ul>
</div></blockquote>
<p>In terms of writing scripts, we follow the following conventions:</p>
<blockquote>
<div><ul>
<li><p class="first">Each script should define the <tt class="docutils literal"><span class="pre">-h</span></tt> and <tt class="docutils literal"><span class="pre">--help</span></tt> options to
give command line help usage.</p>
</li>
<li><p class="first">For tabular output, scripts should output <a class="reference internal" href="glossary.html#term-tsv"><em class="xref std std-term">tsv</em></a> formatted
tables. In these tables, records are separated by new-line
characters and fields by tab characters. Lines with comments are started
by the <tt class="docutils literal"><span class="pre">#</span></tt> character and are ignored. The first uncommented line
should contain the column headers. For example:</p>
<div class="highlight-python"><div class="highlight"><pre># This is a comment
gene_id length
gene1   1000
gene2   2000
# Another comment
</pre></div>
</div>
</li>
<li><p class="first">Scripts should follow the
<a class="reference external" href="http://en.wikipedia.org/wiki/Unix_philosophy">unix philosophy</a>.
They should concentrate on one task and do it well. Ideally,
the major input and output can be read from and written to standard
input and standard output, respectively.</p>
</li>
<li><p class="first">The names of scripts should be meaningful. Most of our scripts
perform data transformation of one kind of another, these are
often called <tt class="docutils literal"><span class="pre">a2b.py</span></tt>. The distinctions can be subtle.
Examples are:</p>
<dl class="docutils">
<dt><a class="reference internal" href="scripts/gtf2gtf.html"><em>gtf2gtf.py - manipulate transcript models</em></a></dt>
<dd><p class="first last">Input is <a class="reference internal" href="glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a>, output is <a class="reference internal" href="glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a>. This script
manipulates gene sets (filtering, merging, ...).</p>
</dd>
<dt><a class="reference internal" href="scripts/gtf2gff.html"><em>gtf2gff.py - convert a transcript set to genomic features</em></a></dt>
<dd><p class="first last">Input is <a class="reference internal" href="glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a>, output is <a class="reference internal" href="glossary.html#term-gff"><em class="xref std std-term">gff</em></a>. This script
takes gene sets and changes the hierarchical description
within a <a class="reference internal" href="glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> file to the flat description of features
in a <a class="reference internal" href="glossary.html#term-gff"><em class="xref std std-term">gff</em></a> file. For example, this script can define
gene territories, regulatory domains or genomic annotations
based on a gene set.</p>
</dd>
<dt><a class="reference internal" href="scripts/bed2gff.html"><em>bed2gff.py - convert bed to gff/gtf</em></a></dt>
<dd><p class="first last">Input is <a class="reference internal" href="glossary.html#term-bed"><em class="xref std std-term">bed</em></a>, output is <a class="reference internal" href="glossary.html#term-gff"><em class="xref std std-term">gff</em></a>. As both
formats describe intervals in the genome, this script
basically does a conversion between the two formats.</p>
</dd>
</dl>
<p>Quite a few scripts contain the <tt class="docutils literal"><span class="pre">2table</span></tt> or <tt class="docutils literal"><span class="pre">2stats</span></tt>. These
compute, respectively, properties or summary statistics for
entries in a file. For example:</p>
<dl class="docutils">
<dt><a class="reference internal" href="scripts/gtf2table.html"><em>gtf2table.py - annotate genes/transcripts</em></a></dt>
<dd><p class="first last">Input is <a class="reference internal" href="glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a>. For each gene or transcript, compute
selected properties. If there are 10,000 genes in the input,
the output table will contain 10,000 rows.</p>
</dd>
<dt><a class="reference internal" href="scripts/gff2stats.html"><em>gff2stats.py - count features, etc. in gff file</em></a></dt>
<dd><p class="first last">Input is <a class="reference internal" href="glossary.html#term-gff"><em class="xref std std-term">gff</em></a>. Compute summary statistics across
all features in the file. Here, aggregate sizes or similar
by feature type or name per chromosome. No matter if there
are 10,000 or 100,000 interval is the input, the output
will be have the same number of rows.</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="where-to-put-code">
<h2>Where to put code<a class="headerlink" href="#where-to-put-code" title="Permalink to this headline">¶</a></h2>
<p>Different parts of the code base go into separate directories.</p>
<dl class="docutils">
<dt>Scripts</dt>
<dd>Scripts are python code that contains a main() function and
are intended to be executed. Scripts go into the directory
<tt class="file docutils literal"><span class="pre">/scripts</span></tt></dd>
<dt>Modules</dt>
<dd>Modules contain supporting code and are imported by scripts or
other modules. Modules go into the directory <tt class="file docutils literal"><span class="pre">/CGAT</span></tt>.</dd>
<dt>Pipelines</dt>
<dd>Pipeline scripts and modules go into the directory <tt class="file docutils literal"><span class="pre">/CGATPipelines</span></tt>.</dd>
</dl>
</div>
<div class="section" id="pipelines">
<h2>Pipelines<a class="headerlink" href="#pipelines" title="Permalink to this headline">¶</a></h2>
<p>All components of a pipeline should go into the <tt class="file docutils literal"><span class="pre">CGATPipelines</span></tt>
directory. The basic layout of a pipeline is:</p>
<div class="highlight-python"><div class="highlight"><pre>CGATPipelines/pipeline_example.py
             /PipelineExample.py
             /PipelineExample.R
             /pipeline_example/pipeline.ini
                              /conf.py
                              /sphinxreport.ini
</pre></div>
</div>
<dl class="docutils">
<dt>pipeline_example.py</dt>
<dd>The main pipeline code. Pipelines start with the word <tt class="docutils literal"><span class="pre">pipeline</span></tt>
and follow the conventions for <em>script names</em>, all lower case with
underscores separating words.</dd>
<dt>pipeline_example/pipeline.ini</dt>
<dd>Default values for pipeline configuration values.</dd>
<dt>pipeline_example/conf.py</dt>
<dd>Configuration script for sphinxreport.</dd>
<dt>pipeline_example/sphinxreport.ini</dt>
<dd>Configuration script for sphinxreport.</dd>
<dt>pipeline_docs/pipeline_example</dt>
<dd>Sphinxreport for pipeline.</dd>
<dt>PipelineExample.py</dt>
<dd>Python utility methods and classes specific to this pipeline. Once
methods and classes are shared between pipelines, consider moving
them to a separate module.</dd>
<dt>PipelineExample.R</dt>
<dd>R utility functions specific to this pipeline.</dd>
</dl>
<ul class="simple">
<li>Make sure that the pipeline.ini file exists and contains example/default
values with annotation.</li>
<li>Make sure that the pipeline can be imported from any directory,
especially those not containing any data files or configuration
files. This is important for the documentation of the pipeline
to be built.</li>
</ul>
</div>
<div class="section" id="other-guidelines">
<h2>Other guidelines<a class="headerlink" href="#other-guidelines" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Only add source code and required data to the repository. Do
not add .pyc files, backup files created by your editor or other
files.</li>
<li>In order to build documentation, each script, module and pipeline needs to
be importable. Thus, make sure that when your pipeline depends on
specific files, it does not fail when imported but not executed.</li>
</ul>
</div>
<div class="section" id="documentation">
<h2>Documentation<a class="headerlink" href="#documentation" title="Permalink to this headline">¶</a></h2>
<div class="section" id="writing-doc-strings">
<h3>Writing doc-strings<a class="headerlink" href="#writing-doc-strings" title="Permalink to this headline">¶</a></h3>
<p>Functions should be documented through their doc-string using
restructured text. For example:</p>
<div class="highlight-python"><div class="highlight"><pre>def computeValue( name, method, accuracy=2):

    :param name: The name to use.
    :type name: str.
    :param method: method to use.
    :type state: choice of (&#39;empirical&#39;, &#39;parametric&#39;)
    :param accuracy:
    :type accuracy: integer
    :returns:  int -- the value
    :raises: AttributeError, KeyError
</pre></div>
</div>
</div>
<div class="section" id="writing-documentation-for-scripts">
<h3>Writing documentation for scripts<a class="headerlink" href="#writing-documentation-for-scripts" title="Permalink to this headline">¶</a></h3>
<p>There is a minimum standard for documentation to maintain clarity of
tools and code.  The documentation for any given script should follow
the basic outline in <a class="reference internal" href="scripts/cgat_script_template.html"><em>&lt;no title&gt;</em></a>.</p>
<p>Three main headers exist:</p>
<dl class="docutils">
<dt><cite>Purpose</cite></dt>
<dd><p class="first">Describe the overall purpose and function of the script and the
input and output formats.  This can be extensive and include
sub-headers to further describe script functionality.
For example:</p>
<div class="last highlight-python"><div class="highlight"><pre>Purpose
-------
This script takes a :term:`gtf` formatted file and computes meta-gene profiles
over various annotations derived from the :term:`gtf` file.

A meta-gene profile is an abstract genomic entity over which reads stored in a
:term:`bam` formatted file have been counted. A meta-gene might be an idealized
eukaryotic gene (upstream, exonic sequence, downstream) or any other genomic landmark
of interest such as transcription start sites.
</pre></div>
</div>
</dd>
<dt><cite>Usage</cite></dt>
<dd><p class="first">Describe example use cases for the script with one or more options.  In addition provide
the head of both example input and example output files.
Example input and output:</p>
<div class="highlight-python"><div class="highlight"><pre>Usage
-----
    samtools view example.bam

    READ1    163    1     13040   15      76M     =       13183   219     ...
    READ1    83     1     13183   7       76M     =       13040   -219    ...
    READ2    147    1     13207   0       76M     =       13120   -163    ...

    python bam2bed.py example.bam

    1       13039   13115   READ1     15      +
    1       13119   13195   READ2     0       +
    1       13182   13258   READ1     7       -
    1       13206   13282   READ2     0       -
</pre></div>
</div>
<p>Example usage:</p>
<div class="last highlight-python"><div class="highlight"><pre>python example_script.py
      --infile=example.bam
      --option1=choice
      --method=method1
</pre></div>
</div>
</dd>
<dt><cite>Options</cite></dt>
<dd><p class="first">Describe all of the options for the script.  If necessary provide extensive detail of the
methods of each option and how they are combined to provide the intended functionality
of the script.  This should include all <cite>choice</cite> for options with a verbose description
of what that <cite>choice</cite> does.
For example:</p>
<div class="last highlight-python"><div class="highlight"><pre>Profiles
--------
Different profiles are accessible through the ``--method`` option. Multiple
methods can be applied at the same time. While ``upstream`` and ``downstream``
typically have a fixed size, the other regions such as ``CDS``, ``UTR`` will be
scaled to a common size.

utrprofile
    UPSTREAM - UTR5 - CDS - UTR3 - DOWNSTREAM
    gene models with UTR. Separate the coding section from the non-coding part.
</pre></div>
</div>
</dd>
</dl>
<p>There is a fourth template-specific header; the command line options that are automatically
generated for every CGAT script:</p>
<dl class="docutils">
<dt><cite>Command line options</cite></dt>
<dd>These are automatically generated from <a class="reference internal" href="scripts/cgat_script_template.html"><em>&lt;no title&gt;</em></a> and
detail each option specified within the script.  No further details need to be
added to this section.</dd>
</dl>
<p>In addition, please pay attention to the following:</p>
<ul>
<li><p class="first">Declare input data types for genomic data sets in optparse using
the <cite>metavar</cite> keyword. For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">parser</span><span class="o">.</span><span class="n">add_option</span><span class="p">(</span> <span class="s">&quot;--extra-intervals&quot;</span><span class="p">,</span> <span class="n">dest</span> <span class="o">=</span> <span class="s">&quot;extra_intervals&quot;</span><span class="p">,</span>
                <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;bed&quot;</span><span class="p">,</span> <span class="n">help</span> <span class="o">=</span> <span class="s">&quot;...&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<p>Setting the type permits the script to be integrated into workflow
sytemns such as <a class="reference external" href="https://main.g2.bx.psu.edu/">galaxy</a>.</p>
</li>
<li><p class="first">Please provide a meaningful example in the command line help (see above for
minimum requirements).</p>
</li>
<li><p class="first">Be verbose. Something that is not documented within a script
will not be used.</p>
</li>
<li><p class="first">Add meaningful tags to your scripts (<tt class="docutils literal"><span class="pre">:Tags:</span></tt>) so that they can be grouped into
categories. Please choose from the following controlled
vocabulary. If needed, additional terms can be added to this list.</p>
<ul>
<li><p class="first">Broad Themes</p>
<blockquote>
<div><ul class="simple">
<li>Genomics</li>
<li>NGS</li>
<li>MultipleAlignment</li>
<li>GenomeAlignment</li>
<li>Intervals</li>
<li>Genesets</li>
<li>Sequences</li>
<li>Variants</li>
<li>Protein</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Formats</p>
<blockquote>
<div><ul class="simple">
<li>BAM</li>
<li>BED</li>
<li>GFF</li>
<li>GTF</li>
<li>FASTA</li>
<li>FASTQ</li>
<li>WIGGLE</li>
<li>PSL</li>
<li>CHAIN</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Actions</p>
<blockquote>
<div><ul class="simple">
<li>Summary - summarizing entities within a file, such as
counting the number of intervals within a file, etc.</li>
<li>Annotation - annotating individual entities within a file,
such as adding length, composition, etc. to intervals.</li>
<li>Comparison - comparing the same type of entities, such as
overlapping to sets of intervals.</li>
<li>Conversion - converting between different formats for the
similar types of objects (Intervals in gff/bed format).</li>
<li>Transformation - transforming one entity into another, such
as transforming intervals into sequences.</li>
<li>Manipulation - changing entities within a file, such as
filtering sequences.</li>
</ul>
</div></blockquote>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Style Guide</a><ul>
<li><a class="reference internal" href="#coding-style">Coding style</a></li>
<li><a class="reference internal" href="#where-to-put-code">Where to put code</a></li>
<li><a class="reference internal" href="#pipelines">Pipelines</a></li>
<li><a class="reference internal" href="#other-guidelines">Other guidelines</a></li>
<li><a class="reference internal" href="#documentation">Documentation</a><ul>
<li><a class="reference internal" href="#writing-doc-strings">Writing doc-strings</a></li>
<li><a class="reference internal" href="#writing-documentation-for-scripts">Writing documentation for scripts</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="testing.html"
                        title="previous chapter">Testing</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="documenting.html"
                        title="next chapter">Documentation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/styleguide.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="documenting.html" title="Documentation"
             >next</a> |</li>
        <li class="right" >
          <a href="testing.html" title="Testing"
             >previous</a> |</li>
        <li><a href="index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>