
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PipelineLncRNA.py - functions and classes for use with the lincRNA pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Modules" href="../modules.html" />
    <link rel="next" title="PipelineMapping.py - Utility functions for mapping short reads" href="PipelineMapping.html" />
    <link rel="prev" title="PipelineKegg.py - Pipeline components related to KEGG analysis" href="PipelineKEGG.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PipelineMapping.html" title="PipelineMapping.py - Utility functions for mapping short reads"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="PipelineKEGG.html" title="PipelineKegg.py - Pipeline components related to KEGG analysis"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" accesskey="U">Modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-PipelineLncRNA"></span><div class="section" id="pipelinelncrna-py-functions-and-classes-for-use-with-the-lincrna-pipeline">
<h1>PipelineLncRNA.py - functions and classes for use with the lincRNA pipeline<a class="headerlink" href="#pipelinelncrna-py-functions-and-classes-for-use-with-the-lincrna-pipeline" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="PipelineLncRNA.buildCodingGeneSet">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">buildCodingGeneSet</tt><span class="sig-paren">(</span><em>abinitio_coding</em>, <em>reference</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.buildCodingGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>takes the output from cuffcompare of a transcript
assembly and filters for annotated protein coding
genes.</p>
<p>NB &#8220;pruned&#8221; refers to nomenclature in the transcript
building pipeline - transcripts that appear in at least
two samples.</p>
<p>Because an abinitio assembly will often contain
fragments of known transcripts and describe them as 
novel, the default behaviour is to produce a set that
is composed of &#8216;complete&#8217; transcripts</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.buildRefcodingGeneSet">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">buildRefcodingGeneSet</tt><span class="sig-paren">(</span><em>coding_set</em>, <em>refcoding_set</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.buildRefcodingGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>takes genes from an ab initio assembly and filters a reference coding set
for these genes. Allows for comparisons of known transcripts for those genes
that are assembled ab initio. Does this by gene name</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.buildRefnoncodingGeneSet">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">buildRefnoncodingGeneSet</tt><span class="sig-paren">(</span><em>reference</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.buildRefnoncodingGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>filter the refnoncoding geneset for things that are described in ensembl
as being:
Ambiguous_orf
Retained_intron
Sense_intronic
antisense
Sense_overlapping
Processed transcript</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.buildLncRNAGeneSet">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">buildLncRNAGeneSet</tt><span class="sig-paren">(</span><em>abinitio_lincrna</em>, <em>reference</em>, <em>refnoncoding</em>, <em>pseudogenes_gtf</em>, <em>numts_gtf</em>, <em>outfile</em>, <em>min_length</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.buildLncRNAGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build lncRNA gene set.</p>
<p>In contrast to the pipeline, this lincRNA set does not contain
the reference noncoding gene set. It is transcripts in the abinitio set that
do not overlap at any protein coding, processed or pseudogene transcripts 
(exons+introns) in a reference gene set.
lincRNA genes are often expressed at low level and thus the resultant transcript
models are fragmentory. To avoid some double counting in downstream analyses
transcripts overlapping on the same strand are merged - this does not neccessarily seem to work well if 
not using the reference set.</p>
<p>Transcripts need to have a length of at least 200 bp.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.buildFilteredLncRNAGeneSet">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">buildFilteredLncRNAGeneSet</tt><span class="sig-paren">(</span><em>flagged_gtf</em>, <em>outfile</em>, <em>genesets_previous</em>, <em>filter_se='transcripts'</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.buildFilteredLncRNAGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>creates a filtered lincRNA geneset. This geneset will not include any
single exon lincRNA unless they have been seen previously i.e. it overlaps
a previously identified lincRNA</p>
<p>At this point we add a flag for whether the gene is a novel lncRNA or not
NB this is a large function and should be modified in the future</p>
<p>note genesets_previous provided as a list - priority is
placed on the first in the list</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.buildFinalLncRNAGeneSet">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">buildFinalLncRNAGeneSet</tt><span class="sig-paren">(</span><em>filteredLncRNAGeneSet</em>, <em>cpc_table</em>, <em>outfile</em>, <em>filter_cpc=False</em>, <em>cpc_threshold=1</em>, <em>rename_lncRNA=False</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.buildFinalLncRNAGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>filters lncRNA set based on the coding potential as output from 
the CPC</p>
</dd></dl>

<dl class="class">
<dt id="PipelineLncRNA.CounterExons">
<em class="property">class </em><tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">CounterExons</tt><span class="sig-paren">(</span><em>gtffile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.CounterExons" title="Permalink to this definition">¶</a></dt>
<dd><p>various classes for counting features in a gtf file</p>
</dd></dl>

<dl class="class">
<dt id="PipelineLncRNA.CounterExonsPerTranscript">
<em class="property">class </em><tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">CounterExonsPerTranscript</tt><span class="sig-paren">(</span><em>gtffile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.CounterExonsPerTranscript" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#PipelineLncRNA.CounterExons" title="PipelineLncRNA.CounterExons"><tt class="xref py py-class docutils literal"><span class="pre">PipelineLncRNA.CounterExons</span></tt></a></p>
<p>returns the average number of exons per transcript</p>
</dd></dl>

<dl class="class">
<dt id="PipelineLncRNA.CounterExonsPerGene">
<em class="property">class </em><tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">CounterExonsPerGene</tt><span class="sig-paren">(</span><em>gtffile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.CounterExonsPerGene" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#PipelineLncRNA.CounterExons" title="PipelineLncRNA.CounterExons"><tt class="xref py py-class docutils literal"><span class="pre">PipelineLncRNA.CounterExons</span></tt></a></p>
<p>returns the average number of exons per transcript</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.flagExonStatus">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">flagExonStatus</tt><span class="sig-paren">(</span><em>gtf_file</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.flagExonStatus" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds two attributes to a gtf, the first species transcript exon status,
the second specifies gene exon status. 
It is possible for genes to contain single-exon transcripts but not the 
reciprocal.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.classifyLncRNA">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">classifyLncRNA</tt><span class="sig-paren">(</span><em>lincRNA_gtf</em>, <em>reference</em>, <em>outfile</em>, <em>dist=2</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.classifyLncRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Classify lincRNA in terms of their proximity to 
protein coding genes - creates indices for intervals on the 
fly - maybe should be creating additional annotations:</p>
<p>antisense - GENE overlapping protein coding exons or introns on opposite strand
antisense_upstream - GENE &lt; Xkb from tss on opposite strand
antisense_downstream - GENE &lt; Xkb from gene end on opposite strand
sense_upstream - GENE &lt; Xkb from tss on same strand
sense_downstream - GENE &lt; Xkb from gene end on same strand
intergenic - &gt;Xkb from any protein coding gene
intronic - overlaps protein coding gene intron on same strand
antisense_intronic - overlaps protein coding intron on opposite strand</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.reClassifyLncRNAGenes">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">reClassifyLncRNAGenes</tt><span class="sig-paren">(</span><em>lncRNA_gtf</em>, <em>reference_gtf</em>, <em>outfile</em>, <em>upstr_dist=5</em>, <em>dstr_dist=5</em>, <em>wdir='.'</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.reClassifyLncRNAGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>This re-write of classifyLncRNAGenes() does not throw out intronic loci, but 
labels them as either sense-intronic or sense-overlap.
It also fixes the bug that cause sense gene-models encompassing reference 
exons to be output as antisense. 
Because lncRNA boundaries intersect multiple intervals in indexes, rather
than classifying each lncRNA multiple times, lncRNA strand is instead 
compared to a list of interval strand values, if any of these are sense, 
then the lncRNA is classified as sense etc.</p>
<p>Sense-intronic: when lncRNA loci start and end are contained within a single
intron. 
Sense-overlap: when lncRNA loci start and end are in different introns. Note 
that different introns do not necessarily come from different gene-models.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.get_spliced_region_alignment">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">get_spliced_region_alignment</tt><span class="sig-paren">(</span><em>index</em>, <em>primary_species</em>, <em>chrom</em>, <em>starts</em>, <em>ends</em>, <em>strand='+'</em>, <em>species=None</em>, <em>mincols=0</em>, <em>overwrite_with_gaps=True</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.get_spliced_region_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a filled spliced region alignment for specified region with start
and end lists.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.gtfToBed12">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">gtfToBed12</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>model</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.gtfToBed12" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a gtf file to bed12 format.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.filterMAF">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">filterMAF</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>removed</em>, <em>filter_alignments=False</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.filterMAF" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates through the MAF file. If filter_alignments == int, then will remove MAF
blocks for which length &lt; int.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.extractGeneBlocks">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">extractGeneBlocks</tt><span class="sig-paren">(</span><em>bedfile</em>, <em>maf_file</em>, <em>outfile</em>, <em>primary_species</em>, <em>secondary_species</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.extractGeneBlocks" title="Permalink to this definition">¶</a></dt>
<dd><p>Is based on Dan Blankenburg&#8217;s interval_to_maf_merged_fasta.py 
Receives a bed12 containing intervals of interest, and maf containing 
pairwise genomic alignment. Iterates through bed intervals and outputs 
intervals in fasta format.
See comments in maf_utilities.py: maf blocks are always extracted in a 
positive strand orientation relative to the src alignment, reverse 
complementing is then done using method AlignedSequence method 
get_sequence_reverse_complement.
MAF file is indexed on the fly using bx.align.maf.MultiIndexed</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.complement">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">complement</tt><span class="sig-paren">(</span><em>template</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates the reverse complement of a template sequence</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.extractMAFGeneBlocks">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">extractMAFGeneBlocks</tt><span class="sig-paren">(</span><em>bedfile</em>, <em>maf_file</em>, <em>genome_file</em>, <em>outfile</em>, <em>primary_species</em>, <em>secondary_species</em>, <em>keep_gaps=True</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.extractMAFGeneBlocks" title="Permalink to this definition">¶</a></dt>
<dd><p>Is based on Dan Blankenburg&#8217;s interval_to_maf_merged_fasta.py 
Receives a bed12 containing intervals of interest, and maf containing 
pairwise genomic alignment. Iterates through bed intervals and outputs 
intervals in fasta format.
See comments in maf_utilities.py: maf blocks are always extracted in a 
positive strand orientation relative to the src alignment, reverse 
complementing is then done using method AlignedSequence method 
get_sequence_reverse_complement.
MAF file is indexed on the fly using bx.align.maf.MultiIndexed</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.splitAlignedFasta">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">splitAlignedFasta</tt><span class="sig-paren">(</span><em>infile</em>, <em>out_stub</em>, <em>name_dict</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.splitAlignedFasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Receives a fasta file containing multiple sequence alignments. Splits into 
multiple outfiles, each containing sequence with the same interval_id.
Identifiers must be in format &gt;species.interval_id
name_dict specifies the format for the outfile identifiers</p>
</dd></dl>

<dl class="function">
<dt id="PipelineLncRNA.removeGapsFromAlignedFasta">
<tt class="descclassname">PipelineLncRNA.</tt><tt class="descname">removeGapsFromAlignedFasta</tt><span class="sig-paren">(</span><em>in_dir</em>, <em>out_dir</em>, <em>min_length=0</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineLncRNA.removeGapsFromAlignedFasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Receives a fasta file containing two or more aligned sequences, removes gaps
from the reference (first) alignment in file and removes corresponding gaps
intervals from subsequence sequence.
Any region of sequence flanked by gaps that is smaller than a specified min
length may also be removed 
WARNING: using this function will likely cause frame shifts in resulting 
sequence.</p>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h4>Previous topic</h4>
  <p class="topless"><a href="PipelineKEGG.html"
                        title="previous chapter">PipelineKegg.py - Pipeline components related to KEGG analysis</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="PipelineMapping.html"
                        title="next chapter">PipelineMapping.py - Utility functions for mapping short reads</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelinemodules/PipelineLncRNA.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PipelineMapping.html" title="PipelineMapping.py - Utility functions for mapping short reads"
             >next</a> |</li>
        <li class="right" >
          <a href="PipelineKEGG.html" title="PipelineKegg.py - Pipeline components related to KEGG analysis"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" >Modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>