
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PipelineChipseq.py - ChIP-Seq tasks associated with intervals. &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Modules" href="../modules.html" />
    <link rel="next" title="PipelineDatabase.py - utility functions for working with a database" href="PipelineDatabase.html" />
    <link rel="prev" title="PipelineBiomart.py - utility tasks for importing data from biomart" href="PipelineBiomart.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PipelineDatabase.html" title="PipelineDatabase.py - utility functions for working with a database"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="PipelineBiomart.html" title="PipelineBiomart.py - utility tasks for importing data from biomart"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" accesskey="U">Modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-PipelineChipseq"></span><div class="section" id="pipelinechipseq-py-chip-seq-tasks-associated-with-intervals">
<h1>PipelineChipseq.py - ChIP-Seq tasks associated with intervals.<a class="headerlink" href="#pipelinechipseq-py-chip-seq-tasks-associated-with-intervals" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="PipelineChipseq.getPeakShiftFromMacs">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">getPeakShiftFromMacs</tt><span class="sig-paren">(</span><em>infile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.getPeakShiftFromMacs" title="Permalink to this definition">¶</a></dt>
<dd><p>get peak shift for filename infile (.macs output file).</p>
<p>returns None if no shift found</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.getPeakShiftFromZinba">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">getPeakShiftFromZinba</tt><span class="sig-paren">(</span><em>infile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.getPeakShiftFromZinba" title="Permalink to this definition">¶</a></dt>
<dd><p>get peak shift for filename infile (.zinba output file).</p>
<p>returns None if no shift found</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.getPeakShift">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">getPeakShift</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.getPeakShift" title="Permalink to this definition">¶</a></dt>
<dd><p>get peak shift for a track.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.getMappedReads">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">getMappedReads</tt><span class="sig-paren">(</span><em>infile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.getMappedReads" title="Permalink to this definition">¶</a></dt>
<dd><p>return number of reads mapped.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.getMinimumMappedReads">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">getMinimumMappedReads</tt><span class="sig-paren">(</span><em>infiles</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.getMinimumMappedReads" title="Permalink to this definition">¶</a></dt>
<dd><p>find the minimum number of mapped reads in infiles.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.getExonLocations">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">getExonLocations</tt><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.getExonLocations" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of exon locations as Bed entries
from a file contain a one ensembl gene ID per line</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.getBedLocations">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">getBedLocations</tt><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.getBedLocations" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of regions as (contig,start,end) tuples
from a bed file</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.buildQuicksectMask">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">buildQuicksectMask</tt><span class="sig-paren">(</span><em>bed_file</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.buildQuicksectMask" title="Permalink to this definition">¶</a></dt>
<dd><p>return Quicksect object containing the regions specified
takes a bed file listing the regions to mask</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.buildBAMforPeakCalling">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">buildBAMforPeakCalling</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em>, <em>dedup</em>, <em>mask</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.buildBAMforPeakCalling" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a BAM file suitable for peak calling.</p>
<p>Infiles are merged and unmapped reads removed.</p>
<p>If specificied duplicate reads are removed. 
This method use Picard.</p>
<p>If a mask is specified, reads falling within
the mask are filtered out.</p>
<p>This uses bedtools.</p>
<p>The mask is a quicksect object containing
the regions from which reads are to be excluded.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.buildSimpleNormalizedBAM">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">buildSimpleNormalizedBAM</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em>, <em>nreads</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.buildSimpleNormalizedBAM" title="Permalink to this definition">¶</a></dt>
<dd><p>normalize a bam file to given number of counts
by random sampling</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.buildNormalizedBAM">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">buildNormalizedBAM</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em>, <em>normalize=True</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.buildNormalizedBAM" title="Permalink to this definition">¶</a></dt>
<dd><p>build a normalized BAM file.</p>
<p>Infiles are merged and duplicated reads are removed. 
If <em>normalize</em> is set, reads are removed such that all 
files will have approximately the same number of reads.</p>
<p>Note that the duplication here is wrong as there
is no sense of strandedness preserved.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.buildBAMStats">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">buildBAMStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.buildBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>calculate bamfile statistics - currently only single-ended
duplicates.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.exportIntervalsAsBed">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">exportIntervalsAsBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.exportIntervalsAsBed" title="Permalink to this definition">¶</a></dt>
<dd><p>export macs peaks as bed files.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.exportMacsIntervalsAsBed">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">exportMacsIntervalsAsBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>foldchange</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.exportMacsIntervalsAsBed" title="Permalink to this definition">¶</a></dt>
<dd><p>export sequences for all intervals.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.exportPeaksAsBed">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">exportPeaksAsBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.exportPeaksAsBed" title="Permalink to this definition">¶</a></dt>
<dd><p>export peaks as bed files.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.mergeBedFiles">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">mergeBedFiles</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.mergeBedFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>generic method for merging bed files.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.mergeIntervalsWithScores">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">mergeIntervalsWithScores</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>dist</em>, <em>method</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.mergeIntervalsWithScores" title="Permalink to this definition">¶</a></dt>
<dd><p>merge adjacent intervals (within dist) and integrate scores (using method) from merged peaks.
Assume bed file sorted by position and score in column 5.
Methods: mean, max, length weighted mean</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.intersectBedFiles">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">intersectBedFiles</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.intersectBedFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>merge <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted <em>infiles</em> by intersection
and write to <em>outfile</em>.</p>
<p>Only intervals that overlap in all files are retained.
Interval coordinates are given by the first file in <em>infiles</em>.</p>
<p>Bed files are normalized (overlapping intervals within 
a file are merged) before intersection.</p>
<p>Intervals are renumbered starting from 1.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.subtractBedFiles">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">subtractBedFiles</tt><span class="sig-paren">(</span><em>infile</em>, <em>subtractfile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.subtractBedFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>subtract intervals in <em>subtractfile</em> from <em>infile</em>
and store in <em>outfile</em>.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.summarizeMACS">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">summarizeMACS</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.summarizeMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>run MACS for peak detection.</p>
<p>This script parses the MACS logfile to extract 
peak calling parameters and results.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.summarizeMACSsolo">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">summarizeMACSsolo</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.summarizeMACSsolo" title="Permalink to this definition">¶</a></dt>
<dd><p>run MACS for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.summarizeMACSFDR">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">summarizeMACSFDR</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.summarizeMACSFDR" title="Permalink to this definition">¶</a></dt>
<dd><p>compile table with peaks that would remain after filtering
by fdr.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.loadMACS">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">loadMACS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>bamfile</em>, <em>tablename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.loadMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>load MACS results in <em>tablename</em></p>
<p>This method loads only positive peaks. It filters peaks by p-value,
q-value and fold change and loads the diagnostic data and
re-calculates peakcenter, peakval, ... using the supplied bamfile.</p>
<p>If <em>tablename</em> is not given, it will be <tt class="file docutils literal"><span class="pre">&lt;track&gt;_intervals</span></tt>
where track is derived from <tt class="docutils literal"><span class="pre">infile</span></tt> and assumed to end
in <tt class="file docutils literal"><span class="pre">.macs</span></tt>.</p>
<p>This method creates two optional additional files:</p>
<ul class="simple">
<li>if the file <tt class="file docutils literal"><span class="pre">&lt;track&gt;_diag.xls</span></tt> is present, load MACS 
diagnostic data into the table <tt class="file docutils literal"><span class="pre">&lt;track&gt;_macsdiag</span></tt>.</li>
<li>if the file <tt class="file docutils literal"><span class="pre">&lt;track&gt;_model.r</span></tt> is present, call R to
create a MACS peak-shift plot and save it as <tt class="file docutils literal"><span class="pre">&lt;track&gt;_model.pdf</span></tt>
in the <tt class="file docutils literal"><span class="pre">export/MACS</span></tt> directory.</li>
</ul>
<p>This method creates <tt class="file docutils literal"><span class="pre">&lt;outfile&gt;.tsv.gz</span></tt> with the results
of the filtering.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.runMACS">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">runMACS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>controlfile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.runMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>run MACS for peak detection from BAM files.</p>
<p>The output bed files contain the P-value as their score field.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.getCounts">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">getCounts</tt><span class="sig-paren">(</span><em>contig</em>, <em>start</em>, <em>end</em>, <em>samfiles</em>, <em>offsets=</em><span class="optional">[</span><span class="optional">]</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.getCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>count reads per position.</p>
<p>If offsets are given, shift tags by offset / 2 and extend
by offset / 2.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.countPeaks">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">countPeaks</tt><span class="sig-paren">(</span><em>contig</em>, <em>start</em>, <em>end</em>, <em>samfiles</em>, <em>offsets=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.countPeaks" title="Permalink to this definition">¶</a></dt>
<dd><p>update peak values within interval contig:start-end.</p>
<p>If offsets is given, tags are moved by the offset
before summarizing.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.runZinba">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">runZinba</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>controlfile</em>, <em>action='full'</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.runZinba" title="Permalink to this definition">¶</a></dt>
<dd><p>run Zinba for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.loadZinba">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">loadZinba</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>bamfile</em>, <em>tablename=None</em>, <em>controlfile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.loadZinba" title="Permalink to this definition">¶</a></dt>
<dd><p>load Zinba results in <em>tablename</em></p>
<p>This method loads only positive peaks. It filters peaks by p-value,
q-value and fold change and loads the diagnostic data and
re-calculates peakcenter, peakval, ... using the supplied bamfile.</p>
<p>If <em>tablename</em> is not given, it will be <tt class="file docutils literal"><span class="pre">&lt;track&gt;_intervals</span></tt>
where track is derived from <tt class="docutils literal"><span class="pre">infile</span></tt> and assumed to end
in <tt class="file docutils literal"><span class="pre">.zinba</span></tt>.</p>
<p>If no peaks were predicted, an empty table is created.</p>
<p>This method creates <tt class="file docutils literal"><span class="pre">&lt;outfile&gt;.tsv.gz</span></tt> with the results
of the filtering.</p>
<p>This method uses the refined peak locations.</p>
<p>Zinba peaks can be overlapping. This method does not merge
overlapping intervals.</p>
<p>Zinba calls peaks in regions where there are many reads inside
the control. Thus this method applies a filtering step 
removing all intervals in which there is a peak of
more than readlength / 2 height in the control.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.makeIntervalCorrelation">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">makeIntervalCorrelation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em>, <em>field</em>, <em>reference</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.makeIntervalCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>compute correlation of interval properties between sets</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.buildIntervalCounts">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">buildIntervalCounts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>track</em>, <em>fg_replicates</em>, <em>bg_replicates</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.buildIntervalCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>count read density in bed files comparing stimulated versus unstimulated binding.</p>
</dd></dl>

<dl class="function">
<dt id="PipelineChipseq.loadIntervalsFromBed">
<tt class="descclassname">PipelineChipseq.</tt><tt class="descname">loadIntervalsFromBed</tt><span class="sig-paren">(</span><em>bedfile</em>, <em>track</em>, <em>outfile</em>, <em>bamfiles</em>, <em>offsets</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineChipseq.loadIntervalsFromBed" title="Permalink to this definition">¶</a></dt>
<dd><p>load intervals from <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted files into database.</p>
<p>Re-evaluate the intervals by counting reads within
the interval. In contrast to the initial pipeline, the
genome is not binned. In particular, the meaning of the
columns in the table changes to:</p>
<p>nProbes: number of reads in interval
PeakCenter: position with maximum number of reads in interval
AvgVal: average coverage within interval</p>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h4>Previous topic</h4>
  <p class="topless"><a href="PipelineBiomart.html"
                        title="previous chapter">PipelineBiomart.py - utility tasks for importing data from biomart</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="PipelineDatabase.html"
                        title="next chapter">PipelineDatabase.py - utility functions for working with a database</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelinemodules/PipelineChipseq.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PipelineDatabase.html" title="PipelineDatabase.py - utility functions for working with a database"
             >next</a> |</li>
        <li class="right" >
          <a href="PipelineBiomart.html" title="PipelineBiomart.py - utility tasks for importing data from biomart"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" >Modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>