
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PipelineTracks.py - Definition of tracks in pipelines &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Modules" href="../modules.html" />
    <link rel="next" title="PipelineTransfacMatch.py - Classes and functions used in pipeline_trasfacmatch.py" href="PipelineTransfacMatch.html" />
    <link rel="prev" title="Rnaseq.py - Tools for RNAseq analysis" href="PipelineRnaseq.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PipelineTransfacMatch.html" title="PipelineTransfacMatch.py - Classes and functions used in pipeline_trasfacmatch.py"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="PipelineRnaseq.html" title="Rnaseq.py - Tools for RNAseq analysis"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" accesskey="U">Modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-PipelineTracks"></span><div class="section" id="pipelinetracks-py-definition-of-tracks-in-pipelines">
<h1>PipelineTracks.py - Definition of tracks in pipelines<a class="headerlink" href="#pipelinetracks-py-definition-of-tracks-in-pipelines" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="motivation">
<h2>Motivation<a class="headerlink" href="#motivation" title="Permalink to this headline">¶</a></h2>
<p>A pipeline typically processes the data streams from several
experimental data sources. These data streams are usually processed
separately (processing, quality control) and as aggregates. For
example, consider the following experimental layout:</p>
<table border="1" class="docutils">
<colgroup>
<col width="38%" />
<col width="63%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Filename</em></td>
<td><em>Content</em></td>
</tr>
<tr class="row-even"><td>liver-stimulated-R1</td>
<td>liver, stimulated, replicate 1</td>
</tr>
<tr class="row-odd"><td>liver-stimulated-R2</td>
<td>liver, stimulated, replicate 2</td>
</tr>
<tr class="row-even"><td>liver-unstimulated-R1</td>
<td>liver, unstimulated, replicate 1</td>
</tr>
<tr class="row-odd"><td>liver-unstimulated-R2</td>
<td>liver, unstimulated, replicate 2</td>
</tr>
<tr class="row-even"><td>heart-stimulated-R1</td>
<td>heart, stimulated, replicate 1</td>
</tr>
<tr class="row-odd"><td>heart-stimulated-R2</td>
<td>heart, stimulated, replicate 2</td>
</tr>
<tr class="row-even"><td>heart-unstimulated-R1</td>
<td>heart, unstimulated, replicate 1</td>
</tr>
<tr class="row-odd"><td>heart-unstimulated-R2</td>
<td>heart, unstimulated, replicate 2</td>
</tr>
</tbody>
</table>
<p>The experiment measured in two tissues with two conditions with two
replicates each giving eight data streams. During the analysis, the
streams are merged in a variety of combinations:</p>
<blockquote>
<div><ul class="simple">
<li>unmerged for initial processing, QC, etc.</li>
<li>by replicates to assess reproducibility of measurements</li>
<li>by condition to assess the size of the response to the stimulus</li>
<li>by tissue to assess differences between tissue and address the
biological question.</li>
</ul>
</div></blockquote>
<p>The crossing of data streams complicates the building of pipelines,
especially as no two experiments are the same. The
<a class="reference internal" href="#module-PipelineTracks" title="PipelineTracks"><tt class="xref py py-mod docutils literal"><span class="pre">PipelineTracks</span></tt></a> module assists in controlling these data
streams. This module provides some tools to map tracks to different
representations and to group them in flexible ways in order to provide
convenient short-cuts in pipelines.</p>
<p>There are three class within <a class="reference internal" href="#module-PipelineTracks" title="PipelineTracks"><tt class="xref py py-mod docutils literal"><span class="pre">PipelineTracks</span></tt></a>: <a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">Sample</span></tt></a>,
<a class="reference internal" href="#PipelineTracks.Tracks" title="PipelineTracks.Tracks"><tt class="xref py py-class docutils literal"><span class="pre">Tracks</span></tt></a> and <tt class="xref py py-class docutils literal"><span class="pre">Aggregate</span></tt>.</p>
<div class="section" id="a-track">
<h3>A Track<a class="headerlink" href="#a-track" title="Permalink to this headline">¶</a></h3>
<p>The basic atomic data structure is a <a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">Sample</span></tt></a> or
<a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a>. A <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a> is a single measurement that can be
combined with other tracks. A track identifier consists of a tuple of
attributes. Each track in and experimental design has the same number
of labels in the same order.  In the example above, there are three
attributes: tissue, condition and replicate. Identifiers are thus
<tt class="docutils literal"><span class="pre">('liver',</span> <span class="pre">'stimulated','R1')</span></tt> or <tt class="docutils literal"><span class="pre">('heart','unstimulated','R2')</span></tt>.</p>
<p>The same track can be represented by different names depending on
context, for example when it is used as a filename or a database
table. As filename, the track <tt class="docutils literal"><span class="pre">('heart','unstimulated','R2')</span></tt> is
rendered as <tt class="docutils literal"><span class="pre">heart-unstimulated-R2</span></tt> (avoiding spaces), while as a
table, it reads <tt class="docutils literal"><span class="pre">heart-unstimulated-R2</span></tt>, avoiding <tt class="docutils literal"><span class="pre">-+.</span></tt>. The
<a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">Sample</span></tt></a> class provides convenience methods to convert names
from one context to another.</p>
</div>
<div class="section" id="track-containers">
<h3>Track containers<a class="headerlink" href="#track-containers" title="Permalink to this headline">¶</a></h3>
<p>A container of type <a class="reference internal" href="#PipelineTracks.Tracks" title="PipelineTracks.Tracks"><tt class="xref py py-class docutils literal"><span class="pre">Tracks</span></tt></a> stores one or more objects of type
<a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">Sample</span></tt></a>.</p>
</div>
<div class="section" id="aggregates">
<h3>Aggregates<a class="headerlink" href="#aggregates" title="Permalink to this headline">¶</a></h3>
<p>Tracks can be combined into aggregates. Aggregation is indicated by
the <tt class="docutils literal"><span class="pre">agg</span></tt> keyword.</p>
<p>For example, the <tt class="docutils literal"><span class="pre">liver-stimulated-agg</span></tt> aggregate combines the
tracks <tt class="docutils literal"><span class="pre">liver-stimulated-R1</span></tt> and <tt class="docutils literal"><span class="pre">liver-stimulated-R2</span></tt>. The
aggregate <tt class="docutils literal"><span class="pre">agg-stimulated-agg</span></tt> combines all replicates and all
tissues (<tt class="docutils literal"><span class="pre">liver-stimulated-R1</span></tt>, <tt class="docutils literal"><span class="pre">liver-stimulated-R2</span></tt>,
<tt class="docutils literal"><span class="pre">heart-stimulated-R1</span></tt>, <tt class="docutils literal"><span class="pre">heart-stimulated-R2</span></tt>)</p>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<div class="section" id="defining-tracks-and-aggregates">
<h3>Defining tracks and aggregates<a class="headerlink" href="#defining-tracks-and-aggregates" title="Permalink to this headline">¶</a></h3>
<p>To use tracks, you need to first define a new <a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">Sample</span></tt></a>. In the
example above with the attributes tissue, condition and replicate, the
<a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">Sample</span></tt></a> could be:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">PipelineTracks</span>

<span class="k">class</span> <span class="nc">MySample</span><span class="p">(</span> <span class="n">PipelineTracks</span><span class="o">.</span><span class="n">Sample</span> <span class="p">):</span>
     <span class="n">attributes</span> <span class="o">=</span> <span class="p">(</span> <span class="s">&quot;tissue&quot;</span><span class="p">,</span> <span class="s">&quot;condition&quot;</span><span class="p">,</span> <span class="s">&quot;replicate&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<p>Once defined, you can add tracks to a <tt class="xref py py-class docutils literal"><span class="pre">tracks</span></tt> container. For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">TRACKS</span> <span class="o">=</span> <span class="n">PipelineTracks</span><span class="o">.</span><span class="n">Tracks</span><span class="p">(</span> <span class="n">MySample</span> <span class="p">)</span><span class="o">.</span><span class="n">loadFromDirectory</span><span class="p">(</span>
         <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s">&quot;*.fastq.gz&quot;</span><span class="p">),</span>
         <span class="n">pattern</span> <span class="o">=</span> <span class="s">&quot;(\S+).fastq.gz&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>will collect all files ending in <tt class="docutils literal"><span class="pre">.fastq.gz</span></tt>. The track identifiers
will be derived by removing the <tt class="docutils literal"><span class="pre">fastq.gz</span></tt> suffix. The variable
<tt class="docutils literal"><span class="pre">TRACKS</span></tt> contains all the tracks derived from files ending in
<tt class="docutils literal"><span class="pre">*.fastq.gz</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">TRACKS</span>
<span class="go">[liver-stimulated-R2, heart-stimulated-R2, liver-stimulated-R1,</span>
<span class="go"> liver-unstimulated-R1, heart-unstimulated-R2, heart-stimulated-R1,</span>
<span class="go"> heart-unstimulated-R1, liver-unstimulated-R2]</span>
</pre></div>
</div>
<p>To build aggregates, use <tt class="xref py py-class docutils literal"><span class="pre">PipelineTracks.Aggregate</span></tt>. The
following combines replicates for each experiment:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">EXPERIMENTS</span> <span class="o">=</span> <span class="n">PipelineTracks</span><span class="o">.</span><span class="n">Aggregate</span><span class="p">(</span>
      <span class="n">TRACKS</span><span class="p">,</span>
      <span class="n">labels</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;condition&quot;</span><span class="p">,</span> <span class="s">&quot;tissue&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>Aggregates are simply containers of associated data sets. To get a
list of experiments, type:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">EXPERIMENTS</span> <span class="o">=</span> <span class="n">PipelineTracks</span><span class="o">.</span><span class="n">Aggregate</span><span class="p">(</span>
<span class="go">          TRACKS,</span>
<span class="go">          labels=(&quot;condition&quot;, &quot;tissue&quot;))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="nb">list</span><span class="p">(</span><span class="n">EXPERIMENT</span><span class="p">)</span>
<span class="go">[heart-stimulated-agg, heart-unstimulated-agg, liver-stimulated-agg, liver-unstimulated-agg]</span>
</pre></div>
</div>
<p>or:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">EXPERIMENT</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[heart-stimulated-agg, heart-unstimulated-agg, liver-stimulated-agg, liver-unstimulated-agg]</span>
</pre></div>
</div>
<p>To obtain all replicates in the experiment <tt class="docutils literal"><span class="pre">heart-stimulated</span></tt>, use
dictionary access:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">EXPERIMENTS</span><span class="p">[</span><span class="s">&#39;heart-stimulated-agg&#39;</span><span class="p">]</span>
<span class="go">[heart-stimulated-R2, heart-stimulated-R1]</span>
</pre></div>
</div>
<p>The returned objects are tracks. To use a <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a> as a tablename
or as a file, use data access functions <a class="reference internal" href="#PipelineTracks.Sample.asTable" title="PipelineTracks.Sample.asTable"><tt class="xref py py-meth docutils literal"><span class="pre">Sample.asTable()</span></tt></a> or
<a class="reference internal" href="#PipelineTracks.Sample.asFile" title="PipelineTracks.Sample.asFile"><tt class="xref py py-meth docutils literal"><span class="pre">Sample.asFile()</span></tt></a>, respectively:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">asFile</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">EXPERIMENTS</span><span class="p">[</span><span class="s">&#39;heart-stimulated-agg&#39;</span><span class="p">]</span> <span class="p">]</span>
<span class="go">[&#39;heart-stimulated-R2&#39;, &#39;heart-stimulated-R1&#39;]</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">EXPERIMENTS</span><span class="p">[</span><span class="s">&#39;heart-stimulated-agg&#39;</span><span class="p">]</span> <span class="p">]</span>
<span class="go">[&#39;heart-stimulated-R2&#39;, &#39;heart-stimulated-R1&#39;]</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">asTable</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">EXPERIMENTS</span><span class="p">[</span><span class="s">&#39;heart-stimulated-agg&#39;</span><span class="p">]</span> <span class="p">]</span>
<span class="go">[&#39;heart_stimulated_R2&#39;, &#39;heart_stimulated_R1&#39;]</span>
</pre></div>
</div>
<p>Note how the <tt class="docutils literal"><span class="pre">-</span></tt> is converted to <tt class="docutils literal"><span class="pre">_</span></tt> as the former are illegal as
SQL table names.</p>
<p>The default representation is file-based. By using the class method:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">MySample</span><span class="o">.</span><span class="n">setDefault</span><span class="p">(</span> <span class="s">&quot;asTable&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<p>the default representation can be changed for all tracks simultaneously.</p>
<p>You can have multiple aggregates. For example, some tasks might
require all conditions or all tissues:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">CONDITIONS</span> <span class="o">=</span> <span class="n">PipelineTracks</span><span class="o">.</span><span class="n">Aggregate</span><span class="p">(</span> <span class="n">TRACKS</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="p">(</span><span class="s">&quot;condition&quot;</span><span class="p">,</span> <span class="p">)</span> <span class="p">)</span>
<span class="n">TISSUES</span> <span class="o">=</span> <span class="n">PipelineTracks</span><span class="o">.</span><span class="n">Aggregate</span><span class="p">(</span> <span class="n">TRACKS</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="p">(</span><span class="s">&quot;tissue&quot;</span><span class="p">,</span> <span class="p">)</span> <span class="p">)</span>
</pre></div>
</div>
<p>You can have several <a class="reference internal" href="#PipelineTracks.Tracks" title="PipelineTracks.Tracks"><tt class="xref py py-class docutils literal"><span class="pre">Tracks</span></tt></a> within a
directory. <a class="reference internal" href="#PipelineTracks.Tracks" title="PipelineTracks.Tracks"><tt class="xref py py-class docutils literal"><span class="pre">Tracks</span></tt></a> are simply containers and as such do not
have any actions associated with them.</p>
</div>
<div class="section" id="using-tracks-in-pipelines">
<h3>Using tracks in pipelines<a class="headerlink" href="#using-tracks-in-pipelines" title="Permalink to this headline">¶</a></h3>
<p>Unfortunately, tracks and aggregates do not work yet directly as
<a class="reference external" href="http://www.ruffus.org.uk/">ruffus</a> task lists. Instead, they need to be converted to files
explicitely using list comprehensions.</p>
<p>If you wanted to process all tracks separately, use:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="p">[</span> <span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s">.fastq.gz&quot;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">asFile</span><span class="p">(),</span>
            <span class="s">&quot;</span><span class="si">%s</span><span class="s">.qc&quot;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">asFile</span><span class="p">())</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">TRACKS</span> <span class="p">]</span> <span class="p">)</span>
<span class="k">def</span> <span class="nf">performQC</span><span class="p">(</span> <span class="n">infile</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>
   <span class="o">....</span>
</pre></div>
</div>
<p>The above statement will create the following list of input/output
files for the <tt class="docutils literal"><span class="pre">performQC</span></tt> task:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span> <span class="p">(</span> <span class="s">&quot;liver-stimulated-R1.fastq.gz&quot;</span><span class="p">,</span> <span class="s">&quot;liver-stimulated-R1.qc&quot;</span> <span class="p">)</span>
  <span class="p">(</span> <span class="s">&quot;liver-stimulated-R2.fastq.gz&quot;</span> <span class="p">,</span> <span class="s">&quot;liver-stimulated-R2.qc&quot;</span> <span class="p">),</span>
  <span class="o">...</span>
<span class="p">]</span>
</pre></div>
</div>
<p>Using aggregates works similarly, though you will need to create the file
lists yourself using nested list comprehensions. The following creates
an analysis per experimemnt:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files</span><span class="p">(</span> <span class="p">[(</span> <span class="p">([</span> <span class="s">&quot;</span><span class="si">%s</span><span class="s">.fastq.gz&quot;</span> <span class="o">%</span> <span class="n">y</span><span class="o">.</span><span class="n">asFile</span><span class="p">()</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">EXPERIMENTS</span><span class="p">[</span><span class="n">x</span><span class="p">]]),</span> 
                  <span class="s">&quot;</span><span class="si">%s</span><span class="s">.out&quot;</span> <span class="o">%</span> <span class="n">x</span><span class="o">.</span><span class="n">asFile</span><span class="p">())</span> 
                  <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">EXPERIMENTS</span> <span class="p">]</span> <span class="p">)</span>
<span class="k">def</span> <span class="nf">checkReproducibility</span><span class="p">(</span> <span class="n">infiles</span><span class="p">,</span> <span class="n">outfile</span> <span class="p">):</span>
   <span class="o">....</span>
</pre></div>
</div>
<p>The above statement will create the following list of input/output files:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span> <span class="p">(</span> <span class="p">(</span> <span class="s">&quot;liver-stimulated-R1.fastq.gz&quot;</span><span class="p">,</span> <span class="s">&quot;liver-stimulated-R2.fastq.gz&quot;</span> <span class="p">),</span> <span class="s">&quot;liver-stimulated-agg.out&quot;</span> <span class="p">),</span>
  <span class="p">(</span> <span class="p">(</span> <span class="s">&quot;liver-unstimulated-R1.fastq.gz&quot;</span><span class="p">,</span> <span class="s">&quot;liver-unstimulated-R2.fastq.gz&quot;</span> <span class="p">),</span> <span class="s">&quot;liver-unstimulated-agg.out&quot;</span> <span class="p">),</span>
  <span class="p">(</span> <span class="p">(</span> <span class="s">&quot;heart-stimulated-R1.fastq.gz&quot;</span><span class="p">,</span> <span class="s">&quot;heart-stimulated-R2.fastq.gz&quot;</span> <span class="p">),</span> <span class="s">&quot;heart-stimulated-agg.out&quot;</span> <span class="p">),</span>
  <span class="p">(</span> <span class="p">(</span> <span class="s">&quot;heart-unstimulated-R1.fastq.gz&quot;</span><span class="p">,</span> <span class="s">&quot;heart-unstimulated-R2.fastq.gz&quot;</span> <span class="p">),</span> <span class="s">&quot;heart-unstimulated-agg.out&quot;</span> <span class="p">),</span>
<span class="p">]</span>
</pre></div>
</div>
<p>The above code makes sure that the file dependencies are
observed. Thus, if <tt class="docutils literal"><span class="pre">heart-stimulated-R1.fastq.gz</span></tt> changes, only
<tt class="docutils literal"><span class="pre">heart-stimulated-agg.out</span></tt> will be re-computed.</p>
<p>Tracks and aggregates can be used within a task. The following code
will collect all replicates for the experiment
<tt class="docutils literal"><span class="pre">liver-stimulated-agg</span></tt></p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">track</span> <span class="o">=</span> <span class="n">TRACKS</span><span class="o">.</span><span class="n">factory</span><span class="p">(</span> <span class="n">filename</span> <span class="o">=</span> <span class="s">&quot;liver-stimulated-agg&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">replicates</span> <span class="o">=</span> <span class="n">PipelineTracks</span><span class="o">.</span><span class="n">getSamplesInTrack</span><span class="p">(</span> <span class="n">track</span><span class="p">,</span> <span class="n">TRACKS</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">replicates</span>
<span class="go">[liver-stimulated-R2, liver-stimulated-R1]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="api">
<h2>API<a class="headerlink" href="#api" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="PipelineTracks.Sample">
<em class="property">class </em><tt class="descclassname">PipelineTracks.</tt><tt class="descname">Sample</tt><span class="sig-paren">(</span><em>filename=None</em>, <em>tablename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>a sample/track with one attribute called <tt class="docutils literal"><span class="pre">experiment</span></tt>.</p>
<p>create a new Sample.</p>
<p>If filename is given, the sample name will be derived from <em>filename</em>.
Similarly, from <em>tablename</em>.</p>
<dl class="method">
<dt id="PipelineTracks.Sample.clone">
<tt class="descname">clone</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.clone" title="Permalink to this definition">¶</a></dt>
<dd><p>return a copy of self.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample.asFile">
<tt class="descname">asFile</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.asFile" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as a filename</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample.asTable">
<tt class="descname">asTable</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.asTable" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as a tablename</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample.asR">
<tt class="descname">asR</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.asR" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as valid R label</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample.fromFile">
<tt class="descname">fromFile</tt><span class="sig-paren">(</span><em>fn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.fromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from filename <em>fn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample.fromTable">
<tt class="descname">fromTable</tt><span class="sig-paren">(</span><em>tn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.fromTable" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from tablename <em>tn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample.fromR">
<tt class="descname">fromR</tt><span class="sig-paren">(</span><em>rn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.fromR" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from R name <em>rn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample.asAggregate">
<tt class="descname">asAggregate</tt><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.asAggregate" title="Permalink to this definition">¶</a></dt>
<dd><p>return a new aggregate Sample.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample.toLabels">
<tt class="descname">toLabels</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.toLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>return attributes that this track is an aggregate of.</p>
</dd></dl>

<dl class="classmethod">
<dt id="PipelineTracks.Sample.setDefault">
<em class="property">classmethod </em><tt class="descname">setDefault</tt><span class="sig-paren">(</span><em>representation=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample.setDefault" title="Permalink to this definition">¶</a></dt>
<dd><p>set default representation for tracks to <em>representation</em>.
If <em>represenation</em> is None, the representation will be set to
the library default (asFile()).</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="PipelineTracks.Sample3">
<em class="property">class </em><tt class="descclassname">PipelineTracks.</tt><tt class="descname">Sample3</tt><span class="sig-paren">(</span><em>filename=None</em>, <em>tablename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">PipelineTracks.Sample</span></tt></a></p>
<p>a sample/track with three attributes: tissue, condition and replicate.</p>
<p>create a new Sample.</p>
<p>If filename is given, the sample name will be derived from <em>filename</em>.
Similarly, from <em>tablename</em>.</p>
<dl class="method">
<dt id="PipelineTracks.Sample3.asAggregate">
<tt class="descname">asAggregate</tt><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.asAggregate" title="Permalink to this definition">¶</a></dt>
<dd><p>return a new aggregate Sample.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample3.asFile">
<tt class="descname">asFile</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.asFile" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as a filename</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample3.asR">
<tt class="descname">asR</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.asR" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as valid R label</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample3.asTable">
<tt class="descname">asTable</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.asTable" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as a tablename</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample3.clone">
<tt class="descname">clone</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.clone" title="Permalink to this definition">¶</a></dt>
<dd><p>return a copy of self.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample3.fromFile">
<tt class="descname">fromFile</tt><span class="sig-paren">(</span><em>fn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.fromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from filename <em>fn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample3.fromR">
<tt class="descname">fromR</tt><span class="sig-paren">(</span><em>rn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.fromR" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from R name <em>rn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample3.fromTable">
<tt class="descname">fromTable</tt><span class="sig-paren">(</span><em>tn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.fromTable" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from tablename <em>tn</em></p>
</dd></dl>

<dl class="classmethod">
<dt id="PipelineTracks.Sample3.setDefault">
<em class="property">classmethod </em><tt class="descname">setDefault</tt><span class="sig-paren">(</span><em>representation=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.setDefault" title="Permalink to this definition">¶</a></dt>
<dd><p>set default representation for tracks to <em>representation</em>.
If <em>represenation</em> is None, the representation will be set to
the library default (asFile()).</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample3.toLabels">
<tt class="descname">toLabels</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample3.toLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>return attributes that this track is an aggregate of.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="PipelineTracks.Sample4">
<em class="property">class </em><tt class="descclassname">PipelineTracks.</tt><tt class="descname">Sample4</tt><span class="sig-paren">(</span><em>filename=None</em>, <em>tablename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">PipelineTracks.Sample</span></tt></a></p>
<p>a sample/track with four attributes: experiment, tissue, condition and replicate.</p>
<p>create a new Sample.</p>
<p>If filename is given, the sample name will be derived from <em>filename</em>.
Similarly, from <em>tablename</em>.</p>
<dl class="method">
<dt id="PipelineTracks.Sample4.asAggregate">
<tt class="descname">asAggregate</tt><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.asAggregate" title="Permalink to this definition">¶</a></dt>
<dd><p>return a new aggregate Sample.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample4.asFile">
<tt class="descname">asFile</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.asFile" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as a filename</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample4.asR">
<tt class="descname">asR</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.asR" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as valid R label</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample4.asTable">
<tt class="descname">asTable</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.asTable" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as a tablename</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample4.clone">
<tt class="descname">clone</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.clone" title="Permalink to this definition">¶</a></dt>
<dd><p>return a copy of self.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample4.fromFile">
<tt class="descname">fromFile</tt><span class="sig-paren">(</span><em>fn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.fromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from filename <em>fn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample4.fromR">
<tt class="descname">fromR</tt><span class="sig-paren">(</span><em>rn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.fromR" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from R name <em>rn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample4.fromTable">
<tt class="descname">fromTable</tt><span class="sig-paren">(</span><em>tn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.fromTable" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from tablename <em>tn</em></p>
</dd></dl>

<dl class="classmethod">
<dt id="PipelineTracks.Sample4.setDefault">
<em class="property">classmethod </em><tt class="descname">setDefault</tt><span class="sig-paren">(</span><em>representation=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.setDefault" title="Permalink to this definition">¶</a></dt>
<dd><p>set default representation for tracks to <em>representation</em>.
If <em>represenation</em> is None, the representation will be set to
the library default (asFile()).</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Sample4.toLabels">
<tt class="descname">toLabels</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Sample4.toLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>return attributes that this track is an aggregate of.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="PipelineTracks.AutoSample">
<em class="property">class </em><tt class="descclassname">PipelineTracks.</tt><tt class="descname">AutoSample</tt><span class="sig-paren">(</span><em>filename=None</em>, <em>tablename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">PipelineTracks.Sample</span></tt></a></p>
<p>a sample/track with unknown number of attributes.</p>
<p>The number of attributes will be guessed from <em>filename</em>,
<em>tablename</em> or a given list of <em>attributes</em>.</p>
<dl class="method">
<dt id="PipelineTracks.AutoSample.asAggregate">
<tt class="descname">asAggregate</tt><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.asAggregate" title="Permalink to this definition">¶</a></dt>
<dd><p>return a new aggregate Sample.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.AutoSample.asFile">
<tt class="descname">asFile</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.asFile" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as a filename</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.AutoSample.asR">
<tt class="descname">asR</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.asR" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as valid R label</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.AutoSample.asTable">
<tt class="descname">asTable</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.asTable" title="Permalink to this definition">¶</a></dt>
<dd><p>return sample as a tablename</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.AutoSample.clone">
<tt class="descname">clone</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.clone" title="Permalink to this definition">¶</a></dt>
<dd><p>return a copy of self.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.AutoSample.fromFile">
<tt class="descname">fromFile</tt><span class="sig-paren">(</span><em>fn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.fromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from filename <em>fn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.AutoSample.fromR">
<tt class="descname">fromR</tt><span class="sig-paren">(</span><em>rn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.fromR" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from R name <em>rn</em></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.AutoSample.fromTable">
<tt class="descname">fromTable</tt><span class="sig-paren">(</span><em>tn</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.fromTable" title="Permalink to this definition">¶</a></dt>
<dd><p>build sample from tablename <em>tn</em></p>
</dd></dl>

<dl class="classmethod">
<dt id="PipelineTracks.AutoSample.setDefault">
<em class="property">classmethod </em><tt class="descname">setDefault</tt><span class="sig-paren">(</span><em>representation=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.setDefault" title="Permalink to this definition">¶</a></dt>
<dd><p>set default representation for tracks to <em>representation</em>.
If <em>represenation</em> is None, the representation will be set to
the library default (asFile()).</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.AutoSample.toLabels">
<tt class="descname">toLabels</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.AutoSample.toLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>return attributes that this track is an aggregate of.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="PipelineTracks.Tracks">
<em class="property">class </em><tt class="descclassname">PipelineTracks.</tt><tt class="descname">Tracks</tt><span class="sig-paren">(</span><em>factory=&lt;class 'PipelineTracks.Sample'&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Tracks" title="Permalink to this definition">¶</a></dt>
<dd><p>a collection of tracks.</p>
<p>create a new container.</p>
<p>New tracks are derived using <em>factory</em>.</p>
<dl class="attribute">
<dt id="PipelineTracks.Tracks.factory">
<tt class="descname">factory</tt><a class="headerlink" href="#PipelineTracks.Tracks.factory" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="#PipelineTracks.Sample" title="PipelineTracks.Sample"><tt class="xref py py-class docutils literal"><span class="pre">Sample</span></tt></a></p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Tracks.loadFromDirectory">
<tt class="descname">loadFromDirectory</tt><span class="sig-paren">(</span><em>files</em>, <em>pattern</em>, <em>exclude=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Tracks.loadFromDirectory" title="Permalink to this definition">¶</a></dt>
<dd><p>load tracks from a list of files, applying pattern.</p>
<p>Pattern is a regular expression with at at least one group,
for example <tt class="docutils literal"><span class="pre">(.*).gz</span></tt>.</p>
<p>If set, exclude files matching regular expression in <em>exclude</em>.</p>
<p>Only the filenames are being used. Any path is removed.</p>
</dd></dl>

<dl class="method">
<dt id="PipelineTracks.Tracks.getTracks">
<tt class="descname">getTracks</tt><span class="sig-paren">(</span><em>pattern=None</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.Tracks.getTracks" title="Permalink to this definition">¶</a></dt>
<dd><p>return all tracks in container.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="PipelineTracks.getSamplesInTrack">
<tt class="descclassname">PipelineTracks.</tt><tt class="descname">getSamplesInTrack</tt><span class="sig-paren">(</span><em>track</em>, <em>tracks</em><span class="sig-paren">)</span><a class="headerlink" href="#PipelineTracks.getSamplesInTrack" title="Permalink to this definition">¶</a></dt>
<dd><p>return all tracks in <em>tracks</em> that constitute <em>track</em>.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">PipelineTracks.py - Definition of tracks in pipelines</a><ul>
<li><a class="reference internal" href="#motivation">Motivation</a><ul>
<li><a class="reference internal" href="#a-track">A Track</a></li>
<li><a class="reference internal" href="#track-containers">Track containers</a></li>
<li><a class="reference internal" href="#aggregates">Aggregates</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#defining-tracks-and-aggregates">Defining tracks and aggregates</a></li>
<li><a class="reference internal" href="#using-tracks-in-pipelines">Using tracks in pipelines</a></li>
</ul>
</li>
<li><a class="reference internal" href="#api">API</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="PipelineRnaseq.html"
                        title="previous chapter">Rnaseq.py - Tools for RNAseq analysis</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="PipelineTransfacMatch.html"
                        title="next chapter">PipelineTransfacMatch.py - Classes and functions used in pipeline_trasfacmatch.py</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelinemodules/PipelineTracks.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PipelineTransfacMatch.html" title="PipelineTransfacMatch.py - Classes and functions used in pipeline_trasfacmatch.py"
             >next</a> |</li>
        <li class="right" >
          <a href="PipelineRnaseq.html" title="Rnaseq.py - Tools for RNAseq analysis"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" >Modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>