==============
Tool reference
==============

This page summarizes prominent tools within the CGAT Code
collection.

Genomics
========

:doc:`scripts/bed2gff`
    Convert a :term:`bed` formatted file to a :term:`gff` or
    :term:`gtf` formatted file.

:doc:`scripts/beds2counts`
    Compute overlap statistics of multiple :term:`bed` files.

:doc:`scripts/diff_bed`
    Compare multiple sets of bed files in pairwise fashion.
    Permits incremental updates of similarity table.

:doc:`scripts/diff_fasta`
    Compare two sets of sequences. Outputs missing, identical
    and fragment sequences.

:doc:`scripts/diff_gtf`
    Compare two gene sets - output common and unique genes by overlap.

:doc:`scripts/gff2gff`
    Work on :term:`gff` formatted files with genomic features. This 
    tools sorts/renames feature files, reconciles chromosome names,
    and more.

:doc:`scripts/gtf2gff`
    Translate a gene set into genomic annotations such as introns,
    intergenic regions, regulatory domains, etc.

:doc:`scripts/gtf2table`
    Annotate transcripts in a :term:`gtf` formatted file. Annotations
    can be in reference to a second gene set (fragments, extensions), 
    aligned reads (coverage, intron overrun, ...) or densities.

:doc:`scripts/index_fasta`
    Build an index for a fasta file. Pre-requisite for many CGAT tools.
	
:doc:`fasta2kmercontent`
    Count kmer content in a set of :term:`fasta` sequences.


NGS
===

:doc:`scripts/bam2geneprofile`
     Compute and plot a meta-gene profile from aligned reads.

:doc:`scripts/fastqs2fasta`
    Interleave paired reads from two fastq files into a single fasta file.


Unsorted
=========

:doc:`scripts/bam2bam`

:doc:`scripts/bam2bed`

:doc:`scripts/bam2fastq`

:doc:`scripts/bam2peakshape`

:doc:`scripts/bam2stats`

:doc:`scripts/bam2UniquePairs`

:doc:`scripts/bam2wiggle`

:doc:`scripts/bed2bed`

:doc:`scripts/bed2fasta`

:doc:`scripts/bed2graph`

:doc:`scripts/bed2summary`

:doc:`scripts/bed2table`

:doc:`scripts/beds2beds`

:doc:`scripts/concatenate_sequences`

:doc:`scripts/diff_bed`

:doc:`scripts/diff_chains`

:doc:`scripts/diff_gtfs`

:doc:`scripts/fasta2bed`

:doc:`scripts/fasta2counts`

:doc:`scripts/fasta2gaps`

:doc:`scripts/fasta2gff`

:doc:`scripts/fasta2properties`

:doc:`scripts/fasta2spliced`

:doc:`scripts/fasta2table`

:doc:`scripts/fasta2variants`

:doc:`scripts/fastq2fastq`

:doc:`scripts/fastq2N`

:doc:`scripts/fastq2table`

:doc:`scripts/fastqs2fastq`

:doc:`scripts/gff2bed`

:doc:`scripts/gff2chunks`

:doc:`scripts/gff2coverage`

:doc:`scripts/gff2fasta`

:doc:`scripts/gff2histogram`

:doc:`scripts/gff2plot`

:doc:`scripts/gff2psl`

:doc:`scripts/gff2stats`

:doc:`scripts/gff2table`

:doc:`scripts/gff2wiggle_stats`

:doc:`scripts/go2plot`

:doc:`scripts/go2svg`

:doc:`scripts/gtf2fasta`

:doc:`scripts/gtf2gff`

:doc:`scripts/gtf2gtf`

:doc:`scripts/gtf2tab`

:doc:`scripts/gtfs2graph`

:doc:`scripts/rnaseq_bam_vs_bed`

:doc:`scripts/rnaseq_bam_vs_exons`

:doc:`scripts/rnaseq_bams_vs_bams`

:doc:`scripts/rnaseq_bams2bam`

:doc:`scripts/rnaseq_junction_bam2bam`

:doc:`scripts/shuffle_fasta`

:doc:`scripts/softmask`

:doc:`scripts/vcf2vcf`

