
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>RNA-Seq Transcript Build pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="NGS Pipelines" href="../CGATPipelines.html" />
    <link rel="next" title="RNA-Seq Differential expression pipeline" href="pipeline_rnaseqdiffexpression.html" />
    <link rel="prev" title="Read mapping pipeline" href="pipeline_mapping.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseqdiffexpression.html" title="RNA-Seq Differential expression pipeline"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pipeline_mapping.html" title="Read mapping pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" accesskey="U">NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pipeline_rnaseqtranscripts"></span><div class="section" id="rna-seq-transcript-build-pipeline">
<h1>RNA-Seq Transcript Build pipeline<a class="headerlink" href="#rna-seq-transcript-build-pipeline" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>The rnaseq transcript build pipeline attempts build a variety of gene sets
from reads mapped to a reference genome.</p>
<p>This pipeline works on a single genome.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>h_The pipeline assumes the data derive from multiple tissues/conditions (<a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>) 
with one or more biological and/or technical replicates (<a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a>). A <a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a>
within each <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a> is a <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a>.</p>
<p>The pipeline builds the following genesets for each input track:</p>
<blockquote>
<div><ul>
<li><p class="first">Direct cufflinks output:</p>
<blockquote>
<div><ol class="arabic simple">
<li>gene set per replicate</li>
<li>gene set per experiment. Data from individual replicates are combined.</li>
<li>a complete gene set (<tt class="file docutils literal"><span class="pre">agg-agg-agg.gtf.gz</span></tt>). 
Data from all experiments are combined</li>
</ol>
</div></blockquote>
</li>
<li><p class="first">Derive gene sets:</p>
<blockquote>
<div><ol class="arabic simple">
<li>full geneset (<tt class="file docutils literal"><span class="pre">full.gtf.gz</span></tt>)
all transcripts predicted by cuffcompare. Derived from <tt class="file docutils literal"><span class="pre">agg-agg-agg.gtf.gz</span></tt></li>
<li>pruned geneset (<tt class="file docutils literal"><span class="pre">pruned.gtf.gz</span></tt>)
only transfrags are kept that are present in at least two samples. 
Derived from <tt class="file docutils literal"><span class="pre">full.gtf.gz</span></tt>.</li>
<li>novel geneset (<tt class="file docutils literal"><span class="pre">novel.gtf.gz</span></tt>)
only transfrags that do not overlap any of the transcripts in the reference
gene set. This data set is derived from <tt class="file docutils literal"><span class="pre">pruned.gtf.gz</span></tt>.</li>
<li>lincRNA gene set (<tt class="file docutils literal"><span class="pre">lincrna.gtf.gz</span></tt>)</li>
</ol>
</div></blockquote>
</li>
</ul>
</div></blockquote>
<div class="section" id="novel-gene-set">
<h3>Novel gene set<a class="headerlink" href="#novel-gene-set" title="Permalink to this headline">¶</a></h3>
<p>The novel gene set is build from the pruned gene set.</p>
<p>Transcripts are removed based on features in the reference gene set. By default, these
are features called <tt class="docutils literal"><span class="pre">protein_coding</span></tt>, <tt class="docutils literal"><span class="pre">lincRNA</span></tt> or <tt class="docutils literal"><span class="pre">processed_transcript</span></tt>.
Transcripts that lie exclusively in repetetive sequence are removed, too.</p>
<p>Removal is aggressive  - as soon as one transcript of a gene/locus overlaps, all 
transcripts of that gene/locus are gone.</p>
<p>The resultant set contains a number of novel transcripts. However, these
transcripts will still overlap some known genomic features like pseudogenes.</p>
</div>
<div class="section" id="lincrna">
<h3>LincRNA<a class="headerlink" href="#lincrna" title="Permalink to this headline">¶</a></h3>
<p>One of the main benefits of RNASeq over microarrays is that novel transcripts can be detected. A particular
interest are currently novel long non-coding RNA. Unfortunately, it seems that these transcripts
are often expressed at very low levels and possibly in a highly regulated manner, for example only in
certain tissues. On top of their low abundance, they frequently seem to be co-localized with protein
coding genes, making it hard to distinguish them from transcription artifacts.</p>
<p>Success in identifying lincRNA will depend a lot on your input data. Long, paired-end reads are likely 
to lead to success. Unfortunately, many exploratory studies go for single-ended, short read data.
With such data, identification of novel spliced transcripts will be rare and the set of novel transcripts
is likely to contain many false-positives.</p>
<dl class="docutils">
<dt>The pipeline constructs a set of novel lncRNA in the following manner:</dt>
<dd><ol class="first last arabic simple">
<li>All transcript models overlapping protein coding transcripts are removed.</li>
<li>Overlapping lncRNA on the same strand are merged.</li>
</ol>
</dd>
</dl>
<div class="section" id="artifacts-in-lncrna-analysis">
<h4>Artifacts in lncRNA analysis<a class="headerlink" href="#artifacts-in-lncrna-analysis" title="Permalink to this headline">¶</a></h4>
<p>There are several sources of artifacts in lncRNA analysis</p>
<div class="section" id="read-mapping-errors">
<h5>Read mapping errors<a class="headerlink" href="#read-mapping-errors" title="Permalink to this headline">¶</a></h5>
<p>Mapping errors are identifyable as sharp peaks in the
coverage profile. Mapping errors occur if the true location of a read has more mismatches
than the original location or it maps across an undetected splice-site. Most of the 
highly-expressed lncRNA are due to mapping errors. Secondary locations very often overlap
highly-expressed protein-coding genes. These errors are annoying for two reasons: they
provide false positives, but at the same time prevent the reads to be counted towards
the expression of the true gene.</p>
<p>They can be detected in two ways:</p>
<p>1. via a peak-like distribution of reads which should result in a low entropy of start position 
density. Note that this possibly can remove transcripts that are close to the length of a single
read.</p>
<p>2. via mapping against known protein coding transcripts. However, getting this mapping right
is hard for two reasons. Firstly, mapping errors usually involve reads aligned with mismatches.
Thus, the mapping has to be done either on the read-level (computationally expensive), or 
on the transcript level after variant calling. (tricky, and also computationally expensive). 
Secondly, as cufflinks extends transcripts generously, only a part of a transcript might actually
be a mismapped part. Distinguishing partial true matches from random matches will be tricky.</p>
<p>Read mapping errors can also be avoided by</p>
<ol class="arabic simple">
<li>Longer read lengths</li>
<li>Strict alignment criteria</li>
<li>A two-stage mapping process.</li>
</ol>
</div>
<div class="section" id="fragments">
<h5>Fragments<a class="headerlink" href="#fragments" title="Permalink to this headline">¶</a></h5>
<p>As lncRNA are expressed at low levels, it is likely that only a partial transcript
can be observed.</p>
</div>
</div>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="../UsingPipelines.html#pipelinesettingup"><em>Setting up a pipeline</em></a> and <a class="reference internal" href="../UsingPipelines.html#pipelinerunning"><em>Running a pipeline</em></a> on general information how to use CGAT pipelines.</p>
<div class="section" id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires a configured <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt> file.</p>
<p>The sphinxreport report requires a <tt class="file docutils literal"><span class="pre">conf.py</span></tt> and <tt class="file docutils literal"><span class="pre">sphinxreport.ini</span></tt> file 
(see <a class="reference internal" href="../UsingPipelines.html#pipelinereporting"><em>Building pipeline reports</em></a>). To start with, use the files supplied with the
<a class="reference internal" href="#example">Example</a> data.</p>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="section" id="mapped-reads">
<h4>Mapped reads<a class="headerlink" href="#mapped-reads" title="Permalink to this headline">¶</a></h4>
<p>The principal input of this pipeline is a collection of reads mapped to a reference genome.
Mapped reads are imported by placing files are linking to files in the <a class="reference internal" href="../make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
<p>The default file format assumes the following convention:</p>
<blockquote>
<div>&lt;sample&gt;-&lt;condition&gt;-&lt;replicate&gt;.bam</div></blockquote>
<p><tt class="docutils literal"><span class="pre">sample</span></tt> and <tt class="docutils literal"><span class="pre">condition</span></tt> make up an <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>, while <tt class="docutils literal"><span class="pre">replicate</span></tt> denotes
the <a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a> within an <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>.</p>
</div>
<div class="section" id="optional-inputs">
<h4>Optional inputs<a class="headerlink" href="#optional-inputs" title="Permalink to this headline">¶</a></h4>
</div>
</div>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires the results from <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a>. Set the configuration variable 
<tt class="xref py py-data docutils literal"><span class="pre">annotations_database</span></tt> and <tt class="xref py py-data docutils literal"><span class="pre">annotations_dir</span></tt>.</p>
<p>On top of the default CGAT setup, the pipeline requires the following software to be in the 
path:</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="22%" />
<col width="55%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Program</em></td>
<td><em>Version</em></td>
<td><em>Purpose</em></td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a></td>
<td>&gt;=1.3.0</td>
<td>transcription levels</td>
</tr>
<tr class="row-odd"><td>samtools</td>
<td>&gt;=0.1.16</td>
<td>bam/sam files</td>
</tr>
<tr class="row-even"><td>bedtools</td>
<td>&nbsp;</td>
<td>working with intervals</td>
</tr>
<tr class="row-odd"><td>R/DESeq</td>
<td>&nbsp;</td>
<td>differential expression</td>
</tr>
<tr class="row-even"><td>picard</td>
<td>&gt;=1.42</td>
<td><dl class="first last docutils">
<dt>bam/sam files. The .jar files need to be in your</dt>
<dd>CLASSPATH environment variable.</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="pipeline-output">
<h2>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this headline">¶</a></h2>
<p>The major output is in the database file <tt class="file docutils literal"><span class="pre">csvdb</span></tt>.</p>
<p>The pipeline will build <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted files for each of
the geneset builds.</p>
<p>For each <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a> there will be the following tables:</p>
<dl class="docutils">
<dt>&lt;track&gt;_cuffcompare_benchmark </dt>
<dd>results from comparing gene models against reference gene set
primary key: track</dd>
<dt>&lt;track&gt;_cuffcompare_transcripts</dt>
<dd>transcript expression values (FPKMs)
primary key: track+transfrag_id
foreign key: transfrag_id</dd>
<dt>&lt;track&gt;_cuffcompare_tracking</dt>
<dd>tracking information linking loci against transcripts.
primary key: transfrag_id, 
foreign key: locus_id</dd>
<dt>&lt;track&gt;_cuffcompare_tracking</dt>
<dd><dl class="first last docutils">
<dt>locus information (number of transcripts within locus per track)</dt>
<dd>primary key: locus_id</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>Example data is available at <a class="reference external" href="http://www.cgat.org/~andreas/sample_data/pipeline_rnaseqtranscripts.tgz">http://www.cgat.org/~andreas/sample_data/pipeline_rnaseqtranscripts.tgz</a>.
To run the example, simply unpack and untar:</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://www.cgat.org/~andreas/sample_data/pipeline_rnaseqtranscripts.tgz
tar -xvzf pipeline_rnaseqtranscripts.tgz
cd pipeline_rnaseq
python &lt;srcdir&gt;/pipeline_rnaseqtranscripts.py make full
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For the pipeline to run, install the <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a> as well.</p>
</div>
</div>
<div class="section" id="glossary">
<h2>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h2>
<dl class="glossary docutils">
<dt id="term-cufflinks">cufflinks</dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a> - transcriptome analysis</dd>
</dl>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pipeline_rnaseqtranscripts.connect">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">connect</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.connect" title="Permalink to this definition">¶</a></dt>
<dd><p>connect to database.</p>
<p>This method also attaches to helper databases.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.writePrunedGTF">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">writePrunedGTF</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.writePrunedGTF" title="Permalink to this definition">¶</a></dt>
<dd><p>remove various gene models from a gtf file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.mergeAndFilterGTF">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">mergeAndFilterGTF</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>logfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.mergeAndFilterGTF" title="Permalink to this definition">¶</a></dt>
<dd><p>sanitize transcripts file for cufflinks analysis.</p>
<p>Merge exons separated by small introns (&lt; 5bp).</p>
<dl class="docutils">
<dt>Transcript will be ignored that</dt>
<dd><ul class="first last simple">
<li>have very long introns (max_intron_size) (otherwise, cufflinks complains)</li>
<li>are located on contigs to be ignored (usually: chrM, _random, ...)</li>
</ul>
</dd>
</dl>
<p>This method preserves all features in a gtf file (exon, CDS, ...).</p>
<p>returns a dictionary of all gene_ids that have been kept.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildReferenceGeneSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildReferenceGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildReferenceGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>sanitize ENSEMBL transcripts file for cufflinks analysis.</p>
<p>Merge exons separated by small introns (&lt; 5bp).</p>
<p>Removes unwanted contigs according to configuration
value <tt class="docutils literal"><span class="pre">geneset_remove_contigs</span></tt>.</p>
<p>Removes transcripts overlapping ribosomal genes if 
<tt class="docutils literal"><span class="pre">geneset_remove_repetitive_rna</span></tt> is set. Protein
coding transcripts are not removed.</p>
<dl class="docutils">
<dt>Transcript will be ignored that</dt>
<dd><ul class="first last simple">
<li>have very long introns (max_intron_size) (otherwise, cufflinks complains)</li>
<li>are located on contigs to be ignored (usually: chrM, _random, ...)</li>
</ul>
</dd>
</dl>
<p>The result is run through cuffdiff in order to add the p_id and tss_id tags
required by cuffdiff.</p>
<p>This will only keep sources of the type &#8216;exon&#8217;. It will also remove
any transcripts not in the reference genome.</p>
<p>Cuffdiff requires overlapping genes to have different tss_id tags.</p>
<p>This gene is the source for most other gene sets in the pipeline.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildNoncodingGeneSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildNoncodingGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildNoncodingGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a new gene set without protein coding 
transcripts.</p>
<p># Nick
Also remove transcripts that are ambiguous in
biotype in terms of their non-codingness</p>
<p>filter the refnoncoding geneset for things that are described in ensembl
as being:
Ambiguous_orf
Retained_intron
Sense_intronic
antisense
Sense_overlapping
Processed transcript</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildReferenceGeneSetWithCDS">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildReferenceGeneSetWithCDS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildReferenceGeneSetWithCDS" title="Permalink to this definition">¶</a></dt>
<dd><p>build a new gene set without protein coding 
transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildCodingGeneSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildCodingGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildCodingGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a gene set with only protein coding 
transcripts.</p>
<p>Genes are selected via their gene biotype in the GTF file.
Note that this set will contain all transcripts of protein
coding genes, including processed transcripts.</p>
<p>This set includes UTR and CDS.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildCodingTranscriptSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildCodingTranscriptSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildCodingTranscriptSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a gene set with only protein coding transcripts.</p>
<p>Protein coding transcripts are selected via the ensembl
transcript biotype</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildMaskGtf">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildMaskGtf</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildMaskGtf" title="Permalink to this definition">¶</a></dt>
<dd><p>This takes ensembl annotations (geneset_all.gtf.gz) and writes out all entries that 
have a &#8216;source&#8217; match to &#8220;rRNA&#8221; or &#8216;contig&#8217; match to &#8220;chrM&#8221;. for use with cufflinks</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildTranscriptsWithCufflinks">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildTranscriptsWithCufflinks</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildTranscriptsWithCufflinks" title="Permalink to this definition">¶</a></dt>
<dd><p>build transcript models for each track separately.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.METHODTARGET">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">METHODTARGET</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.METHODTARGET" title="Permalink to this definition">¶</a></dt>
<dd><p>build transcript models for each track separately.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.runCuffCompare">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">runCuffCompare</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em>, <em>reffile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.runCuffCompare" title="Permalink to this definition">¶</a></dt>
<dd><p>run cuffcompare.</p>
<p>Will create a .tmap and .refmap input file for each input file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.compareTranscriptsPerExperiment">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">compareTranscriptsPerExperiment</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.compareTranscriptsPerExperiment" title="Permalink to this definition">¶</a></dt>
<dd><p>compare transcript models between replicates within each experiment.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.compareTranscriptsBetweenExperiments">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">compareTranscriptsBetweenExperiments</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.compareTranscriptsBetweenExperiments" title="Permalink to this definition">¶</a></dt>
<dd><p>compare transcript models between replicates in all experiments.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.mergeUsingCuffmerge">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">mergeUsingCuffmerge</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.mergeUsingCuffmerge" title="Permalink to this definition">¶</a></dt>
<dd><p>use cuffmerge to reassemble transcripts from independent assemblies
on each sample
CURRENTLY THIS IS NOT WIRED INTO THE MAIN PIPELINE FLOW</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.loadTranscriptComparison">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">loadTranscriptComparison</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.loadTranscriptComparison" title="Permalink to this definition">¶</a></dt>
<dd><p>load data from transcript comparison.</p>
<p>This task creates two tables:</p>
<p>&lt;track&gt;_benchmark
&lt;track&gt;_loci</p>
<p>The following tables are only present if there are
multiple replicates in a sample:</p>
<p>&lt;track&gt;_tracking</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildFullGeneSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildFullGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildFullGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>builds ab-initio gene set.</p>
<p>The ab-initio gene set is derived from the cuffcompare result.</p>
<p>The following transfrags are removed at this stage:</p>
<blockquote>
<div><ul class="simple">
<li>transfrags overlapping RNA genes</li>
<li>transfrags on certain contigs (usually: mitochondrial genes)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildAndLoadFullGeneSetTracking">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildAndLoadFullGeneSetTracking</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildAndLoadFullGeneSetTracking" title="Permalink to this definition">¶</a></dt>
<dd><p>take those transcripts accepted in the &#8220;Full&#8221; set and retrive the 
records for them from the agg-agg-agg_tracking set</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildPrunedGeneSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildPrunedGeneSet</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildPrunedGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>builds a gene set by merging the ab-initio gene set and
the reference gene set.</p>
<p>The gene set is cleaned in order to permit differential expression
analysis.</p>
<p>Only transfrags are kept that are:</p>
<ol class="arabic">
<li><dl class="first docutils">
<dt>observed in at least 2 samples to remove partial transfrags that</dt>
<dd><p class="first last">are the result of low coverage observations in one sample</p>
</dd>
</dl>
</li>
</ol>
<p>see also: <a class="reference external" href="http://seqanswers.com/forums/showthread.php?t=3967">http://seqanswers.com/forums/showthread.php?t=3967</a></p>
<p>Will also build removed.gtf.gz of removed transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildNovelGeneSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildNovelGeneSet</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildNovelGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a gene set of novel genes by merging the ab-initio gene set and
the reference gene set.</p>
<p>Ab-initio transcripts are removed based on features in the reference gene set.</p>
<p>Removal is aggressive  - as soon as one transcript of a 
gene/locus overlaps, all transcripts of that gene/locus are gone.</p>
<p>Transcripts that lie exclusively in repetetive sequence are removed, too.</p>
<p>The resultant set contains a number of novel transcripts. However, these
transcripts will still overlap some known genomic features like pseudogenes.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildLincRNAGeneSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildLincRNAGeneSet</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildLincRNAGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build lincRNA gene set.</p>
<p>The lincRNA gene set contains all known lincRNA transcripts from
the reference gene set plus all transcripts in the novel set that
do not overlap at any protein coding, processed or pseudogene transcripts 
(exons+introns) in the reference gene set.</p>
<p>Transcripts that lie exclusively in repetetive sequence are removed, too.</p>
<p>lincRNA genes are often expressed at low level and thus the resultant transcript
models are fragmentory. To avoid some double counting in downstream analyses, 
transcripts overlapping on the same strand are merged.</p>
<p>Transcripts need to have a length of at least 200 bp.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildAbinitioLincRNAGeneSet">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildAbinitioLincRNAGeneSet</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildAbinitioLincRNAGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build ab initio lincRNA gene set. 
In contrast to the buildLincRNAGeneSet this lincRNA set does not contain
the reference noncoding gene set. It is transcripts in the abinitio (pruned) set that
do not overlap at any protein coding, processed or pseudogene transcripts 
(exons+introns) in the reference gene set.</p>
<p>Transcripts overlapping protein coding transcripts on opposite transcripts are retained.
It also does not filter out repeats as LincRNA transcripts seem to be found in such regions on occasion</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildGeneSetStats">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildGeneSetStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildGeneSetStats" title="Permalink to this definition">¶</a></dt>
<dd><p>compile gene set statistics.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.loadGeneSetStats">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">loadGeneSetStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.loadGeneSetStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load geneset statisticts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.classifyTranscripts">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">classifyTranscripts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.classifyTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>classify transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.classifyTranscriptsCuffcompare">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">classifyTranscriptsCuffcompare</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.classifyTranscriptsCuffcompare" title="Permalink to this definition">¶</a></dt>
<dd><p>classify transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildTranscriptLevelReadCounts">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildTranscriptLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildTranscriptLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>count reads falling into transcripts of protein coding </dt>
<dd>gene models.</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In paired-end data sets each mate will be counted. Thus
the actual read counts are approximately twice the fragment
counts.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.hasReplicates">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">hasReplicates</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.hasReplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>indicator function - return true if track has replicates</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.buildReproducibility">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">buildReproducibility</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.buildReproducibility" title="Permalink to this definition">¶</a></dt>
<dd><p>all-vs-all comparison between samples.</p>
<p>Compute correlation between expressed transfrags. Transfrags missing
from another set are ignored.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.loadReproducibility">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">loadReproducibility</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.loadReproducibility" title="Permalink to this definition">¶</a></dt>
<dd><p>load reproducibility results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.build_report">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">build_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>build report from scratch.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.update_report">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">update_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.update_report" title="Permalink to this definition">¶</a></dt>
<dd><p>update report.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqtranscripts.publish">
<tt class="descclassname">pipeline_rnaseqtranscripts.</tt><tt class="descname">publish</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqtranscripts.publish" title="Permalink to this definition">¶</a></dt>
<dd><p>publish files.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">RNA-Seq Transcript Build pipeline</a><ul>
<li><a class="reference internal" href="#overview">Overview</a><ul>
<li><a class="reference internal" href="#novel-gene-set">Novel gene set</a></li>
<li><a class="reference internal" href="#lincrna">LincRNA</a><ul>
<li><a class="reference internal" href="#artifacts-in-lncrna-analysis">Artifacts in lncRNA analysis</a><ul>
<li><a class="reference internal" href="#read-mapping-errors">Read mapping errors</a></li>
<li><a class="reference internal" href="#fragments">Fragments</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
<li><a class="reference internal" href="#input">Input</a><ul>
<li><a class="reference internal" href="#mapped-reads">Mapped reads</a></li>
<li><a class="reference internal" href="#optional-inputs">Optional inputs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pipeline-output">Pipeline output</a></li>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#glossary">Glossary</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pipeline_mapping.html"
                        title="previous chapter">Read mapping pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pipeline_rnaseqdiffexpression.html"
                        title="next chapter">RNA-Seq Differential expression pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelines/pipeline_rnaseqtranscripts.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseqdiffexpression.html" title="RNA-Seq Differential expression pipeline"
             >next</a> |</li>
        <li class="right" >
          <a href="pipeline_mapping.html" title="Read mapping pipeline"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" >NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>