
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Peak calling pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="NGS Pipelines" href="../CGATPipelines.html" />
    <link rel="next" title="Long non-coding RNA pipeline" href="pipeline_rnaseqlncrna.html" />
    <link rel="prev" title="Transcriptome analysis pipeline" href="pipeline_transcriptome.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseqlncrna.html" title="Long non-coding RNA pipeline"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pipeline_transcriptome.html" title="Transcriptome analysis pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" accesskey="U">NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pipeline_peakcalling"></span><div class="section" id="peak-calling-pipeline">
<h1>Peak calling pipeline<a class="headerlink" href="#peak-calling-pipeline" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>The peak calling pipeline applies various peak calling algorithms
on mapped reads.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>pipeline_peakcalling takes as input reads aligned to genomic sequence
as <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted files and calls peaks. The pipeline
implements several peak callers:</p>
<dl class="docutils">
<dt><a class="reference external" href="http://liulab.dfci.harvard.edu/MACS/00README.html">macs</a></dt>
<dd>Model-based Analysis of ChIP-Seq (MACS), for identifying
transcript factor binding sites. MACS captures the influence of
genome complexity to evaluate the significance of enriched ChIP
regions, and MACS improves the spatial resolution of binding sites
through combining the information of both sequencing tag position
and orientation. MACS can be easily used for ChIP-Seq data alone,
or with control sample with the increase of specificity.</dd>
<dt><a class="reference external" href="http://liulab.dfci.harvard.edu/MACS/00README.html">macs2</a> </dt>
<dd>MACS 2 is the new release of the MACS peak caller. Among other
improvements it adds support for handling paired end reads.</dd>
<dt><a class="reference external" href="http://compbio.med.harvard.edu/Supplements/ChIP-seq/tutorial.html">spp</a> </dt>
<dd>SPP is a R package especially designed for the analysis of
Chip-Seq data from Illummina platform. The package was developed by
Peter Park&#8217;s group from Harvard Medical School.</dd>
<dt><a class="reference external" href="http://code.google.com/p/zinba/">zinba</a> </dt>
<dd>ZINBA (Zero Inflated Negative Binomial Algorithm) is a
computational and statistical framework used to call regions of the
genome enriched for sequencing reads originating from a diverse
array of biological experiments. We collectively refer to the
sequencing data derived from these experiments as DNA-seq,
including FAIRE-seq, ChIP-seq, and DNAase-seq experiments</dd>
<dt>sicer_narrow </dt>
<dd>A clustering approach for identification of enriched
domains from histone modification ChIP-Seq data.  The types of
region called by the sicer alogrithm reflect the paramaters it is
run with. In this pipeline sicer is run by default with two
different parameter sets, to allow the simultaneous detection of
narrower and broader regions of enrichmnet.</dd>
<dt>sicer_broad</dt>
<dd>(See above)</dd>
<dt>peakranger_ranger </dt>
<dd>PeakRanger is a multi-purpose, ultrafast ChIP Seq
peak caller. It is used in the modENCODE project and included in
the iPlant pipeline system.  PeakRanger v1.02 was developed in
Dr.Lincoln Stein&#8217;s lab at OICR and is now in continual development
at Dr.Helen Hobbs&#8217;s lab of the McDermott Center of UT
Southwestern.  PeakRanger can run two separate algorithms for peak
detection, &#8220;ranger&#8221; for point-source binding event detection and
&#8220;ccat&#8221; for the detection of broader regions of enrichment.  In
this pipeline, both alorighms are presented as separate peak
callers for convience.</dd>
<dt>peakranger_ccat </dt>
<dd>PeakRanger is here run using the CCAT alogorithm (See:
Xu, H., L. Handoko, et al. (2010).A signal-noise model for
significance analysis of ChIP-seq with negative
control.Bioinformatics 26(9): 1199-1204)</dd>
<dt>scripture</dt>
<dd>As of version 2, scripture has added a ChIP-Seq module. The current
version (5/11/2013) is still marked beta. Scripture implements a
continuous segmentation algorithm. It scans windows of a fixed size
across the genome, identifies significant windows and merges/trims
to obtain peaks. Significance is assessed using the scan statistic
(Wallenstein and Neff, 1987).
Scripture in the pipeline is used according to Garber et al. (2012)
(PMID:22940246). Significant windows are filtered against input or
other control data (Garber et al.: whole cell extract) by computing
an enrichment scrore. Only significant windows above a certain 
enrichment score are kept.</dd>
</dl>
<p>Peak callers have different strengths and weaknesses. Some might work
well on broad peaks such as some histone marks, others work better for
narrow, sharp peaks. Many callers these days attempt to call both
types of peaks.  This pipeline implements the following nomenclature
for peaks:</p>
<dl class="glossary docutils">
<dt id="term-region">region</dt>
<dd>A broad region defined by a peak caller. Regions are usually
created in the first step of peak calling, before peak
refinement or detection of subpeaks is performed.</dd>
<dt id="term-summit">summit</dt>
<dd>A narrow region defined by a caller. These are the output of any
peak refinement or subpeak detection steps</dd>
<dt id="term-interval">interval</dt>
<dd>Generic term for an interval in which the read density is higher
then expected. Both a <a class="reference internal" href="#term-region"><em class="xref std std-term">region</em></a> or a <a class="reference internal" href="#term-summit"><em class="xref std std-term">summit</em></a> are an
<a class="reference internal" href="#term-interval"><em class="xref std std-term">interval</em></a>.</dd>
<dt id="term-peak">peak</dt>
<dd>Within a <a class="reference internal" href="#term-region"><em class="xref std std-term">region</em></a> or <a class="reference internal" href="#term-summit"><em class="xref std std-term">summit</em></a> the position with the
highest base coverage.</dd>
</dl>
<p>The pipeline computes some basic measures to validate peak calling. In
order to fully annotate peaks, use <a class="reference internal" href="pipeline_intervals.html"><em>Interval pipeline</em></a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The pipeline currently expects that mulit-mapping reads (reads
mapping to multiple locations) have been removed.</p>
</div>
<div class="section" id="qc">
<h3>QC<a class="headerlink" href="#qc" title="Permalink to this headline">¶</a></h3>
<p>The pipeline implements the following QC measures. See <a href="#id1"><span class="problematic" id="id2">:pmid:`22955991`</span></a>.</p>
<dl class="docutils">
<dt>NSC</dt>
<dd>normalized strand coefficient.</dd>
<dt>RSC</dt>
<dd>relative strand correlacion.</dd>
</dl>
<p>The pipeline will also do and IDR analysis (see
<a class="reference external" href="https://sites.google.com/site/anshulkundaje/projects/idr">https://sites.google.com/site/anshulkundaje/projects/idr</a>) for spp
called peaks.</p>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="../UsingPipelines.html#pipelinesettingup"><em>Setting up a pipeline</em></a> and <a class="reference internal" href="../UsingPipelines.html#pipelinerunning"><em>Running a pipeline</em></a> on general
information how to use CGAT pipelines.</p>
<div class="section" id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires a configured <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt> file.</p>
<p>The sphinxreport report requires a <tt class="file docutils literal"><span class="pre">conf.py</span></tt> and
<tt class="file docutils literal"><span class="pre">sphinxreport.ini</span></tt> file (see <a class="reference internal" href="../UsingPipelines.html#pipelinereporting"><em>Building pipeline reports</em></a>). To start
with, use the files supplied with the <a class="reference internal" href="#example">Example</a> data.</p>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="section" id="mapped-reads">
<h4>Mapped reads<a class="headerlink" href="#mapped-reads" title="Permalink to this headline">¶</a></h4>
<p>The principal input of this pipeline is a collection of reads mapped to a reference genome.
Mapped reads are imported by placing files are linking to files in the <a class="reference internal" href="../make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
<p>The default file format assumes the following convention:</p>
<blockquote>
<div>&lt;sample&gt;-&lt;condition&gt;-&lt;replicate&gt;.genome.bam</div></blockquote>
<p><tt class="docutils literal"><span class="pre">sample</span></tt> and <tt class="docutils literal"><span class="pre">condition</span></tt> make up an <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>, while <tt class="docutils literal"><span class="pre">replicate</span></tt> denotes
the <a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a> within an <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>.</p>
<p>Please not the suffix <tt class="docutils literal"><span class="pre">genome.bam</span></tt> which is required to distinguish the input <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a>
formatted files from those that are created in the pipeline after duplication removal (<tt class="docutils literal"><span class="pre">.prep.bam</span></tt>
and <tt class="docutils literal"><span class="pre">.call.bam</span></tt>)</p>
</div>
<div class="section" id="optional-inputs">
<h4>Optional inputs<a class="headerlink" href="#optional-inputs" title="Permalink to this headline">¶</a></h4>
</div>
</div>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires the results from <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a>. Set the configuration variable 
<tt class="xref py py-data docutils literal"><span class="pre">annotations_database</span></tt> and <tt class="xref py py-data docutils literal"><span class="pre">annotations_dir</span></tt>.</p>
<p>On top of the default CGAT setup, the pipeline requires the following software to be in the 
path:</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="22%" />
<col width="55%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Program</em></td>
<td><em>Version</em></td>
<td><em>Purpose</em></td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://compbio.med.harvard.edu/Supplements/ChIP-seq/tutorial.html">spp</a></td>
<td>1.11</td>
<td>R package</td>
</tr>
<tr class="row-odd"><td><a class="reference external" href="http://liulab.dfci.harvard.edu/MACS/00README.html">macs</a></td>
<td>&gt;1.4</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://code.google.com/p/zinba/">zinba</a></td>
<td>2.02.03</td>
<td>R package</td>
</tr>
<tr class="row-odd"><td>peakranger</td>
<td>1.15</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>sicer</td>
<td>1.1</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="pipeline-output">
<h2>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this headline">¶</a></h2>
<p>The major output is in the database file <tt class="file docutils literal"><span class="pre">csvdb</span></tt>. For each peak caller there are tables
called:</p>
<p>&lt;track&gt;_&lt;caller&gt;_regions
&lt;track&gt;_&lt;caller&gt;_summits</p>
<p>Each of these tables contains the following columns:</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="78%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Column</em></td>
<td><em>Content</em></td>
</tr>
<tr class="row-even"><td>avgval</td>
<td>Average read depth in interval</td>
</tr>
<tr class="row-odd"><td>contig</td>
<td>Contig</td>
</tr>
<tr class="row-even"><td>control_avgval</td>
<td>Average read depth in control within inter val</td>
</tr>
<tr class="row-odd"><td>control_length</td>
<td>Interval length</td>
</tr>
<tr class="row-even"><td>control_npeaks</td>
<td>Number of peaks in control</td>
</tr>
<tr class="row-odd"><td>control_nreads</td>
<td>Number of control reads in interval</td>
</tr>
<tr class="row-even"><td>control_peakcenter</td>
<td>Peak center of control</td>
</tr>
<tr class="row-odd"><td>control_peakval</td>
<td>Number of reads at peak in control</td>
</tr>
<tr class="row-even"><td>end</td>
<td>End coordinate of interval</td>
</tr>
<tr class="row-odd"><td>length</td>
<td>Length of interval</td>
</tr>
<tr class="row-even"><td>npeaks</td>
<td>Number of peaks in interval</td>
</tr>
<tr class="row-odd"><td>nreads</td>
<td>Number of reads in interval</td>
</tr>
<tr class="row-even"><td>peakcenter</td>
<td>Peak center in interval</td>
</tr>
<tr class="row-odd"><td>peakval</td>
<td>Number of reads at peak</td>
</tr>
<tr class="row-even"><td>start</td>
<td>246251</td>
</tr>
</tbody>
</table>
<p>The unprocessed output files created by the peak callers are in individual subdirectories
for each caller (<tt class="file docutils literal"><span class="pre">macs.dir</span></tt>, <tt class="file docutils literal"><span class="pre">zinba.dir</span></tt>, etc.).</p>
<div class="section" id="idr-analysis">
<h3>IDR analysis<a class="headerlink" href="#idr-analysis" title="Permalink to this headline">¶</a></h3>
<p>The output of the IDR analysis is in the <tt class="file docutils literal"><span class="pre">idr.dir</span></tt> directory.</p>
</div>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>Example data is available at <a class="reference external" href="http://www.cgat.org/~andreas/sample_data/pipeline_mapping.tgz">http://www.cgat.org/~andreas/sample_data/pipeline_mapping.tgz</a>.
To run the example, simply unpack and untar:</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://www.cgat.org/~andreas/sample_data/pipeline_mapping.tgz
tar -xvzf pipeline_mapping.tgz
cd pipeline_mapping
python &lt;srcdir&gt;/pipeline_mapping.py make full
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For the pipeline to run, install the <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a> as well.</p>
</div>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pipeline_peakcalling.getControl">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">getControl</tt><span class="sig-paren">(</span><em>track</em>, <em>suffix='.genome.bam'</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.getControl" title="Permalink to this definition">¶</a></dt>
<dd><p>return appropriate control(s) for a track.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.getUnstimulated">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">getUnstimulated</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.getUnstimulated" title="Permalink to this definition">¶</a></dt>
<dd><p>return unstimulated condition for a track</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.getSubtracted">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">getSubtracted</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.getSubtracted" title="Permalink to this definition">¶</a></dt>
<dd><p>return &#8220;subtracted&#8221; condition for a track</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.getUnsubtracted">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">getUnsubtracted</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.getUnsubtracted" title="Permalink to this definition">¶</a></dt>
<dd><p>return &#8220;unsubtracted&#8221; condition for a track</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.getBamFiles">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">getBamFiles</tt><span class="sig-paren">(</span><em>infile</em>, <em>suffix</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.getBamFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>return associated bamfiles with infiles.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.connect">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">connect</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.connect" title="Permalink to this definition">¶</a></dt>
<dd><p>connect to database.</p>
<p>This method also attaches to helper databases.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.checkInput">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">checkInput</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.checkInput" title="Permalink to this definition">¶</a></dt>
<dd><p>perform a check if every track has a control/input.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.makeMask">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">makeMask</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.makeMask" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a mask for filtering reads if required.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.prepareBAMForPeakCalling">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">prepareBAMForPeakCalling</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.prepareBAMForPeakCalling" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare BAM files for peak calling.</p>
<ul>
<li><p class="first">unmapped reads are removed.</p>
</li>
<li><dl class="first docutils">
<dt>if the option &#8220;calling_deduplicate&#8221; is Picard.MarkDuplicates is run </dt>
<dd><p class="first last">to remove duplicate reads</p>
</dd>
</dl>
</li>
<li><p class="first">reads may be filtered by exon or location</p>
<blockquote>
<div><ul class="simple">
<li>to remove reads by exon, the option &#8220;calling_filter_exons&#8221; should specify a file containing 
a list of ensembl gene identifiers (one per line)</li>
<li>to remove reads by location, the option &#8220;calling_filter_regions&#8221; should specify a bed file&#8217;&#8216;</li>
</ul>
</div></blockquote>
</li>
</ul>
<p>The resulting bam file has a .prep.bam extension. Merging infiles is currently untested and the 
methods only consider single end reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadDuplicationStats">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadDuplicationStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadDuplicationStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load output from Picard Deduplication step.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.buildBAMStats">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">buildBAMStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.buildBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadBAMStats">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadBAMStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>import bam statisticis.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.checkDataQuality">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">checkDataQuality</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.checkDataQuality" title="Permalink to this definition">¶</a></dt>
<dd><p>uses peakranger to check data quality.</p>
<p>nr is the signal/noise ratio.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadDataQuality">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadDataQuality</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadDataQuality" title="Permalink to this definition">¶</a></dt>
<dd><p>load data quality information.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.checkLibraryComplexity">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">checkLibraryComplexity</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.checkLibraryComplexity" title="Permalink to this definition">¶</a></dt>
<dd><p>uses peakranger to check library complexity.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callPeaksWithMACS">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callPeaksWithMACS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callPeaksWithMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>run MACS for peak detection.</p>
<p>output bed files are compressed and indexed.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadMACS">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadMACS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.cleanMACS">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">cleanMACS</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.cleanMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>clean up MACS - build bigwig file and remove wig files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.summarizeMACS">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">summarizeMACS</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.summarizeMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize MACS results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.summarizeMACSFDR">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">summarizeMACSFDR</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.summarizeMACSFDR" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize MACS results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadMACSSummary">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadMACSSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadMACSSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadMACSSummaryFDR">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadMACSSummaryFDR</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadMACSSummaryFDR" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callPeaksWithMACS2">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callPeaksWithMACS2</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callPeaksWithMACS2" title="Permalink to this definition">¶</a></dt>
<dd><p>run MACS 2 for peak detection.
output bed files are compressed and indexed.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadMACS2">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadMACS2</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadMACS2" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.summarizeMACS2">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">summarizeMACS2</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.summarizeMACS2" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize MACS results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.summarizeMACS2FDR">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">summarizeMACS2FDR</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.summarizeMACS2FDR" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize MACS results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadMACS2Summary">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadMACS2Summary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadMACS2Summary" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs2 summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadMACS2SummaryFDR">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadMACS2SummaryFDR</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadMACS2SummaryFDR" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs2 summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callPeaksWithZinba">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callPeaksWithZinba</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callPeaksWithZinba" title="Permalink to this definition">¶</a></dt>
<dd><p>run Zinba for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadZinba">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadZinba</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadZinba" title="Permalink to this definition">¶</a></dt>
<dd><p>load Zinba results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callNarrowerPeaksWithSICER">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callNarrowerPeaksWithSICER</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callNarrowerPeaksWithSICER" title="Permalink to this definition">¶</a></dt>
<dd><p>run SICER for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callBroaderPeaksWithSICER">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callBroaderPeaksWithSICER</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callBroaderPeaksWithSICER" title="Permalink to this definition">¶</a></dt>
<dd><p>run SICER for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadSICER">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadSICER</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadSICER" title="Permalink to this definition">¶</a></dt>
<dd><p>load sicer results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.summarizeSICER">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">summarizeSICER</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.summarizeSICER" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize SICER results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadSICERSummary">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadSICERSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadSICERSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>load sicer summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callPeaksWithPeakRanger">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callPeaksWithPeakRanger</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callPeaksWithPeakRanger" title="Permalink to this definition">¶</a></dt>
<dd><p>run PeakRanger Ranger for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadPeakRanger">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadPeakRanger</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadPeakRanger" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.summarizePeakRanger">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">summarizePeakRanger</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.summarizePeakRanger" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize peakranger results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadPeakRangerSummary">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadPeakRangerSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadPeakRangerSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>load Peakranger summarys.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callPeaksWithPeakRangerCCAT">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callPeaksWithPeakRangerCCAT</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callPeaksWithPeakRangerCCAT" title="Permalink to this definition">¶</a></dt>
<dd><p>run Peak Ranger CCAT for broad peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadPeakRangerCCAT">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadPeakRangerCCAT</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadPeakRangerCCAT" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.summarizePeakRangerCCAT">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">summarizePeakRangerCCAT</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.summarizePeakRangerCCAT" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize peakranger results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadPeakRangerSummaryCCAT">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadPeakRangerSummaryCCAT</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadPeakRangerSummaryCCAT" title="Permalink to this definition">¶</a></dt>
<dd><p>load Peakranger summarys.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callPeaksWithSPP">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callPeaksWithSPP</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callPeaksWithSPP" title="Permalink to this definition">¶</a></dt>
<dd><p>run SICER for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadSPP">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadSPP</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadSPP" title="Permalink to this definition">¶</a></dt>
<dd><p>load spp results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.summarizeSPP">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">summarizeSPP</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.summarizeSPP" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize SPP results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadSPPSummary">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadSPPSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadSPPSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>load sicer summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.estimateSPPQualityMetrics">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">estimateSPPQualityMetrics</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.estimateSPPQualityMetrics" title="Permalink to this definition">¶</a></dt>
<dd><p>estimate ChIP-Seq quality metrics using SPP</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadSPPQualityMetrics">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadSPPQualityMetrics</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadSPPQualityMetrics" title="Permalink to this definition">¶</a></dt>
<dd><p>load spp quality metrics.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callPeaksWithSPPForIDR">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callPeaksWithSPPForIDR</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callPeaksWithSPPForIDR" title="Permalink to this definition">¶</a></dt>
<dd><p>run SICER for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.applyIDR">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">applyIDR</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.applyIDR" title="Permalink to this definition">¶</a></dt>
<dd><p>apply IDR analysis.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.plotIDR">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">plotIDR</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.plotIDR" title="Permalink to this definition">¶</a></dt>
<dd><p>plot IDR results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.callPeaksWithScripture">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">callPeaksWithScripture</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.callPeaksWithScripture" title="Permalink to this definition">¶</a></dt>
<dd><p>run SICER for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadScripture">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadScripture</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadScripture" title="Permalink to this definition">¶</a></dt>
<dd><p>load scripture data.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.exportIntervalsAsBed">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">exportIntervalsAsBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.exportIntervalsAsBed" title="Permalink to this definition">¶</a></dt>
<dd><p>export all intervals as bed files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.flattenBedFiles">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">flattenBedFiles</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.flattenBedFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>dummy target - merge all files in exportIntervalsAsBed</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.getPeakShift">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">getPeakShift</tt><span class="sig-paren">(</span><em>track</em>, <em>method</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.getPeakShift" title="Permalink to this definition">¶</a></dt>
<dd><p>return peak shift for track and method.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.buildPeakShapeTable">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">buildPeakShapeTable</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.buildPeakShapeTable" title="Permalink to this definition">¶</a></dt>
<dd><p>build a table with peak shape parameters.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadPeakShapeTable">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadPeakShapeTable</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadPeakShapeTable" title="Permalink to this definition">¶</a></dt>
<dd><p>load peak shape information.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.makeReproducibilityOfMethods">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">makeReproducibilityOfMethods</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.makeReproducibilityOfMethods" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlap between intervals.</p>
<p>Note the exon percentages are approximations assuming that there are
not more than one intervals in one set overlapping one in the other set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.makeReproducibilityOfReplicates">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">makeReproducibilityOfReplicates</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.makeReproducibilityOfReplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlap between intervals.</p>
<p>Note the exon percentages are approximations assuming that there are
not more than one intervals in one set overlapping one in the other set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.loadReproducibility">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">loadReproducibility</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.loadReproducibility" title="Permalink to this definition">¶</a></dt>
<dd><p>load Reproducibility results</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.build_report">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">build_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>build report from scratch.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.update_report">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">update_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.update_report" title="Permalink to this definition">¶</a></dt>
<dd><p>update report.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_peakcalling.publish">
<tt class="descclassname">pipeline_peakcalling.</tt><tt class="descname">publish</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_peakcalling.publish" title="Permalink to this definition">¶</a></dt>
<dd><p>publish files.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Peak calling pipeline</a><ul>
<li><a class="reference internal" href="#overview">Overview</a><ul>
<li><a class="reference internal" href="#qc">QC</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
<li><a class="reference internal" href="#input">Input</a><ul>
<li><a class="reference internal" href="#mapped-reads">Mapped reads</a></li>
<li><a class="reference internal" href="#optional-inputs">Optional inputs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pipeline-output">Pipeline output</a><ul>
<li><a class="reference internal" href="#idr-analysis">IDR analysis</a></li>
</ul>
</li>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pipeline_transcriptome.html"
                        title="previous chapter">Transcriptome analysis pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pipeline_rnaseqlncrna.html"
                        title="next chapter">Long non-coding RNA pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelines/pipeline_peakcalling.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseqlncrna.html" title="Long non-coding RNA pipeline"
             >next</a> |</li>
        <li class="right" >
          <a href="pipeline_transcriptome.html" title="Transcriptome analysis pipeline"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" >NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>