
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Transcriptome analysis pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="NGS Pipelines" href="../CGATPipelines.html" />
    <link rel="next" title="Peak calling pipeline" href="pipeline_peakcalling.html" />
    <link rel="prev" title="RNA-Seq Differential expression pipeline" href="pipeline_rnaseqdiffexpression.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_peakcalling.html" title="Peak calling pipeline"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseqdiffexpression.html" title="RNA-Seq Differential expression pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" accesskey="U">NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pipeline_transcriptome"></span><div class="section" id="transcriptome-analysis-pipeline">
<h1>Transcriptome analysis pipeline<a class="headerlink" href="#transcriptome-analysis-pipeline" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>The transcriptome analysis pipeline examines various gene/transcript sets
and attempts to interprete and compare them. Among the quantities that it 
computes are:</p>
<blockquote>
<div><ul>
<li><p class="first">coding potential (using CPC)</p>
</li>
<li><p class="first">overlap with repeats</p>
</li>
<li><p class="first">substitution rates</p>
</li>
<li><dl class="first docutils">
<dt>conservation </dt>
<dd><ul class="first last simple">
<li>with respect to ancestral repeat rates)</li>
<li>TODO: conservation (with respect to UCSC tracks)</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>annotation with respect to a reference gene set, for example</dt>
<dd><ul class="first last simple">
<li>location (intronic, exonic, intergenic, ...)</li>
<li>distance to closest gene</li>
<li>intron overrun</li>
</ul>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="../UsingPipelines.html#pipelinesettingup"><em>Setting up a pipeline</em></a> and <a class="reference internal" href="../UsingPipelines.html#pipelinerunning"><em>Running a pipeline</em></a> on general information how to use CGAT pipelines.</p>
<div class="section" id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires a configured <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt> file.</p>
<p>The sphinxreport report requires a <tt class="file docutils literal"><span class="pre">conf.py</span></tt> and <tt class="file docutils literal"><span class="pre">sphinxreport.ini</span></tt> file 
(see <a class="reference internal" href="../UsingPipelines.html#pipelinereporting"><em>Building pipeline reports</em></a>). To start with, use the files supplied with the
<a class="reference internal" href="#example">Example</a> data.</p>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>The pipeline works from gene sets in <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> format within the <a class="reference internal" href="../make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
</div>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires the results from <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a> and <a class="reference internal" href="pipeline_ancestral_repeats.html"><em>Ancestral repeats pipeline</em></a>.</p>
<p>On top of the default CGAT setup, the pipeline requires the following software to be in the 
path:</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="22%" />
<col width="55%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Program</em></td>
<td><em>Version</em></td>
<td><em>Purpose</em></td>
</tr>
<tr class="row-even"><td>CPC</td>
<td>&gt;=0.9-r2</td>
<td>prediction of coding potential</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>Example data is available at <a class="reference external" href="http://www.cgat.org/~andreas/sample_data/pipeline_transcriptome.tgz">http://www.cgat.org/~andreas/sample_data/pipeline_transcriptome.tgz</a>.
To run the example, simply unpack and untar:</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://www.cgat.org/~andreas/sample_data/pipeline_transcriptome.tgz
tar -xvzf pipeline_transcriptome.tgz
cd pipeline_transcriptome
python &lt;srcdir&gt;/pipeline_transcriptome.py make full
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For the pipeline to run, install the <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a> and <a class="reference internal" href="pipeline_ancestral_repeats.html"><em>Ancestral repeats pipeline</em></a>
as well.</p>
</div>
</div>
<div class="section" id="glossary">
<h2>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h2>
<dl class="glossary docutils">
</dl>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pipeline_transcriptome.getSourceTrack">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">getSourceTrack</tt><span class="sig-paren">(</span><em>track</em>, <em>all_tracks</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.getSourceTrack" title="Permalink to this definition">¶</a></dt>
<dd><p>if track is in derived tracks, get the source track.</p>
<p>returns None if track is not a derived track</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.getRelatedTracks">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">getRelatedTracks</tt><span class="sig-paren">(</span><em>track</em>, <em>all_tracks</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.getRelatedTracks" title="Permalink to this definition">¶</a></dt>
<dd><p>return tracks in <tt class="docutils literal"><span class="pre">all_tracks</span></tt> that are related to <tt class="docutils literal"><span class="pre">track</span></tt> (including itself)</p>
<p>related tracks are build by merging or slicing another track.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.buildRepeatTrack">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">buildRepeatTrack</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.buildRepeatTrack" title="Permalink to this definition">¶</a></dt>
<dd><p>build a repeat track as negative control.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.buildIntronTrack">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">buildIntronTrack</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.buildIntronTrack" title="Permalink to this definition">¶</a></dt>
<dd><p>build an intron track as negative control.</p>
<p>The <tt class="docutils literal"><span class="pre">intron</span></tt> track is build from ancestral repeats. Here,
repeats are taken as exons, and artifical genes are constructed
from 2 to 5 ancestral repeats at least 100 bases apart, but
within 10 kb of each other.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadRepeatInformation">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadRepeatInformation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadRepeatInformation" title="Permalink to this definition">¶</a></dt>
<dd><p>load genome information.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.buildMergedTracks">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">buildMergedTracks</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.buildMergedTracks" title="Permalink to this definition">¶</a></dt>
<dd><p>merge gtf transcripts across all data sets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.buildFilteredAlignment">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">buildFilteredAlignment</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.buildFilteredAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>build a genomic alignment without exons from the reference gene set.</p>
<p>This alignment is used for rate computation within introns.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.buildAlignmentStats">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">buildAlignmentStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.buildAlignmentStats" title="Permalink to this definition">¶</a></dt>
<dd><p>build alignment statistics.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadGTF">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadGTF</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadGTF" title="Permalink to this definition">¶</a></dt>
<dd><p>load gtf files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.buildAnnotations">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">buildAnnotations</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.buildAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcripts by location (intergenic, intronic, ...)</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadAnnotations">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadAnnotations</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>load annotations</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeOverrun">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeOverrun</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeOverrun" title="Permalink to this definition">¶</a></dt>
<dd><p>compute intron overrun.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadOverrun">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadOverrun</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadOverrun" title="Permalink to this definition">¶</a></dt>
<dd><p>load annotations</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeDistances">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeDistances</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeDistances" title="Permalink to this definition">¶</a></dt>
<dd><p>compute intron overrun.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadDistances">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadDistances</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadDistances" title="Permalink to this definition">¶</a></dt>
<dd><p>load annotations</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeSegments">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeSegments</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeSegments" title="Permalink to this definition">¶</a></dt>
<dd><p>compute intron overrun.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadSegments">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadSegments</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadSegments" title="Permalink to this definition">¶</a></dt>
<dd><p>load segments</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeRepeats">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeRepeats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeRepeats" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlap with repeats</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadRepeats">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadRepeats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadRepeats" title="Permalink to this definition">¶</a></dt>
<dd><p>load repeat overlap</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeDifference">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeDifference</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeDifference" title="Permalink to this definition">¶</a></dt>
<dd><p>compute the difference between two sets</p>
<p>This target computes three differences
1. exon overlap %.diff.{diff,overlap,total,genes_ovl,genes_total,genes_uniq1,genes_uniq2}
An intronic transcript will no overlap with an exonic transrcipt.
2. overlap statistics on each gene in set 1: %.diff
3. intron/exon overlap %_genes.diff.{diff,overlap,total,genes_ovl,genes_total,genes_uniq1,genes_uniq2}
An intronic transcript will overlap with an exonic transcript.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadDifference">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadDifference</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadDifference" title="Permalink to this definition">¶</a></dt>
<dd><p>load overlap</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeOverlapAll">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeOverlapAll</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeOverlapAll" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlaps between sets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeOverlapUnknown">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeOverlapUnknown</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeOverlapUnknown" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlaps between sets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeOverlapAmbiguous">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeOverlapAmbiguous</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeOverlapAmbiguous" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlaps between sets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeOverlapKnown">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeOverlapKnown</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeOverlapKnown" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlaps between sets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeOverlapPC">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeOverlapPC</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeOverlapPC" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlaps between sets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadOverlap">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadOverlap</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadOverlap" title="Permalink to this definition">¶</a></dt>
<dd><p>load results of overlap computation.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeMergedOverlap">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeMergedOverlap</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeMergedOverlap" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlap of a track with the merged set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.buildRepeatMaskedSequences">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">buildRepeatMaskedSequences</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.buildRepeatMaskedSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>output sequences masking repeats.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.runAgainstProteinDatabase">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">runAgainstProteinDatabase</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.runAgainstProteinDatabase" title="Permalink to this definition">¶</a></dt>
<dd><p>run blastx.</p>
<p>search on both strands. Note that the CPC default is: only forward strand</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.runFrameFinder">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">runFrameFinder</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.runFrameFinder" title="Permalink to this definition">¶</a></dt>
<dd><p>run FrameFinder</p>
<p>search on both strands (-r TRUE). Note that CPC default is: only forward strand.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.buildCodingPotential">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">buildCodingPotential</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.buildCodingPotential" title="Permalink to this definition">¶</a></dt>
<dd><p>run CPC analysis as in the cpc script.</p>
<p>This module runs framefinder and blastx on both strands.
It seems to work, but I have not thoroughly tested it.
I expect that the false positive rate increases (i.e.,
predicting non-coding as coding) in cases where the best 
framefinder match and the best blast match are on opposite
strands. In the original CPC, these would be separated.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadCodingPotential">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadCodingPotential</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadCodingPotential" title="Permalink to this definition">¶</a></dt>
<dd><p>load annotations</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.exportSequences">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">exportSequences</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.exportSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>collect sequences from a gtf file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeRates">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeRates</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeRates" title="Permalink to this definition">¶</a></dt>
<dd><p>compute nucleotide substitution rates for transcripts from a gtf file - 
this applies only for transcripts mapped onto the reference genome.</p>
<p>Sequences from the transcripts are mapped onto the rate genome.</p>
<p>Softmasked sequence will be ignored unless track is in CONTROL_TRACKS.</p>
<p>The longest contiguous block is selected ignoring matches to other parts of the genome.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadRates">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadRates</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadRates" title="Permalink to this definition">¶</a></dt>
<dd><p>load rates.</p>
<p>Select the longest stretch for each transcript.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.makeRepeatsRates">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">makeRepeatsRates</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.makeRepeatsRates" title="Permalink to this definition">¶</a></dt>
<dd><p>collect ancestral repeats around transcripts and
estimate neutral rate.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadRepeatsRates">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadRepeatsRates</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadRepeatsRates" title="Permalink to this definition">¶</a></dt>
<dd><p>load repeat overlap</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.loadSummary">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">loadSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.loadSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>load several rates into a single convenience table.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_transcriptome.build_report">
<tt class="descclassname">pipeline_transcriptome.</tt><tt class="descname">build_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_transcriptome.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>build report from scratch.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Transcriptome analysis pipeline</a><ul>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
<li><a class="reference internal" href="#input">Input</a></li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
</ul>
</li>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#glossary">Glossary</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pipeline_rnaseqdiffexpression.html"
                        title="previous chapter">RNA-Seq Differential expression pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pipeline_peakcalling.html"
                        title="next chapter">Peak calling pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelines/pipeline_transcriptome.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_peakcalling.html" title="Peak calling pipeline"
             >next</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseqdiffexpression.html" title="RNA-Seq Differential expression pipeline"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" >NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>