
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>RNA-Seq Differential expression pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="NGS Pipelines" href="../CGATPipelines.html" />
    <link rel="next" title="Transcriptome analysis pipeline" href="pipeline_transcriptome.html" />
    <link rel="prev" title="RNA-Seq Transcript Build pipeline" href="pipeline_rnaseqtranscripts.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_transcriptome.html" title="Transcriptome analysis pipeline"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseqtranscripts.html" title="RNA-Seq Transcript Build pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" accesskey="U">NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pipeline_rnaseqdiffexpression"></span><div class="section" id="rna-seq-differential-expression-pipeline">
<h1>RNA-Seq Differential expression pipeline<a class="headerlink" href="#rna-seq-differential-expression-pipeline" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>The RNA-Seq differential expression pipeline performs differential
expression analysis. It requires three inputs:</p>
<blockquote>
<div>1 one or more genesets in <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted files 
2 mapped reads in <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted files
3 design files as <a class="reference internal" href="../glossary.html#term-tsv"><em class="xref std std-term">tsv</em></a>-separated format</div></blockquote>
<p>This pipeline works on a single genome.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The pipeline assumes the data derive from multiple tissues/conditions (<a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>) 
with one or more biological and/or technical replicates (<a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a>). A <a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a>
within each <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a> is a <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a>.</p>
<dl class="docutils">
<dt>The pipeline performs the following:</dt>
<dd><ul class="first last">
<li><p class="first">compute tag counts on an exon, transcript and gene level for each of the genesets</p>
</li>
<li><p class="first">estimate FPKM using cufflinks for each of the gene sets</p>
</li>
<li><dl class="first docutils">
<dt>perform differential expression analysis. The methods currently implemented are:</dt>
<dd><ul class="first last simple">
<li>DESeq</li>
<li>EdgeR</li>
<li>Cuffdiff</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="background">
<h2>Background<a class="headerlink" href="#background" title="Permalink to this headline">¶</a></h2>
<p>The quality of the rnaseq data (read-length, paired-end) determines
the quality of transcript models. For instance, if reads are short
(35bp) and/or reads are not paired-ended, transcript models will be
short and truncated.  In these cases it might be better to concentrate
the analysis on only previously known transcript models.</p>
<p>Transcripts are the natural choice to measure expression of. However
other quantities might be of interest. Some quantities are biological
meaningful, for example differential expression from a promotor shared
by several trancripts. Other quantities might no biologically
meaningful but are necessary as a technical comprise.  For example,
the overlapping transcripts might be hard to resolve and thus might
need to be aggregated per gene. Furthermore, functional annotation is
primarily associated with genes and not individual transcripts. The
pipeline attempts to measure transcription and differential expression
for a variety of entities following the classification laid down by
<a class="reference internal" href="#term-cuffdiff"><em class="xref std std-term">cuffdiff</em></a>:</p>
<dl class="docutils">
<dt>isoform</dt>
<dd>Transcript level</dd>
<dt>gene</dt>
<dd>Gene level, aggregates several isoform/transcripts</dd>
<dt>tss</dt>
<dd>Transcription start site. Aggregate all isoforms starting from the same <a class="reference internal" href="../glossary.html#term-tss"><em class="xref std std-term">tss</em></a>.</dd>
<dt>cds</dt>
<dd>Coding sequence expression. Ignore reads overlapping non-coding parts of transcripts (UTRs, etc.). Requires
annotation of the cds and thus only available for <tt class="file docutils literal"><span class="pre">reference.gtf.gz</span></tt>.</dd>
</dl>
<p>Methods differ in their ability to measure transcription on all
levels.</p>
<div class="admonition-todo admonition" id="index-0">
<p class="first admonition-title">Todo</p>
<p class="last">add promoters and splicing output</p>
</div>
<p>Overprediction of differential expression for low-level expressed
transcripts with <a class="reference internal" href="#term-cuffdiff"><em class="xref std std-term">cuffdiff</em></a> is a <a class="reference external" href="http://seqanswers.com/forums/showthread.php?t=6283&amp;highlight=fpkm">known problem</a>.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="../UsingPipelines.html#pipelinesettingup"><em>Setting up a pipeline</em></a> and <a class="reference internal" href="../UsingPipelines.html#pipelinerunning"><em>Running a pipeline</em></a> on general
information how to use CGAT pipelines.</p>
<div class="section" id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires a configured <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt> file.</p>
<p>The sphinxreport report requires a <tt class="file docutils literal"><span class="pre">conf.py</span></tt> and
<tt class="file docutils literal"><span class="pre">sphinxreport.ini</span></tt> file (see <a class="reference internal" href="../UsingPipelines.html#pipelinereporting"><em>Building pipeline reports</em></a>). To start
with, use the files supplied with the <a class="reference internal" href="#example">Example</a> data.</p>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="section" id="reads">
<h4>Reads<a class="headerlink" href="#reads" title="Permalink to this headline">¶</a></h4>
<p>Reads are imported by placing <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted files are linking
to files in the <a class="reference internal" href="../make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
<p>The default file format assumes the following convention:</p>
<blockquote>
<div>&lt;sample&gt;-&lt;condition&gt;-&lt;replicate&gt;.&lt;suffix&gt;</div></blockquote>
<p><tt class="docutils literal"><span class="pre">sample</span></tt> and <tt class="docutils literal"><span class="pre">condition</span></tt> make up an <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>, while
<tt class="docutils literal"><span class="pre">replicate</span></tt> denotes the <a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a> within an
<a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>.</p>
</div>
<div class="section" id="genesets">
<h4>Genesets<a class="headerlink" href="#genesets" title="Permalink to this headline">¶</a></h4>
<p>Genesets are imported by placing <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted files in the
<a class="reference internal" href="../make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
</div>
<div class="section" id="design-matrices">
<h4>Design matrices<a class="headerlink" href="#design-matrices" title="Permalink to this headline">¶</a></h4>
<p>Design matrices are imported by placing <a class="reference internal" href="../glossary.html#term-tsv"><em class="xref std std-term">tsv</em></a> formatted files
into the <a class="reference internal" href="../make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.  The file has four columns:</p>
<div class="highlight-python"><div class="highlight"><pre>track   include group   pair
CW-CD14-R1      0       CD14    1
CW-CD14-R2      0       CD14    1
CW-CD14-R3      1       CD14    1
CW-CD4-R1       1       CD4     1
FM-CD14-R1      1       CD14    2
FM-CD4-R2       0       CD4     2
FM-CD4-R3       0       CD4     2
FM-CD4-R4       0       CD4     2
</pre></div>
</div>
<dl class="docutils">
<dt>track</dt>
<dd>name of track - should correspond to column header in <em>infile</em></dd>
<dt>include</dt>
<dd>flag to indicate whether or not to include this data</dd>
<dt>group</dt>
<dd>group indicator - experimental group</dd>
<dt>pair</dt>
<dd>pair that sample belongs to (for paired tests) - set to 0 if the
design is not paired.</dd>
</dl>
</div>
<div class="section" id="optional-inputs">
<h4>Optional inputs<a class="headerlink" href="#optional-inputs" title="Permalink to this headline">¶</a></h4>
</div>
</div>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires the results from
<a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a>. Set the configuration variable
<tt class="xref py py-data docutils literal"><span class="pre">annotations_database</span></tt> and <tt class="xref py py-data docutils literal"><span class="pre">annotations_dir</span></tt>.</p>
<p>On top of the default CGAT setup, the pipeline requires the following
software to be in the path:</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="22%" />
<col width="55%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Program</em></td>
<td><em>Version</em></td>
<td><em>Purpose</em></td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a></td>
<td>&gt;=1.3.0</td>
<td>transcription levels</td>
</tr>
<tr class="row-odd"><td>samtools</td>
<td>&gt;=0.1.16</td>
<td>bam/sam files</td>
</tr>
<tr class="row-even"><td>bedtools</td>
<td>&nbsp;</td>
<td>working with intervals</td>
</tr>
<tr class="row-odd"><td>R/DESeq</td>
<td>&nbsp;</td>
<td>differential expression</td>
</tr>
<tr class="row-even"><td>sra-tools</td>
<td>&nbsp;</td>
<td>extracting reads from .sra files</td>
</tr>
<tr class="row-odd"><td>picard</td>
<td>&gt;=1.42</td>
<td><dl class="first last docutils">
<dt>bam/sam files. The .jar files need to be in your</dt>
<dd>CLASSPATH environment variable.</dd>
</dl>
</td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://www.agf.liv.ac.uk/454/sabkea/samStats_13-01-2011">bamstats</a></td>
<td>&gt;=1.22</td>
<td>from CGR, Liverpool</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="pipeline-output">
<h2>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this headline">¶</a></h2>
<div class="section" id="fpkm-values">
<h3>FPKM values<a class="headerlink" href="#fpkm-values" title="Permalink to this headline">¶</a></h3>
<p>The directory <tt class="file docutils literal"><span class="pre">fpkm.dir</span></tt> contains the output of running
cufflinks against each set of reads for each geneset.</p>
<p>The syntax of the output files is <tt class="docutils literal"><span class="pre">&lt;geneset&gt;_&lt;track&gt;.cufflinks</span></tt>. The
FPKM values are uploaded into tables as <tt class="docutils literal"><span class="pre">&lt;geneset&gt;_&lt;track&gt;_fpkm</span></tt> for
per-transcript FPKM values and <tt class="docutils literal"><span class="pre">&lt;geneset&gt;_&lt;track&gt;_genefpkm</span></tt> for
per-gene FPKM values.</p>
</div>
<div class="section" id="differential-gene-expression-results">
<h3>Differential gene expression results<a class="headerlink" href="#differential-gene-expression-results" title="Permalink to this headline">¶</a></h3>
<p>Differential expression is estimated for different genesets with a
variety of methods. Results are stored per method in subdirectories
such as <tt class="file docutils literal"><span class="pre">deseq.dir</span></tt>, <tt class="file docutils literal"><span class="pre">edger.dir</span></tt> or <tt class="file docutils literal"><span class="pre">cuffdiff.dir</span></tt>.</p>
<p>The syntax of the output files is <tt class="docutils literal"><span class="pre">&lt;design&gt;_&lt;geneset&gt;_...</span></tt>.</p>
<dl class="docutils">
<dt><tt class="file docutils literal"><span class="pre">.diff</span></tt> files:</dt>
<dd><p class="first">The .diff files are <a class="reference internal" href="../glossary.html#term-tsv"><em class="xref std std-term">tsv</em></a> formatted tables with the result of the differential
expression analysis.</p>
<p>The column names are:</p>
<dl class="last docutils">
<dt>test_id </dt>
<dd>gene_id/transcript_id/tss_id, etc</dd>
<dt>treatment_name  </dt>
<dd>name of the treatment</dd>
<dt>control_name    </dt>
<dd>name of the control</dd>
<dt>!=_mean  </dt>
<dd>mean expression value of treatment/control</dd>
<dt>!=_std   </dt>
<dd>standard deviation of treatment/control</dd>
<dt>pvalue  </dt>
<dd>pvalue of test</dd>
<dt>qvalue  </dt>
<dd>qvalue of test (FDR if the threshold was set at pvalue)</dd>
<dt>l2fold  </dt>
<dd>log2 of fold change</dd>
<dt>fold    </dt>
<dd>fold change, treatment / control</dd>
<dt>significant     </dt>
<dd>flag: 1 if called significant</dd>
<dt>status</dt>
<dd>status of test. Values are 
OK: test ok, FAIL: test failed, NOCALL: no test (insufficient data, etc.).</dd>
</dl>
</dd>
</dl>
<p>The results are uploaded into the database as tables called
<tt class="docutils literal"><span class="pre">&lt;design&gt;_&lt;geneset&gt;_&lt;method&gt;_&lt;level&gt;_&lt;section&gt;</span></tt>.</p>
<p>Level denotes the biological entity that the differential analysis was
performed on.  Level can be one of <tt class="docutils literal"><span class="pre">cds</span></tt>,``isoform``,``tss`` and
<tt class="docutils literal"><span class="pre">gene</span></tt>. The later should always be present.</p>
<p>Section is <tt class="docutils literal"><span class="pre">diff</span></tt> for differential expression results
(<tt class="file docutils literal"><span class="pre">.diff`</span></tt> files) and <tt class="docutils literal"><span class="pre">levels</span></tt> for expression levels.</p>
</div>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>Example data is available at
<a class="reference external" href="http://www.cgat.org/~andreas/sample_data/pipeline_rnaseqdiffexpression.tgz">http://www.cgat.org/~andreas/sample_data/pipeline_rnaseqdiffexpression.tgz</a>.
To run the example, simply unpack and untar:</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://www.cgat.org/~andreas/sample_data/pipeline_rnaseqdiffexpression.tgz
tar -xvzf pipeline_rnaseq.tgz
cd pipeline_rnaseq
python &lt;srcdir&gt;/pipeline_rnaseq.py make full
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For the pipeline to run, install the <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a> as well.</p>
</div>
</div>
<div class="section" id="glossary">
<h2>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h2>
<dl class="glossary docutils">
<dt id="term-cufflinks">cufflinks</dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a> - transcriptome analysis</dd>
<dt id="term-tophat">tophat</dt>
<dd><a class="reference external" href="http://tophat.cbcb.umd.edu/">tophat</a> - a read mapper to detect splice-junctions</dd>
<dt id="term-deseq">deseq</dt>
<dd><a class="reference external" href="http://www-huber.embl.de/users/anders/DESeq/">deseq</a> - differential expression analysis</dd>
<dt id="term-cuffdiff">cuffdiff</dt>
<dd>find differentially expressed transcripts. Part of <a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a>.</dd>
<dt id="term-cuffcompare">cuffcompare</dt>
<dd>compare transcriptomes. Part of <a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a>.</dd>
</dl>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.connect">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">connect</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.connect" title="Permalink to this definition">¶</a></dt>
<dd><p>connect to database.</p>
<p>This method also attaches to helper databases.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildMaskGtf">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildMaskGtf</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildMaskGtf" title="Permalink to this definition">¶</a></dt>
<dd><p>This takes ensembl annotations (geneset_all.gtf.gz) and writes out all entries that 
have a &#8216;source&#8217; match to &#8220;rRNA&#8221; or &#8216;contig&#8217; match to &#8220;chrM&#8221;. for use with cufflinks</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.runCufflinks">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">runCufflinks</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.runCufflinks" title="Permalink to this definition">¶</a></dt>
<dd><p>estimate expression levels in each set using cufflinks.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.loadCufflinks">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">loadCufflinks</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.loadCufflinks" title="Permalink to this definition">¶</a></dt>
<dd><p>load expression level measurements.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.mergeCufflinksGeneFPKM">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">mergeCufflinksGeneFPKM</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.mergeCufflinksGeneFPKM" title="Permalink to this definition">¶</a></dt>
<dd><p>build aggregate table with cufflinks FPKM values.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.mergeCufflinksIsoformFPKM">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">mergeCufflinksIsoformFPKM</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.mergeCufflinksIsoformFPKM" title="Permalink to this definition">¶</a></dt>
<dd><p>build aggregate table with cufflinks FPKM values.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.loadCufflinksFPKM">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">loadCufflinksFPKM</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.loadCufflinksFPKM" title="Permalink to this definition">¶</a></dt>
<dd><p>load fkpm data into table.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.runCuffdiff">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">runCuffdiff</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.runCuffdiff" title="Permalink to this definition">¶</a></dt>
<dd><p>perform differential expression analysis using cuffdiff.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.loadCuffdiff">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">loadCuffdiff</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.loadCuffdiff" title="Permalink to this definition">¶</a></dt>
<dd><p>load results from differential expression analysis and produce
summary plots.</p>
<p>Note: converts from ln(fold change) to log2 fold change.</p>
<p>The cuffdiff output is parsed.</p>
<p>Pairwise comparisons in which one gene is not expressed (fpkm &lt; fpkm_silent)
are set to status &#8216;NOCALL&#8217;. These transcripts might nevertheless be significant.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildExpressionStats">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildExpressionStats</tt><span class="sig-paren">(</span><em>tables</em>, <em>method</em>, <em>outfile</em>, <em>outdir</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildExpressionStats" title="Permalink to this definition">¶</a></dt>
<dd><p>build expression summary statistics.</p>
<p>Creates some diagnostic plots in</p>
<p>&lt;exportdir&gt;/&lt;method&gt; directory.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildCuffdiffPlots">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildCuffdiffPlots</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildCuffdiffPlots" title="Permalink to this definition">¶</a></dt>
<dd><p>create summaries of cufflinks results (including some diagnostic plots)</p>
<p>Plots are created in the &lt;exportdir&gt;/cuffdiff directory.</p>
<p>Plots are:</p>
<dl class="docutils">
<dt>&lt;geneset&gt;_&lt;method&gt;_&lt;level&gt;_&lt;track1&gt;_vs_&lt;track2&gt;_significance.png</dt>
<dd>fold change against expression level</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.loadCuffdiffStats">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">loadCuffdiffStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.loadCuffdiffStats" title="Permalink to this definition">¶</a></dt>
<dd><p>import cuffdiff results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildCodingExons">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildCodingExons</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildCodingExons" title="Permalink to this definition">¶</a></dt>
<dd><p>compile set of protein coding exons.</p>
<p>This set is used for splice-site validation</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildUnionIntersectionExons">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildUnionIntersectionExons</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildUnionIntersectionExons" title="Permalink to this definition">¶</a></dt>
<dd><p>build union/intersection genes according to Bullard et al. (2010) BMC Bioinformatics.</p>
<p>Builds a single-segment bed file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildUnionExons">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildUnionExons</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildUnionExons" title="Permalink to this definition">¶</a></dt>
<dd><p>build union genes.</p>
<p>Exons across all transcripts of a gene are merged.
They are then intersected between genes to remove any overlap.</p>
<p>Builds a single-segment bed file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildGeneLevelReadCounts">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildGeneLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildGeneLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>compute coverage of exons with reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.aggregateGeneLevelReadCounts">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">aggregateGeneLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.aggregateGeneLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a matrix of counts with genes and tracks dimensions</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildGeneLevelReadExtension">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildGeneLevelReadExtension</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildGeneLevelReadExtension" title="Permalink to this definition">¶</a></dt>
<dd><p>compute extension of cds.</p>
<p>Known UTRs are counted as well.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildTranscriptLevelReadCounts">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildTranscriptLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildTranscriptLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>count reads falling into transcripts of protein coding </dt>
<dd>gene models.</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In paired-end data sets each mate will be counted. Thus
the actual read counts are approximately twice the fragment
counts.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.buildFeatureCounts">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">buildFeatureCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.buildFeatureCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>counts reads falling into &#8220;features&#8221;, which by default are genes.</p>
<p>A read overlaps if at least one bp overlaps.</p>
<p>Pairs and strandedness can be used to resolve reads falling into
more than one feature. Reads that cannot be resolved to a single
feature are ignored.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.aggregateFeatureCounts">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">aggregateFeatureCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.aggregateFeatureCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a matrix of counts with genes and tracks dimensions.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.summarizeFeatureCountsAll">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">summarizeFeatureCountsAll</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.summarizeFeatureCountsAll" title="Permalink to this definition">¶</a></dt>
<dd><p>perform summarization of read counts</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.summarizeFeatureCountsPerDesign">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">summarizeFeatureCountsPerDesign</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.summarizeFeatureCountsPerDesign" title="Permalink to this definition">¶</a></dt>
<dd><p>perform summarization of read counts within experiments.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.loadTagCountSummary">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">loadTagCountSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.loadTagCountSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>load windows summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.runDESeq">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">runDESeq</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.runDESeq" title="Permalink to this definition">¶</a></dt>
<dd><p>perform differential expression analysis using deseq.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.loadDESeq">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">loadDESeq</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.loadDESeq" title="Permalink to this definition">¶</a></dt>
<dd><p>load differential expression results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.runEdgeR">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">runEdgeR</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.runEdgeR" title="Permalink to this definition">¶</a></dt>
<dd><p>perform differential expression analysis using edger.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.loadEdgeR">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">loadEdgeR</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.loadEdgeR" title="Permalink to this definition">¶</a></dt>
<dd><p>load differential expression results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.loadDEStats">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">loadDEStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.loadDEStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load DE stats into table.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.plotRNASEQTagData">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">plotRNASEQTagData</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.plotRNASEQTagData" title="Permalink to this definition">¶</a></dt>
<dd><p>perform differential expression analysis using deseq.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.plotRNASEQDEData">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">plotRNASEQDEData</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.plotRNASEQDEData" title="Permalink to this definition">¶</a></dt>
<dd><p>plot differential expression stats</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.build_report">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">build_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>build report from scratch.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.update_report">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">update_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.update_report" title="Permalink to this definition">¶</a></dt>
<dd><p>update report.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseqdiffexpression.publish">
<tt class="descclassname">pipeline_rnaseqdiffexpression.</tt><tt class="descname">publish</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseqdiffexpression.publish" title="Permalink to this definition">¶</a></dt>
<dd><p>publish files.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">RNA-Seq Differential expression pipeline</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#background">Background</a></li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
<li><a class="reference internal" href="#input">Input</a><ul>
<li><a class="reference internal" href="#reads">Reads</a></li>
<li><a class="reference internal" href="#genesets">Genesets</a></li>
<li><a class="reference internal" href="#design-matrices">Design matrices</a></li>
<li><a class="reference internal" href="#optional-inputs">Optional inputs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pipeline-output">Pipeline output</a><ul>
<li><a class="reference internal" href="#fpkm-values">FPKM values</a></li>
<li><a class="reference internal" href="#differential-gene-expression-results">Differential gene expression results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#glossary">Glossary</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pipeline_rnaseqtranscripts.html"
                        title="previous chapter">RNA-Seq Transcript Build pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pipeline_transcriptome.html"
                        title="next chapter">Transcriptome analysis pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelines/pipeline_rnaseqdiffexpression.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_transcriptome.html" title="Transcriptome analysis pipeline"
             >next</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseqtranscripts.html" title="RNA-Seq Transcript Build pipeline"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" >NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>