
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>RNA-Seq pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="NGS Pipelines" href="../CGATPipelines.html" />
    <link rel="next" title="Lecgacy pipelines" href="../MakePipelines.html" />
    <link rel="prev" title="ChIP-Seq pipeline" href="pipeline_chipseq.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../MakePipelines.html" title="Lecgacy pipelines"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pipeline_chipseq.html" title="ChIP-Seq pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" accesskey="U">NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pipeline_rnaseq"></span><div class="section" id="rna-seq-pipeline">
<h1>RNA-Seq pipeline<a class="headerlink" href="#rna-seq-pipeline" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>The RNA-Seq pipeline imports unmapped reads from one or more
RNA-Seq experiments and performs the basic RNA-Seq analysis steps:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Map reads to genome</li>
<li>Build transcript models</li>
<li>Estimate differential expression</li>
</ol>
</div></blockquote>
<p>This pipeline works on a single genome.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The pipeline assumes the data derive from multiple tissues/conditions (<a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>) 
with one or more biological and/or technical replicates (<a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a>). A <a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a>
within each <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a> is a <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a>.</p>
<p>The pipeline performs the following tasks:</p>
<blockquote>
<div><ul>
<li><dl class="first docutils">
<dt>analyse each experiment:</dt>
<dd><ul class="first last">
<li><dl class="first docutils">
<dt>for each replicate</dt>
<dd><ul class="first last simple">
<li>map reads using tophat for each term:<cite>replicate</cite> separately.</li>
<li>predict splice isoforms and expression levels with <a class="reference internal" href="pipeline_rnaseqtranscripts.html#term-cufflinks"><em class="xref std std-term">cufflinks</em></a>.</li>
<li>estimate expression levels of reference gene set with <a class="reference internal" href="pipeline_rnaseqtranscripts.html#term-cufflinks"><em class="xref std std-term">cufflinks</em></a>.</li>
<li>annotate isoforms in replicates with genomic annotations</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>compare isoforms in replicates within each <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a> (<a class="reference internal" href="pipeline_rnaseqdiffexpression.html#term-cuffcompare"><em class="xref std std-term">cuffcompare</em></a>) </dt>
<dd><p class="first last">and to reference gene set.</p>
</dd>
</dl>
</li>
<li><p class="first">summary statistics on reproducibility within each experiment</p>
</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>build a combined gene set including the reference gene set and isoforms predicted by <a class="reference internal" href="pipeline_rnaseqtranscripts.html#term-cufflinks"><em class="xref std std-term">cufflinks</em></a>.</dt>
<dd><ul class="first last">
<li><dl class="first docutils">
<dt>compare all isoforms in all experiments+isoforms (<a class="reference internal" href="pipeline_rnaseqdiffexpression.html#term-cuffcompare"><em class="xref std std-term">cuffcompare</em></a>) to each other </dt>
<dd><p class="first last">and the reference gene set
* summary statistics on isoforms with respect to gene set</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>estimate differential expression levels of transcripts</dt>
<dd><ul class="first last">
<li><dl class="first docutils">
<dt>different gene sets</dt>
<dd><ul class="first last simple">
<li>reference gene set</li>
<li>combined gene set</li>
<li>novel gene set</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>different methods</dt>
<dd><ul class="first last simple">
<li><a class="reference internal" href="pipeline_rnaseqdiffexpression.html#term-deseq"><em class="xref std std-term">DESeq</em></a> (tag counting)</li>
<li><a class="reference internal" href="pipeline_rnaseqdiffexpression.html#term-cuffdiff"><em class="xref std std-term">cuffdiff</em></a></li>
</ul>
</dd>
</dl>
</li>
<li><p class="first">summary statistics on differential expression</p>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<div class="section" id="mapping-strategy">
<h3>Mapping strategy<a class="headerlink" href="#mapping-strategy" title="Permalink to this headline">¶</a></h3>
<p>The best strategy for mapping and transcriptome assembly depends on the length of your reads.
With short reads, detecting novel splice-junctions is a difficult task. In this case it will be
best to rely on a set of known splice-junctions. Longer reads map more easily across splice-junctions.</p>
<p>From the tophat manual:</p>
<div class="highlight-python"><div class="highlight"><pre>TopHat finds splice junctions without a reference annotation. TopHat version 1.4 maps RNA-seq reads
first to a reference transcriptome. Only those reads that don&#39;t map in this initial process are 
mapped against the genome.

Through the second stage of genome mapping, TopHat identifies novel splice junctions and then confirms
these through mapping to known junctions.

Short read sequencing machines can currently produce reads 100bp or longer, but many exons are 
shorter than this, and so would be missed in the initial mapping. TopHat solves this problem 
by splitting all input reads into smaller segments, and then mapping them independently. The segment 
alignments are &quot;glued&quot; back together in a final step of the program to produce the end-to-end read alignments.

TopHat generates its database of possible splice junctions from three sources of evidence. The first 
source is pairings of &quot;coverage islands&quot;, which are distinct regions of piled up reads in the 
initial mapping. Neighboring islands are often spliced together in the transcriptome, so 
TopHat looks for ways to join these with an intron. The second source is only used when 
TopHat is run with paired end reads. When reads in a pair come from different exons of a 
transcript, they will generally be mapped far apart in the genome coordinate space. When 
this happens, TopHat tries to &quot;close&quot; the gap between them by looking for subsequences of 
the genomic interval between mates with a total length about equal to the expected distance 
between mates. The &quot;introns&quot; in this subsequence are added to the database. The third, and 
strongest, source of evidence for a splice junction is when two segments from the same read #
are mapped far apart, or when an internal segment fails to map. With long (&gt;=75bp) reads, 
&quot;GT-AG&quot;, &quot;GC-AG&quot; and &quot;AT-AC&quot; introns be found ab initio. With shorter reads, TopHat only 
reports alignments across &quot;GT-AG&quot; introns
</pre></div>
</div>
<p>Thus, in order to increase the sensitivity of splice-site detection, it might be best to derive a set of 
splice-junctions using all reads. This is not done automatically, but can be done manually by 
adding a file with junctions to the <tt class="docutils literal"><span class="pre">tophat_options</span></tt> entry in the configuration file.</p>
<p>The pipeline supplies tophat with a list of all coding exons to facilitate mapping across known
splice-junctions. If they are prioritized, I do not know.</p>
<p>Transcripts are built individually for each <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a>. This seems to be the most rigorous way
as there might be conflicting transcripts between replicates and merging the sets might confuse transcript
reconstruction. Also, conflicting transcripts between replicates give an idea of the variability of the data. 
However, if there are only few reads,  there might be a case for building transcript models using reads 
from all replicates of an experiment. However, there is no reason to merge reads between experiments.</p>
</div>
<div class="section" id="lincrna">
<h3>LincRNA<a class="headerlink" href="#lincrna" title="Permalink to this headline">¶</a></h3>
<p>One of the main benefits of RNASeq over microarrays is that novel transcripts can be detected. A particular
interest are currently novel long non-coding RNA. Unfortunately, it seems that these transcripts
are often expressed at very low levels and possibly in a highly regulated manner, for example only in
certain tissues. On top of their low abundance, they frequently seem to be co-localized with protein
coding genes, making it hard to distinguish them from transcription artifacts.</p>
<p>Success in identifying lincRNA will depend a lot on your input data. Long, paired-end reads are likely 
to lead to success. Unfortunately, many exploratory studies go for single-ended, short read data.
With such data, identification of novel spliced transcripts will be rare and the set of novel transcripts
is likely to contain many false-positives.</p>
<dl class="docutils">
<dt>The pipeline constructs a set of novel lncRNA in the following manner:</dt>
<dd><ol class="first last arabic simple">
<li>All transcript models overlapping protein coding transcripts are removed.</li>
<li>Overlapping lncRNA on the same strand are merged.</li>
</ol>
</dd>
</dl>
<div class="section" id="artifacts-in-lncrna-analysis">
<h4>Artifacts in lncRNA analysis<a class="headerlink" href="#artifacts-in-lncrna-analysis" title="Permalink to this headline">¶</a></h4>
<p>There are several sources of artifacts in lncRNA analysis</p>
<div class="section" id="read-mapping-errors">
<h5>Read mapping errors<a class="headerlink" href="#read-mapping-errors" title="Permalink to this headline">¶</a></h5>
<p>Mapping errors are identifyable as sharp peaks in the
coverage profile. Mapping errors occur if the true location of a read has more mismatches
than the original location or it maps across an undetected splice-site. Most of the 
highly-expressed lncRNA are due to mapping errors. Secondary locations very often overlap
highly-expressed protein-coding genes. These errors are annoying for two reasons: they
provide false positives, but at the same time prevent the reads to be counted towards
the expression of the true gene.</p>
<p>They can be detected in two ways:</p>
<p>1. via a peak-like distribution of reads which should result in a low entropy of start position 
density. Note that this possibly can remove transcripts that are close to the length of a single
read.</p>
<p>2. via mapping against known protein coding transcripts. However, getting this mapping right
is hard for two reasons. Firstly, mapping errors usually involve reads aligned with mismatches.
Thus, the mapping has to be done either on the read-level (computationally expensive), or 
on the transcript level after variant calling. (tricky, and also computationally expensive). 
Secondly, as cufflinks extends transcripts generously, only a part of a transcript might actually
be a mismapped part. Distinguishing partial true matches from random matches will be tricky.</p>
<p>Read mapping errors can also be avoided by</p>
<ol class="arabic simple">
<li>Longer read lengths</li>
<li>Strict alignment criteria</li>
<li>A two-stage mapping process.</li>
</ol>
</div>
<div class="section" id="fragments">
<h5>Fragments<a class="headerlink" href="#fragments" title="Permalink to this headline">¶</a></h5>
<p>As lncRNA are expressed at low levels, it is likely that only a partial transcript
can be observed.</p>
</div>
</div>
</div>
<div class="section" id="differential-expression">
<h3>Differential expression<a class="headerlink" href="#differential-expression" title="Permalink to this headline">¶</a></h3>
<p>The quality of the rnaseq data (read-length, paired-end) determines the
quality of transcript models. For instance, if reads are short (35bp) and/or 
reads are not paired-ended, transcript models will be short and truncated.
In these cases it might be better to concentrate the analysis on only previously
known transcript models.</p>
<p>The pipeline offers various sets for downstream analysis of differential expression.</p>
<ol class="arabic simple">
<li>A set of previously known transcripts (<tt class="file docutils literal"><span class="pre">reference.gtf.gz</span></tt>). Use this set if 
only interested in the transcription of previously known transcripts or read length
does not permit transcript assembly. This does include all transcripts within the
ENSEMBL gene set, including processed but untranscribed transcripts, transcripts with
retained introns, pseudogenes, etc.</li>
<li>A set of previously known protein coding transcripts (<tt class="file docutils literal"><span class="pre">refcoding.gtf.gz</span></tt>).
This set is derived from (<tt class="file docutils literal"><span class="pre">reference.gtf.gz</span></tt>) but only includes exons of 
transcripts that are protein coding.</li>
<li>An ab-initio gene set (<tt class="file docutils literal"><span class="pre">abinitio.gtf.gz</span></tt>). The ab-initio set is built by running <a class="reference internal" href="pipeline_rnaseqdiffexpression.html#term-cuffcompare"><em class="xref std std-term">cuffcompare</em></a> on
the combined individual <a class="reference internal" href="pipeline_rnaseqtranscripts.html#term-cufflinks"><em class="xref std std-term">cufflinks</em></a> results. Transcripts that have been observed in only
one <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a> are removed (removed transcripts end up in <tt class="file docutils literal"><span class="pre">removed.gtf.gz</span></tt>) in order
to exclude partial transcripts. Use this set if reads are of good length and/or are paired-ended.</li>
<li>A set of novel transcribed loci (<tt class="file docutils literal"><span class="pre">novel.gtf.gz</span></tt>). This gene set is derived from the
set of ab-initio transcripts. All ab-initio transcripts overlapping protein coding transcripts 
in <tt class="file docutils literal"><span class="pre">refcoding.gtf.gz</span></tt> are removed. Overlapping transcripts are merged into a single 
transcript/gene. This removes individual transcript structure, but retains constitutive 
introns. This set retains transcripts that are only observed in a single experiment. It also
includes known non-coding transcripts, so a locus might not necessarily be &#8220;novel&#8221;.</li>
</ol>
<p>Transcripts are the natural choice to measure expression of. However other quantities
might be of interest. Some quantities are biological meaningful, for example differential 
expression from a promotor shared by several trancripts. Other quantities might no biologically
meaningful but are necessary as a technical comprise.
For example, the overlapping transcripts might be hard to resolve and thus might need to be
aggregated per gene. Furthermore, functional annotation is primarily associated with genes
and not individual transcripts. The pipeline attempts to measure transcription and differential
expression for a variety of entities following the classification laid down by <a class="reference internal" href="pipeline_rnaseqdiffexpression.html#term-cuffdiff"><em class="xref std std-term">cuffdiff</em></a>:</p>
<dl class="docutils">
<dt>isoform</dt>
<dd>Transcript level</dd>
<dt>gene</dt>
<dd>Gene level, aggregates several isoform/transcripts</dd>
<dt>tss</dt>
<dd>Transcription start site. Aggregate all isoforms starting from the same <a class="reference internal" href="../glossary.html#term-tss"><em class="xref std std-term">tss</em></a>.</dd>
<dt>cds</dt>
<dd>Coding sequence expression. Ignore reads overlapping non-coding parts of transcripts (UTRs, etc.). Requires
annotation of the cds and thus only available for <tt class="file docutils literal"><span class="pre">reference.gtf.gz</span></tt>.</dd>
</dl>
<p>Methods differ in their ability to measure transcription on all levels.</p>
<div class="admonition-todo admonition" id="index-0">
<p class="first admonition-title">Todo</p>
<p class="last">add promoters and splicing output</p>
</div>
<p>Overprediction of differential expression for low-level expressed transcripts with <a class="reference internal" href="pipeline_rnaseqdiffexpression.html#term-cuffdiff"><em class="xref std std-term">cuffdiff</em></a>
is a <a class="reference external" href="http://seqanswers.com/forums/showthread.php?t=6283&amp;highlight=fpkm">known problem</a>.</p>
</div>
<div class="section" id="estimating-coverage">
<h3>Estimating coverage<a class="headerlink" href="#estimating-coverage" title="Permalink to this headline">¶</a></h3>
<p>An important question in RNASeq analysis is if the sequencing has been done to sufficient depth.
The questions split into two parts:</p>
<blockquote>
<div><ul class="simple">
<li>What is the minimum abundant transcript that should be detectable with the number of reads
mapped? See for example <cite>PMID: 20565853 &lt;http://www.ncbi.nlm.nih.gov/pubmed/20565853&gt;</cite></li>
<li>What is the minimum expression change between two conditions that can be reliably inferred?
See for examples <cite>PMID: 21498551 &lt;http://www.ncbi.nlm.nih.gov/pubmed/21498551?dopt=Abstract&gt;</cite></li>
</ul>
</div></blockquote>
<p>These questions are difficult to answer due to the complexity of RNASeq data: Genes have multiple
transcripts, transcript/gene expression varies by orders of magnitude and a large fraction of
reads might stem from repetetive RNA.</p>
<p>See <a class="reference external" href="http://www.nature.com/nbt/journal/v28/n5/full/nbt.1621.html">figure 4</a> from the cufflinks paper
to get an idea about the reliability of transcript construction with varying sequencing depth.</p>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="../UsingPipelines.html#pipelinesettingup"><em>Setting up a pipeline</em></a> and <a class="reference internal" href="../UsingPipelines.html#pipelinerunning"><em>Running a pipeline</em></a> on general information how to use CGAT pipelines.</p>
<div class="section" id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires a configured <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt> file.</p>
<p>The sphinxreport report requires a <tt class="file docutils literal"><span class="pre">conf.py</span></tt> and <tt class="file docutils literal"><span class="pre">sphinxreport.ini</span></tt> file 
(see <a class="reference internal" href="../UsingPipelines.html#pipelinereporting"><em>Building pipeline reports</em></a>). To start with, use the files supplied with the
<a class="reference internal" href="#example">Example</a> data.</p>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="section" id="reads">
<h4>Reads<a class="headerlink" href="#reads" title="Permalink to this headline">¶</a></h4>
<p>Reads are imported by placing files are linking to files in the <a class="reference internal" href="../make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
<p>The default file format assumes the following convention:</p>
<blockquote>
<div>&lt;sample&gt;-&lt;condition&gt;-&lt;replicate&gt;.&lt;suffix&gt;</div></blockquote>
<p><tt class="docutils literal"><span class="pre">sample</span></tt> and <tt class="docutils literal"><span class="pre">condition</span></tt> make up an <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>, while <tt class="docutils literal"><span class="pre">replicate</span></tt> denotes
the <a class="reference internal" href="../glossary.html#term-replicate"><em class="xref std std-term">replicate</em></a> within an <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a>. The <tt class="docutils literal"><span class="pre">suffix</span></tt> determines the file type.
The following suffixes/file types are possible:</p>
<dl class="docutils">
<dt>sra</dt>
<dd>Short-Read Archive format. Reads will be extracted using the <tt class="file docutils literal"><span class="pre">fastq-dump</span></tt> tool.</dd>
<dt>fastq.gz</dt>
<dd>Single-end reads in fastq format.</dd>
<dt>fastq.1.gz, fastq2.2.gz</dt>
<dd>Paired-end reads in fastq format. The two fastq files must be sorted by read-pair.</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Quality scores need to be of the same scale for all input files. Thus it might be
difficult to mix different formats.</p>
</div>
</div>
<div class="section" id="optional-inputs">
<h4>Optional inputs<a class="headerlink" href="#optional-inputs" title="Permalink to this headline">¶</a></h4>
</div>
</div>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires the results from <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a>. Set the configuration variable 
<tt class="xref py py-data docutils literal"><span class="pre">annotations_database</span></tt> and <tt class="xref py py-data docutils literal"><span class="pre">annotations_dir</span></tt>.</p>
<p>On top of the default CGAT setup, the pipeline requires the following software to be in the 
path:</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="22%" />
<col width="55%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Program</em></td>
<td><em>Version</em></td>
<td><em>Purpose</em></td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">bowtie</a></td>
<td>&gt;=0.12.7</td>
<td>read mapping</td>
</tr>
<tr class="row-odd"><td><a class="reference external" href="http://tophat.cbcb.umd.edu/">tophat</a></td>
<td>&gt;=1.4.0</td>
<td>read mapping</td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a></td>
<td>&gt;=1.3.0</td>
<td>transcription levels</td>
</tr>
<tr class="row-odd"><td>samtools</td>
<td>&gt;=0.1.16</td>
<td>bam/sam files</td>
</tr>
<tr class="row-even"><td>bedtools</td>
<td>&nbsp;</td>
<td>working with intervals</td>
</tr>
<tr class="row-odd"><td>R/DESeq</td>
<td>&nbsp;</td>
<td>differential expression</td>
</tr>
<tr class="row-even"><td>sra-tools</td>
<td>&nbsp;</td>
<td>extracting reads from .sra files</td>
</tr>
<tr class="row-odd"><td>picard</td>
<td>&gt;=1.42</td>
<td><dl class="first last docutils">
<dt>bam/sam files. The .jar files need to be in your</dt>
<dd>CLASSPATH environment variable.</dd>
</dl>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="pipeline-output">
<h2>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this headline">¶</a></h2>
<p>The major output is in the database file <tt class="file docutils literal"><span class="pre">csvdb</span></tt>.</p>
<p>For each <a class="reference internal" href="../glossary.html#term-experiment"><em class="xref std std-term">experiment</em></a> there will be the following tables:</p>
<dl class="docutils">
<dt>&lt;track&gt;_cuffcompare_benchmark </dt>
<dd>results from comparing gene models against reference gene set
primary key: track</dd>
<dt>&lt;track&gt;_cuffcompare_transcripts</dt>
<dd>transcript expression values (FPKMs)
primary key: track+transfrag_id
foreign key: transfrag_id</dd>
<dt>&lt;track&gt;_cuffcompare_tracking</dt>
<dd>tracking information linking loci against transcripts.
primary key: transfrag_id, 
foreign key: locus_id</dd>
<dt>&lt;track&gt;_cuffcompare_tracking</dt>
<dd><dl class="first last docutils">
<dt>locus information (number of transcripts within locus per track)</dt>
<dd>primary key: locus_id</dd>
</dl>
</dd>
</dl>
<div class="section" id="differential-gene-expression-results">
<h3>Differential gene expression results<a class="headerlink" href="#differential-gene-expression-results" title="Permalink to this headline">¶</a></h3>
<p>Differential expression is estimated for different genesets
with a variety of methods. Differential expression can be defined
for various levels.</p>
<dl class="docutils">
<dt>&lt;geneset&gt;_&lt;method&gt;_&lt;level&gt;_diff</dt>
<dd>Results of the pairwise tests for differential expression
primary keys: track1, track2</dd>
<dt>&lt;geneset&gt;_&lt;method&gt;_&lt;level&gt;_levels</dt>
<dd>Expression levels</dd>
</dl>
</div>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>Example data is available at <a class="reference external" href="http://www.cgat.org/~andreas/sample_data/pipeline_rnaseq.tgz">http://www.cgat.org/~andreas/sample_data/pipeline_rnaseq.tgz</a>.
To run the example, simply unpack and untar:</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://www.cgat.org/~andreas/sample_data/pipeline_rnaseq.tgz
tar -xvzf pipeline_rnaseq.tgz
cd pipeline_rnaseq.dir
python &lt;srcdir&gt;/pipeline_rnaseq.py make full
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For the pipeline to run, install the <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a> as well.</p>
</div>
</div>
<div class="section" id="glossary">
<h2>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h2>
<dl class="glossary docutils">
<dt id="term-cufflinks">cufflinks</dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a> - transcriptome analysis</dd>
<dt id="term-tophat">tophat</dt>
<dd><a class="reference external" href="http://tophat.cbcb.umd.edu/">tophat</a> - a read mapper to detect splice-junctions</dd>
<dt id="term-deseq">deseq</dt>
<dd><a class="reference external" href="http://www-huber.embl.de/users/anders/DESeq/">deseq</a> - differential expression analysis</dd>
<dt id="term-cuffdiff">cuffdiff</dt>
<dd>find differentially expressed transcripts. Part of <a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a>.</dd>
<dt id="term-cuffcompare">cuffcompare</dt>
<dd>compare transcriptomes. Part of <a class="reference external" href="http://cufflinks.cbcb.umd.edu/index.html">cufflinks</a>.</dd>
</dl>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pipeline_rnaseq.connect">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">connect</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.connect" title="Permalink to this definition">¶</a></dt>
<dd><p>connect to database.</p>
<p>This method also attaches to helper databases.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.writePrunedGTF">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">writePrunedGTF</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.writePrunedGTF" title="Permalink to this definition">¶</a></dt>
<dd><p>remove various gene models from a gtf file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.mergeAndFilterGTF">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">mergeAndFilterGTF</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>logfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.mergeAndFilterGTF" title="Permalink to this definition">¶</a></dt>
<dd><p>sanitize transcripts file for cufflinks analysis.</p>
<p>Merge exons separated by small introns (&lt; 5bp).</p>
<dl class="docutils">
<dt>Transcript will be ignored that</dt>
<dd><ul class="first last simple">
<li>have very long introns (max_intron_size) (otherwise, cufflinks complains)</li>
<li>are located on contigs to be ignored (usually: chrM, _random, ...)</li>
</ul>
</dd>
</dl>
<p>This method preserves all features in a gtf file (exon, CDS, ...).</p>
<p>returns a dictionary of all gene_ids that have been kept.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildReferenceGeneSet">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildReferenceGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildReferenceGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>sanitize ENSEMBL transcripts file for cufflinks analysis.</p>
<p>Merge exons separated by small introns (&lt; 5bp).</p>
<p>Removes unwanted contigs according to configuration
value <tt class="docutils literal"><span class="pre">geneset_remove_contigs</span></tt>.</p>
<p>Removes transcripts overlapping ribosomal genes if 
<tt class="docutils literal"><span class="pre">geneset_remove_repetitive_rna</span></tt> is set. Protein
coding transcripts are not removed.</p>
<dl class="docutils">
<dt>Transcript will be ignored that</dt>
<dd><ul class="first last simple">
<li>have very long introns (max_intron_size) (otherwise, cufflinks complains)</li>
<li>are located on contigs to be ignored (usually: chrM, _random, ...)</li>
</ul>
</dd>
</dl>
<p>The result is run through cuffdiff in order to add the p_id and tss_id tags
required by cuffdiff.</p>
<p>This will only keep sources of the type &#8216;exon&#8217;. It will also remove
any transcripts not in the reference genome.</p>
<p>Cuffdiff requires overlapping genes to have different tss_id tags.</p>
<p>This gene is the source for most other gene sets in the pipeline.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildNoncodingGeneSet">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildNoncodingGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildNoncodingGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a new gene set without protein coding 
transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildReferenceGeneSetWithCDS">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildReferenceGeneSetWithCDS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildReferenceGeneSetWithCDS" title="Permalink to this definition">¶</a></dt>
<dd><p>build a new gene set without protein coding 
transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildCodingGeneSet">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildCodingGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildCodingGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a gene set with only protein coding 
transcripts.</p>
<p>Genes are selected via their gene biotype in the GTF file.
Note that this set will contain all transcripts of protein
coding genes, including processed transcripts.</p>
<p>This set includes UTR and CDS.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildCodingTranscriptSet">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildCodingTranscriptSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildCodingTranscriptSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a gene set with only protein coding transcripts.</p>
<p>Protein coding transcripts are selected via the ensembl
transcript biotype</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildGeneRegions">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildGeneRegions</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildGeneRegions" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate genomic regions with reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildTerminalExons">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildTerminalExons</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildTerminalExons" title="Permalink to this definition">¶</a></dt>
<dd><p>output terminal truncated exons.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildIntronGeneModels">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildIntronGeneModels</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildIntronGeneModels" title="Permalink to this definition">¶</a></dt>
<dd><p>build protein-coding intron-transcipts.</p>
<p>Intron-transcripts are the reverse complement of transcripts.</p>
<p>Only protein coding genes are taken.</p>
<p>10 bp are truncated on either end of an intron and need
to have a minimum length of 100.</p>
<p>Introns from nested genes might overlap, but all exons
are removed.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildJunctions">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildJunctions</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildJunctions" title="Permalink to this definition">¶</a></dt>
<dd><p>build file with splice junctions from gtf file.</p>
<p>A junctions file is a better option than supplying a GTF
file, as parsing the latter often fails. See:</p>
<p><a class="reference external" href="http://seqanswers.com/forums/showthread.php?t=7563">http://seqanswers.com/forums/showthread.php?t=7563</a></p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildReferenceTranscriptome">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildReferenceTranscriptome</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildReferenceTranscriptome" title="Permalink to this definition">¶</a></dt>
<dd><p>build reference transcriptome.</p>
<p>The reference transcriptome contains all known 
protein coding transcripts.</p>
<p>The sequences include both UTR and CDS.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildMaskGtf">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildMaskGtf</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildMaskGtf" title="Permalink to this definition">¶</a></dt>
<dd><p>This takes ensembl annotations (geneset_all.gtf.gz) and writes out all entries that 
have a &#8216;source&#8217; match to &#8220;rRNA&#8221; or &#8216;contig&#8217; match to &#8220;chrM&#8221;. for use with cufflinks</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.mapReadsWithBowtieAgainstTranscriptome">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">mapReadsWithBowtieAgainstTranscriptome</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.mapReadsWithBowtieAgainstTranscriptome" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads from short read archive sequence using bowtie against
transcriptome data.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.mapReadsWithTophat">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">mapReadsWithTophat</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.mapReadsWithTophat" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads from .fastq or .sra files.</p>
<p>A list with known splice junctions is supplied.</p>
<p>If tophat fails with an error such as:</p>
<div class="highlight-python"><div class="highlight"><pre>Error: segment-based junction search failed with err =-6
what():  std::bad_alloc
</pre></div>
</div>
<p>it means that it ran out of memory.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildJunctionsDB">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildJunctionsDB</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildJunctionsDB" title="Permalink to this definition">¶</a></dt>
<dd><p>build a database of all junctions.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildFastQCReport">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildFastQCReport</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildFastQCReport" title="Permalink to this definition">¶</a></dt>
<dd><p>run fastqc on aligned reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildBAMs">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildBAMs</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildBAMs" title="Permalink to this definition">¶</a></dt>
<dd><p>reconcile genomic and transcriptome matches.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildMismappedBAMs">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildMismappedBAMs</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildMismappedBAMs" title="Permalink to this definition">¶</a></dt>
<dd><p>pseudo target - update the mismapped bam files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildPicardInsertSize">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildPicardInsertSize</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildPicardInsertSize" title="Permalink to this definition">¶</a></dt>
<dd><p>build alignment stats using picard.</p>
<p>Note that picards counts reads but they are in fact alignments.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildPicardStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildPicardStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildPicardStats" title="Permalink to this definition">¶</a></dt>
<dd><p>build alignment stats using picard.</p>
<p>Note that picards counts reads but they are in fact alignments.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildBAMReports">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildBAMReports</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildBAMReports" title="Permalink to this definition">¶</a></dt>
<dd><p>build alignment stats using bamstats</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadPicardStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadPicardStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadPicardStats" title="Permalink to this definition">¶</a></dt>
<dd><p>merge alignment stats into single tables.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildBAMStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildBAMStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildTranscriptBAMStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildTranscriptBAMStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildTranscriptBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadBAMStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadBAMStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>import bam statisticis.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildContextStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildContextStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildContextStats" title="Permalink to this definition">¶</a></dt>
<dd><p>build mapping context stats.</p>
<p>Examines the genomic context to where reads align.</p>
<p>A read is assigned to the genomic context that it
overlaps by at least 50%. Thus some reads mapping
several contexts might be dropped.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadContextStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadContextStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadContextStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load context mapping statistics.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildGeneModels">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildGeneModels</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildGeneModels" title="Permalink to this definition">¶</a></dt>
<dd><p>build transcript models for each track separately.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.oldClassifyTranscripts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">oldClassifyTranscripts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.oldClassifyTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>classify transcripts against a reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.estimateExpressionLevelsInReference">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">estimateExpressionLevelsInReference</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.estimateExpressionLevelsInReference" title="Permalink to this definition">¶</a></dt>
<dd><p>estimate expression levels against a set of reference gene models.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadExpressionLevels">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadExpressionLevels</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadExpressionLevels" title="Permalink to this definition">¶</a></dt>
<dd><p>load expression level measurements.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.runCuffCompare">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">runCuffCompare</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em>, <em>reffile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.runCuffCompare" title="Permalink to this definition">¶</a></dt>
<dd><p>run cuffcompare.</p>
<p>Will create a .tmap and .refmap input file for each input file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.compareTranscriptsPerExperiment">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">compareTranscriptsPerExperiment</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.compareTranscriptsPerExperiment" title="Permalink to this definition">¶</a></dt>
<dd><p>compare transcript models between replicates within each experiment.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.compareTranscriptsBetweenExperiments">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">compareTranscriptsBetweenExperiments</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.compareTranscriptsBetweenExperiments" title="Permalink to this definition">¶</a></dt>
<dd><p>compare transcript models between replicates in all experiments.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadTranscriptComparison">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadTranscriptComparison</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadTranscriptComparison" title="Permalink to this definition">¶</a></dt>
<dd><p>load data from transcript comparison.</p>
<p>This task creates two tables:</p>
<p>&lt;track&gt;_benchmark
&lt;track&gt;_loci</p>
<p>The following tables are only present if there are
multiple replicates in a sample:</p>
<p>&lt;track&gt;_tracking</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildAbinitioGeneSet">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildAbinitioGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildAbinitioGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>builds ab-initio gene set.</p>
<p>The ab-initio gene set is derived from the cuffcompare result.</p>
<p>The following transfrags are removed at this stage:</p>
<blockquote>
<div><ul class="simple">
<li>transfrags overlapping RNA genes</li>
<li>transfrags on certain contigs (usually: mitochondrial genes)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildFullGeneSet">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildFullGeneSet</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildFullGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>builds a gene set by merging the ab-initio gene set and
the reference gene set.</p>
<p>The gene set is cleaned in order to permit differential expression
analysis.</p>
<p>Only transfrags are kept that are:</p>
<ol class="arabic">
<li><dl class="first docutils">
<dt>observed in at least 2 samples to remove partial transfrags that</dt>
<dd><p class="first last">are the result of low coverage observations in one sample</p>
</dd>
</dl>
</li>
</ol>
<p>see also: <a class="reference external" href="http://seqanswers.com/forums/showthread.php?t=3967">http://seqanswers.com/forums/showthread.php?t=3967</a></p>
<p>Will also build removed.gtf.gz of removed transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildNovelGeneSet">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildNovelGeneSet</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildNovelGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a gene set of novel genes by merging the ab-initio gene set and
the reference gene set.</p>
<p>Ab-initio transcripts are removed based on features in the reference gene set.
Removal is aggressive  - as soon as one transcript of a 
gene/locus overlaps, all transcripts of that gene/locus are gone.</p>
<p>Transcripts that lie exclusively in repetetive sequence are removed, too.</p>
<p>The resultant set contains a number of novel transcripts. However, these
transcripts will still overlap some known genomic features like pseudogenes.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildLincRNAGeneSet">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildLincRNAGeneSet</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildLincRNAGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build lincRNA gene set.</p>
<p>The lincRNA gene set contains all known lincRNA transcripts from
the reference gene set plus all transcripts in the novel set that
do not overlap at any protein coding, processed or pseudogene transcripts 
(exons+introns) in the reference gene set.</p>
<p>Transcripts that lie exclusively in repetetive sequence are removed, too.</p>
<p>lincRNA genes are often expressed at low level and thus the resultant transcript
models are fragmentory. To avoid some double counting in downstream analyses, 
transcripts overlapping on the same strand are merged.</p>
<p>Transcripts need to have a length of at least 200 bp.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.annotateLincRNA">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">annotateLincRNA</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.annotateLincRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>align lincRNA against reference transcriptome
in order to spot pseudogenes.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadGeneSetsBuildInformation">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadGeneSetsBuildInformation</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadGeneSetsBuildInformation" title="Permalink to this definition">¶</a></dt>
<dd><p>load results from gene set filtering into database.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.classifyTranscripts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">classifyTranscripts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.classifyTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>classify transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.classifyTranscriptsCuffcompare">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">classifyTranscriptsCuffcompare</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.classifyTranscriptsCuffcompare" title="Permalink to this definition">¶</a></dt>
<dd><p>classify transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildGeneSetStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildGeneSetStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildGeneSetStats" title="Permalink to this definition">¶</a></dt>
<dd><p>compile gene set statistics.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadGeneSetStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadGeneSetStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadGeneSetStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load geneset statisticts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.annotateTranscriptsMappability">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">annotateTranscriptsMappability</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.annotateTranscriptsMappability" title="Permalink to this definition">¶</a></dt>
<dd><p>classify transcripts with respect to the gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadTranscriptsMappability">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadTranscriptsMappability</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadTranscriptsMappability" title="Permalink to this definition">¶</a></dt>
<dd><p>load interval annotations: genome architecture</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.annotateTranscripts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">annotateTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.annotateTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>classify transcripts with respect to the gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadAnnotations">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadAnnotations</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>load interval annotations: genome architecture</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.hasReplicates">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">hasReplicates</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.hasReplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>indicator function - return true if track has replicates</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildReproducibility">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildReproducibility</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildReproducibility" title="Permalink to this definition">¶</a></dt>
<dd><p>all-vs-all comparison between samples.</p>
<p>Compute correlation between expressed transfrags. Transfrags missing
from another set are ignored.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadReproducibility">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadReproducibility</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadReproducibility" title="Permalink to this definition">¶</a></dt>
<dd><p>load reproducibility results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.runCuffdiff">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">runCuffdiff</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.runCuffdiff" title="Permalink to this definition">¶</a></dt>
<dd><p>estimate differential expression using cuffdiff.</p>
<p>Replicates are grouped.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadCuffdiff">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadCuffdiff</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadCuffdiff" title="Permalink to this definition">¶</a></dt>
<dd><p>load results from differential expression analysis and produce
summary plots.</p>
<p>Note: converts from ln(fold change) to log2 fold change.</p>
<p>The cuffdiff output is parsed.</p>
<p>Pairwise comparisons in which one gene is not expressed (fpkm &lt; fpkm_silent)
are set to status &#8216;NOCALL&#8217;. These transcripts might nevertheless be significant.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildExpressionStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildExpressionStats</tt><span class="sig-paren">(</span><em>tables</em>, <em>method</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildExpressionStats" title="Permalink to this definition">¶</a></dt>
<dd><p>build expression summary statistics.</p>
<p>Creates some diagnostic plots in</p>
<p>&lt;exportdir&gt;/&lt;method&gt; directory.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildCuffdiffPlots">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildCuffdiffPlots</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildCuffdiffPlots" title="Permalink to this definition">¶</a></dt>
<dd><p>create summaries of cufflinks results (including some diagnostic plots)</p>
<p>Plots are created in the &lt;exportdir&gt;/cuffdiff directory.</p>
<p>Plots are:</p>
<dl class="docutils">
<dt>&lt;geneset&gt;_&lt;method&gt;_&lt;level&gt;_&lt;track1&gt;_vs_&lt;track2&gt;_significance.png</dt>
<dd>fold change against expression level</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadCuffdiffStats">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadCuffdiffStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadCuffdiffStats" title="Permalink to this definition">¶</a></dt>
<dd><p>import cuffdiff results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildFPKMGeneLevelTagCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildFPKMGeneLevelTagCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildFPKMGeneLevelTagCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>build tag counts using normalized counts from tophat.</p>
<p>These are gene-length normalized count levels.</p>
<p>They are obtained by multiplying the FPKM value
by the median library size.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildCodingExons">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildCodingExons</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildCodingExons" title="Permalink to this definition">¶</a></dt>
<dd><p>compile set of protein coding exons.</p>
<p>This set is used for splice-site validation</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildExonValidation">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildExonValidation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildExonValidation" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadExonValidation">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadExonValidation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadExonValidation" title="Permalink to this definition">¶</a></dt>
<dd><p>merge alignment stats into single tables.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildUnionIntersectionExons">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildUnionIntersectionExons</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildUnionIntersectionExons" title="Permalink to this definition">¶</a></dt>
<dd><p>build union/intersection genes according to Bullard et al. (2010) BMC Bioinformatics.</p>
<p>Builds a single-segment bed file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildUnionExons">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildUnionExons</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildUnionExons" title="Permalink to this definition">¶</a></dt>
<dd><p>build union genes.</p>
<p>Exons across all transcripts of a gene are merged.
They are then intersected between genes to remove any overlap.</p>
<p>Builds a single-segment bed file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildExonLevelReadCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildExonLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildExonLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>compute coverage of exons with reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadExonLevelReadCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadExonLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadExonLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>load exon level read counts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildGeneLevelReadCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildGeneLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildGeneLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>compute coverage of exons with reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildAggregateGeneLevelReadCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildAggregateGeneLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildAggregateGeneLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>count reads falling into transcripts of protein coding 
gene models.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In paired-end data sets each mate will be counted. Thus
the actual read counts are approximately twice the fragment
counts.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildIntronLevelReadCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildIntronLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildIntronLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>compute coverage of exons with reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildGeneLevelReadExtension">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildGeneLevelReadExtension</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildGeneLevelReadExtension" title="Permalink to this definition">¶</a></dt>
<dd><p>compute extension of cds.</p>
<p>Known UTRs are counted as well.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.plotGeneLevelReadExtension">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">plotGeneLevelReadExtension</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.plotGeneLevelReadExtension" title="Permalink to this definition">¶</a></dt>
<dd><p>plot reads extending beyond last exon.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildUTRExtension">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildUTRExtension</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildUTRExtension" title="Permalink to this definition">¶</a></dt>
<dd><p>build new utrs.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildUTRs">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildUTRs</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildUTRs" title="Permalink to this definition">¶</a></dt>
<dd><p>build new utrs by merging estimated UTR extensions
from all data sets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildTranscriptLevelReadCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildTranscriptLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildTranscriptLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>count reads falling into transcripts of protein coding </dt>
<dd>gene models.</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In paired-end data sets each mate will be counted. Thus
the actual read counts are approximately twice the fragment
counts.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildAggregateTranscriptLevelReadCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildAggregateTranscriptLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildAggregateTranscriptLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>count reads falling into transcripts of protein coding </dt>
<dd>gene models.</dd>
</dl>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In paired-end data sets each mate will be counted. Thus
the actual read counts are approximately twice the fragment
counts.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This step takes very long if multiple bam-files are supplied.
It has thus been taken out of the pipeline. The aggregate can be derived from summing
the individual counts anyways.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.aggregateExonLevelReadCounts">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">aggregateExonLevelReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.aggregateExonLevelReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>aggregate exon level tag counts for each gene.</p>
<p>coverageBed adds the following four columns:</p>
<ol class="arabic simple">
<li>The number of features in A that overlapped (by at least one base pair) the B interval.</li>
<li>The number of bases in B that had non-zero coverage from features in A.</li>
<li>The length of the entry in B.</li>
<li>The fraction of bases in B that had non-zero coverage from features in A.</li>
</ol>
<p>For bed6: use column 7
For bed12: use column 13</p>
<p>This method uses the maximum number of reads
found in any exon as the tag count.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.runDESeq">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">runDESeq</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.runDESeq" title="Permalink to this definition">¶</a></dt>
<dd><p>estimate differential expression using DESeq.</p>
<p>The final output is a table. It is slightly edited such that
it contains a similar output and similar fdr compared 
cuffdiff.</p>
<p>Plots are:</p>
<dl class="docutils">
<dt>&lt;geneset&gt;_&lt;method&gt;_&lt;level&gt;_&lt;track1&gt;_vs_&lt;track2&gt;_significance.png</dt>
<dd>fold change against expression level</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.loadDESeq">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">loadDESeq</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.loadDESeq" title="Permalink to this definition">¶</a></dt>
<dd><p>load differential expression results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.buildGeneSetsOfInterest">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">buildGeneSetsOfInterest</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.buildGeneSetsOfInterest" title="Permalink to this definition">¶</a></dt>
<dd><p>export gene sets of interest</p>
<ul class="simple">
<li>differentially expressed genes</li>
</ul>
<p>Regions of interest are exported as <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.createViewMapping">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">createViewMapping</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.createViewMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>create view in database for alignment stats.</p>
<p>This view aggregates all information on a per-track basis.</p>
<p>The table is built from the following tracks:</p>
<p>tophat_stats: .genome
mapping_stats: .accepted
bam_stats: .accepted
context_stats: .accepted
picard_stats: .accepted</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.build_report">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">build_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>build report from scratch.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.update_report">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">update_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.update_report" title="Permalink to this definition">¶</a></dt>
<dd><p>update report.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_rnaseq.publish">
<tt class="descclassname">pipeline_rnaseq.</tt><tt class="descname">publish</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_rnaseq.publish" title="Permalink to this definition">¶</a></dt>
<dd><p>publish files.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">RNA-Seq pipeline</a><ul>
<li><a class="reference internal" href="#overview">Overview</a><ul>
<li><a class="reference internal" href="#mapping-strategy">Mapping strategy</a></li>
<li><a class="reference internal" href="#lincrna">LincRNA</a><ul>
<li><a class="reference internal" href="#artifacts-in-lncrna-analysis">Artifacts in lncRNA analysis</a><ul>
<li><a class="reference internal" href="#read-mapping-errors">Read mapping errors</a></li>
<li><a class="reference internal" href="#fragments">Fragments</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#differential-expression">Differential expression</a></li>
<li><a class="reference internal" href="#estimating-coverage">Estimating coverage</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
<li><a class="reference internal" href="#input">Input</a><ul>
<li><a class="reference internal" href="#reads">Reads</a></li>
<li><a class="reference internal" href="#optional-inputs">Optional inputs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pipeline-output">Pipeline output</a><ul>
<li><a class="reference internal" href="#differential-gene-expression-results">Differential gene expression results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#glossary">Glossary</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pipeline_chipseq.html"
                        title="previous chapter">ChIP-Seq pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../MakePipelines.html"
                        title="next chapter">Lecgacy pipelines</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelines/pipeline_rnaseq.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../MakePipelines.html" title="Lecgacy pipelines"
             >next</a> |</li>
        <li class="right" >
          <a href="pipeline_chipseq.html" title="ChIP-Seq pipeline"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" >NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>