
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Annotation pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="NGS Pipelines" href="../CGATPipelines.html" />
    <link rel="next" title="Ancestral repeats pipeline" href="pipeline_ancestral_repeats.html" />
    <link rel="prev" title="NGS Pipelines" href="../CGATPipelines.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_ancestral_repeats.html" title="Ancestral repeats pipeline"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../CGATPipelines.html" title="NGS Pipelines"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" accesskey="U">NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pipeline_annotations"></span><div class="section" id="annotation-pipeline">
<h1>Annotation pipeline<a class="headerlink" href="#annotation-pipeline" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>The annotation pipeline imports various annotations and organizes them
for use in other pipelines.</p>
<blockquote>
<div><ul class="simple">
<li>a geneset (from ENSEMBL gene sets)</li>
<li>repeats (from UCSC repeatmasker tracks)</li>
</ul>
</div></blockquote>
<p>This pipeline works on a single genome. Annotations are often shared
between versions within the same project or even between projects, hence
this separate pipeline. The output of this pipeline is used by various
other pipelines, for example the the <a class="reference internal" href="pipeline_rnaseq.html"><em>RNA-Seq pipeline</em></a> and the 
<a class="reference internal" href="pipeline_chipseq.html"><em>ChIP-Seq pipeline</em></a>.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The pipeline takes as input an ENSEMBL gene set and builds various gene sets
of interest.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="../UsingPipelines.html#pipelinesettingup"><em>Setting up a pipeline</em></a> and <a class="reference internal" href="../UsingPipelines.html#pipelinerunning"><em>Running a pipeline</em></a> on general information how to use CGAT pipelines.</p>
<div class="section" id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt> needs to be edited so that it points to the
appropriate locations of the auxiliary files. See especially:</p>
<dl class="docutils">
<dt>1 section <tt class="docutils literal"><span class="pre">[ensembl]</span></tt> with the location of the ENSEMBL dump</dt>
<dd>files (<tt class="docutils literal"><span class="pre">filename_gtf</span></tt>, filename_pep``, <tt class="docutils literal"><span class="pre">filename_cdna</span></tt>)</dd>
<dt>2 section <tt class="docutils literal"><span class="pre">[general]</span></tt> with the location of the indexed genomic</dt>
<dd>fasta files to use and the name of the genome (default=``hg19``),
see <a class="reference internal" href="../modules/IndexedFasta.html"><em>IndexedFasta.py - fast random access in fasta files</em></a>.</dd>
</dl>
<p>3 section <tt class="docutils literal"><span class="pre">[ucsc]</span></tt> with the name of the database to use (default=``hg19``).</p>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>This script requires no input within the <a class="reference internal" href="../make_pipelines/Gpipe.html#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
</div>
</div>
<div class="section" id="pipeline-output">
<h2>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this headline">¶</a></h2>
<p>The results of the computation are all stored in an sqlite relational
database file. Additional files are output in the working directory
for use in other pipelines. These are:</p>
<dl class="docutils">
<dt>csvdb</dt>
<dd>An sqlite database with most of the information created by this pipeline</dd>
<dt>annotation_gff.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file annotating the genome with respect to the geneset.
Annotations are non-overlapping and are based only on protein coding transcripts.</dd>
<dt>geneset_all.gtf.gz</dt>
<dd>The full gene set after reconciling with assembly. Chromosomes names are
renamed to be consistent with the assembly.</dd>
<dt>geneset_cds.gtf.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file with only the CDS parts of transcripts.
This set will naturally include only coding transcripts.</dd>
<dt>geneset_exons.gtf.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file with only the exon parts of transcripts.
This set includes both coding and non-coding transcripts. Coding 
transcripts span both the UTR and the CDS.</dd>
<dt>geneset_flat.gtf.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted flattened gene models. All overlapping transcripts
have been merged. This set includes both coding and non-coding transcripts.</dd>
<dt>pseudogenes.gtf.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file with pseudogenes. Pseudogenes are either taken from
the ENSEMBL annotation or processed transcripts with similarity to protein coding
sequence. As some protein coding genes contain processed transcripts without an ORF,
Pseudogenes might overlap with protein coding transcripts.</dd>
<dt>peptides.fasta</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> formatted file of peptide sequences of coding transcripts.</dd>
<dt>cds.fasta</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> formatted file of coding sequence of coding transcripts.</dd>
<dt>cdna.fasta</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> formatted file of transcripts including both coding and non-coding parts.</dd>
<dt>contigs.tsv</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-tsv"><em class="xref std std-term">tsv</em></a> formatted table with contig sizes</dd>
<dt>tss.bed.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file with transcription start sites.</dd>
<dt>tts.bed.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file with transcription termination sites.</dd>
<dt>promotors.bed.gs</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file with promotor regions (fixed witdth segments upstream of 
transcription start sites).</dd>
<dt>repeats.bed.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file of repetitive sequences (obtained from UCSC repeatmasker tracks).</dd>
<dt>rna.gff.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file of repetitive RNA sequences in the genome
(obtained from UCSC repeatmasker tracks).</dd>
<dt>go.tsv.gz</dt>
<dd>A list of <a class="reference internal" href="../glossary.html#term-go"><em class="xref std std-term">GO</em></a> assignments for each gene.</dd>
<dt>goslim.tsv.gz</dt>
<dd>A list of <a class="reference internal" href="../glossary.html#term-goslim"><em class="xref std std-term">GOSlim</em></a> assignments for each gene.</dd>
<dt>territories.gff.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file of non-overlapping gene territories.</dd>
<dt>tssterritories.gff.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file of non-overlapping tss territories.</dd>
<dt>greatdomains.gff.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file of great regulatory domains (basal+extension rule).</dd>
<dt>gc_segmentation.bed.gz</dt>
<dd>A <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file with the genome segmented in regions
of different G+C content.</dd>
<dt>cpg.bed.gz</dt>
<dd>A list of all CpGs in the genome sequence</dd>
<dt>gwas_catalog.bed.gz</dt>
<dd>Bed intervals with GWAS intervals from the gwas catalog.</dd>
<dt>gwas_distild.bed.gz</dt>
<dd>Bed intervals with GWAS intervals from the DistiLD database.</dd>
</dl>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>Example data is available at <a class="reference external" href="http://www.cgat.org/~andreas/sample_data/pipeline_annotations.tgz">http://www.cgat.org/~andreas/sample_data/pipeline_annotations.tgz</a>.
To run the example, simply unpack and untar:</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://www.cgat.org/~andreas/sample_data/pipeline_annotations.tgz
tar -xvzf pipeline_annotations.tgz
cd pipeline_annotations.dir
python &lt;srcdir&gt;/pipeline_annotations.py make full
</pre></div>
</div>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pipeline_annotations.connect">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">connect</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.connect" title="Permalink to this definition">¶</a></dt>
<dd><p>connect to database.</p>
<p>This method also attaches to helper databases.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildContigSizes">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildContigSizes</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildContigSizes" title="Permalink to this definition">¶</a></dt>
<dd><p>output contig sizes.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildContigBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildContigBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildContigBed" title="Permalink to this definition">¶</a></dt>
<dd><p>output bed file with contigs</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildUngappedContigBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildUngappedContigBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildUngappedContigBed" title="Permalink to this definition">¶</a></dt>
<dd><p>output bed file with contigs.</p>
<p>Genomic regions with gaps are excluded.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGenomeInformation">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGenomeInformation</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGenomeInformation" title="Permalink to this definition">¶</a></dt>
<dd><p>compute genome composition information.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadGenomeInformation">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadGenomeInformation</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadGenomeInformation" title="Permalink to this definition">¶</a></dt>
<dd><p>load genome information.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.calculateMappability">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">calculateMappability</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.calculateMappability" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate mappability using GEM</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.countMappableBases">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">countMappableBases</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.countMappableBases" title="Permalink to this definition">¶</a></dt>
<dd><p>Count mappable bases in genome</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadMappableBases">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadMappableBases</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadMappableBases" title="Permalink to this definition">¶</a></dt>
<dd><p>load count of mappable bases in genome</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.splitMappabiliyFileByContig">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">splitMappabiliyFileByContig</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.splitMappabiliyFileByContig" title="Permalink to this definition">¶</a></dt>
<dd><p>Count mappable bases in genome</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.countMappableBasesPerContig">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">countMappableBasesPerContig</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.countMappableBasesPerContig" title="Permalink to this definition">¶</a></dt>
<dd><p>Count mappable bases for each contig</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadMappableBasesPerContig">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadMappableBasesPerContig</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadMappableBasesPerContig" title="Permalink to this definition">¶</a></dt>
<dd><p>load count of mappable bases per contig</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGeneSet">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a gene set - firstly, reconciles chromosome names by 
removing those that do not occur in the specified genome assembly; 
secondly, removes chromosome names specified in pipeline.ini</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.annotateGenome">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">annotateGenome</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.annotateGenome" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate genomic regions with reference gene set.</p>
<p>Only considers protein coding genes. 
Processed_transcripts tend to cover larger genomic regions
and often overlap between adjacent protein coding genes.</p>
<p>In case of overlapping genes, only take the longest 
(in genomic coordinates).</p>
<p>Genes not on UCSC contigs are removed.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.annotateGeneStructure">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">annotateGeneStructure</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.annotateGeneStructure" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate genome with gene structures.</p>
<p>Only considers protein coding genes.</p>
<p>Processed_transcripts tend to cover larger genomic regions
and often overlap between adjacent protein coding genes.</p>
<p>In case of overlapping genes, only take the longest 
(in genomic coordinates).</p>
<p>Genes not on UCSC contigs are removed.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildFlatGeneSet">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildFlatGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildFlatGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a flattened gene set.</p>
<p>All transcripts in a gene are merged into a single transcript.</p>
<p><em>infile</em> is an ENSEMBL gtf file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadGeneInformation">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadGeneInformation</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadGeneInformation" title="Permalink to this definition">¶</a></dt>
<dd><p>load the transcript set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadGeneStats">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadGeneStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadGeneStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load the transcript set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildCDSTranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildCDSTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildCDSTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a collection of transcripts from the protein-coding
section of the ENSEMBL gene set.</p>
<p>Only CDS parts of exons are output - UTR&#8217;s are removed.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildExonTranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildExonTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildExonTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a collection of transcripts from the protein-coding
section of the ENSEMBL gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadExonStats">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadExonStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadExonStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load the transcript set stats.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildCodingExonTranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildCodingExonTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildCodingExonTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a collection of transcripts from the protein-coding
section of the ENSEMBL gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildNonCodingExonTranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildNonCodingExonTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildNonCodingExonTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a collection of transcripts from the protein-coding
section of the ENSEMBL gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildLincRNAExonTranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildLincRNAExonTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildLincRNAExonTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a collection of transcripts from the protein-coding
section of the ENSEMBL gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadCDSTranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadCDSTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadCDSTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>load the transcript set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadCDSStats">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadCDSStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadCDSStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load the transcript set stats.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadTranscriptInformation">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadTranscriptInformation</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadTranscriptInformation" title="Permalink to this definition">¶</a></dt>
<dd><p>load transcript information.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadEntrezToEnsembl">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadEntrezToEnsembl</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadEntrezToEnsembl" title="Permalink to this definition">¶</a></dt>
<dd><p>load table to convert from ENSEMBL gene ids to entrez gene ids</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadTranscriptSynonyms">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadTranscriptSynonyms</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadTranscriptSynonyms" title="Permalink to this definition">¶</a></dt>
<dd><p>load table with synonyms for transcript identifiers.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildPeptideFasta">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildPeptideFasta</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildPeptideFasta" title="Permalink to this definition">¶</a></dt>
<dd><p>load ENSEMBL peptide file</p>
<p><em>infile</em> is an ENSEMBL .pep.all.fa.gz file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildCDNAFasta">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildCDNAFasta</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildCDNAFasta" title="Permalink to this definition">¶</a></dt>
<dd><p>load ENSEMBL peptide file</p>
<p><em>infile</em> is an ENSEMBL .cdna.all.fa.gz file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildCDSFasta">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildCDSFasta</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildCDSFasta" title="Permalink to this definition">¶</a></dt>
<dd><p>build cds sequences from peptide and cds file.</p>
<p><em>infile</em> is an ENSEMBL .cdna.all.fa.gz file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadProteinStats">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadProteinStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadProteinStats" title="Permalink to this definition">¶</a></dt>
<dd><p>load the transcript set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildPseudogenes">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildPseudogenes</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildPseudogenes" title="Permalink to this definition">¶</a></dt>
<dd><p>build set of pseudogenes.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildNUMTs">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildNUMTs</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildNUMTs" title="Permalink to this definition">¶</a></dt>
<dd><p>build list of NUMTs.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildSelenoList">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildSelenoList</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildSelenoList" title="Permalink to this definition">¶</a></dt>
<dd><p>export a list of seleno cysteine transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGeneTerritories">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGeneTerritories</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGeneTerritories" title="Permalink to this definition">¶</a></dt>
<dd><p>build gene territories from protein coding genes.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildTSSTerritories">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildTSSTerritories</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildTSSTerritories" title="Permalink to this definition">¶</a></dt>
<dd><p>build gene territories from protein coding genes.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGREATRegulatoryDomains">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGREATRegulatoryDomains</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGREATRegulatoryDomains" title="Permalink to this definition">¶</a></dt>
<dd><p>build gene territories from protein coding genes.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildPromotorRegions">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildPromotorRegions</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildPromotorRegions" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate promotor regions from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildTranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcripts from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildTranscriptsBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildTranscriptsBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildTranscriptsBed" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcripts from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildNoncodingTranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildNoncodingTranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildNoncodingTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcripts from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildNoncodingTranscriptsBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildNoncodingTranscriptsBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildNoncodingTranscriptsBed" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcripts from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildNoncodingGenesBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildNoncodingGenesBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildNoncodingGenesBed" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcripts from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildLincRNATranscripts">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildLincRNATranscripts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildLincRNATranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcripts from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildLincRNATranscriptsBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildLincRNATranscriptsBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildLincRNATranscriptsBed" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcripts from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildTranscriptTSS">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildTranscriptTSS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildTranscriptTSS" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcription start sites from reference gene set.</p>
<p>Similar to promotors, except that the witdth is set to 1.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGeneTSS">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGeneTSS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGeneTSS" title="Permalink to this definition">¶</a></dt>
<dd><p>create a single TSS for each gene</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGeneTSSInterval">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGeneTSSInterval</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGeneTSSInterval" title="Permalink to this definition">¶</a></dt>
<dd><p>create a single interval that encompasses all annotated TSSs for a given gene</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildNoncodingGeneTSS">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildNoncodingGeneTSS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildNoncodingGeneTSS" title="Permalink to this definition">¶</a></dt>
<dd><p>Assign a TSS for each non-coding gene</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.extendRegion">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">extendRegion</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.extendRegion" title="Permalink to this definition">¶</a></dt>
<dd><p>convert bed to gtf</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.convertToGTF">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">convertToGTF</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.convertToGTF" title="Permalink to this definition">¶</a></dt>
<dd><p>convert bed to gtf</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildTranscriptTTS">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildTranscriptTTS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildTranscriptTTS" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcription termination sites from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGeneTTS">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGeneTTS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGeneTTS" title="Permalink to this definition">¶</a></dt>
<dd><p>annotate transcription termination sites from reference gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.importRNAAnnotationFromUCSC">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">importRNAAnnotationFromUCSC</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.importRNAAnnotationFromUCSC" title="Permalink to this definition">¶</a></dt>
<dd><p>import repetetive RNA from a UCSC formatted file.</p>
<p>The repetetive RNA are taken from the repeat-masker track.</p>
<p>The results are stored as a <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.importRepeatsFromUCSC">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">importRepeatsFromUCSC</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.importRepeatsFromUCSC" title="Permalink to this definition">¶</a></dt>
<dd><p>import repeats from a UCSC formatted file.</p>
<p>The repeats are stored as a <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.importCpGIslandsFromUCSC">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">importCpGIslandsFromUCSC</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.importCpGIslandsFromUCSC" title="Permalink to this definition">¶</a></dt>
<dd><p>import cpg islands from UCSC</p>
<p>The repeats are stored as a <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadRepeats">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadRepeats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadRepeats" title="Permalink to this definition">¶</a></dt>
<dd><p>load total repeats length</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.countTotalRepeatLength">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">countTotalRepeatLength</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.countTotalRepeatLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Count total repeat length and add to database</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.importAllRepeatsFromUCSC">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">importAllRepeatsFromUCSC</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.importAllRepeatsFromUCSC" title="Permalink to this definition">¶</a></dt>
<dd><p>import repeats from a UCSC formatted file.</p>
<p>The repeats are stored as a <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildMapableRegions">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildMapableRegions</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildMapableRegions" title="Permalink to this definition">¶</a></dt>
<dd><p>build bed files with mappable regions.</p>
<p>Convert bigwig tracks with mappability information to a
bed-formatted file that contains only mappable regions of the
genome.</p>
<p>A mapable region is more permissive than a mapable position.</p>
<p>This method assumes that files use the <tt class="docutils literal"><span class="pre">CRG</span> <span class="pre">Alignability</span>
<span class="pre">tracks</span></tt>.</p>
<p>UCSC says:</p>
<blockquote>
<div>The CRG Alignability tracks display how uniquely k-mer sequences
align to a region of the genome. To generate the data, the
GEM-mappability program has been employed. The method is
equivalent to mapping sliding windows of k-mers (where k has been
set to 36, 40, 50, 75 or 100 nts to produce these tracks) back to
the genome using the GEM mapper aligner (up to 2 mismatches were
allowed in this case). For each window, a mapability score was
computed (S = 1/(number of matches found in the genome): S=1 means
one match in the genome, S=0.5 is two matches in the genome, and
so on). The CRG Alignability tracks were generated independently
of the ENCODE project, in the framework of the GEM (GEnome
Multitool) project.</div></blockquote>
<p>For the purpose of these tracks, a region is defined to be un-mapable
if its maximum mapability score is less than 0.5.</p>
<p>Unmapable regions that are less than half kmer size are mapable, as
reads from the left/right mapable positions will extend into the region.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.filterMapableRegions">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">filterMapableRegions</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.filterMapableRegions" title="Permalink to this definition">¶</a></dt>
<dd><p>remove small windows from a mapability track.</p>
<p>Too many fragmented regions will cause gat to fail as it
fragments the workspace into too many individual segments.</p>
<p>The filtering works by merging all segments that are
within mapability_merge_distance and removing all those
that are larger than mapabpility_min_segment_size</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.createGO">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">createGO</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.createGO" title="Permalink to this definition">¶</a></dt>
<dd><p>get GO assignments from ENSEMBL</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.createGOSlim">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">createGOSlim</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.createGOSlim" title="Permalink to this definition">¶</a></dt>
<dd><p>get GO assignments from ENSEMBL</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGOPaths">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGOPaths</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGOPaths" title="Permalink to this definition">¶</a></dt>
<dd><p>compute a file with paths of each GO term to the ancestral node.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGOTable">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGOTable</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGOTable" title="Permalink to this definition">¶</a></dt>
<dd><p>build a simple table with GO descriptions in obo.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.loadGOTable">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">loadGOTable</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.loadGOTable" title="Permalink to this definition">¶</a></dt>
<dd><p>load GO descriptions into database.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.importKEGGAssignments">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">importKEGGAssignments</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.importKEGGAssignments" title="Permalink to this definition">¶</a></dt>
<dd><p>import the KEGG annotations from the R KEGG.db 
annotations package. Note that since KEGG is no longer
publically availible, this is not up-to-date and maybe removed
from bioconductor in future releases</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.createGOFromGeneOntology">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">createGOFromGeneOntology</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.createGOFromGeneOntology" title="Permalink to this definition">¶</a></dt>
<dd><p>get GO assignments from ENSEMBL</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGenomicFunctionalAnnotation">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGenomicFunctionalAnnotation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGenomicFunctionalAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>output a bed file with genomic regions with functional annotations.</p>
<p>Each bed entry is a gene territory. Bed entries are labeled
by functional annotations associated with a gene.</p>
<p>Ambiguities in territories are resolved by outputting 
annotations for all genes within a territory.</p>
<p>The output file contains annotations for both GO and GOSlim. These
are prefixed by <tt class="docutils literal"><span class="pre">go:</span></tt> and <tt class="docutils literal"><span class="pre">goslim:</span></tt>.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGenomicContext">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGenomicContext</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGenomicContext" title="Permalink to this definition">¶</a></dt>
<dd><p>build a file with genomic context.</p>
<p>The output is a bed formatted file, annotating genomic segments
according to whether they are any of the ENSEMBL annotations.</p>
<p>It also adds the RNA and repeats annotations from the UCSC.</p>
<p>The annotations can be partially or fully overlapping.</p>
<p>Adjacent features (less than 10 bp apart) of the same type are merged.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGenomicContextStats">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGenomicContextStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGenomicContextStats" title="Permalink to this definition">¶</a></dt>
<dd><p>analysis overlap of genomic contexts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGeneIntervals">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGeneIntervals</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGeneIntervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge all transcripts per gene (including utr) to get start and stop 
coordinates for every protein-coding gene and store in a GTF file</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGeneBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGeneBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGeneBed" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert genic GTF to BED</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildUpstreamFlankBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildUpstreamFlankBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildUpstreamFlankBed" title="Permalink to this definition">¶</a></dt>
<dd><p>build interval upstream of gene start for each entry in bed file</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildDownstreamFlankBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildDownstreamFlankBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildDownstreamFlankBed" title="Permalink to this definition">¶</a></dt>
<dd><p>build interval downstream of gene start for each entry in bed file</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildIntergenicBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildIntergenicBed</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildIntergenicBed" title="Permalink to this definition">¶</a></dt>
<dd><p>Genomic regions not associated with any other features</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.convertBedToGtf">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">convertBedToGtf</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.convertBedToGtf" title="Permalink to this definition">¶</a></dt>
<dd><p>convert bed files to GTF</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGenomeGCSegmentation">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGenomeGCSegmentation</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGenomeGCSegmentation" title="Permalink to this definition">¶</a></dt>
<dd><p>segment the genome into windows according to G+C content.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.runGenomeGCProfile">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">runGenomeGCProfile</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.runGenomeGCProfile" title="Permalink to this definition">¶</a></dt>
<dd><p>segment the genome into windows according to G+C content.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGenomeGCProfile">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGenomeGCProfile</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGenomeGCProfile" title="Permalink to this definition">¶</a></dt>
<dd><p>aggregate windows with similar G+C content into bins.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildCpGBed">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildCpGBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildCpGBed" title="Permalink to this definition">¶</a></dt>
<dd><p>bulid bed file with CpG locations.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.downloadGWASCatalog">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">downloadGWASCatalog</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.downloadGWASCatalog" title="Permalink to this definition">¶</a></dt>
<dd><p>download GWAS catalog.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.downloadDistiLD">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">downloadDistiLD</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.downloadDistiLD" title="Permalink to this definition">¶</a></dt>
<dd><p>download GWAS data from distild database.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildDistiLDTracks">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildDistiLDTracks</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildDistiLDTracks" title="Permalink to this definition">¶</a></dt>
<dd><p>build bed tracks from DistiLD database.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildGFFSummary">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildGFFSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildGFFSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize genomic coverage of gff file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.buildBedSummary">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">buildBedSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.buildBedSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize genomic coverage of bed file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.genome">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">genome</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.genome" title="Permalink to this definition">¶</a></dt>
<dd><p>import information on geneset.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.geneset">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">geneset</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.geneset" title="Permalink to this definition">¶</a></dt>
<dd><p>import information on geneset.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.repeats">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">repeats</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.repeats" title="Permalink to this definition">¶</a></dt>
<dd><p>import repeat annotations.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.fasta">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">fasta</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>build fasta files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.tss">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">tss</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.tss" title="Permalink to this definition">¶</a></dt>
<dd><p>build promotors.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.promotors">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">promotors</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.promotors" title="Permalink to this definition">¶</a></dt>
<dd><p>build promotors.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.ontologies">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">ontologies</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.ontologies" title="Permalink to this definition">¶</a></dt>
<dd><p>create and load ontologies</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.GenicBedFiles">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">GenicBedFiles</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.GenicBedFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>build genic bed files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.summary">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">summary</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.summary" title="Permalink to this definition">¶</a></dt>
<dd><p>summary targets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.gemMappability">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">gemMappability</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.gemMappability" title="Permalink to this definition">¶</a></dt>
<dd><p>Count mappable bases in genome</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.full">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">full</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.full" title="Permalink to this definition">¶</a></dt>
<dd><p>build all targets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.build_report">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">build_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>build report from scratch.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.update_report">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">update_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.update_report" title="Permalink to this definition">¶</a></dt>
<dd><p>update report.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_annotations.publish_report">
<tt class="descclassname">pipeline_annotations.</tt><tt class="descname">publish_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_annotations.publish_report" title="Permalink to this definition">¶</a></dt>
<dd><p>publish report.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Annotation pipeline</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
<li><a class="reference internal" href="#input">Input</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pipeline-output">Pipeline output</a></li>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../CGATPipelines.html"
                        title="previous chapter">NGS Pipelines</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pipeline_ancestral_repeats.html"
                        title="next chapter">Ancestral repeats pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelines/pipeline_annotations.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_ancestral_repeats.html" title="Ancestral repeats pipeline"
             >next</a> |</li>
        <li class="right" >
          <a href="../CGATPipelines.html" title="NGS Pipelines"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" >NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>