
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Benchmark: RNASeq-Mapping &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="NGS Pipelines" href="../CGATPipelines.html" />
    <link rel="next" title="Optimizing cufflinks parameters" href="pipeline_cufflinks_optimization.html" />
    <link rel="prev" title="Pipeline template" href="pipeline_template.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_cufflinks_optimization.html" title="Optimizing cufflinks parameters"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pipeline_template.html" title="Pipeline template"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" accesskey="U">NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pipeline_benchmark_rnaseqmappers"></span><div class="section" id="benchmark-rnaseq-mapping">
<h1>Benchmark: RNASeq-Mapping<a class="headerlink" href="#benchmark-rnaseq-mapping" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger, Nicholas Ilott</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>This pipeline takes one or more fastq formatted file from an RNASeq experiment
and rus various mappers on it. The focus is here on sensitivity versus specificity. 
The pipeline will not</p>
<blockquote>
<div><ul class="simple">
<li>rigorously time the tools</li>
<li>do an exhaustive parameter search</li>
<li>perform validation from simulated data sets (TODO)</li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This pipeline is currently configured for mapping in SOLiD colourspace, 
single end reads and mapping against a mammalian (human) genome.</p>
</div>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<div class="section" id="mapping">
<h3>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h3>
<p>The pipeline aligns reads to a genome using a variety of mappers. Currently implemented are:</p>
<blockquote>
<div><ul class="simple">
<li>tophat</li>
<li>shrimp</li>
<li>bfast</li>
<li>bwa</li>
<li>bowtie</li>
<li>novoalignCS</li>
</ul>
</div></blockquote>
<p>Mappers are configured to only report &#8220;best&#8221; matches&#8217;, usually defined as those
with a minimum number of mismatches.</p>
<p>Mappers can be tuned in the configuration file. By default, no tuning takes place.</p>
</div>
<div class="section" id="validation">
<h3>Validation<a class="headerlink" href="#validation" title="Permalink to this headline">¶</a></h3>
<p>Validation with simulated data is straight-forward as the origin of a read is known.
Validation of real data needs to use indirect proxies to assess mapping quality.
Here, we use known protein coding genes as reference points on the genome. The expectation
is that a good mapper will predominantly map to protein coding exons.</p>
<p>This section provides a list of the criteria that can be used for validation</p>
<div class="section" id="number-of-mapped-reads">
<h4>Number of mapped reads<a class="headerlink" href="#number-of-mapped-reads" title="Permalink to this headline">¶</a></h4>
<p>The number of mapped reads is a simple measure of the sensitivity of a mapper.</p>
<blockquote>
<div><ul>
<li><p class="first">Number of reads mapped: More mapped reads is better as it increases coverage.</p>
</li>
<li><dl class="first docutils">
<dt>Number of uniquely mapped reads: A higher proportion of uniquely mapped reads</dt>
<dd><p class="first last">is better.</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<p>These criteria are confounded. For example, if a mapper tolerates more mismatches,
the number of mapped reads will increase while the number of uniquely 
mapped reads might drop.</p>
<p>See table <tt class="file docutils literal"><span class="pre">view_mapping</span></tt>.</p>
</div>
<div class="section" id="spliced-reads">
<h4>Spliced reads<a class="headerlink" href="#spliced-reads" title="Permalink to this headline">¶</a></h4>
<p>For a splicing mapper, the more reads that are spliced and mapped, the better.
Splicing can also be used to measure selectivity. Both ends of a spliced read
should overlap with exons.</p>
<p>See table <tt class="file docutils literal"><span class="pre">exon_validation</span></tt>.</p>
</div>
<div class="section" id="using-strandedness">
<h4>Using strandedness<a class="headerlink" href="#using-strandedness" title="Permalink to this headline">¶</a></h4>
<p>If the library is stranded, sense matches in protein coding transcripts should
vastly outnumber antisense matches. A measure of specificity is thus the number
of antisense reads in protein coding exons.</p>
<p>See tables <tt class="file docutils literal"><span class="pre">&lt;track&gt;_exon_coverage</span></tt> and <tt class="file docutils literal"><span class="pre">&lt;track&gt;_region_coverage</span></tt>.</p>
</div>
<div class="section" id="using-exon-boundaries">
<h4>Using exon boundaries<a class="headerlink" href="#using-exon-boundaries" title="Permalink to this headline">¶</a></h4>
<p>For transcript assembly, exon boundaries should be delineated as accurately as possible.</p>
<p>See tables <tt class="file docutils literal"><span class="pre">&lt;track&gt;_overrun`</span></tt>.</p>
</div>
<div class="section" id="identifying-mismapped-reads">
<h4>Identifying mismapped reads<a class="headerlink" href="#identifying-mismapped-reads" title="Permalink to this headline">¶</a></h4>
<p>By mapping reads against both the genome and the transcriptome and reconciling the locations,
mismapped reads can be identified as those that map to a transcript, but map to the genome in
a location that is not annotated as that particular transcript.</p>
<p>See table <tt class="file docutils literal"><span class="pre">transcriptome_validation</span></tt>.</p>
</div>
<div class="section" id="comparison-between-mappers">
<h4>Comparison between mappers<a class="headerlink" href="#comparison-between-mappers" title="Permalink to this headline">¶</a></h4>
<p>Consistency between mappers can also be used to gauge the effect of read quality onto
mapping.</p>
<p>See table <tt class="file docutils literal"><span class="pre">read_correspondence</span></tt>.</p>
</div>
</div>
</div>
<div class="section" id="configuration">
<h2>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<p>The pipeline expects one or more <tt class="docutils literal"><span class="pre">.fastq.gz</span></tt> files in the
working directory.</p>
<div class="section" id="optional-inputs">
<h3>Optional inputs<a class="headerlink" href="#optional-inputs" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="notes">
<h3>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="code">
<h3>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.connect">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">connect</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.connect" title="Permalink to this definition">¶</a></dt>
<dd><p>connect to database.</p>
<p>Use this method to connect to additional databases.</p>
<p>Returns a database connection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mapReadsWithBFAST">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mapReadsWithBFAST</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mapReadsWithBFAST" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads with bfast</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mapReadsWithShrimp">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mapReadsWithShrimp</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mapReadsWithShrimp" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads with shrimp</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mapReadsWithNovoalign">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mapReadsWithNovoalign</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mapReadsWithNovoalign" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads with shrimp</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mapReadsWithBWA">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mapReadsWithBWA</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mapReadsWithBWA" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads with shrimp</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mapReadsWithTophat">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mapReadsWithTophat</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mapReadsWithTophat" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads with tophat</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mapReadsWithBowtie">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mapReadsWithBowtie</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mapReadsWithBowtie" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads with bowtie</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildBAMs">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildBAMs</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildBAMs" title="Permalink to this definition">¶</a></dt>
<dd><p>build BAM files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildCodingGeneSet">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildCodingGeneSet</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildCodingGeneSet" title="Permalink to this definition">¶</a></dt>
<dd><p>build a gene set with only protein coding 
transcripts.</p>
<p>Genes are selected via their gene biotype in the GTF file.
Note that this set will contain all transcripts of protein
coding genes, including processed transcripts.</p>
<p>This set includes UTR and CDS.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildReferenceTranscriptome">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildReferenceTranscriptome</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildReferenceTranscriptome" title="Permalink to this definition">¶</a></dt>
<dd><p>build reference transcriptome.</p>
<p>The reference transcriptome contains all known 
protein coding transcripts.</p>
<p>The sequences include both UTR and CDS.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mapReadsWithBowtieAgainstTranscriptome">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mapReadsWithBowtieAgainstTranscriptome</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mapReadsWithBowtieAgainstTranscriptome" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads from short read archive sequence using bowtie against
transcriptome data.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.checkMappedReadsAgainstTranscriptome">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">checkMappedReadsAgainstTranscriptome</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.checkMappedReadsAgainstTranscriptome" title="Permalink to this definition">¶</a></dt>
<dd><p>reconcile genomic and transcriptome matches.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.loadTranscriptomeValidation">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">loadTranscriptomeValidation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.loadTranscriptomeValidation" title="Permalink to this definition">¶</a></dt>
<dd><p>load transcriptome validation data into database.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildListOfMappedReadsGenome">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildListOfMappedReadsGenome</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildListOfMappedReadsGenome" title="Permalink to this definition">¶</a></dt>
<dd><p>compile list of reads mapped to transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildJunctionsDB">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildJunctionsDB</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildJunctionsDB" title="Permalink to this definition">¶</a></dt>
<dd><p>build a database of all junctions.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mapReadsWithBowtieAgainstJunctions">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mapReadsWithBowtieAgainstJunctions</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mapReadsWithBowtieAgainstJunctions" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads from short read archive sequence using bowtie against
splice junctions.</p>
<p>The reads are converted to genomic coordinates.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildListOfMappedReadsTranscriptome">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildListOfMappedReadsTranscriptome</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildListOfMappedReadsTranscriptome" title="Permalink to this definition">¶</a></dt>
<dd><p>compile list of reads mapped to transcripts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildMissedTranscriptomeReads">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildMissedTranscriptomeReads</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildMissedTranscriptomeReads" title="Permalink to this definition">¶</a></dt>
<dd><p>compile list of reads mapped to transcriptome that fail to match
to the genome.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildMissedJunctionsReads">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildMissedJunctionsReads</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildMissedJunctionsReads" title="Permalink to this definition">¶</a></dt>
<dd><p>compile list of reads mapped to junctions.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.loadMissedReadCounts">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">loadMissedReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.loadMissedReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>load summary table of numbers of missed reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildFastQCReport">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildFastQCReport</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildFastQCReport" title="Permalink to this definition">¶</a></dt>
<dd><p>run fastqc on aligned reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.countReads">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">countReads</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.countReads" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads in fastq file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildBAMStats">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildBAMStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.loadBAMStats">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">loadBAMStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.loadBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>import bam statisticis.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildExonValidation">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildExonValidation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildExonValidation" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildReadStats">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildReadStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildReadStats" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildReadCorrespondence">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildReadCorrespondence</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildReadCorrespondence" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.loadReadCorrespondence">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">loadReadCorrespondence</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.loadReadCorrespondence" title="Permalink to this definition">¶</a></dt>
<dd><p>load read correspondence data into database.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.mergeAndLoad">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">mergeAndLoad</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em>, <em>suffix</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.mergeAndLoad" title="Permalink to this definition">¶</a></dt>
<dd><p>load categorical tables (two columns) into a database.</p>
<p>The tables are merged and entered row-wise.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.loadExonValidation">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">loadExonValidation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.loadExonValidation" title="Permalink to this definition">¶</a></dt>
<dd><p>merge alignment stats into single tables.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildExonCoverage">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildExonCoverage</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildExonCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildRegionCoverage">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildRegionCoverage</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildRegionCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.buildAlignmentStats">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">buildAlignmentStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.buildAlignmentStats" title="Permalink to this definition">¶</a></dt>
<dd><p>build alignment stats using picard.</p>
<p>Note that picards counts reads but they are in fact alignments.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.loadAlignmentStats">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">loadAlignmentStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.loadAlignmentStats" title="Permalink to this definition">¶</a></dt>
<dd><p>merge alignment stats into single tables.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.createViewMapping">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">createViewMapping</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.createViewMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>create view in database for alignment stats.</p>
<p>This view aggregates all information on a per-track basis.</p>
<p>The table is built from the following tracks:</p>
<p>mapping_stats
bam_stats</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.build_report">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">build_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>build report from scratch.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.update_report">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">update_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.update_report" title="Permalink to this definition">¶</a></dt>
<dd><p>update report.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_benchmark_rnaseqmappers.publish">
<tt class="descclassname">pipeline_benchmark_rnaseqmappers.</tt><tt class="descname">publish</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_benchmark_rnaseqmappers.publish" title="Permalink to this definition">¶</a></dt>
<dd><p>publish files.</p>
</dd></dl>

</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Benchmark: RNASeq-Mapping</a><ul>
<li><a class="reference internal" href="#overview">Overview</a><ul>
<li><a class="reference internal" href="#mapping">Mapping</a></li>
<li><a class="reference internal" href="#validation">Validation</a><ul>
<li><a class="reference internal" href="#number-of-mapped-reads">Number of mapped reads</a></li>
<li><a class="reference internal" href="#spliced-reads">Spliced reads</a></li>
<li><a class="reference internal" href="#using-strandedness">Using strandedness</a></li>
<li><a class="reference internal" href="#using-exon-boundaries">Using exon boundaries</a></li>
<li><a class="reference internal" href="#identifying-mismapped-reads">Identifying mismapped reads</a></li>
<li><a class="reference internal" href="#comparison-between-mappers">Comparison between mappers</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
<li><a class="reference internal" href="#input">Input</a><ul>
<li><a class="reference internal" href="#optional-inputs">Optional inputs</a></li>
<li><a class="reference internal" href="#notes">Notes</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pipeline_template.html"
                        title="previous chapter">Pipeline template</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pipeline_cufflinks_optimization.html"
                        title="next chapter">Optimizing cufflinks parameters</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelines/pipeline_benchmark_rnaseqmappers.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_cufflinks_optimization.html" title="Optimizing cufflinks parameters"
             >next</a> |</li>
        <li class="right" >
          <a href="pipeline_template.html" title="Pipeline template"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" >NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>