
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>ChIP-Seq pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="NGS Pipelines" href="../CGATPipelines.html" />
    <link rel="next" title="RNA-Seq pipeline" href="pipeline_rnaseq.html" />
    <link rel="prev" title="CAPseq pipeline" href="pipeline_fastqToBigWig.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseq.html" title="RNA-Seq pipeline"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pipeline_fastqToBigWig.html" title="CAPseq pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" accesskey="U">NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-pipeline_chipseq"></span><div class="section" id="chip-seq-pipeline">
<h1>ChIP-Seq pipeline<a class="headerlink" href="#chip-seq-pipeline" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id: pipeline_chipseq.py 2900 2010-04-13 14:38:00Z andreas $</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>The ChIP-Seq pipeline imports reads from one or more ChIP-Seq experiments and
performs the following tasks:</p>
<blockquote>
<div><ul class="simple">
<li>align reads to the genome</li>
<li>call peaks</li>
<li>annotate intervals with respect to a reference gene set</li>
<li>describe de-novo motifs</li>
<li>find motifs within intervals</li>
</ul>
</div></blockquote>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="../UsingPipelines.html#pipelinesettingup"><em>Setting up a pipeline</em></a> and <a class="reference internal" href="../UsingPipelines.html#pipelinerunning"><em>Running a pipeline</em></a> on general information how to use CGAT pipelines.</p>
<div class="section" id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires a configured <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt> file. The pipeline looks for a configuration file in several places:</p>
<blockquote>
<div><ol class="arabic simple">
<li>The default configuration in the <a class="reference internal" href="../glossary.html#term-code-directory"><em class="xref std std-term">code directory</em></a>.</li>
<li>A shared configuration file <tt class="file docutils literal"><span class="pre">../pipeline.ini</span></tt>.</li>
<li>A local configuration <tt class="file docutils literal"><span class="pre">pipeline.ini</span></tt>.</li>
</ol>
</div></blockquote>
<p>The order is as above. Thus, a local configuration setting will
override a shared configuration setting and a default configuration
setting.</p>
<p>Configuration files follow the ini format (see the python
<cite>ConfigParser &lt;http://docs.python.org/library/configparser.html&gt;</cite> documentation).
The configuration file is organized by section and the variables are documented within 
the file. In order to get a local configuration file in the current directory, type:</p>
<div class="highlight-python"><div class="highlight"><pre>python &lt;codedir&gt;/pipeline_chipseq.py config
</pre></div>
</div>
<p>The following sections and parameters probably should be changed from the default 
values:</p>
<div class="admonition-todo admonition" id="index-0">
<p class="first admonition-title">Todo</p>
<p class="last">describe important parameters</p>
</div>
<p>The sphinxreport report requires a <tt class="file docutils literal"><span class="pre">conf.py</span></tt> and <tt class="file docutils literal"><span class="pre">sphinxreport.ini</span></tt> file 
(see <a class="reference internal" href="../UsingPipelines.html#pipelinereporting"><em>Building pipeline reports</em></a>). To start with, use the files supplied with the
<a class="reference internal" href="#example">Example</a> data.</p>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="section" id="reads">
<h4>Reads<a class="headerlink" href="#reads" title="Permalink to this headline">¶</a></h4>
<p>Input are <tt class="file docutils literal"><span class="pre">.export.txt.gz</span></tt>-formatted files from Illumina, <tt class="file docutils literal"><span class="pre">fastq.gz</span></tt> files,
or <tt class="file docutils literal"><span class="pre">csfasta.gz</span></tt> files.</p>
<p>The files should be labeled in the following way:</p>
<div class="highlight-python"><div class="highlight"><pre>sample-condition-replicate.&lt;suffix&gt;.gz
</pre></div>
</div>
<p>For example:</p>
<div class="highlight-python"><div class="highlight"><pre>GM00855-D3-R1.&lt;suffix&gt;.gz
GM00855-D3-R2.&lt;suffix&gt;.gz
GM00855-input-R1.&lt;suffix&gt;.gz
GM00855-unstim-R1.&lt;suffix&gt;.gz
GM00855-unstim-R2.&lt;suffix&gt;.gz
</pre></div>
</div>
<p>Note that neither <tt class="docutils literal"><span class="pre">sample</span></tt>, <tt class="docutils literal"><span class="pre">condition</span></tt> or <tt class="docutils literal"><span class="pre">replicate</span></tt> should contain 
<tt class="docutils literal"><span class="pre">_</span></tt> (underscore) and <tt class="docutils literal"><span class="pre">.</span></tt> (dot) characters as these are used by the pipeline
to delineate tasks.</p>
</div>
<div class="section" id="optional-inputs">
<h4>Optional inputs<a class="headerlink" href="#optional-inputs" title="Permalink to this headline">¶</a></h4>
<p>Optinally, peaks can be supplied as <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted files. These peak files
will then be processed in the same way as peaks called within the pipeline. Use the
option <tt class="docutils literal"><span class="pre">tracks_extra</span></tt> to declare any additional tracks.</p>
<p>Additional peak files can be associated with one of the <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> files created
by the pipeline. This permits counting the number of tags inside peaks, finding the
peak summit, etc. In order to associated a peak file with a <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted
file, define a section in the pipeline.ini file. For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="n">mycalls</span><span class="o">.</span><span class="n">bed</span><span class="p">]</span>
<span class="n">track</span><span class="o">=</span><span class="n">tissue</span><span class="o">-</span><span class="n">sample</span><span class="o">-</span><span class="n">agg</span>
</pre></div>
</div>
<p>will process the file <tt class="docutils literal"><span class="pre">mycalls.bed</span></tt> exactly the same way as the track <tt class="docutils literal"><span class="pre">tissue-sample-agg</span></tt>.
Replicates can be specified explicitely:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="n">mycalls</span><span class="o">.</span><span class="n">bed</span><span class="p">]</span>
<span class="n">replicates</span><span class="o">=</span><span class="n">tissue1</span><span class="o">-</span><span class="n">sample1</span><span class="o">-</span><span class="n">R1</span><span class="p">,</span><span class="n">tissue1</span><span class="o">-</span><span class="n">sample1</span><span class="o">-</span><span class="n">R2</span>
</pre></div>
</div>
<p>will associate the file <tt class="docutils literal"><span class="pre">mycalls.bed</span></tt> with the replicates <tt class="docutils literal"><span class="pre">tissue1-sample1-R1</span></tt> 
and <tt class="docutils literal"><span class="pre">tissue1-sample1-R2</span></tt>. Note that globs don&#8217;t work for this yet, all
replicates have to be specified explicitely.</p>
</div>
<div class="section" id="reference-motifs">
<h4>Reference motifs<a class="headerlink" href="#reference-motifs" title="Permalink to this headline">¶</a></h4>
<p>Reference motifs are described by fasta-formatted multiple alignments, see for
example Jaspar download. The motifs are build by running MEME on the file.</p>
<p>Reference motifs should end in the suffix &#8221;.motif.fasta&#8221;, for example,
<tt class="file docutils literal"><span class="pre">rxrvdr.motif.fasta</span></tt>.</p>
</div>
</div>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>The pipeline requires the information from the following pipelines:</p>
<dl class="docutils">
<dt><a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a></dt>
<dd>set the configuration variable <tt class="xref py py-data docutils literal"><span class="pre">annotations_database</span></tt> and 
<tt class="xref py py-data docutils literal"><span class="pre">annotations_dir</span></tt>.</dd>
</dl>
<p>On top of the default CGAT setup, the pipeline requires the following software to be in the 
path:</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="22%" />
<col width="55%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Program</em></td>
<td><em>Version</em></td>
<td><em>Purpose</em></td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">bowtie</a></td>
<td>&gt;=0.12.7</td>
<td>read mapping</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="pipline-output">
<h2>Pipline Output<a class="headerlink" href="#pipline-output" title="Permalink to this headline">¶</a></h2>
<p>The results of the computation are all stored in an sqlite relational
database <tt class="file docutils literal"><span class="pre">csvdb</span></tt>.</p>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>Example data is available at <a class="reference external" href="http://www.cgat.org/~andreas/sample_data/pipeline_chipseq.tgz">http://www.cgat.org/~andreas/sample_data/pipeline_chipseq.tgz</a>.
To run the example, simply unpack and untar:</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://www.cgat.org/~andreas/sample_data/pipeline_chipseq.tgz
tar -xvzf pipeline_chipseq.tgz
cd pipeline_chipseq
python &lt;srcdir&gt;/pipeline_chipseq.py make full
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For the pipeline to run, install the <a class="reference internal" href="pipeline_annotations.html"><em>Annotation pipeline</em></a> as well.</p>
</div>
</div>
<div class="section" id="glossary">
<h2>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h2>
<dl class="glossary docutils">
<dt id="term-bowtie">bowtie</dt>
<dd><a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">bowtie</a> - a read mapper</dd>
</dl>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pipeline_chipseq.getControl">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">getControl</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.getControl" title="Permalink to this definition">¶</a></dt>
<dd><p>return appropriate control for a track</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.getUnstimulated">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">getUnstimulated</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.getUnstimulated" title="Permalink to this definition">¶</a></dt>
<dd><p>return unstimulated condition for a track</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.getSubtracted">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">getSubtracted</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.getSubtracted" title="Permalink to this definition">¶</a></dt>
<dd><p>return &#8220;subtracted&#8221; condition for a track</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.getUnsubtracted">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">getUnsubtracted</tt><span class="sig-paren">(</span><em>track</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.getUnsubtracted" title="Permalink to this definition">¶</a></dt>
<dd><p>return &#8220;unsubtracted&#8221; condition for a track</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.connect">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">connect</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.connect" title="Permalink to this definition">¶</a></dt>
<dd><p>connect to database.</p>
<p>This method also attaches to helper databases.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.countReads">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">countReads</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.countReads" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads in input files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadReadCounts">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadReadCounts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadReadCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>load read counts into database.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildBAM">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildBAM</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildBAM" title="Permalink to this definition">¶</a></dt>
<dd><p>re-map eland formatted reads with bowtie</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.makeMask">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">makeMask</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.makeMask" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a mask for filtering reads if required.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.prepareBAMForPeakCalling">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">prepareBAMForPeakCalling</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.prepareBAMForPeakCalling" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare BAM files for peak calling.</p>
<ul>
<li><p class="first">unmapped reads are removed.</p>
</li>
<li><dl class="first docutils">
<dt>if the option &#8220;calling_deduplicate&#8221; is Picard.MarkDuplicates is run </dt>
<dd><p class="first last">to remove duplicate reads</p>
</dd>
</dl>
</li>
<li><p class="first">reads may be filtered by exon or location</p>
<blockquote>
<div><ul class="simple">
<li>to remove reads by exon, the option &#8220;calling_filter_exons&#8221; should specify a file containing 
a list of ensembl gene identifiers (one per line)</li>
<li>to remove reads by location, the option &#8220;calling_filter_regions&#8221; should specify a bed file&#8217;&#8216;</li>
</ul>
</div></blockquote>
</li>
</ul>
<p>The resulting bam file has a .prep.bam extension. Merging infiles is currently untested and the 
methods only consider single end reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.makeReadCorrelation">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">makeReadCorrelation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.makeReadCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>compute correlation between reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.makeReadCorrelationTable">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">makeReadCorrelationTable</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.makeReadCorrelationTable" title="Permalink to this definition">¶</a></dt>
<dd><p>compute correlation between reads.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildBAMStats">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildBAMStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of reads mapped, duplicates, etc.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadBAMStats">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadBAMStats</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadBAMStats" title="Permalink to this definition">¶</a></dt>
<dd><p>import bam statistics.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.runMACS">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">runMACS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.runMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>run MACS for peak detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.summarizeMACS">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">summarizeMACS</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.summarizeMACS" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize MACS results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.summarizeMACSFDR">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">summarizeMACSFDR</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.summarizeMACSFDR" title="Permalink to this definition">¶</a></dt>
<dd><p>summarize MACS results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadMACSSummary">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadMACSSummary</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadMACSSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadMACSSummaryFDR">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadMACSSummaryFDR</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadMACSSummaryFDR" title="Permalink to this definition">¶</a></dt>
<dd><p>load macs summary.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.combineExperiment">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">combineExperiment</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.combineExperiment" title="Permalink to this definition">¶</a></dt>
<dd><p>combine replicates between experiments.</p>
<p>The replicates are combined using intersection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.combineCondition">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">combineCondition</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.combineCondition" title="Permalink to this definition">¶</a></dt>
<dd><p>combine conditions between cell lines.</p>
<p>The conditions are merged via intersection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.combineTissue">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">combineTissue</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.combineTissue" title="Permalink to this definition">¶</a></dt>
<dd><p>combine conditions between cell lines.</p>
<p>The conditions are merged via intersection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.subtractUnstimulated">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">subtractUnstimulated</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.subtractUnstimulated" title="Permalink to this definition">¶</a></dt>
<dd><p>remove the unstimulated data sets from the individual tracks.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadIntervalsFromBed">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadIntervalsFromBed</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadIntervalsFromBed" title="Permalink to this definition">¶</a></dt>
<dd><p>load intervals from <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted files into database.</p>
<p>These <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted intervals are derived by 
merging/intersecting the various tracks or have been
placed explicitely into the directory.</p>
<p>Also, re-evaluate the intervals by counting reads within
the interval. In contrast to the initial pipeline, the
genome is not binned. In particular, the meaning of the
columns in the table changes to:</p>
<p>nProbes: number of reads in interval
PeakCenter: position with maximum number of reads in interval
AvgVal: average coverage within interval</p>
<p>If <em>replicates</em> is true, only replicates will be considered
for the counting. Otherwise the counts aggregate both replicates
and conditions.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildMergedIntervals">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildMergedIntervals</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildMergedIntervals" title="Permalink to this definition">¶</a></dt>
<dd><p>combine all experiments.</p>
<p>The replicates are combined using a merge.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.getBamPeakPairParams">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">getBamPeakPairParams</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.getBamPeakPairParams" title="Permalink to this definition">¶</a></dt>
<dd><p>retrieves filenames for all vs. all comparsion (bed vs bam)</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildReadProfileOfTranscripts">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildReadProfileOfTranscripts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildReadProfileOfTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a table with peak shape parameters.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildIntervalProfileOfTranscripts">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildIntervalProfileOfTranscripts</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildIntervalProfileOfTranscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>build a table with peak shape parameters.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildPeakShapeTable">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildPeakShapeTable</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildPeakShapeTable" title="Permalink to this definition">¶</a></dt>
<dd><p>build a table with peak shape parameters.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.peakCoverage">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">peakCoverage</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.peakCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>uses coverageBed to count the total number of reads under peaks</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildReadCoverageTable">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildReadCoverageTable</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildReadCoverageTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts reads from .coverage files and writes out a matrix of the results from all .bed vs. all .bam files</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadReadCoverageTable">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadReadCoverageTable</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadReadCoverageTable" title="Permalink to this definition">¶</a></dt>
<dd><p>load read coverage table.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.exportBigwig">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">exportBigwig</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.exportBigwig" title="Permalink to this definition">¶</a></dt>
<dd><p>convert BAM to bigwig file.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildBigwigInfo">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildBigwigInfo</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildBigwigInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>collate summary of bigwig files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.exportMotifSequences">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">exportMotifSequences</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.exportMotifSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>export sequences for all intervals.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.exportMotifControlSequences">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">exportMotifControlSequences</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.exportMotifControlSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>for each interval, export the left and right 
sequence segment of the same size.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildOverlap">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildOverlap</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildOverlap" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlap between intervals.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadOverlap">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadOverlap</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadOverlap" title="Permalink to this definition">¶</a></dt>
<dd><p>load overlap results.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.makeReproducibility">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">makeReproducibility</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.makeReproducibility" title="Permalink to this definition">¶</a></dt>
<dd><p>compute overlap between intervals.</p>
<p>Note the exon percentages are approximations assuming that there are
not more than one intervals in one set overlapping one in the other set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadReproducibility">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadReproducibility</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadReproducibility" title="Permalink to this definition">¶</a></dt>
<dd><p>load Reproducibility results</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.makePeakvalCorrelation">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">makePeakvalCorrelation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.makePeakvalCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>compute correlation of interval properties between sets for field peakval.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.makeAvgvalCorrelation">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">makeAvgvalCorrelation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.makeAvgvalCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>compute correlation of interval properties between sets for field peakval.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.makeLengthCorrelation">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">makeLengthCorrelation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.makeLengthCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>compute correlation of interval properties between sets for field peakval.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadCorrelation">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadCorrelation</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>load correlation data.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildReferenceMotifs">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildReferenceMotifs</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildReferenceMotifs" title="Permalink to this definition">¶</a></dt>
<dd><p>build motif from jasper sites by running MEME.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.runMEME">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">runMEME</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.runMEME" title="Permalink to this definition">¶</a></dt>
<dd><p>run MEME to find motifs.</p>
<p>In order to increase the signal/noise ratio,
MEME is not run on all intervals but only the 
top 10% of intervals (peakval) are used. 
Also, only the segment of 200 bp around the peak
is used and not the complete interval.</p>
<ul class="simple">
<li>Softmasked sequence is converted to hardmasked
sequence to avoid the detection of spurious motifs.</li>
<li>Sequence is run through dustmasker</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.runGLAM2">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">runGLAM2</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.runGLAM2" title="Permalink to this definition">¶</a></dt>
<dd><p>run glam2 on all intervals and motifs.</p>
<p>In order to increase the signal/noise ratio,
MEME is not run on all intervals but only the 
top 10% of intervals (peakval) are used. 
Also, only the segment of 200 bp around the peak
is used and not the complete interval.</p>
<ul class="simple">
<li>Softmasked sequence is converted to hardmasked
sequence to avoid the detection of spurious motifs.</li>
<li>Sequence is run through dustmasker</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.runTomTom">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">runTomTom</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.runTomTom" title="Permalink to this definition">¶</a></dt>
<dd><p>compare ab-initio motifs against tomtom.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadTomTom">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadTomTom</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadTomTom" title="Permalink to this definition">¶</a></dt>
<dd><p>compare ab-initio motifs against tomtom.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.filterMotifs">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">filterMotifs</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.filterMotifs" title="Permalink to this definition">¶</a></dt>
<dd><p>scan motifs found with MEME against master motif using tomtom.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.collectMotifs">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">collectMotifs</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.collectMotifs" title="Permalink to this definition">¶</a></dt>
<dd><p>dummy target - merge motifs.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadMotifInformation">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadMotifInformation</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadMotifInformation" title="Permalink to this definition">¶</a></dt>
<dd><p>load information about motifs into database.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadMotifSequenceComposition">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadMotifSequenceComposition</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadMotifSequenceComposition" title="Permalink to this definition">¶</a></dt>
<dd><p>compute sequence composition of sequences used for ab-initio search.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.runMAST">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">runMAST</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.runMAST" title="Permalink to this definition">¶</a></dt>
<dd><p>run mast on all intervals and motifs.</p>
<p>Collect all results for an E-value up to 10000 so that
all sequences are output and MAST curves can be computed.</p>
<p>10000 is a heuristic.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.runBioProspector">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">runBioProspector</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.runBioProspector" title="Permalink to this definition">¶</a></dt>
<dd><p>run bioprospector for motif discovery.</p>
<p>Bioprospector is run on only the top 10% of peaks.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadBioProspector">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadBioProspector</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadBioProspector" title="Permalink to this definition">¶</a></dt>
<dd><p>load results from bioprospector.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadMAST">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadMAST</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadMAST" title="Permalink to this definition">¶</a></dt>
<dd><p>parse mast file and load into database.</p>
<p>Parse several motif runs and add them to the same
table.</p>
<p>Add columns for the control data as well.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.runGLAM2SCAN">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">runGLAM2SCAN</tt><span class="sig-paren">(</span><em>infiles</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.runGLAM2SCAN" title="Permalink to this definition">¶</a></dt>
<dd><p>run glam2scan on all intervals and motifs.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadGLAM2SCAN">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadGLAM2SCAN</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadGLAM2SCAN" title="Permalink to this definition">¶</a></dt>
<dd><p>parse mast file and load into database.</p>
<p>Parse several motif runs and add them to the same
table.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.annotateIntervals">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">annotateIntervals</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.annotateIntervals" title="Permalink to this definition">¶</a></dt>
<dd><p>classify chipseq intervals according to their location 
with respect to the gene set.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.annotateTSS">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">annotateTSS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.annotateTSS" title="Permalink to this definition">¶</a></dt>
<dd><p>compute distance to TSS</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.annotateRepeats">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">annotateRepeats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.annotateRepeats" title="Permalink to this definition">¶</a></dt>
<dd><p>count the overlap between intervals and repeats.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadAnnotations">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadAnnotations</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>load interval annotations: genome architecture</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadTSS">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadTSS</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadTSS" title="Permalink to this definition">¶</a></dt>
<dd><p>load interval annotations: distance to transcription start sites</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadRepeats">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadRepeats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadRepeats" title="Permalink to this definition">¶</a></dt>
<dd><p>load interval annotations: repeats</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.buildIntervalCounts">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">buildIntervalCounts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.buildIntervalCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>count read density in bed files comparing stimulated versus unstimulated binding.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.loadIntervalCounts">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">loadIntervalCounts</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.loadIntervalCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>load interval counts.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.mapping">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">mapping</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>map reads</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.intervals">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">intervals</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>compute binding intervals.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.export">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">export</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.export" title="Permalink to this definition">¶</a></dt>
<dd><p>export files.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.discover_motifs">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">discover_motifs</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.discover_motifs" title="Permalink to this definition">¶</a></dt>
<dd><p>run motif discovery.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.detect_motifs">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">detect_motifs</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.detect_motifs" title="Permalink to this definition">¶</a></dt>
<dd><p>run motif detection.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.correlation">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">correlation</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>run the correlation targets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.annotation">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">annotation</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.annotation" title="Permalink to this definition">¶</a></dt>
<dd><p>run the annotation targets.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.createViewMapping">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">createViewMapping</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.createViewMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>create view in database for alignment stats.</p>
<p>This view aggregates all information on a per-track basis.</p>
<p>The table is built from the following tracks:</p>
<p>bam_stats: .call</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.full">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">full</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.full" title="Permalink to this definition">¶</a></dt>
<dd><p>run the full pipeline.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.build_report">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">build_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.build_report" title="Permalink to this definition">¶</a></dt>
<dd><p>build report from scratch.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.update_report">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">update_report</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.update_report" title="Permalink to this definition">¶</a></dt>
<dd><p>update report.</p>
</dd></dl>

<dl class="function">
<dt id="pipeline_chipseq.publish">
<tt class="descclassname">pipeline_chipseq.</tt><tt class="descname">publish</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pipeline_chipseq.publish" title="Permalink to this definition">¶</a></dt>
<dd><p>publish files.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">ChIP-Seq pipeline</a><ul>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
<li><a class="reference internal" href="#input">Input</a><ul>
<li><a class="reference internal" href="#reads">Reads</a></li>
<li><a class="reference internal" href="#optional-inputs">Optional inputs</a></li>
<li><a class="reference internal" href="#reference-motifs">Reference motifs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pipline-output">Pipline Output</a></li>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#glossary">Glossary</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pipeline_fastqToBigWig.html"
                        title="previous chapter">CAPseq pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pipeline_rnaseq.html"
                        title="next chapter">RNA-Seq pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pipelines/pipeline_chipseq.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pipeline_rnaseq.html" title="RNA-Seq pipeline"
             >next</a> |</li>
        <li class="right" >
          <a href="pipeline_fastqToBigWig.html" title="CAPseq pipeline"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../CGATPipelines.html" >NGS Pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>