
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>align_transcripts.py - multiply align transcripts &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="peptides2cds.py - map peptide sequences onto nucleic acid sequences" href="peptides2cds.html" />
    <link rel="prev" title="tree_species2genes.py - collapse leaves with the same species." href="tree_species2genes.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="peptides2cds.html" title="peptides2cds.py - map peptide sequences onto nucleic acid sequences"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tree_species2genes.html" title="tree_species2genes.py - collapse leaves with the same species."
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-align_transcripts"></span><div class="section" id="align-transcripts-py-multiply-align-transcripts">
<h1>align_transcripts.py - multiply align transcripts<a class="headerlink" href="#align-transcripts-py-multiply-align-transcripts" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script will multiply align transcripts. Transcripts
are aligned as amino acid sequences and later back-translated
to codon sequences.</p>
<p>The script is aware of gene structures will correctly align
exons in sequence.</p>
<p>If there are multiple transcripts for a gene, concatenate the
exons into one pseudo-transcript and submit that to the multiple
alignment in order to avoid incompatible exons to be aligned.
After alignment, the pseudo-transcripts are split into separate
parts again.</p>
<p>If the peptide sequence contains at least one character <tt class="docutils literal"><span class="pre">U</span></tt>, the
sequence is assumed to be selonoprotein.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>python align_transcripts.py --help
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python align_transcripts.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: align_transcripts.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -m MASTER, --master=MASTER
                        master sequence.
  -p MASTER_PATTERN, --master-pattern=MASTER_PATTERN
                        master pattern.
  --master-species=MASTER_SPECIES
                        species to use as master sequences.
  -t FILENAME_TRANSLATION, --translate=FILENAME_TRANSLATION
                        filename on where to store translated sequences.
  -e FILENAME_EXONS, --exons=FILENAME_EXONS
                        filename on where to exon information.
  -g FILENAME_GTF, --gtf=FILENAME_GTF
                        filename with exon information in gtf format.
  -c, --mark-codons     mark codons.
  --remove-stops        remove stop codons.
  --mask-stops          mask stop codons.
  --mask-char=MASK_CHAR
                        masking character to use.
  -f, --remove-frameshifts
                        remove columns corresponding to frameshifts.
  -s, --split-exons     split columns aligned to different exons in the same
                        gene.
  -a TARGET, --target=TARGET
                        perform cleaning up for certain targets.
  --stop-at=STOP_AT     stop at intermediate stage and dump out that
                        alignment.
  --force-map           force mapping of sequences that have changed to
                        previous sequence.
  --cds=FILENAME_CDS    filename with cds - useful if you expect pseudogenes
                        in your set. The peptide
                        sequences will be aligned to the cds sequences. This
                        produces better coordinates.
  --output=OUTPUT       which alignment to output: aligned=aligned sequences,
                        but before untangling exons;  final=final multiple
                        alignment; unaligned=unaligned sequences;
                        coords=genomic coordinates (corresponding to
                        &#39;final_na&#39;).
  --output-filename-pattern=OUTPUT_FILENAME_PATTERN
                        filename pattern for multiple alignment output files.
                        If no --output option is given, stdout is used.
  --output-filename-coords=OUTPUT_FILENAME_COORDS
                        filename to output coordinates to.
  --output-format=OUTPUT_FORMAT
                        output format of multiple alignments.
  --strict              apply strict checking.
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">align_transcripts.py - multiply align transcripts</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tree_species2genes.html"
                        title="previous chapter">tree_species2genes.py - collapse leaves with the same species.</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="peptides2cds.html"
                        title="next chapter">peptides2cds.py - map peptide sequences onto nucleic acid sequences</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/align_transcripts.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="peptides2cds.html" title="peptides2cds.py - map peptide sequences onto nucleic acid sequences"
             >next</a> |</li>
        <li class="right" >
          <a href="tree_species2genes.html" title="tree_species2genes.py - collapse leaves with the same species."
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>