
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>bam2bam.py - modify bam files &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="cgat_script_template.py - template for CGAT scripts" href="bam2bed.html" />
    <link rel="prev" title="bam2UniquePairs.py - filter/report uniquely mapped read pairs from a (bwa!) bam-file" href="bam2UniquePairs.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="bam2bed.html" title="cgat_script_template.py - template for CGAT scripts"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="bam2UniquePairs.html" title="bam2UniquePairs.py - filter/report uniquely mapped read pairs from a (bwa!) bam-file"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-bam2bam"></span><div class="section" id="bam2bam-py-modify-bam-files">
<h1>bam2bam.py - modify bam files<a class="headerlink" href="#bam2bam-py-modify-bam-files" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Genomics NGS BAM Manipulation</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script reads a <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file from stdin and outputs a
modified <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file on stdout.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">You need to redirect logging information to a file or turn it off
via -v 0 in order to get a valid sam/bam file.</p>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>For example:</p>
<div class="highlight-python"><div class="highlight"><pre>python bam2bam.py --filter=mapped &lt; in.bam &gt; out.bam
</pre></div>
</div>
<p>will remove all unmapped reads from the bam-file. To remove
unmapped reads from multiple bam-files, try:</p>
<div class="highlight-python"><div class="highlight"><pre>python bam2bam.py --inplace --filter=mapped *.bam
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python bam2bam.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="documentation">
<h2>Documentation<a class="headerlink" href="#documentation" title="Permalink to this headline">¶</a></h2>
<p>The script implements the following methods:</p>
<dl class="docutils">
<dt><tt class="docutils literal"><span class="pre">set-nh</span></tt></dt>
<dd>set the NH flag. Some tools (bowtie, bwa) do not set the NH flag. 
If set, this option will set the NH flag (for mapped reads).
This option requires the bam/sam file to be sorted by read name.</dd>
<dt><tt class="docutils literal"><span class="pre">unset-unmapped_mapq</span></tt> </dt>
<dd>some tools set the mapping quality of unmapped reads. This
causes a violation in the Picard tools.</dd>
<dt><tt class="docutils literal"><span class="pre">remove-better</span></tt></dt>
<dd>remove alignments that are worse than alignments present in
an alternative <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file (<tt class="docutils literal"><span class="pre">--filter-bam</span></tt>).</dd>
<dt><tt class="docutils literal"><span class="pre">filter</span></tt></dt>
<dd>remove alignments based on a variety of flags.</dd>
<dt><tt class="docutils literal"><span class="pre">strip</span></tt> </dt>
<dd>remove the sequence and/or quality scores from all reads in
a bam-file. If <tt class="docutils literal"><span class="pre">match</span></tt>, sequence and quality are stripped 
for alignments without mismatches. The latter requires the
<tt class="docutils literal"><span class="pre">NM</span></tt> tag. If not present, nothing will be stripped. Note 
that stripping is not reversible if the read names are not
unique.</dd>
<dt><tt class="docutils literal"><span class="pre">set-sequence</span></tt></dt>
<dd>set the sequence and quality scores in the bam file to some 
dummy sequence. Necessary for some tools that can not work
with bam-files without sequence.</dd>
<dt><tt class="docutils literal"><span class="pre">unstrip</span></tt></dt>
<dd>add sequence and quality scores back to a bam file. The
argument to this option is a <a class="reference internal" href="../glossary.html#term-36"><em class="xref std std-term">fastq</em></a> formatted file
with the sequences and quality scores to insert.</dd>
</dl>
<p>By default, the script works from stdin and outputs to stdout.
If the <tt class="docutils literal"><span class="pre">--inplace</span> <span class="pre">option</span></tt> is given, the script will modify
bam-files given as command line arguments in-place - a temporary
copy will be written to <tt class="file docutils literal"><span class="pre">/tmp</span></tt> and once completed copied
over the original file. The script will automatically re-index
the modified files.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: bam2bam.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  --set-nh              sets the NH flag. The file needs to be sorted by
                        readname [False]
  --unset-unmapped-mapq
                        sets the mapping quality of unmapped reads to 0
                        [False]
  --set-sequence        sets the sequence to &#39;A&#39;s (a valid base) and the
                        quality to &#39;F&#39;s,which is defined in all fastq scoring
                        schemes [False]
  --strip=STRIP         remove parts of the bam-file. Note that stripping the
                        sequence will also strip the quality values [none]
  --unstrip             add sequence and quality into bam file [none]
  --filter=FILTER       filter bam file. The option denotes the property that
                        is  used to determine better match [[]]
  --reference-bam=REFERENCE_BAM
                        bam-file to filter with [none]
  --sam                 output in sam format [False]
  --inplace             modify bam files in-place. Bam files need to be given
                        as arguments. Temporary bam files are written to /tmp
                        [False]
  -1 FASTQ_PAIR1, --fastq1=FASTQ_PAIR1
                        fastq file with read information for first in pair or
                        unpaired [none]
  -2 FASTQ_PAIR2, --fastq2=FASTQ_PAIR2
                        fastq file with read information for second in pair
                        [none]
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">bam2bam.py - modify bam files</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#documentation">Documentation</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="bam2UniquePairs.html"
                        title="previous chapter">bam2UniquePairs.py - filter/report uniquely mapped read pairs from a (bwa!) bam-file</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="bam2bed.html"
                        title="next chapter">cgat_script_template.py - template for CGAT scripts</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/bam2bam.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="bam2bed.html" title="cgat_script_template.py - template for CGAT scripts"
             >next</a> |</li>
        <li class="right" >
          <a href="bam2UniquePairs.html" title="bam2UniquePairs.py - filter/report uniquely mapped read pairs from a (bwa!) bam-file"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>