
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>bam2peakshape.py - compute peak shape features from a bam-file &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="bam2stats.py - compute stats from a bam-file" href="bam2stats.html" />
    <link rel="prev" title="bam2fastq.py - output fastq files from a bam-file" href="bam2fastq.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="bam2stats.html" title="bam2stats.py - compute stats from a bam-file"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="bam2fastq.html" title="bam2fastq.py - output fastq files from a bam-file"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-bam2peakshape"></span><div class="section" id="bam2peakshape-py-compute-peak-shape-features-from-a-bam-file">
<h1>bam2peakshape.py - compute peak shape features from a bam-file<a class="headerlink" href="#bam2peakshape-py-compute-peak-shape-features-from-a-bam-file" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Genomics NGS Intervals BAM BED Summary</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script takes a <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file with regions of interest,
for example binding intervals from a ChIP-Seq experiment. Using a collection
of aligned reads is a <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file, the script outputs a collection
of features describing the peak shape.</p>
<p>This script is designed with a slight emphasis on ChIP-Seq datasets.
The main reason that bam2peakshape.py is better suited for ChIP-Seq is that: 
(1) It is able to center the counting window at the summit of every individual peak;
(2) bam2peakshape.py is also able to use the control ChIP-Seq library to enable 
side-by-side comparison of treatment vs control;
(3) It can randomly shift the set of input regions to generate a background 
set of regions in order to provide a peaks profile that can be used as the 
control.
For example, given the peaks regions defined by analyzing some ChIP-Seq dataset
(e.g. by using MACS), and without using any additional genomic annotations 
(e.g. ENSEMBL, refseq), we can visualise the binding profiles of transcription 
factors ChIP-Seq data relative to the center of the peak regions.</p>
<p>Alternatively, you may consider using <a class="reference internal" href="bam2geneprofile.html"><em>bam2geneprofile.py - build meta-gene profile for a set of transcripts/genes</em></a>, which is
designed with a slight emphasis on analysing RNA-Seq datasets. Because it takes 
care of spliced reads by using the CIGAR string in the BAM file to accurately
count the covered bases (when the &#8211;base-accurate is specified).</p>
<p>The script outputs a tab-separated table on stdout containing features for each interval.
A peak is defined as the location of the highest density in an interval. The width of 
the peak (peak_width) is defined as the region around the peak in which the density does
not drop below a threshold of peak_heigt * 90%.</p>
<p>Among the features output are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="27%" />
<col width="73%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Column</em></td>
<td><em>Content</em></td>
</tr>
<tr class="row-even"><td>peak_height</td>
<td>number of reads at peak</td>
</tr>
<tr class="row-odd"><td>peak_median</td>
<td>median coverage compared to peak height</td>
</tr>
<tr class="row-even"><td>interval_width</td>
<td>width of interval</td>
</tr>
<tr class="row-odd"><td>peak_width</td>
<td>width of peak</td>
</tr>
<tr class="row-even"><td>bins</td>
<td>bins for a histogram of densities within the interval.</td>
</tr>
<tr class="row-odd"><td>npeaks</td>
<td>number of density peaks in interval.</td>
</tr>
<tr class="row-even"><td>peak_center</td>
<td>point of highest density in interval</td>
</tr>
<tr class="row-odd"><td>peak_relative_pos</td>
<td>point of highest density in interval coordinates</td>
</tr>
<tr class="row-even"><td>counts</td>
<td>counts for a histogram of densities within the interval</td>
</tr>
<tr class="row-odd"><td>furthest_half_height</td>
<td>Distance of peak center to furthest half-height position</td>
</tr>
<tr class="row-even"><td>closest_half_height</td>
<td>Distance of peak center to closest half-height position</td>
</tr>
</tbody>
</table>
<p>Additionally, the script outputs a set of matrixes with densities over intervals that
can be used for plotting. The default filenames are <tt class="docutils literal"><span class="pre">(matrix|control)_&lt;sortorder&gt;.tsv.gz</span></tt>,
The names can be controlled with the <tt class="docutils literal"><span class="pre">--output-filename-pattern</span></tt> option.</p>
<div class="admonition-todo admonition" id="index-0">
<p class="first admonition-title">Todo</p>
<p class="last">paired-endedness is not fully implemented.</p>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>python bam2peakshape.py bamfile bedfile
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python bam2peakshape.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: bam2peakshape.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -f FORMAT, --format=FORMAT
                        format of genomic input files for densities [bam]
  -o, --only-interval   only count tags strictly in interval. Otherwise, use
                        window around peak center  [none]
  -w WINDOW_SIZE, --window-size=WINDOW_SIZE
                        window size to use[1000]
  -b BIN_SIZE, --bin-size=BIN_SIZE
                        bin-size for histogram of read depth. [10]
  -s SORT, --sort=SORT  output sort order for matrices. [[]]
  -c CONTROL_FILE, --control-file=CONTROL_FILE
                        control file. If given, two peakshapes are computed,
                        one for the primary data and one for the control data.
                        The control file is centered around the same base as
                        the primary file and output in the same sort order as
                        the primary profile to all side-by-side. comparisons.
                        [none]
  -r, --random-shift    shift intervals in random direction up/downstream of
                        interval [False]
  -e CENTRING_METHOD, --centring-method=CENTRING_METHOD
                        centring method. Available are: reads=use density to
                        determine peak, middle=use middle of interval [reads]
  -n NORMALIZATION, --normalization=NORMALIZATION
                        normalisation to perform. [none]
  --use-strand          use strand information in intervals. Intervals on the
                        negative strand are flipped [False]
  -i SHIFT, --shift=SHIFT
                        shift for reads. When processing bam files, reads will
                        be shifted upstream/downstream by this amount. [0]
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -P OUTPUT_FILENAME_PATTERN, --output-filename-pattern=OUTPUT_FILENAME_PATTERN
                        OUTPUT filename pattern for various methods [%s].
  -F, --force           force over-writing of existing files.
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">bam2peakshape.py - compute peak shape features from a bam-file</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="bam2fastq.html"
                        title="previous chapter">bam2fastq.py - output fastq files from a bam-file</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="bam2stats.html"
                        title="next chapter">bam2stats.py - compute stats from a bam-file</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/bam2peakshape.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="bam2stats.html" title="bam2stats.py - compute stats from a bam-file"
             >next</a> |</li>
        <li class="right" >
          <a href="bam2fastq.html" title="bam2fastq.py - output fastq files from a bam-file"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>