
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>bed2bed.py - manipulate bed files &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="bed2gff.py - convert bed to gff/gtf" href="bed2gff.html" />
    <link rel="prev" title="bam2geneprofile.py - build meta-gene profile for a set of transcripts/genes" href="bam2geneprofile.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="bed2gff.html" title="bed2gff.py - convert bed to gff/gtf"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="bam2geneprofile.html" title="bam2geneprofile.py - build meta-gene profile for a set of transcripts/genes"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-bed2bed"></span><div class="section" id="bed2bed-py-manipulate-bed-files">
<h1>bed2bed.py - manipulate bed files<a class="headerlink" href="#bed2bed-py-manipulate-bed-files" title="Permalink to this headline">¶</a></h1>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script provides various methods for merging (by position, by name
or by score), filtering and moving bed formatted intervals and
outputting the results as a bed file</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div>python bed2bed.py &#8211;method=[METHOD] [options]</div></blockquote>
<p>This script provides several methods:</p>
<div class="section" id="merge">
<h3>merge<a class="headerlink" href="#merge" title="Permalink to this headline">¶</a></h3>
<p>Merge together overlapping or adjacent intervals. The basic
functionality is similar to bedtools merge, but with some additions:</p>
<ul>
<li><p class="first">Merging by name: specifying the &#8211;merge-by-name option will mean
that only overlaping (or adjacent intervals) with the same value in
the 4th column of the bed will be merged</p>
</li>
<li><dl class="first docutils">
<dt>Removing overlapping intervals with inconsistent names: set the</dt>
<dd><p class="first last"><tt class="docutils literal"><span class="pre">--remove-inconsistent</span></tt> option.</p>
</dd>
</dl>
</li>
</ul>
<div class="admonition caution">
<p class="first admonition-title">Caution</p>
<p class="last">Intervals of the same name will only be merged if they 
are consecutive in the bed file.</p>
</div>
<ul class="simple">
<li>Only output merged intervals: By specifiying the &#8211;merge-min-intervals=n
options, only those intervals that were created by merging at least n
intervals together will be output</li>
</ul>
</div>
<div class="section" id="bins">
<h3>bins<a class="headerlink" href="#bins" title="Permalink to this headline">¶</a></h3>
<p>Merges together overlapping or adjecent intervals only if they have
&#8220;similar&#8221; scores. Score similarity is assessed by creating a number of
score bins and assigning each interval to a bin. If too adjacent
intervals are in the same bin, the intervals are merged. Note that in
contrast to merge-by-name above, two intervals do not need to be
overlapping or within a certain distance to be merged.</p>
<p>There are several methods to create the bins:</p>
<ul class="simple">
<li>equal-bases: Bins are created to that they contain the same number of bases.  Specified
by passing &#8220;equal-bases&#8221; to &#8211;binning-method. This is the default.</li>
<li>equal-intervals: Score bins are create so that each bin contains the
same number of intervals. Specified by passing &#8220;equal-intervals&#8221; to
&#8211;binning-method.</li>
<li>equal-range: Score bins are created so that
each bin covers the same fraction of the total range of
scores. Specified by passing &#8220;equal-range&#8221; to &#8211;binning-method.</li>
<li>bin-edges: Score binds can be specified by manually passing a comma
seperated list of bin edges to &#8211;bin-edges.</li>
</ul>
<p>The number of bins is specified by the &#8211;num-bins options, and the
default is 5.</p>
</div>
<div class="section" id="block">
<h3>block<a class="headerlink" href="#block" title="Permalink to this headline">¶</a></h3>
<p>Creates blocked bed12 outputs from a bed6, where intervals with the
same name are merged together to create a single bed12 entry.</p>
<div class="admonition caution">
<p class="first admonition-title">Caution</p>
<p class="last">Input must be sorted so that entries of the same</p>
</div>
<p>name are together.</p>
</div>
<div class="section" id="filter-genome">
<h3>filter-genome<a class="headerlink" href="#filter-genome" title="Permalink to this headline">¶</a></h3>
<p>Removes intervals that are on unknown contigs or extend off the 3&#8217; or
5&#8217; end of the contig.  Requires a tab seperated input file to -g which
lists the contigs in the genome, plus their lengths.</p>
</div>
<div class="section" id="sanitize-genome">
<h3>sanitize-genome<a class="headerlink" href="#sanitize-genome" title="Permalink to this headline">¶</a></h3>
<p>As above, but instead of removing intervals overlapping the ends of
contigs, truncates them.  Also removes empty intervals.</p>
</div>
<div class="section" id="shift">
<h3>shift<a class="headerlink" href="#shift" title="Permalink to this headline">¶</a></h3>
<p>Moves intervals by the specified amount, but will not allow them to be
shifted off the end of contigs. Thus if a shift will shift the start
of end of the contig, the interval is only moved as much as is
possible without doing this.</p>
</div>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: bed2bed.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -m METHODS, --method=METHODS
                        method to apply [default=[]]
  --num-bins=NUM_BINS   number of bins into which to merge (used for method
                        `bins) [default=5]
  --bin-edges=BIN_EDGES
                        bin_edges for binning method [default=none]
  --binning-method=BINNING_METHOD
                        method used for binning (used for method `bins` if no
                        bin_edges is given) [default=equal-bases]
  --merge-distance=MERGE_DISTANCE
                        distance in bases over which to merge that are not
                        directly adjacent [default=0]
  --merge-min-intervals=MERGE_MIN_INTERVALS
                        only output merged intervals that are build from at
                        least x intervals [default=1]
  --merge-by-name       only merge intervals with the same name
                        [default=False]
  --remove-inconsistent
                        when merging, do not output intervals where the names
                        of overlapping intervals do not match [default=False]
  --offset=OFFSET       offset for shifting intervals [default=10000]
  -g GENOME_FILE, --genome-file=GENOME_FILE
                        filename with genome.
  -b BAM_FILE, --bam-file=BAM_FILE
                        bam-formatted filename with genome.
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">bed2bed.py - manipulate bed files</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#merge">merge</a></li>
<li><a class="reference internal" href="#bins">bins</a></li>
<li><a class="reference internal" href="#block">block</a></li>
<li><a class="reference internal" href="#filter-genome">filter-genome</a></li>
<li><a class="reference internal" href="#sanitize-genome">sanitize-genome</a></li>
<li><a class="reference internal" href="#shift">shift</a></li>
</ul>
</li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="bam2geneprofile.html"
                        title="previous chapter">bam2geneprofile.py - build meta-gene profile for a set of transcripts/genes</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="bed2gff.html"
                        title="next chapter">bed2gff.py - convert bed to gff/gtf</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/bed2bed.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="bed2gff.html" title="bed2gff.py - convert bed to gff/gtf"
             >next</a> |</li>
        <li class="right" >
          <a href="bam2geneprofile.html" title="bam2geneprofile.py - build meta-gene profile for a set of transcripts/genes"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>