
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>gtf2fasta.py - annotate genomic bases from a gene set &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="snp2table.py - annotate variants" href="snp2table.html" />
    <link rel="prev" title="gtfs2graph.py - compute overlap graph between genes" href="gtfs2graph.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="snp2table.html" title="snp2table.py - annotate variants"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="gtfs2graph.html" title="gtfs2graph.py - compute overlap graph between genes"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-gtf2fasta"></span><div class="section" id="gtf2fasta-py-annotate-genomic-bases-from-a-gene-set">
<h1>gtf2fasta.py - annotate genomic bases from a gene set<a class="headerlink" href="#gtf2fasta-py-annotate-genomic-bases-from-a-gene-set" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id: gtf2fasta.py 2861 2010-02-23 17:36:32Z andreas $</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Genomics Genesets Sequences GTF FASTA Transformation</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script takes a <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file from ENSEMBL and annotates each base 
in the genome according to its <em>function</em>. The script multiplexes both strands with lower-
case characters referring to the forward strand and upper-case characters
referring to the reverse strand.</p>
<p>The codes and their meaning are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="93%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>code</td>
<td>description</td>
</tr>
<tr class="row-even"><td>a</td>
<td>first codon position within a complete codon</td>
</tr>
<tr class="row-odd"><td>b</td>
<td>second codon position within a complete codon</td>
</tr>
<tr class="row-even"><td>c</td>
<td>third codon position within a complete codon</td>
</tr>
<tr class="row-odd"><td>d</td>
<td>coding base, but in multiple frames or strands</td>
</tr>
<tr class="row-even"><td>e</td>
<td>non-coding base in exon</td>
</tr>
<tr class="row-odd"><td>f</td>
<td>frame-shifted base</td>
</tr>
<tr class="row-even"><td>g</td>
<td>intergenic base</td>
</tr>
<tr class="row-odd"><td>i</td>
<td>intronic base</td>
</tr>
<tr class="row-even"><td>l</td>
<td>base in other RNA</td>
</tr>
<tr class="row-odd"><td>m</td>
<td>base in miRNA</td>
</tr>
<tr class="row-even"><td>n</td>
<td>base in snRNA</td>
</tr>
<tr class="row-odd"><td>o</td>
<td>base in snoRNA</td>
</tr>
<tr class="row-even"><td>r</td>
<td>base in rRNA (both genomic and mitochondrial)</td>
</tr>
<tr class="row-odd"><td>p</td>
<td>base in pseudogene (including transcribed, unprocessed and processed)</td>
</tr>
<tr class="row-even"><td>q</td>
<td>base in retrotransposon</td>
</tr>
<tr class="row-odd"><td>s</td>
<td>base within a splice signal (GT/AG)</td>
</tr>
<tr class="row-even"><td>t</td>
<td>base in tRNA (both genomic and mitochondrial)</td>
</tr>
<tr class="row-odd"><td>u</td>
<td>base in 5&#8217; UTR</td>
</tr>
<tr class="row-even"><td>v</td>
<td>base in 3&#8217; UTR</td>
</tr>
<tr class="row-odd"><td>x</td>
<td>ambiguous base with multiple functions.</td>
</tr>
<tr class="row-even"><td>y</td>
<td>unknown base</td>
</tr>
</tbody>
</table>
<p>This script can be used for a quick-and-dirty annotation of variants in a genome.
For a better prediction of variant effects in coding sequences, see <a class="reference internal" href="gtf2alleles.html"><em>gtf2alleles.py - predict effects of variants on gene set</em></a>.</p>
<div class="section" id="output-files">
<h3>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h3>
<p>The annotated genome is output on stdout.</p>
<p>The script creates the following additional output files:</p>
<dl class="docutils">
<dt>counts</dt>
<dd>Counts for each annotations</dd>
<dt>junctions</dt>
<dd><p class="first">Splice junctions. This is a tab separated table linking residues that are
joined via features. The coordinates are forward/reverse coordinates.</p>
<p>The columns are:</p>
<dl class="last docutils">
<dt>contig</dt>
<dd>the contig</dd>
<dt>strand</dt>
<dd>direction of linkage</dd>
<dt>end</dt>
<dd>last base of exon in direction of strand</dd>
<dt>start</dt>
<dd>first base of exon in direction of strand</dd>
<dt>frame</dt>
<dd>frame base at second coordinate (for coding sequences)</dd>
</dl>
</dd>
</dl>
</div>
</div>
<div class="section" id="known-problems">
<h2>Known problems<a class="headerlink" href="#known-problems" title="Permalink to this headline">¶</a></h2>
<p>The stop-codon is part of the UTR. This has the following effects:</p>
<blockquote>
<div><ul class="simple">
<li>On the mitochondrial chromosome, the stop-codon might be used for
ncRNA transcripts and thus the base is recorded as ambiguous.</li>
<li>On the mitochondrial chromosome, alternative transcripts might
read through a stop-codon (RNA editing). The codon itself will be
recorded as ambiguous.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>For example:</p>
<div class="highlight-python"><div class="highlight"><pre>zcat hg19.gtf.gz | python gtf2fasta.py --genome-file=hg19
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python gtf2fasta.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: gtf2fasta.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -g GENOME_FILE, --genome-file=GENOME_FILE
                        filename with genome [default=none].
  -i, --ignore-missing  Ignore transcripts on contigs that are not in the
                        genome-file [default=False].
  --min-intron-length=MIN_INTRON_LENGTH
                        minimum intron length. If the distance between two
                        consecutive exons is smaller, the region will be
                        marked &#39;unknown&#39; [default=30].
  -m METHOD, --method=METHOD
                        method to apply [default=full].
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -P OUTPUT_FILENAME_PATTERN, --output-filename-pattern=OUTPUT_FILENAME_PATTERN
                        OUTPUT filename pattern for various methods [%s].
  -F, --force           force over-writing of existing files.
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">gtf2fasta.py - annotate genomic bases from a gene set</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a><ul>
<li><a class="reference internal" href="#output-files">Output files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#known-problems">Known problems</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="gtfs2graph.html"
                        title="previous chapter">gtfs2graph.py - compute overlap graph between genes</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="snp2table.html"
                        title="next chapter">snp2table.py - annotate variants</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/gtf2fasta.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="snp2table.html" title="snp2table.py - annotate variants"
             >next</a> |</li>
        <li class="right" >
          <a href="gtfs2graph.html" title="gtfs2graph.py - compute overlap graph between genes"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>