
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>gtf2table.py - annotate genes/transcripts &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="gtfs2graph.py - compute overlap graph between genes" href="gtfs2graph.html" />
    <link rel="prev" title="gtf2reads.py - sample reads from genes" href="gtf2reads.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="gtfs2graph.html" title="gtfs2graph.py - compute overlap graph between genes"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="gtf2reads.html" title="gtf2reads.py - sample reads from genes"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-gtf2table"></span><div class="section" id="gtf2table-py-annotate-genes-transcripts">
<h1>gtf2table.py - annotate genes/transcripts<a class="headerlink" href="#gtf2table-py-annotate-genes-transcripts" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Genomics Genesets GTF Annotation</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>compute properties of genes or transcripts given in a <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a>
formatted file and output them in tabular format.</p>
<p>Quantities can be either computed per gene (all exons across all transcripts)
or per transcript. The input needs to be sorted accordingly.</p>
<p>The following methods (see option <tt class="docutils literal"><span class="pre">--counter</span></tt>) are available:</p>
<dl class="docutils">
<dt>bigwig-counts</dt>
<dd>collect density values from bigwig file and output
summary statistics and percentage of bases covered (<tt class="docutils literal"><span class="pre">pcovered</span></tt>)
by value in bigwig file.</dd>
<dt>binding-pattern</dt>
<dd>given a list of intervals, determine the binding
pattern within and surrounding the gene. For each gene, intervals
overlapping the CDS, introns, UTRs and the flank are collected and
recorded. The binding is summarized with a binding pattern, a
binary pattern indicating overlap/no overlap with 5&#8217; flank, 5&#8217; UTR,
CDS, Introns, 3&#8217; UTR, 3&#8217; flank.</dd>
<dt>classifier</dt>
<dd>classify transcripts according to genomic annotation
Requires a <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> file with genomic annotations
(see <a class="reference internal" href="gtf2gff.html"><em>gtf2gff.py - convert a transcript set to genomic features</em></a>.)</dd>
<dt>classifier-rnaseq</dt>
<dd>classify rnaseq transcripts with respect to a reference
geneset. Requires a <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> file with a reference
gene set.</dd>
<dt>classifier-polii</dt>
<dd>classify according to PolII transcripts. A
gene/transcript is transcribed, if it is covered by large PolII
intervals over 80% of its length. A gene/transript is primed if its
promotor/UTR is covered by 50% of its length, while the rest of the
gene body isn&#8217;t.</dd>
<dt>composition-na</dt>
<dd>output nucleotide composition of gene</dd>
<dt>composition-cgp</dt>
<dd>output cpg composition of gene</dd>
<dt>coverage</dt>
<dd>compute nucleotide coverage of input with segments in another file.
The values are output in 5&#8217; to 3&#8217; order for each nucleotide. Requires
a second <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file with features to cover.</dd>
<dt>distance</dt>
<dd>compute distance of genes to features in a second file. Requires
a second <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file with transcripts. The strand
information of the features is ignored.</dd>
<dt>distance-genes</dt>
<dd>compute distance of genes to genes in a second file. Requires a
second <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file with genes. The counter distinguishes
a variety of cases (closest upstream/downstream).</dd>
<dt>distance-tss</dt>
<dd>compute distance of genes to transcription start sites. Requires a
second <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file with genes.</dd>
<dt>length</dt>
<dd>output exon length summary of gene.</dd>
<dt>neighbours</dt>
<dd>output features in second stream that are in proximity to genes
in input. Requires a <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file with genes.</dd>
<dt>overlap</dt>
<dd>compute overlap of genes in input with features in second stream.
Requires a <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file with gene territories.</dd>
<dt>overlap-stranded</dt>
<dd>count overlap with genomic features in second another file. Outputs
the number of overlapping exons. Records the direction of overlap
(sense/antisense). Requires a <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file with
features.</dd>
<dt>overlap-transcripts</dt>
<dd>count overlap of genes with transcripts in another set.
Requires a <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> formatted file.</dd>
<dt>overrun</dt>
<dd>output intron overrun, exons in the input gene set extending
into the introns of a reference gene set. Requries a <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a>
formatted file with a reference gene set.</dd>
<dt>position</dt>
<dd>output genomic coordinates of gene</dd>
<dt>proximity</dt>
<dd>report summary stats (lengths,values) of features in proximity to
genes input gene set. Requires a <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file with
genomic features.</dd>
<dt>proximity-exclusive</dt>
<dd>as proximity, but exclude any ranges overlapping the gene set.</dd>
<dt>proximity-lengthmatched</dt>
<dd>as proximity-exclusive, but length-match features with genes.</dd>
<dt>quality</dt>
<dd>output base-quality information summary of gene. Needs quality scores.</dd>
<dt>read-coverage</dt>
<dd>output read coverage summary statistics of gene</dd>
</dl>
<p>read-extension</p>
<dl class="docutils">
<dt>read-counts</dt>
<dd>count number of reads overlapping a gene or transcript. Counts
uniquely by read name and counts duplicate and non-duplicate
(reads) separately.</dd>
<dt>readpair-counts</dt>
<dd>count number of read pairs overlapping a gene or
transcript. Summarizes the output of readpair-fullcounts.</dd>
<dt>readpair-fullcounts</dt>
<dd>count number of read pairs overlapping a gene or transcript.
Pairs are counted according to a variety of attributes
(exonic overlap/read pair status/splice status/...).</dd>
<dt>splice</dt>
<dd>output splicing summary of gene</dd>
</dl>
<p>splice-comparison</p>
<p>territories</p>
<div class="section" id="read-counting">
<h3>Read counting<a class="headerlink" href="#read-counting" title="Permalink to this headline">¶</a></h3>
<p>The methods <tt class="docutils literal"><span class="pre">read-counts</span></tt> and <tt class="docutils literal"><span class="pre">readpair-counts</span></tt> count the number
of reads inside a <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file that map inside a
transcript or gene.</p>
<p>These counters proceed on a per-gene or per-transcript basis depending
on the <tt class="docutils literal"><span class="pre">--reporter</span></tt> option. All reads overlapping the exons or
introns of a transcript are collected and counted either individually
(<tt class="docutils literal"><span class="pre">read-counts</span></tt>) or as pairs (<tt class="docutils literal"><span class="pre">readpair-counts</span></tt>).</p>
<p>For paired-read counting, each pair is classified and counted
according to the following for axes:</p>
<ol class="arabic simple">
<li>Pair status: proper pair, unmapped read in pair, ...</li>
<li>Direction: for stranded protocols, sense, antisense, ...</li>
<li>Overlap status: Pair overlaps exons only, introns only, ...</li>
<li>Splice status: Reads are unspliced, spliced consistently with
transcript model, or inconsistently with transcript model</li>
</ol>
<p>Counts are then collated into a smaller set of summaries for convenience:</p>
<table border="1" class="docutils">
<colgroup>
<col width="28%" />
<col width="72%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Column</td>
<td>Number of read pairs</td>
</tr>
<tr class="row-even"><td>counted_all</td>
<td>considered to be correct: they are sense, overlap
with exons fully.</td>
</tr>
<tr class="row-odd"><td>counted_spliced</td>
<td>in sense direction and exonic, spliced and spliced
correctly.</td>
</tr>
<tr class="row-even"><td>counted_unspliced</td>
<td>in sense direction and exonic, not spliced.</td>
</tr>
<tr class="row-odd"><td>sense_intronic</td>
<td>intronic and sense direction.</td>
</tr>
<tr class="row-even"><td>sense_inconsistent</td>
<td>in sense direction, but overlap both introns and
exons or extend beyond the transcript modelse</td>
</tr>
<tr class="row-odd"><td>sense_other</td>
<td>in sense direction, but not counted, intronic or
inconsistent.</td>
</tr>
<tr class="row-even"><td>antisense</td>
<td>in antisense direction.</td>
</tr>
<tr class="row-odd"><td>nonsense</td>
<td>in unexpected orientation</td>
</tr>
<tr class="row-even"><td>notproper</td>
<td>that are not proper pairs</td>
</tr>
<tr class="row-odd"><td>quality_pairs</td>
<td>affected by a the removal of a low-quality read</td>
</tr>
<tr class="row-even"><td>quality_reads</td>
<td>Number of low quality reads removed</td>
</tr>
<tr class="row-odd"><td>total</td>
<td>Total number of pairs considered</td>
</tr>
</tbody>
</table>
<p>The Counter <tt class="docutils literal"><span class="pre">readpair-fullcounts</span></tt> provides the detailed counts
for each transcript model according to the four axes. The output
can be used to implement custom counting schemes.</p>
<p>Reads below a minimum quality score will be ignored
(<tt class="docutils literal"><span class="pre">--min-read-quality</span></tt>). By default, all reads will be counted.</p>
<p>Reads mapping to multiple locations will be downweighted if
the <tt class="docutils literal"><span class="pre">--weight-multi-mapping</span></tt> option is set. This requires
the presence of the <tt class="docutils literal"><span class="pre">NH</span></tt> flag in the <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> file.</p>
<p>For paired read counting, the library type can be specified with the
<tt class="docutils literal"><span class="pre">--library-type</span></tt> option to make use of strand information. Library
types are labelled according to the <a href="#id1"><span class="problematic" id="id2">tophat_</span></a> and <a href="#id3"><span class="problematic" id="id4">cufflinks_</span></a>
convention. A summary is <a class="reference external" href="http://www.nature.com/nprot/journal/v7/n3/fig_tab/nprot.2012.016_T1.html">here</a>.</p>
<table border="1" class="docutils">
<colgroup>
<col width="29%" />
<col width="29%" />
<col width="43%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Library type</td>
<td>RNA-seq protocol</td>
<td>Explanation</td>
</tr>
<tr class="row-even"><td><dl class="first last docutils">
<dt>fr-unstranded</dt>
<dd>(default)</dd>
</dl>
</td>
<td>Illumina TruSeq</td>
<td>Reads from the leftmost end of
the fragment (in transcript
coordinates) map to the
transcript strand, and the
rightmost end maps to the
opposite strand left</td>
</tr>
<tr class="row-odd"><td>fr-firststrand</td>
<td>dUTP, NSR, NNSR39</td>
<td>Same as above except we
enforce the rule that the
rightmost end of the fragment
(in transcript coordinates) is
the first sequenced (or only
sequenced for single-end
reads). Equivalently, it is
assumed that only the strand
generated during first strand
synthesis is sequenced</td>
</tr>
<tr class="row-even"><td>fr-secondstrand</td>
<td>Directional Illumina
(Ligation)
Standard SOLiD</td>
<td>Same as above
except TopHat/Cufflinks
enforce the rule that the
leftmost end of the fragment
(in transcript coordinates) is
the first sequenced (or only
sequenced for single-end
reads). Equivalently, it is
assumed that only the strand
generated during second strand
synthesis is sequenced</td>
</tr>
</tbody>
</table>
<p>For unstranded protocols, all reads and pairs are considered to matching
in the sense direction.</p>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>python gtf2table.py --counter=length &lt; geneset.gtf &gt; geneset.tsv
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python gtf2table.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: gtf2table.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -g GENOME_FILE, --genome-file=GENOME_FILE
                        filename with genome [default=none].
  -q QUALITY_FILE, --quality-file=QUALITY_FILE
                        filename with genomic base quality information
                        [default=none].
  -b bam, --bam-file=bam
                        filename with read mapping information. Multiple files
                        can be submitted in a comma-separated list
                        [default=none].
  -i bigwig, --bigwig-file=bigwig
                        filename with bigwig information [default=none].
  -f bed, --filename-gff=bed
                        filename with extra gff files. The order is important
                        [default=[]].
  --filename-format=FILENAME_FORMAT
                        format of secondary stream [default=none].
  --gff-source=GFF_SOURCES
                        restrict input to this &#39;source&#39; in extra gff file (for
                        counter: overlap) [default=[]].
  --gff-feature=GFF_FEATURES
                        restrict input to this &#39;feature&#39; in extra gff file
                        (for counter: overlap) [default=[]].
  -r REPORTER, --reporter=REPORTER
                        report results for &#39;genes&#39; or &#39;transcripts&#39;
                        [default=genes].
  -s SECTIONS, --section=SECTIONS
                        select range on which counters will operate
                        [default=[]].
  -c COUNTERS, --counter=COUNTERS
                        select counters to apply to input [default=[]].
  --add-gtf-source      add gtf field of source to output [default=False].
  --proximal-distance=PROXIMAL_DISTANCE
                        distance to be considered proximal to an interval
                        [default=10000].
  --weight-multi-mapping
                        weight multi-mapping reads in bam-files. Requires the
                        NH flag to be set by the mapper [default=False].
  --prefix=PREFIXES     add prefix to column headers - prefixes are used in
                        the same order as the counters [default=[]].
  --library-type=LIBRARY_TYPE
                        library type of reads in bam file. [default=fr-
                        unstranded]
  --min-read-quality=MINIMUM_READ_QUALITY
                        minimum read quality. Reads with a quality score of
                        less will be ignored. [default=0]
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -P OUTPUT_FILENAME_PATTERN, --output-filename-pattern=OUTPUT_FILENAME_PATTERN
                        OUTPUT filename pattern for various methods [%s].
  -F, --force           force over-writing of existing files.
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">gtf2table.py - annotate genes/transcripts</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a><ul>
<li><a class="reference internal" href="#read-counting">Read counting</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="gtf2reads.html"
                        title="previous chapter">gtf2reads.py - sample reads from genes</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="gtfs2graph.html"
                        title="next chapter">gtfs2graph.py - compute overlap graph between genes</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/gtf2table.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="gtfs2graph.html" title="gtfs2graph.py - compute overlap graph between genes"
             >next</a> |</li>
        <li class="right" >
          <a href="gtf2reads.html" title="gtf2reads.py - sample reads from genes"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>