
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>gtf2gtf.py - manipulate transcript models &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="gtf2reads.py - sample reads from genes" href="gtf2reads.html" />
    <link rel="prev" title="gtf2gff.py - convert a transcript set to genomic features" href="gtf2gff.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="gtf2reads.html" title="gtf2reads.py - sample reads from genes"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="gtf2gff.html" title="gtf2gff.py - convert a transcript set to genomic features"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-gtf2gtf"></span><div class="section" id="gtf2gtf-py-manipulate-transcript-models">
<h1>gtf2gtf.py - manipulate transcript models<a class="headerlink" href="#gtf2gtf-py-manipulate-transcript-models" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Genomics Genesets GTF Manipulation</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script reads a gene set in <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> format from stdin, applies some
transformation, and outputs a new gene set in <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> format to stdout.</p>
</div>
<div class="section" id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<p>Options available for use in this script can broadly be classified into four
categories:</p>
<ol class="arabic simple">
<li>sorting gene sets</li>
<li>manipulating gene models</li>
<li>filtering gene sets</li>
<li>setting/resetting fields within a gtf file</li>
</ol>
<p>Further options for working with gtf files are available in gff2gff.py,
which can be run with the specification &#8211;is-gtf</p>
<div class="section" id="sort-gene-sets">
<h3>Sort gene sets<a class="headerlink" href="#sort-gene-sets" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt><tt class="docutils literal"><span class="pre">--sort</span></tt></dt>
<dd><p class="first">Sorts entries in gtf file by one or more fields</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="28%" />
<col width="72%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td rowspan="2"><blockquote class="first">
<div>option</div></blockquote>
<blockquote class="last">
<div>gene</div></blockquote>
</td>
<td rowspan="2"><blockquote class="first">
<div>order in which fields are sorted</div></blockquote>
<blockquote class="last">
<div>gene_id, transcript_id, contig, start</div></blockquote>
</td>
</tr>
<tr class="row-even"></tr>
<tr class="row-odd"><td>contig+gene</td>
<td>contig, gene_id, transcript_id, start</td>
</tr>
<tr class="row-even"><td>transcript</td>
<td>transcript_id, contig, start</td>
</tr>
<tr class="row-odd"><td>position</td>
<td>contig, start</td>
</tr>
<tr class="row-even"><td>position+gene</td>
<td>contig( gene_id, start )</td>
</tr>
<tr class="row-odd"><td>gene+position</td>
<td>gene_id, contig, start</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p class="last">N.B. position+gene sorts by gene_id, start, then subsequently sorts
flattened gene lists by contig, start</p>
</dd>
</dl>
</div>
<div class="section" id="manipulate-gene-models">
<h3>Manipulate gene-models<a class="headerlink" href="#manipulate-gene-models" title="Permalink to this headline">¶</a></h3>
<p>Options that can be used to alter the features represented in a <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a>
file. For further detail see command line options.</p>
<p>Input gtfs need to be sorted so that features for a gene or transcript
appear consecutively within the file. This can be achevied using <tt class="docutils literal"><span class="pre">--sort</span></tt>.</p>
<dl class="docutils">
<dt><tt class="docutils literal"><span class="pre">--merge-exons</span></tt></dt>
<dd>Merges overlapping exons for all transcripts of a gene, outputting the
merged exons. Can be used in conjunction with <tt class="docutils literal"><span class="pre">--merge-exons-distance</span></tt>
to set the minimum distance that may appear between two exons before
they are merged.If <tt class="docutils literal"><span class="pre">--with-utr</span></tt> is set, the output interval will also
contain UTR.</dd>
<dt><tt class="docutils literal"><span class="pre">--merge-transcripts</span></tt></dt>
<dd>Merges all transcripts of a gene. Outputs contains a single interval that
spans the original gene (both introns and exons). If <tt class="docutils literal"><span class="pre">--with-utr</span></tt> is
set, the output interval will also contain UTR.</dd>
<dt><tt class="docutils literal"><span class="pre">--merge-genes</span></tt></dt>
<dd>Merges genes that have overlapping exons, outputting a single gene_id and
transcript_id for all exons of overlapping genes.
(Does not merge intervals on different strands)</dd>
<dt><tt class="docutils literal"><span class="pre">--join-exons</span></tt></dt>
<dd>Joins together all exons of a transcript, outputting a single interval that
spans the original transcript (both introns and exons).</dd>
<dt><tt class="docutils literal"><span class="pre">--intersect-transcripts</span></tt></dt>
<dd>Finds regions representing the intersect of all transcripts of a gene.
Output will contain intervals spanning only those bases covered by all
transcripts. If <tt class="docutils literal"><span class="pre">--with-utr</span></tt> is set, the UTR will also be included in the
intersect.</dd>
<dt><tt class="docutils literal"><span class="pre">--merge-introns</span></tt></dt>
<dd>Merges the region spanned by introns for all transcripts of a
gene.  Outputs a single interval that spans the region between the
start and end of the first and last intron, respectively.</dd>
<dt><tt class="docutils literal"><span class="pre">--exons2introns</span></tt></dt>
<dd>Merges overlapping introns for all transcripts of a gene,
outputting the merged introns. Use <tt class="docutils literal"><span class="pre">--intron-min-length</span></tt> to
ignore merged introns below a specified length. Use
<tt class="docutils literal"><span class="pre">--intron-border</span></tt> to specify a number of residues to remove at
either end of output introns (residues are removed prior to
filtering on size when used in conjunction with
<tt class="docutils literal"><span class="pre">--intron-min-length</span></tt>).</dd>
<dt><tt class="docutils literal"><span class="pre">--transcripts2genes</span></tt></dt>
<dd>Currently it doesn&#8217;t work...
pysam.TabProxies.GTFProxy has no attribute &#8216;gene_id&#8217;
May be used in conjunction with <tt class="docutils literal"><span class="pre">--reset-strand</span></tt></dd>
</dl>
<p>The option <tt class="docutils literal"><span class="pre">--permit-duplicates</span></tt> may be specified in order to
allow gene-ids to be duplicated within the input <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> file
(i.e. for the same gene-id to appear non-consecutively within the
input file). However, this option currently only works for
<tt class="docutils literal"><span class="pre">--merge-exons</span></tt>, <tt class="docutils literal"><span class="pre">--merge-transcripts</span></tt>, <tt class="docutils literal"><span class="pre">--merge-introns</span></tt>, and
<tt class="docutils literal"><span class="pre">--intersect-transcripts</span></tt>. It DOES NOT work for <tt class="docutils literal"><span class="pre">--merge-genes</span></tt>,
<tt class="docutils literal"><span class="pre">--join-exons</span></tt>, or <tt class="docutils literal"><span class="pre">--exons2introns</span></tt>.</p>
</div>
<div class="section" id="filter-gene-sets">
<h3>Filter gene sets<a class="headerlink" href="#filter-gene-sets" title="Permalink to this headline">¶</a></h3>
<p>Options that can be used to filter <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> files. For further
detail see command line options.</p>
<dl class="docutils">
<dt>Input gtfs need to be sorted so that features for a gene or transcript</dt>
<dd>appear consecutively within the file. This can be achevied using <tt class="docutils literal"><span class="pre">--sort</span></tt>.</dd>
<dt><tt class="docutils literal"><span class="pre">--filter</span></tt></dt>
<dd><p class="first">When filtering on the basis of &#8216;gene-id&#8217; or &#8216;transcript-id&#8217; a filename
containing ids to be removed may provided using <tt class="docutils literal"><span class="pre">--apply</span></tt>. Alternatively,
a random subsample of genes/transcripts may be retained using
<tt class="docutils literal"><span class="pre">--sample-size</span></tt>. Use <tt class="docutils literal"><span class="pre">--min-exons-length</span></tt> in conjunction with
<tt class="docutils literal"><span class="pre">--sample-size</span></tt> to specify a minimum length for genes/transcripts to be
retained. Use <tt class="docutils literal"><span class="pre">--reset-strand</span></tt> to set strand to &#8216;.&#8217; in output.</p>
<p>Other filter options include longest-gene, longest-transcript,
or representative-transcript.</p>
<p class="last">When filtering on the basis of gene-id, transcript-id or longest-gene,
<tt class="docutils literal"><span class="pre">--invert-filter</span></tt> may be used to invert the selection.</p>
</dd>
<dt><tt class="docutils literal"><span class="pre">--remove-overlapping</span></tt></dt>
<dd>Takes as argument a <a class="reference internal" href="../glossary.html#term-gff"><em class="xref std std-term">gff</em></a> formatted file. Any transcripts that
intersect intervals in the supplied file are removed.
(Does not account for strand.)</dd>
<dt><tt class="docutils literal"><span class="pre">--remove-duplicates</span></tt></dt>
<dd>Setting to &#8216;gene&#8217;, &#8216;transcript&#8217;, or &#8216;coordinates&#8217; will remove any
interval for which non-consecutive occurrances of specified term
appear in input <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> file.  Setting to &#8216;ucsc&#8217;, will remove
any interval for which transcript-id contains &#8216;_dup&#8217;.</dd>
</dl>
</div>
<div class="section" id="set-reset-fields">
<h3>Set/reset fields<a class="headerlink" href="#set-reset-fields" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt>Options for altering fields within <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a>. For further details see command</dt>
<dd>line options.</dd>
<dt><tt class="docutils literal"><span class="pre">--rename</span></tt></dt>
<dd>When a mapping file is provided using <tt class="docutils literal"><span class="pre">--apply</span></tt>, renames either
gene-id or transcript-id. Outputs a <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> file with field
renamed. Any entry in input <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> not appearing in mapping
file is discarded.</dd>
<dt><tt class="docutils literal"><span class="pre">--add-protein-id</span></tt></dt>
<dd>Takes as argument a file that maps transcript-id to protein-id and
appends the protein-id provided to the attributes field.  Any
entry ininput <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> not appearing in mapping file is
discarded.</dd>
<dt><tt class="docutils literal"><span class="pre">--renumber-genes</span></tt></dt>
<dd>Renumber genes from 1 using the pattern provided, e.g. ENSG%s</dd>
<dt><tt class="docutils literal"><span class="pre">--renumber-transcripts</span></tt></dt>
<dd>Renumber transcripts from 1 using the pattern provided, e.g. ENST%s</dd>
<dt><tt class="docutils literal"><span class="pre">--unset-genes</span></tt></dt>
<dd>Renumber genes from 1 using the pattern provided,
e.g. ENSG%s. Transcripts with the same gene-id in input
<a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> file will have different gene-ids in the output
<a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> file.</dd>
<dt><tt class="docutils literal"><span class="pre">--set-transcript-to-gene</span></tt></dt>
<dd>Will reset the transcript-id to the gene-id</dd>
<dt><tt class="docutils literal"><span class="pre">--set-gene-to-transcript</span></tt></dt>
<dd>Will reset the gene-id to the transcript-id</dd>
<dt><tt class="docutils literal"><span class="pre">--set-protein-to-transcript</span></tt></dt>
<dd>Will append transcript to attributes field as &#8216;protein_id&#8217;</dd>
<dt><tt class="docutils literal"><span class="pre">--set-score-to-distance</span></tt></dt>
<dd>Will reset the score field (field 6) of each feature in input
<a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> to be the distance from transcription start site to
the start of the feature.  (Assumes input file is sorted by
transcript-id)</dd>
<dt><tt class="docutils literal"><span class="pre">--rename-duplicates</span></tt></dt>
<dd>Rename duplicate gene_ids and transcript_ids by addition of
numerical suffix</dd>
</dl>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>python gtf2gtf.py             --sort=gene     | python gtf2gtf.py                --intersect-transcripts                --with-utr                --renumber-transcripts=MERGED%s
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python gtf2gtf.py --help
</pre></div>
</div>
</div>
<div class="section" id="command-line-options">
<h2>Command line Options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: gtf2gtf.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -m, --merge-exons     merge overlapping exons of all transcripts within a
                        gene. The merged exons will be output. Input needs to
                        sorted by gene [default=False].
  -t, --merge-transcripts
                        merge all transcripts within a gene. The entry will
                        span the whole gene (exons and introns). The
                        transcript does not include the UTR unless --with-utr
                        is set. [default=False].
  --merge-genes         merge overlapping genes if their exons overlap. A gene
                        with a single transcript containing all exons of the
                        overlapping transcripts will be output. This operation
                        ignores strand information The input needs te sorted
                        by transcript [default=False].
  --merge-exons-distance=MERGE_EXONS_DISTANCE
                        distance in nucleotides between exons to be merged
                        [default=0].
  -j, --join-exons      join all exons per transcript. A new transcript will
                        be output that spans a whole transcript. Input needs
                        to be sorted by transcript [default=False].
  --unset-genes=UNSET_GENES
                        unset gene identifiers, keeping transcripts intact.
                        New gene identifiers are set to the pattern given. For
                        example, &#39;--unset-genes=%06i&#39; [default=none].
  --sort=SORT           sort input data [default=none].
  -u, --with-utr        include utr in merged transcripts [default=False].
  --intersect-transcripts
                        intersect all transcripts within a gene. The entry
                        will only span those bases that are covered by all
                        transcrips.The transcript does not include the UTR
                        unless --with-utr is set. This method will remove all
                        other features (stop_codon, etc.) The input needs to
                        be sorted by gene. [default=False].
  -i, --merge-introns   merge and output all introns within a gene. The output
                        will contain all intronic regions within a gene.
                        Single exon genes are skipped. The input needs to be
                        sorted by gene. [default=none].
  -g, --set-transcript-to-gene, --set-transcript2gene
                        set the transcript_id to the gene_id [default=False].
  --set-protein-to-transcript
                        set the protein_id to the transcript_id
                        [default=False].
  --add-protein-id=ADD_PROTEIN_ID
                        add a protein_id for each transcript_id. The argument
                        is a filename containing a mapping between
                        transcript_id to protein_id [default=none].
  -G, --set-gene-to-transcript, --set-gene2transcript
                        set the gene_id to the transcript_id [default=False].
  -d, --set-score2distance
                        set the score field for each feature to the distance
                        to transcription start site [default=False].
  --exons2introns       for each gene build an &#39;intronic&#39; transcript
                        containing the union of all intronic regions of all
                        transcripts in a gene.The features are labeled as
                        &#39;intron&#39;.The input needs to be sorted by gene.
                        [default=none].
  -f FILTER, --filter=FILTER
                        apply a filter to the input file. Available filters
                        are: &#39;gene&#39;: filter by gene_id, &#39;transcript&#39;: filter
                        by transcript_id, &#39;longest-gene&#39;: output the longest
                        gene for overlapping genes ,&#39;longest-transcript&#39;:
                        output the longest transcript per gene
                        ,&#39;representative-transcript&#39;: output the
                        representative transcript per gene. The representative
                        transcript is the transcript that shares most exons
                        with the other transcripts in a gene. The input needs
                        to be sorted by gene. [default=none].
  -r RENAME, --rename=RENAME
                        rename genes or transcripts with a map given by the
                        option `--apply`. Those that can not be renamed are
                        removed [default=none].
  --renumber-genes=RENUMBER_GENES
                        renumber genes according to the given pattern.
                        [default=none].
  --renumber-transcripts=RENUMBER_TRANSCRIPTS
                        renumber transcripts according to the given pattern.
                        [default=none].
  -a tsv, --apply=tsv   filename of ids to map/filter [default=none].
  --invert-filter       when using --filter, invert selection (like grep -v).
                        [default=False].
  --sample-size=SAMPLE_SIZE
                        extract a random sample of size # if the option &#39;--
                        filter&#39; is set[default=0].
  --intron-min-length=INTRON_MIN_LENGTH
                        minimum length for introns (for --exons2introns)
                        [default=none].
  --min-exons-length=MIN_EXONS_LENGTH
                        minimum length for gene (sum of exons) (--sample-size)
                        [default=0].
  --intron-border=INTRON_BORDER
                        number of residues to exclude at intron at either end
                        (--exons2introns) [default=none].
  --transcripts2genes   cluster overlapping transcripts into genes.
  --reset-strand        remove strandedness of features (set to &#39;.&#39;) when
                        using --transcripts2genes[default=False].
  --remove-overlapping=gff
                        remove all transcripts that overlap intervals in a
                        gff-formatted file.The comparison ignores strand
                        [default=none].
  --permit-duplicates   permit duplicate genes. [default=True]
  --remove-duplicates=REMOVE_DUPLICATES
                        remove duplicates by gene/transcript. If ``ucsc`` is
                        chosen, transcripts ending on _dup# are removed. This
                        is necessary to remove duplicate entries that are next
                        to each other in the sort order [none]
  --rename-duplicates   rename duplicate gene_ids and transcript_ids by
                        addition of a numerical suffix
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">gtf2gtf.py - manipulate transcript models</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#options">Options</a><ul>
<li><a class="reference internal" href="#sort-gene-sets">Sort gene sets</a></li>
<li><a class="reference internal" href="#manipulate-gene-models">Manipulate gene-models</a></li>
<li><a class="reference internal" href="#filter-gene-sets">Filter gene sets</a></li>
<li><a class="reference internal" href="#set-reset-fields">Set/reset fields</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line Options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="gtf2gff.html"
                        title="previous chapter">gtf2gff.py - convert a transcript set to genomic features</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="gtf2reads.html"
                        title="next chapter">gtf2reads.py - sample reads from genes</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/gtf2gtf.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="gtf2reads.html" title="gtf2reads.py - sample reads from genes"
             >next</a> |</li>
        <li class="right" >
          <a href="gtf2gff.html" title="gtf2gff.py - convert a transcript set to genomic features"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>