
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>index_fasta.py - Index fasta formatted files &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="mali2kaks.py - rate analysis with PAML" href="mali2kaks.html" />
    <link rel="prev" title="mask_fasta.py - mask fasta formatted sequences" href="mask_fasta.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="mali2kaks.html" title="mali2kaks.py - rate analysis with PAML"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="mask_fasta.html" title="mask_fasta.py - mask fasta formatted sequences"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-index_fasta"></span><div class="section" id="index-fasta-py-index-fasta-formatted-files">
<h1>index_fasta.py - Index fasta formatted files<a class="headerlink" href="#index-fasta-py-index-fasta-formatted-files" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Genomics Sequences FASTA Manipulation</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script indexes one or more <a class="reference internal" href="../glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> formatted files into a 
database that can be used by other scripts in the CGAT code collection
and <a class="reference internal" href="../modules/IndexedFasta.html#module-IndexedFasta" title="IndexedFasta"><tt class="xref py py-mod docutils literal"><span class="pre">IndexedFasta</span></tt></a> for quick acces to sequence fragments.</p>
<p>By default, the database is itself a <a class="reference internal" href="../glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> formatted file in which
all line breaks and other white space characters have been removed.
Compression methods are available to conserve disk space, though they do 
come at a performance penalty.</p>
<p>See also <a class="reference external" href="http://pypi.python.org/pypi/pyfasta">http://pypi.python.org/pypi/pyfasta</a> for another implementation. 
Samtools provides similar functionality with the <tt class="docutils literal"><span class="pre">samtools</span> <span class="pre">faidx</span></tt> command.</p>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>Index a collection of fasta files:</p>
<div class="highlight-python"><div class="highlight"><pre>python index_fasta.py oa_ornAna1_softmasked ornAna1.fa.gz &gt; oa_ornAna1_softmasked.log
</pre></div>
</div>
<p>To retrieve a segment:</p>
<div class="highlight-python"><div class="highlight"><pre>python index_fasta.py --extract=chr5:1000:2000 oa_ornAna1_softmasked 
</pre></div>
</div>
<p>Indexing from a tar file:</p>
<blockquote>
<div>python index_fasta.py oa_ornAna1_softmasked ornAna1.tar.gz &gt; oa_ornAna1_softmasked.log</div></blockquote>
<p>Indexing from stdin:</p>
<blockquote>
<div>zcat ornAna1.fa.gz | python index_fasta.py oa_ornAna1_softmasked - &gt; oa_ornAna1_softmasked.log</div></blockquote>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python index_fasta.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: index_fasta.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -e EXTRACT, --extract=EXTRACT
                        extract region for testing purposes. Format is
                        contig:strand:from:to. The default coordinates are
                        0-based open/closed coordinates on both strands. For
                        example, chr1:+:10:12 will return bases 11 to 12 on
                        chr1.
  -c COMPRESSION, --compression=COMPRESSION
                        compress database [default=none].
  --random-access-points=RANDOM_ACCESS_POINTS
                        save random access points every # number of
                        nucleotides [default=0].
  -i INPUT_FORMAT, --input-format=INPUT_FORMAT
                        coordinate format of input [default=zero-both-open].
  -s SYNONYMS, --synonyms=SYNONYMS
                        list of synonyms, comma separated with =, for example,
                        chr1=chr1b [default=none]
  -b, --benchmark       benchmark time for read access [default=False].
  --benchmark-num-iterations=BENCHMARK_NUM_ITERATIONS
                        number of iterations for benchmark [default=1000000].
  --benchmark-fragment-size=BENCHMARK_FRAGMENT_SIZE
                        benchmark: fragment size [default=1000].
  --verify=VERIFY       verify against other database [default=none].
  --file-format=FILE_FORMAT
                        file format of input. Supply if data comes from stdin
                        [default=auto].
  -a, --clean-sequence  remove X/x from DNA sequences - they cause errors in
                        exonerate [default=False].
  --allow-duplicates    allow duplicate identifiers. Further occurances of an
                        identifier are suffixed by an &#39;_%i&#39; [default=False].
  --regex-identifier=REGEX_IDENTIFIER
                        regular expression for extracting the identifier from
                        fasta description line [default=none].
  --compress-index      compress index [default=False].
  --force               force overwriting of existing files [default=False].
  -t TRANSLATOR, --translator=TRANSLATOR
                        translate numerical quality scores [default=none].
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">index_fasta.py - Index fasta formatted files</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="mask_fasta.html"
                        title="previous chapter">mask_fasta.py - mask fasta formatted sequences</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="mali2kaks.html"
                        title="next chapter">mali2kaks.py - rate analysis with PAML</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/index_fasta.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="mali2kaks.html" title="mali2kaks.py - rate analysis with PAML"
             >next</a> |</li>
        <li class="right" >
          <a href="mask_fasta.html" title="mask_fasta.py - mask fasta formatted sequences"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>