
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>mali2kaks.py - rate analysis with PAML &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="mali2rates.py - calculate rates for sequences in a multiple alignment" href="mali2rates.html" />
    <link rel="prev" title="index_fasta.py - Index fasta formatted files" href="index_fasta.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="mali2rates.html" title="mali2rates.py - calculate rates for sequences in a multiple alignment"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index_fasta.html" title="index_fasta.py - Index fasta formatted files"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-mali2kaks"></span><div class="section" id="mali2kaks-py-rate-analysis-with-paml">
<h1>mali2kaks.py - rate analysis with PAML<a class="headerlink" href="#mali2kaks-py-rate-analysis-with-paml" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script will do pairwise rate analysis on a set of multiply aligned
sequences in <a class="reference internal" href="../glossary.html#term-fasta"><em class="xref std std-term">fasta</em></a> format.</p>
<p>PAML bails if sequences in a multiple alignment are not overlapping. Thus this
code checks for each pair the overlap. If a pair is not overlapping, it goes
into pairwise mode.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>python mali2kaks.py --help
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python mali2kaks.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: mali2kaks.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  --set-omega=OMEGA     initial omega value.
  --set-kappa=KAPPA     initial kappa value.
  --fix-kappa           do not estimate kappa.
  --fix-omega           do not estimate omega.
  --set-codon-frequencies=CODON_FREQUENCIES
                        set codon frequencies.
  --set-method=PAML_METHOD
                        set paml optimization method [0|1].
  --set-sequence-type=SEQTYPE
                        sequence type.
  --set-clean-data=CLEAN_DATA
                        PAML should cleanup data:  0=only gaps within pair are
                        removed, 1=columns in the mali with gaps are removed.
  --dump                dump raw output [False].
  --set-optimization-threshold=OPTIMIZATION_THRESHOLD
                        set paml optimization threshold [none].
  -i INPUT_FORMAT, --input-format=INPUT_FORMAT
                        input format of multiple alignment [fasta].
  --pairwise            force pairwise comparison [False].
  --iteration=ITERATION
                        iteration mode [all-vs-all].
  --no-clean            do not clean multiple alignment before submitting to
                        codeml. It might take too long for very large
                        sequences.
  --method=METHOD       choose method for rate computation [paml]
  --xrate-model=XRATE_MODEL
                        models to use [f3x4-four].
  -w WRITE, --write=WRITE
                        output sections to write [[]].
  -o OUTPUT_PATTERN, --output-pattern=OUTPUT_PATTERN
                        output pattern for output files [%s.eg].
  --xrate-insert-frequencies
                        estimate codon frequencies from input [False].
  --xrate-uniform-frequencies
                        use uniform codon frequencies [False].
  --xrate-fix-frequencies
                        set initial frequencies to const [False].
  --xrate-estimate-frequencies
                        estimate nucleotide frequencies [False].
  --xrate-fix-rates=FIX_RATES
                        fix rates to specified values. Note that the number of
                        rates has to match the ones
                        in the model. Provide values in a comma-separated list
                        [none].
  --xrate-min-increment=XRATE_MIN_INCREMENT
                        minimum increment to stop iteration in xrate [1e-06].
  --min-overlap=MIN_OVERLAP
                        minimum overlap between a sequence pair in residues
                        [60].
  --with-rho            output rho values (substitution rates per codon). This
                        requires a patched version of PAML [False].
  --with-counts         output counts of aligned positions, transitions and
                        transversions [False].
  --remove-stops        remove stop codons [False].
  --replicates=REPLICATES
                        in benchmarking mode expect ## replicates [none].
  --tree=TREE           use tree for estimation [none].
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">mali2kaks.py - rate analysis with PAML</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index_fasta.html"
                        title="previous chapter">index_fasta.py - Index fasta formatted files</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="mali2rates.html"
                        title="next chapter">mali2rates.py - calculate rates for sequences in a multiple alignment</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/mali2kaks.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="mali2rates.html" title="mali2rates.py - calculate rates for sequences in a multiple alignment"
             >next</a> |</li>
        <li class="right" >
          <a href="index_fasta.html" title="index_fasta.py - Index fasta formatted files"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>