
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>malis2mali.py - concatenate multiple alignments &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="malis2malis.py - extract subsets of malis" href="malis2malis.html" />
    <link rel="prev" title="mali2summary.py - compute summary stats on a mali" href="mali2summary.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="malis2malis.html" title="malis2malis.py - extract subsets of malis"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="mali2summary.html" title="mali2summary.py - compute summary stats on a mali"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-malis2mali"></span><div class="section" id="malis2mali-py-concatenate-multiple-alignments">
<h1>malis2mali.py - concatenate multiple alignments<a class="headerlink" href="#malis2mali-py-concatenate-multiple-alignments" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>concatenate multiple alignments into a single multiple alignment.</p>
<p>This script requires the following input:</p>
<ol class="arabic simple">
<li>Multiple alignments</li>
</ol>
<p>The multiple alignments can be given as separate files (option
&#8211;pattern-mali contains a &#8216;%%s&#8217;) or in a single file.</p>
<ol class="arabic simple" start="2">
<li>A list of components</li>
</ol>
<p>This file maps sequence identifiers in the multiple alignment file(s)
to multiple alignment identifiers. The format for this file is a tab
separated table containing the following fields:</p>
<div class="highlight-python"><div class="highlight"><pre>1  sequence     a sequence identifier

2  input_id     The id under which the alignment is found. This value
                        is substituted for the &#39;%%s&#39; in
                        --pattern-mali.

3 component_id  The component_id or multiple alignment identifier.
                This field is optional. If no third column is specified,
                the input_id is used.
</pre></div>
</div>
<p>Methods:</p>
<p>filter-variants</p>
<blockquote>
<div>only keep variant columns in the multiple alignment. Set
<tt class="docutils literal"><span class="pre">--width</span></tt> to 3 to test for variant codons. Gaps will be ignored.</div></blockquote>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>python malis2mali.py --help
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python malis2mali.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: malis2mali.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -b PATTERN_COMPONENT, --pattern-component=PATTERN_COMPONENT
                        how to extract identifier of mali from component name.
  -c FILENAME_COMPONENTS, --filename-components=FILENAME_COMPONENTS
                        filename of components to choose for each multiple
                        alignment.
  -e PATTERN_FILTER, --pattern-filter=PATTERN_FILTER
                        pattern of how to extract filter from identifier.
  -i INPUT_FORMAT, --input-format=INPUT_FORMAT
                        input format of multiple alignment
  -o OUTPUT_FORMAT, --output-format=OUTPUT_FORMAT
                        output format of multiple alignment
  --filename-mask-regions=FILENAME_MASK_REGIONS
                        mask regions given file. Input format is component_id
                        start   end with coordinates in
                        the multiple alignment.
  --filename-extract-regions=FILENAME_EXTRACT_REGIONS
                        extract regions given file. Input format is
                        component_id        start   end with coordinates in
                        the multiple alignment. The alignment is restricted to
                        the given coordinates by masking everything else.
  --filename-annotate-regions=FILENAME_ANNOTATE_REGIONS
                        annotate regions given file. Input format is
                        component_id       start   end     label with
                        coordinates in
                        the multiple alignment. The label should be a single
                        letter. If the file is empty, the full alignment is
                        output with label N.
  --use-input-id        use input-id instead of component_id in
                        --filename-X-regions
  -a PATTERN_MALI, --pattern-mali=PATTERN_MALI
                        filename pattern for input multiple alignment files.
  --test=TEST           test output with first # components.
  --skip-doubles        skip doubles, takes first entry encountered.
  --ignore-missing      ignore missing alignments [False].
  --remove-all-gaps=REMOVE_ALL_GAPS
                        remove fully masked or gapped positions in a multiple
                        alignment. The integer parameter supplies the frame.
                        Use 3 for codon alignments.
  --remove-any-gaps=REMOVE_ANY_GAPS
                        remove positions in a multiple alignment containing at
                        least one gap. The integer parameter supplies the
                        frame. Use 3 for codon alignments.
  --sample=SAMPLE       sample # components.
  --sample-method=SAMPLE_METHOD
                        sampling method: without-replacement=sample without
                        replacement. length-sample: sample components of same
                        length (without replacement).
  --filename-sample-reference=FILENAME_SAMPLE_REFERENCE
                        filename to with reference components. The components
                        should be part of the file supplied by --filename-
                        components.
  --minimum-mali-length=MINIMUM_MALI_LENGTH
                        minimum multiple alignment length.
  --filename-coordinates=FILENAME_COORDINATES
                        filename of coordinates that constitute the multiple
                        alignment.
  --filename-identifiers=FILENAME_IDENTIFIERS
                        filename with list of identifiers to use.
  -x PATTERN_IDENTIFIER, --pattern-identifier=PATTERN_IDENTIFIER
                        pattern to extract identifier from a sequence header.
  -w WIDTH, --width=WIDTH
                        width of an alignment column (choose 3 for codon
                        alignments) [default=none].
  -m METHODS, --method=METHODS
                        methods to apply
  -p PARAMETERS, --parameters=PARAMETERS
                        parameter stack for methods that require one.
  --mask-acgtn          mask. Anything not [ACGTN] will be N.
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">malis2mali.py - concatenate multiple alignments</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="mali2summary.html"
                        title="previous chapter">mali2summary.py - compute summary stats on a mali</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="malis2malis.html"
                        title="next chapter">malis2malis.py - extract subsets of malis</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/malis2mali.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="malis2malis.html" title="malis2malis.py - extract subsets of malis"
             >next</a> |</li>
        <li class="right" >
          <a href="mali2summary.html" title="mali2summary.py - compute summary stats on a mali"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>