
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>psl2assembly.py - assemble (long) reads on genome &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="psl2psl.py - manipulate psl files" href="psl2psl.html" />
    <link rel="prev" title="gff2bed.py - convert from gff to bed" href="gff2bed.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="psl2psl.html" title="psl2psl.py - manipulate psl files"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="gff2bed.html" title="gff2bed.py - convert from gff to bed"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-psl2assembly"></span><div class="section" id="psl2assembly-py-assemble-long-reads-on-genome">
<h1>psl2assembly.py - assemble (long) reads on genome<a class="headerlink" href="#psl2assembly-py-assemble-long-reads-on-genome" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>Assemble reads on a reference genome. Overlapping reads are merged.</p>
<p>This script is useful for building transcript models from
long aligned reads (for example, 454 reads or ESTs).</p>
<p>The script is aware of splice motifs and modifies alignments 
around intron boundaries to conform to splice site signals.</p>
<p>Input is a psl formatted file sorted by target contig and target start.</p>
<p>Output is a table linking reads to transcripts. Optional
output files are:</p>
<p>.transcripts: a psl formatted output file containing transcripts</p>
<p>.pileup: fasta file with pileup alignments of reads on transcripts</p>
<p>.regions: a tab separated file of transcript coordinates</p>
<p>.coverage: coverage statistics for each transcript</p>
<p>.sbjct_coverage: coverage statistics of the target genome</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>python &lt;script_name&gt;.py --help
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python &lt;script_name&gt;.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: psl2assembly.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -g GENOME_FILE, --genome-file=GENOME_FILE
                        filename with genome [default=none].
  --input-filename-queries=INPUT_FILENAME_QUERIES
                        fasta filename with queries [default=none].
  --input-filename-coverage=INPUT_FILENAME_COVERAGE
                        tabular table with coverage information
                        [default=none].
  -p OUTPUT_FILENAME_PATTERN, --output-filename-pattern=OUTPUT_FILENAME_PATTERN
                        OUTPUT filename with histogram information on
                        aggregate coverages [%s].
  --mali-output-format=MALI_OUTPUT_FORMAT
                        output format to choose [default=fasta].
  --method=METHODS      methods to apply [[]].
  -z, --from-zipped     input is zipped.
  --output-pileup       in the transcript modules, output the pileup alignment
                        with the predicted transcript [default=False].
  --test=TEST           test - stop after # rows of parsing[none].
  --staggered=STAGGERED_ALIGNMENTS
                        how to deal with staggered alignments[none].
  --force-merge=FORCE_MERGE
                        in case of staggered alignments, force merge if there
                        are # alignments. This avoids costly computation of
                        components in large sets. If 0, do not apply the
                        threshold [none].
  --filter=FILTERS      filters to apply [[]].
  --threshold-merge-distance=THRESHOLD_MERGE_DISTANCE
                        distance in nucleotides at which two adjacent reads
                        shall be merged even if they are not overlapping [1].
  --threshold-merge-overlap=THRESHOLD_MERGE_OVERLAP
                        require reads to be overlapping by this amount to be
                        joined into the same transcript [default=0].
  --start-at=START_AT   start numbering ids at number [default=1].
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">psl2assembly.py - assemble (long) reads on genome</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="gff2bed.html"
                        title="previous chapter">gff2bed.py - convert from gff to bed</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="psl2psl.html"
                        title="next chapter">psl2psl.py - manipulate psl files</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/psl2assembly.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="psl2psl.html" title="psl2psl.py - manipulate psl files"
             >next</a> |</li>
        <li class="right" >
          <a href="gff2bed.html" title="gff2bed.py - convert from gff to bed"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>