
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>rnaseq_junction_bams2bam.py - convert mappings against junctions to genomic coordinates &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="gff2chunks.py - split a gff file into chunks" href="split_gff.html" />
    <link rel="prev" title="gtfs2tsv.py - compare two genesets" href="gtfs2tsv.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="split_gff.html" title="gff2chunks.py - split a gff file into chunks"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="gtfs2tsv.html" title="gtfs2tsv.py - compare two genesets"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-rnaseq_junction_bam2bam"></span><div class="section" id="rnaseq-junction-bams2bam-py-convert-mappings-against-junctions-to-genomic-coordinates">
<h1>rnaseq_junction_bams2bam.py - convert mappings against junctions to genomic coordinates<a class="headerlink" href="#rnaseq-junction-bams2bam-py-convert-mappings-against-junctions-to-genomic-coordinates" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Genomics NGS Genesets</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script takes as input a BAM file resulting from reads mapped against
a junction database and outputs a <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> formatted file in genomic
coordinates.</p>
<p>The contigs should be of the format 
&lt;chromosome&gt;|&lt;start&gt;|&lt;exon-end&gt;-&lt;exon-start&gt;|&lt;end&gt;|&lt;splice&gt;|&lt;strand&gt;.</p>
<p>&lt;start&gt; - 0-based coordinate of first base
&lt;exon-end&gt; - 0-based coordinate of last base in exon
&lt;exon-start&gt; - 0-based coordinate of first base in exon
&lt;end&gt; - 0-based coordinate of base after last base</p>
<p>Strand can be either <tt class="docutils literal"><span class="pre">fwd</span></tt> or <tt class="docutils literal"><span class="pre">rev</span></tt>, though sequences in the database
and coordinates are all on the forward strand.</p>
<p>For example <tt class="docutils literal"><span class="pre">chr1|1244933|1244982-1245060|1245110|GTAG|fwd</span></tt> translates to the
intron <tt class="docutils literal"><span class="pre">chr1:1244983-1245060</span></tt> in python coordinates.</p>
<p>The input bam-file is supposed to be sorted by read. Only the best matches
are output for each read, were best is defined both in terms of number 
of mismatches and number of colour mismatches.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>cat input.bam | python rnaseq_junction_bam2bam.py - --log=log &gt; output.bam
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python rnaseq_junction_bam2bam.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: rnaseq_junction_bam2bam.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -t FILENAME_GENOME_BAM, --template-bam=FILENAME_GENOME_BAM
                        input bam file for header information [none]
  -s FILENAME_CONTIGS, --contig-sizes=FILENAME_CONTIGS
                        filename with contig sizes [none]
  -o, --colour          mismatches will use colour differences (CM tag)
                        [False]
  -i, --ignore-mismatches
                        ignore mismatches [False]
  -c REMOVE_CONTIGS, --remove-contigs=REMOVE_CONTIGS
                        &#39;,&#39;-separated list of contigs to remove [none]
  -f, --force           force overwriting of existing files [False]
  -u, --unique          remove reads not matching uniquely [False]
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">rnaseq_junction_bams2bam.py - convert mappings against junctions to genomic coordinates</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="gtfs2tsv.html"
                        title="previous chapter">gtfs2tsv.py - compare two genesets</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="split_gff.html"
                        title="next chapter">gff2chunks.py - split a gff file into chunks</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/rnaseq_junction_bam2bam.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="split_gff.html" title="gff2chunks.py - split a gff file into chunks"
             >next</a> |</li>
        <li class="right" >
          <a href="gtfs2tsv.html" title="gtfs2tsv.py - compare two genesets"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>