
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>snp2counts.py - count number SNPs in geneset &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="snp2map.py - build maf formatted multiple genomic alignments from SNPs" href="snp2maf.html" />
    <link rel="prev" title="maq2psl.py - wonvert the output of maqview to psl format" href="maq2psl.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="snp2maf.html" title="snp2map.py - build maf formatted multiple genomic alignments from SNPs"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="maq2psl.html" title="maq2psl.py - wonvert the output of maqview to psl format"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-snp2counts"></span><div class="section" id="snp2counts-py-count-number-snps-in-geneset">
<h1>snp2counts.py - count number SNPs in geneset<a class="headerlink" href="#snp2counts-py-count-number-snps-in-geneset" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id: snp2counts.py 2872 2010-03-03 10:21:13Z andreas $</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>read a list of genomic point locations (SNPs) and count the number
of SNPs falling in pre-defined windows.</p>
<p>The windows are given in gtf format.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>The script will be able to count snps in disjoint segments
using the gene_id field in gtf format. It will not check 
if these segments are non-overlapping.</p>
<p class="last">In case of a gene set, make sure to first flatten the gene set by combining
all transcript/exons per gene.</p>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python &lt;script_name&gt;.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>The script currently uses <tt class="docutils literal"><span class="pre">variant</span></tt> in two meanings:</p>
<ol class="arabic simple">
<li>a variable site (SNP/INDEL)</li>
<li>a transcript variant (a transcript sequence that differs from the wild type)</li>
</ol>
<p class="last">I have started calling the latter <tt class="docutils literal"><span class="pre">allele</span></tt>, though it is not
consistent across the whole script. However, the output is consistent and calls
the former <tt class="docutils literal"><span class="pre">variant_site</span></tt> and the latter <tt class="docutils literal"><span class="pre">allele</span></tt>.</p>
</div>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: snp2counts.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -g GENOME_FILE, --genome-file=GENOME_FILE
                        filename with genome [default=none].
  -f FILENAME_EXONS, --filename-exons=FILENAME_EXONS
                        filename with exon information (gtf formatted file)
                        [default=none].
  -s FILENAME_SELENO, --filename-seleno=FILENAME_SELENO
                        filename of a list of transcript ids that are
                        selenoproteins [default=none].
  -c FILENAME_VCF, --filename-vcf=FILENAME_VCF
                        vcf file to parse [default=none].
  -m MODULES, --module=MODULES
                        modules to apply [default=[]].
  -i INPUT_FORMAT, --input-format=INPUT_FORMAT
                        input format [default=pileup].
  --vcf-sample=VCF_SAMPLE
                        sample id in vcf file to analyse [default=none].
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -P OUTPUT_FILENAME_PATTERN, --output-filename-pattern=OUTPUT_FILENAME_PATTERN
                        OUTPUT filename pattern for various methods [%s].
  -F, --force           force over-writing of existing files.
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">snp2counts.py - count number SNPs in geneset</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="maq2psl.html"
                        title="previous chapter">maq2psl.py - wonvert the output of maqview to psl format</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="snp2maf.html"
                        title="next chapter">snp2map.py - build maf formatted multiple genomic alignments from SNPs</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/snp2counts.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="snp2maf.html" title="snp2map.py - build maf formatted multiple genomic alignments from SNPs"
             >next</a> |</li>
        <li class="right" >
          <a href="maq2psl.html" title="maq2psl.py - wonvert the output of maqview to psl format"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>