
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>snp2table.py - annotate variants &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Scripts" href="../scripts.html" />
    <link rel="next" title="psl2fasta.py - output alignment pairs from psl formatted alignments" href="psl2fasta.html" />
    <link rel="prev" title="gtf2fasta.py - annotate genomic bases from a gene set" href="gtf2fasta.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="psl2fasta.html" title="psl2fasta.py - output alignment pairs from psl formatted alignments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="gtf2fasta.html" title="gtf2fasta.py - annotate genomic bases from a gene set"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-snp2table"></span><div class="section" id="snp2table-py-annotate-variants">
<h1>snp2table.py - annotate variants<a class="headerlink" href="#snp2table-py-annotate-variants" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id: snp2table.py 2861 2010-02-23 17:36:32Z andreas $</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This script reads a file in samtools pileup -c format and annotates the
SNPs in that file.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This script is still under construction. There are issues with the
phasing of variants, using the wild-type sequence and merging
splice and coding variants.</p>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre>gunzip &lt; snps.pileup.gz 
| python snp2table.py  
     --genome-file=genome 
     --filename-exons=exons
     --filename-annotations=bases
     --filename-junctions=junctions
     --log=log 
&gt; result.out
</pre></div>
</div>
<p>Type:</p>
<div class="highlight-python"><div class="highlight"><pre>python &lt;script_name&gt;.py --help
</pre></div>
</div>
<p>for command line help.</p>
</div>
<div class="section" id="documenation">
<h2>Documenation<a class="headerlink" href="#documenation" title="Permalink to this headline">¶</a></h2>
<p>The following annotations are available:</p>
<dl class="docutils">
<dt>exons</dt>
<dd><p class="first">Given a gtf-file with the output from <tt class="docutils literal"><span class="pre">gtf2gff.py</span> <span class="pre">--method=exons</span></tt>, each
SNP that overlaps an exon is labeled according to its position within a transcript. 
The columns are:</p>
<dl class="last docutils">
<dt>ntranscripts</dt>
<dd>number of transrcipts in the gene that the exon belongs to.</dd>
<dt>nused</dt>
<dd>number of transcripts that contain this exon</dd>
<dt>positions</dt>
<dd>positions of exons within transcipts</dd>
</dl>
</dd>
<dt>annotation</dt>
<dd><p class="first">Given a fasta file with base-level annotations (output from <tt class="docutils literal"><span class="pre">gtf2fasta.py</span></tt>),
each SNP is labeled by its effect. The columns are:</p>
<dl class="docutils">
<dt>code</dt>
<dd>one letter annotation of base. See <tt class="file docutils literal"><span class="pre">gtf2fasta.py</span></tt>.</dd>
<dt>reference_codon</dt>
<dd>the reference codon (in case sequence is coding)</dd>
<dt>reference_aa</dt>
<dd>the reference amina acid (in case sequence is coding)</dd>
<dt>variant_type</dt>
<dd>the variant types (&#8216;I&#8217; for insertion, &#8216;D&#8217; for deletion&#8217;, &#8216;S&#8217; for substitution).</dd>
<dt>variant_codon</dt>
<dd>the variant codons (in case sequence is coding)</dd>
<dt>variant_aa</dt>
<dd>the variant amino acids (in case sequence is coding)</dd>
</dl>
<p class="last">Variants will be separated by <tt class="docutils literal"><span class="pre">,</span></tt> if there are several variants per position.</p>
</dd>
</dl>
<p>The script deconvolutes substitution and indel snps. The following columns are always output:</p>
<dl class="docutils">
<dt>chromosome</dt>
<dd>chromosome</dd>
<dt>position</dt>
<dd>reference base position  (0-based)</dd>
<dt>reference_base</dt>
<dd>reference base. For indels, this is <tt class="docutils literal"><span class="pre">*</span></tt>.</dd>
<dt>genotype</dt>
<dd>consensus base. For indels, this is the genotype</dd>
<dt>consensus_quality</dt>
<dd>consensus quality</dd>
<dt>snp_quality</dt>
<dd>snp quality</dd>
<dt>rms_mapping_qualit</dt>
<dd>rms quality</dd>
<dt>coverage</dt>
<dd>coverage</dd>
<dt>read_bases</dt>
<dd>read bases. For indels, this is the first allele (<tt class="docutils literal"><span class="pre">*</span></tt> for reference).</dd>
<dt>base_qualities</dt>
<dd>base qualities. For indels, this is the second allele (<tt class="docutils literal"><span class="pre">*</span></tt> for reference).</dd>
</dl>
<p>See the samtools documentation for the meaning of these.</p>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="highlight-text"><div class="highlight"><pre>Usage: snp2table.py [options]

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --no-usage            output help without usage information
  -g GENOME_FILE, --genome-file=GENOME_FILE
                        filename with genome [default=none].
  -a FILENAME_ANNOTATIONS, --filename-annotations=FILENAME_ANNOTATIONS
                        filename with base annotations (output from
                        gtf2fasta.py) [default=none].
  -f FILENAME_EXONS, --filename-exons=FILENAME_EXONS
                        filename with exon information (gff formatted file)
                        [default=none].
  -j FILENAME_JUNCTIONS, --filename-junctions=FILENAME_JUNCTIONS
                        filename with junction information (filename with exon
                        junctions)  [default=none].
  -c FILENAME_VCF, --filename-vcf=FILENAME_VCF
                        vcf file to parse [default=none].
  -i INPUT_FORMAT, --input-format=INPUT_FORMAT
                        input format [default=pileup].
  --vcf-sample=VCF_SAMPLE
                        sample id in vcf file to analyse [default=none].
  -v LOGLEVEL, --verbose=LOGLEVEL
                        loglevel [1]. The higher, the more output.
  --timeit=TIMEIT_FILE  store timeing information in file [none].
  --timeit-name=TIMEIT_NAME
                        name in timing file for this class of jobs [all].
  --timeit-header       add header for timing information [none].
  --random-seed=RANDOM_SEED
                        random seed to initialize number generator with
                        [none].
  -I FILE, --stdin=FILE
                        file to read stdin from [default = stdin].
  -L FILE, --log=FILE   file with logging information [default = stdout].
  -E FILE, --error=FILE
                        file with error information [default = stderr].
  -S FILE, --stdout=FILE
                        file where output is to go [default = stdout].
</pre></div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">snp2table.py - annotate variants</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#documenation">Documenation</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="gtf2fasta.html"
                        title="previous chapter">gtf2fasta.py - annotate genomic bases from a gene set</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="psl2fasta.html"
                        title="next chapter">psl2fasta.py - output alignment pairs from psl formatted alignments</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/snp2table.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="psl2fasta.html" title="psl2fasta.py - output alignment pairs from psl formatted alignments"
             >next</a> |</li>
        <li class="right" >
          <a href="gtf2fasta.html" title="gtf2fasta.py - annotate genomic bases from a gene set"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>