
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Testing for Functional enrichment &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="testing-for-functional-enrichment">
<h1>Testing for Functional enrichment<a class="headerlink" href="#testing-for-functional-enrichment" title="Permalink to this headline">Â¶</a></h1>
<p>This tutorial demonstrates the usage of <em>gat</em> with a simple example -
where does a transcription factor bind in the genome?</p>
<p>This tutorial uses the SRF data set described in <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/19160518">Valouev et
al. (2008)</a>. The data sets used in this tutorial are available at:</p>
<p><a class="reference external" href="http://www.cgat.org/~andreas/sample_data/srf.hg19.bed.gz">http://www.cgat.org/~andreas/sample_data/srf.hg19.bed.gz</a></p>
<p>This <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted file contains 556 high confidence peaks
from the analysis of <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/19160518">Valouev et al. (2008)</a> mapped to human
chromosome hg19.</p>
<p>We want to find out, where these binding sites are located in the
genome. First let us download the genomic sequence for hg19 and
index it:</p>
<div class="highlight-python"><div class="highlight"><pre>wget -qO- http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz
| cgat index_fasta --file-format=tar.gz hg19 -
&gt; hg19.log
</pre></div>
</div>
<p>Next, we need to define where intronic and intergenic regions are
located in the genome. We do this by by obtaining the latest geneset
from <a class="reference external" href="http:://www.ensembl.org">ENSEMBL</a> and pushing it through a sequence of commands:</p>
<div class="highlight-python"><div class="highlight"><pre>wget -qO- ftp://ftp.ensembl.org/pub/release-72/gtf/homo_sapiens/Homo_sapiens.GRCh37.72.gtf.gz
| gunzip
| awk &#39;$2 == &quot;protein_coding&quot;&#39;
| cgat gff2gff --genome-file=hg19 --sanitize=genome --skip-missing
| cgat gtf2gtf --sort=gene
| cgat gtf2gtf --merge-exons --with-utr
| cgat gtf2gtf --filter=longest-gene
| cgat gtf2gtf --sort=position
| cgat gtf2gff --genome-file=hg19 --flank=5000 --method=genome
| gzip
&gt; annotations.hg19.gff.gz
</pre></div>
</div>
<p>The commands do the following.</p>
<ol class="arabic">
<li><p class="first">Reconcile the chromosome names in the gene set (ENSEMBL: 1,2,3)
with the UCSC convention (chr1,chr2,chr3):</p>
<div class="highlight-python"><div class="highlight"><pre>| cgat gff2ff --genome-file=hg19 --sanitize=genome --skip-missing
</pre></div>
</div>
</li>
<li><p class="first">Sort the gene set by gene making sure that all exons within a gene
appear in a block:</p>
<div class="highlight-python"><div class="highlight"><pre>| cgat gtf2gtf --merge-exons --with-utr
</pre></div>
</div>
</li>
<li><p class="first">Merge overlapping exons from alternative transcripts of the same gene:</p>
<div class="highlight-python"><div class="highlight"><pre>| cgat gtf2gtf --merge-exons --with-utr
</pre></div>
</div>
</li>
<li><p class="first">Resolve nested genes. In nested genes a genomic region might be
both defined intronic and intergenic. Here, we select the longer
one:</p>
<div class="highlight-python"><div class="highlight"><pre>| cgat gtf2gtf --filter=longest-gene
</pre></div>
</div>
</li>
<li><p class="first">Sort by genomic position:</p>
<div class="highlight-python"><div class="highlight"><pre>| cgat gtf2gtf --sort=position
</pre></div>
</div>
</li>
<li><p class="first">Define intronic, intergenic and other gene set based annotations:</p>
<div class="highlight-python"><div class="highlight"><pre>| cgat gtf2gff --genome-file=hg19 --flank=5000 --method=genome
</pre></div>
</div>
</li>
</ol>
<p>The tool gff2stats</p>
<p>We can now use the file <tt class="file docutils literal"><span class="pre">annotations.hg19.gff.gz</span></tt> to classify
individual peaks with the <a class="reference internal" href="../scripts/bed2table.html"><em>bed2table.py - annotate intervals</em></a> tool:</p>
<div class="highlight-python"><div class="highlight"><pre>zcat srf.hg19.bed.gz
| cgat bed2table --genome-file=hg19 --counter=classify-chipseq --filename-gff=annotations.hg19.gff.gz
| gzip
&gt; srf.hg19.tsv.gz
</pre></div>
</div>
<p>The table <tt class="file docutils literal"><span class="pre">srf.hg19.tsv.gz</span></tt> contains a row for each interval in
the input file <tt class="file docutils literal"><span class="pre">srf.hg19.bed.gz</span></tt> describing which genomic
features it overlaps and assigns it to a category such as introning,
intergenic, etc.. We can upload this file into a database or view in
excel to easily filter and summarize the data.</p>
<p>To get a more global view of where the transcription factor binds,
we make use of the <a class="reference external" href="http://code.google.com/p/genomic-association-tester/">gat</a> tool. Gat tests if two sets of genomic
features are overlapping more - or less - than expected by chance
through simulation. Note that <a class="reference external" href="http://code.google.com/p/genomic-association-tester/">gat</a> needs to be installed separately
(<tt class="docutils literal"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">gat</span></tt>).</p>
<p>For gat, we need the file with genomic annotations
(<tt class="file docutils literal"><span class="pre">annotations.hg19.gff.gz</span></tt>) we created previously and a workspace - a
set of genomic regions that are accessible for simulation. Here, we
will use the full genome for simulation excluding regions that are
gaps as we do not expect to be able to detect transcription factor
binding sites in those in an NGS experiment.  To get these regions, we
use the <a class="reference internal" href="../scripts/fasta2bed.html"><em>&lt;no title&gt;</em></a> tool:</p>
<div class="highlight-python"><div class="highlight"><pre>cat hg19.fasta
| cgat fasta2bed --method=ungapped --min-gap-size=100
| awk &#39;$1 ~ /^chr/&#39;
| cut -f 1,2,3
| gzip
&gt; ungapped.hg19.bed.gz
</pre></div>
</div>
<p>Gat needs <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> formatted input files, so let us quickly convert
<tt class="file docutils literal"><span class="pre">annogations.hg19.gff.gz</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre>zcat annogations.hg19.gff.gz
| cgat gff2bed.py
| gzip
&gt; annotations.bed.gz
</pre></div>
</div>
<p>We are now ready to run gat:</p>
<div class="highlight-python"><div class="highlight"><pre>gat-run.py
   --ignore-segment-tracks
   --segments=srf.hg19.bed.gz
   --annotations=annotations.hg19.bed.gz
   --workspace=ungapped.hg19.bed.gz
   --num-samples=1000
   --log=gat.log
| gzip
&gt; gat.out
</pre></div>
</div>
<p>The option <cite>&#8211;ignore-segment-tracks</cite> tells <em>gat</em> to ignore the fourth
column in the <em class="xref std std-term">tracks</em> file and assume that all intervals in
this file belong to the same <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a>. If not given, each
interval would be treated separately.</p>
<p>The above statement finishes in a few seconds. With large interval
collections or many annotations, <em>gat</em> might take a while. It is thus
good practice to always save the output in a file. The option <cite>&#8211;log</cite>
tells gat to save information or warning messages into a separate log
file.</p>
<p>The first 11 columns of the output file are the most informative:</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="10%" />
<col width="8%" />
<col width="10%" />
<col width="10%" />
<col width="10%" />
<col width="8%" />
<col width="7%" />
<col width="7%" />
<col width="10%" />
<col width="10%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>track</td>
<td>annotation</td>
<td>observed</td>
<td>expected</td>
<td>CI95low</td>
<td>CI95high</td>
<td>stddev</td>
<td>fold</td>
<td>l2fold</td>
<td>pvalue</td>
<td>qvalue</td>
</tr>
<tr class="row-even"><td>merged</td>
<td>telomeric</td>
<td>0</td>
<td>69.7440</td>
<td>0.0000</td>
<td>200.0000</td>
<td>59.6216</td>
<td>0.0141</td>
<td>-6.1445</td>
<td>2.5100e-01</td>
<td>3.9443e-01</td>
</tr>
<tr class="row-odd"><td>merged</td>
<td>intergenic</td>
<td>6200</td>
<td>13909.1770</td>
<td>12989.0000</td>
<td>14800.0000</td>
<td>570.3231</td>
<td>0.4458</td>
<td>-1.1656</td>
<td>1.0000e-03</td>
<td>2.2000e-03</td>
</tr>
<tr class="row-even"><td>merged</td>
<td>intronic</td>
<td>8415</td>
<td>11401.6660</td>
<td>10440.0000</td>
<td>12345.0000</td>
<td>577.7517</td>
<td>0.7381</td>
<td>-0.4382</td>
<td>1.0000e-03</td>
<td>2.2000e-03</td>
</tr>
<tr class="row-odd"><td>merged</td>
<td>UTR3</td>
<td>284</td>
<td>305.5370</td>
<td>114.0000</td>
<td>500.0000</td>
<td>120.2095</td>
<td>0.9297</td>
<td>-0.1051</td>
<td>4.3000e-01</td>
<td>5.2556e-01</td>
</tr>
<tr class="row-even"><td>merged</td>
<td>unknown</td>
<td>0</td>
<td>0.0140</td>
<td>0.0000</td>
<td>0.0000</td>
<td>0.3603</td>
<td>0.9862</td>
<td>-0.0201</td>
<td>9.9800e-01</td>
<td>9.9800e-01</td>
</tr>
<tr class="row-odd"><td>merged</td>
<td>frameshift</td>
<td>0</td>
<td>0.0050</td>
<td>0.0000</td>
<td>0.0000</td>
<td>0.0947</td>
<td>0.9950</td>
<td>-0.0072</td>
<td>9.9700e-01</td>
<td>9.9800e-01</td>
</tr>
<tr class="row-even"><td>merged</td>
<td>3flank</td>
<td>800</td>
<td>699.4930</td>
<td>400.0000</td>
<td>1045.0000</td>
<td>187.2328</td>
<td>1.1435</td>
<td>0.1934</td>
<td>3.0300e-01</td>
<td>4.1662e-01</td>
</tr>
<tr class="row-odd"><td>merged</td>
<td>CDS</td>
<td>758</td>
<td>392.1510</td>
<td>192.0000</td>
<td>611.0000</td>
<td>131.0955</td>
<td>1.9306</td>
<td>0.9490</td>
<td>3.0000e-03</td>
<td>5.5000e-03</td>
</tr>
<tr class="row-even"><td>merged</td>
<td>flank</td>
<td>1335</td>
<td>176.1320</td>
<td>50.0000</td>
<td>350.0000</td>
<td>90.7093</td>
<td>7.5424</td>
<td>2.9150</td>
<td>1.0000e-03</td>
<td>2.2000e-03</td>
</tr>
<tr class="row-odd"><td>merged</td>
<td>5flank</td>
<td>6224</td>
<td>742.0590</td>
<td>450.0000</td>
<td>1071.0000</td>
<td>191.1824</td>
<td>8.3775</td>
<td>3.0665</td>
<td>1.0000e-03</td>
<td>2.2000e-03</td>
</tr>
<tr class="row-even"><td>merged</td>
<td>UTR5</td>
<td>3784</td>
<td>104.0220</td>
<td>0.0000</td>
<td>237.0000</td>
<td>68.5653</td>
<td>36.0401</td>
<td>5.1715</td>
<td>1.0000e-03</td>
<td>2.2000e-03</td>
</tr>
</tbody>
</table>
<p>The first two columns contain the name of the <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a> and
<em class="xref std std-term">annotation</em> that are being compared. The columns
<em class="xref std std-term">observed</em> and <em class="xref std std-term">expected</em> give the observed and expected
nucleotide overlap, respectively, between the <a class="reference internal" href="../glossary.html#term-track"><em class="xref std std-term">track</em></a> and <em class="xref std std-term">annotation</em>.</p>
<p>The following columns CI95low, CI95high, stddev give 95% confidence
intervals and the standard deviation of the sample distribution,
respectively.</p>
<p>The <em class="xref std std-term">fold</em> column is the fold enrichment or depletion and is
computed as the ratio of <em class="xref std std-term">observed</em> over <em class="xref std std-term">expected</em>. The
column <em class="xref std std-term">l2fold</em> is the log2 of this ratio.</p>
<p>The column <em class="xref std std-term">pvalue</em> gives the empirical <em class="xref std std-term">p-value</em>, i.e. in what
proportion of samples was a higher enrichment or lower depletion
found than the one that was observed.</p>
<p>The column <em class="xref std std-term">qvalue</em> lists a multiple testing corrected <em class="xref std std-term">p-value</em>.
Setting a qvalue threshold and accepting only those comparisons with a
qvalue below that threshold corresponds to controlling the false discovery
rate at that particular level.</p>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/recipes/gat.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>