
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>What is the binding profile of NFKB across gene models? &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="what-is-the-binding-profile-of-nfkb-across-gene-models">
<h1>What is the binding profile of NFKB across gene models?<a class="headerlink" href="#what-is-the-binding-profile-of-nfkb-across-gene-models" title="Permalink to this headline">¶</a></h1>
<p>After processing RNA-seq data through alignment, gene/transcript abundance estimation and differential
expression analysis, we are left with an unannotated list of differentially expressed genes. We may want
to know whether there is evidence to suggest that these genes are regulated by a transcription factor
of interest. We can answer this using ChIP-seq data that we ourselves have generated or by using
public resources such as ENCODE.</p>
<p>For example, we have carried out an RNA-seq experiment in lymphoblastoid cell lines (LCLs) looking at
the effect of TNF-a stimulation on gene expression. Using one of the many tools for conducting
differential expression analysis we have arrived at a set of 133 genes that are up-regulated when
LCLs are stimulated with TNF-a.</p>
<p>We know that the main transcription factor that drives expression of inflammatory genes using other
immune stimulators is NFKB. We would therefore like to answer the question:</p>
<p>Is there evidence to support a role for NFKB in the regulation of genes regulated by TNF-a in LCLs?</p>
<p>ENCODE have produced many ChIP-seq data sets and by a stroke of luck they have NFKB ChIP-seq data in
TNF-a stimulated LCLs. In an exploratory phase of the analysis, we would like to see what the profile
of NFKB binding is across genes i.e does it bind predominantly at the TSS, exons or 3&#8217; UTR. We
can do this fairly easily with a few files and a few commands.</p>
<p>The input files that we require are:</p>
<ul>
<li><p class="first">A <a class="reference internal" href="../glossary.html#term-gtf"><em class="xref std std-term">gtf</em></a> file  containing a complete set of known protein coding gene transcripts, which may
be downloaded from ENSEMBL by typing:</p>
<div class="highlight-python"><div class="highlight"><pre>wget ftp://ftp.ensembl.org/pub/release-73/gtf/homo_sapiens/Homo_sapiens.GRCh37.73.gtf.gz -o logfile
</pre></div>
</div>
</li>
<li><p class="first">A file containing aligned NFKB ChIP-seq reads in <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> format, which is available via UCSC:</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://hgdownload-test.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeSydhTfbs/wgEncodeSydhTfbsGm10847NfkbTnfaIggrabAlnRep1.bam -o logfile

wget http://hgdownload-test.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeSydhTfbs/wgEncodeSydhTfbsGm10847NfkbTnfaIggrabAlnRep1.bam.bai -o logfile
</pre></div>
</div>
</li>
</ul>
<p>We can then take all protein coding genes from this geneset with an awk statement:</p>
<div class="highlight-python"><div class="highlight"><pre>zcat Homo_sapiens.GRCh37.73.gtf.gz  | awk &#39;$2==&quot;protein_coding&quot;&#39; | gzip &gt; coding_geneset.gtf.gz
</pre></div>
</div>
<p>Using the CGAT tool bam2geneprofile we can then assess the binding profile of NFKB across gene models:</p>
<div class="highlight-python"><div class="highlight"><pre>cgat bam2geneprofile --bamfile=wgEncodeSydhTfbsGm10847NfkbTnfaIggrabAlnRep1.bam
                     --gtffile=coding_geneset.gtf.gz
                     --method=geneprofile
                     --profile_normalization=counts
                     --output-filename-pattern=nfkb_profile_%s
</pre></div>
</div>
<p>This statement will produce a matrix as an output file named &#8220;nfkb_profile.geneprofile.matrix.tsv.gz&#8221;
with the following format:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="23%" />
<col width="29%" />
<col width="40%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>bin</td>
<td>region</td>
<td>region_bin</td>
<td>counts</td>
</tr>
<tr class="row-even"><td>0</td>
<td>upstream</td>
<td>0</td>
<td>0.22691292876</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>upstream</td>
<td>1</td>
<td>0.224274406332</td>
</tr>
<tr class="row-even"><td>2</td>
<td>upstream</td>
<td>2</td>
<td>0.221635883905</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>upstream</td>
<td>3</td>
<td>0.192612137203</td>
</tr>
<tr class="row-even"><td>4</td>
<td>upstream</td>
<td>4</td>
<td>0.221635883905</td>
</tr>
<tr class="row-odd"><td>5</td>
<td>upstream</td>
<td>5</td>
<td>0.213720316623</td>
</tr>
<tr class="row-even"><td>6</td>
<td>upstream</td>
<td>6</td>
<td>0.213720316623</td>
</tr>
<tr class="row-odd"><td>7</td>
<td>upstream</td>
<td>7</td>
<td>0.200527704485</td>
</tr>
<tr class="row-even"><td>8</td>
<td>upstream</td>
<td>8</td>
<td>0.20580474934</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>These data are amenable to further manipulation and visualisation. For example, we can use R to produce a profile plot
over the gene model. Start R and type:</p>
<div class="highlight-python"><div class="highlight"><pre>R version 2.15.2 (2012-10-26) -- &quot;Trick or Treat&quot;
Copyright (C) 2012 The R Foundation for Statistical Computing
ISBN 3-900051-07-0
   Platform: x86_64-unknown-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type &#39;license()&#39; or &#39;licence()&#39; for distribution details.

R is a collaborative project with many contributors.
Type &#39;contributors()&#39; for more information and
&#39;citation()&#39; on how to cite R or R packages in publications.

Type &#39;demo()&#39; for some demos, &#39;help()&#39; for on-line help, or
&#39;help.start()&#39; for an HTML browser interface to help.
Type &#39;q()&#39; to quit R.

&gt; profile &lt;- read.csv(&quot;nfkb_profile.geneprofile.matrix.tsv.gz&quot;, header = T, stringsAsFactors = F, sep = &quot;\t&quot;)

&gt; plot(profile$bin, profile$counts, cex = 0, xaxt = &quot;none&quot;)

&gt; lines(profile$bin, profile$counts, col = &quot;blue&quot;)

&gt; abline(v = c(1000, 2000), lty = 2)

&gt; mtext(&quot;upstream&quot;, adj = 0.1)

&gt; mtext(&quot;exons&quot;, adj = 0.5)

&gt; mtext(&quot;downstream&quot;, adj = 0.9)
</pre></div>
</div>
<p>This set of commands will produce the figure shown.</p>
<img alt="../_images/nfkb_profile.png" src="../_images/nfkb_profile.png" />
<p>This plot displays the predominance of NFKB binding at transcription start sites of protein coding genes.</p>
<div class="section" id="visulazing-chip-seq-read-coverage-across-nfkb-binding-intervals">
<h2>Visulazing ChIP-seq read coverage across NFKB binding intervals<a class="headerlink" href="#visulazing-chip-seq-read-coverage-across-nfkb-binding-intervals" title="Permalink to this headline">¶</a></h2>
<p>While NFKB binds to the TSSs of protein coding genes, it also binds to many intergenic regions of the genome. In addition
to meta-gene profiles we may also want to know the chromatin state at which NFKB binding occurs. For example, we can
integrate additional histone modification ChIP-seq data from the ENCODE project. H3K4me3 and H3K4me1 mark promoters and
enhancers, respectively. We would like to visualise the profile of these marks at all the genomic locations of
NFKB binding.</p>
<p>For this example we require three further files:</p>
<ul class="simple">
<li>A file containing <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> intervals describing NFKB peaks (<a class="reference external" href="http://www.cgat.org/~jethro/cgat/recipes/nfkb_profile/nfkb.bed">NFKB.bed</a>), which may either be downloaded directly or
created from the earlier <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> file of NFKB ChIP-seq reads using a peak caller such as MACS.</li>
<li>files containing aligned H3K4me1 and H3K4me3 ChIP-seq reads in <a class="reference internal" href="../glossary.html#term-bam"><em class="xref std std-term">bam</em></a> format (<a class="reference external" href="http://www.cgat.org/~jethro/cgat/recipes/nfkb_profile/H3K4me3.bam">H3K4me3.bam</a>, <a class="reference external" href="http://www.cgat.org/~jethro/cgat/recipes/nfkb_profile/H3K4me1.bam">H3K4me1.bam</a>)</li>
</ul>
<p>Using the CGAT tool bam2peakshape it is possible to produce a matrix depicting read coverage across all intervals in
the supplied <a class="reference internal" href="../glossary.html#term-bed"><em class="xref std std-term">bed</em></a> file.</p>
<p>The following command line statement</p>
<ol class="arabic">
<li><p class="first">finds the peak of H3K4me3 read coverage within each interval</p>
</li>
<li><p class="first">calculates coverage across a 1000bp window centered around this peak</p>
</li>
<li><p class="first">outputs a matrix in which intervals are ranked by peak height.</p>
</li>
<li><p class="first">outputs an equivalent matrix depicting H3K4me1 coverage across the same windows:</p>
<div class="highlight-python"><div class="highlight"><pre>cgat bam2peakshape   H3K4me3.bam
                     NFKB.bed
                     --control=H3K4me1.bam
                     --sort=peak-height
                     --output-filename-pattern=peakshape.%s
                     &gt; peakshape.table
</pre></div>
</div>
</li>
</ol>
<p>Two files are produced named peakshape.matrix_peak_height.gz &amp; peakshape.control_peak_height.gz that contain matrices
depicting H3K4me3 coverage and H3K4me1 coverage across intervals, respectively.</p>
<p>Both matrices are amenable to plotting as heatmaps using the R package gplots:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt; library( gplots )

&gt; library( RColorBrewer )

&gt; # read the H3K4me3 matrix into R
&gt; me3 &lt;- read.csv( &quot;peakshape.matrix_peak_height.gz&quot;, header=TRUE, sep=&quot;\t&quot;, row.names=1 )

&gt; # convert to matrix
&gt; me3.matrix &lt;- as.matrix( me3 )

&gt; # A proportion of NFkB intervals have no discernable H3K4me3 or H3K4me1 coverage. These are removed before plotting.
&gt; me3.matrix &lt;- me3.matrix[ c( 4000, 14906 ), ]

&gt; # the remainder are plotted
&gt; cols &lt;- brewer.pal( 9, &quot;Blues&quot; )

&gt; heatmap.2( me3.matrix, col=cols, Rowv=F, Colv=F, labRow=&quot;&quot;, key=FALSE, labCol=&quot;&quot;, trace=&quot;none&quot;, dendrogram=&quot;none&quot;, breaks=seq(0, 1000, 101) )

&gt; # A second plot can be produced for the H3K4me1 data
&gt; me1 &lt;- read.csv( &quot;peakshape.control_peak_height.gz&quot;, header=T, sep=&quot;\t&quot;, row.names=1 )

&gt; me1.matrix &lt;- as.matrix( me3 )

&gt; me1.matrix &lt;- me1.matrix[ c( 4000, 14906 ), ]

&gt; cols &lt;- brewer.pal( 9, &quot;Greens&quot; )

&gt; heatmap.2( me1.matrix, col=cols, Rowv=F, Colv=F, labRow=&quot;&quot;, key=FALSE, labCol=&quot;&quot;, trace=&quot;none&quot;, dendrogram=&quot;none&quot;, breaks=seq(0, 100, 11))
</pre></div>
</div>
<p>The resulting plots indicate that a subset of NFKB binding intervals may be characterised on the basis of their chromatin state:</p>
<div class="figure">
<img alt="../_images/H3K4me3_heatmap.png" src="../_images/H3K4me3_heatmap.png" />
<p class="caption">H3K4me3</p>
</div>
<div class="figure">
<img alt="../_images/H3K4me1_heatmap.png" src="../_images/H3K4me1_heatmap.png" />
<p class="caption">H3K4me1</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">What is the binding profile of NFKB across gene models?</a><ul>
<li><a class="reference internal" href="#visulazing-chip-seq-read-coverage-across-nfkb-binding-intervals">Visulazing ChIP-seq read coverage across NFKB binding intervals</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/recipes/nfkb_profile.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>