
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Expression.py - wrap various differential expression tools &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Modules" href="../modules.html" />
    <link rel="next" title="Fastq.py - methods for dealing with fastq files" href="Fastq.html" />
    <link rel="prev" title="CSV2DB.py - utilities for uploading a table to database" href="CSV2DB.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Fastq.html" title="Fastq.py - methods for dealing with fastq files"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="CSV2DB.html" title="CSV2DB.py - utilities for uploading a table to database"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" accesskey="U">Modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-Expression"></span><div class="section" id="expression-py-wrap-various-differential-expression-tools">
<h1>Expression.py - wrap various differential expression tools<a class="headerlink" href="#expression-py-wrap-various-differential-expression-tools" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>This module provides tools for differential expression analysis
for a variety of methods.</p>
<p>Methods implemented are:</p>
<blockquote>
<div>DESeq
EdgeR
cuffdiff
ttest</div></blockquote>
<p>The aim of this module is to run these individual tools and
output a table in a common format.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="documentation">
<h2>Documentation<a class="headerlink" href="#documentation" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="Expression.buildProbeset2Gene">
<tt class="descclassname">Expression.</tt><tt class="descname">buildProbeset2Gene</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em>, <em>database='hgu133plus2.db'</em>, <em>mapping='hgu133plus2ENSEMBL'</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.buildProbeset2Gene" title="Permalink to this definition">¶</a></dt>
<dd><p>build map relating a probeset to an ENSEMBL gene_id</p>
</dd></dl>

<dl class="class">
<dt id="Expression.GeneExpressionResult">
<em class="property">class </em><tt class="descclassname">Expression.</tt><tt class="descname">GeneExpressionResult</tt><a class="headerlink" href="#Expression.GeneExpressionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">tuple</span></tt></p>
<p>GeneExpressionResult(test_id, treatment_name, treatment_mean, treatment_std, control_name, control_mean, control_std, pvalue, qvalue, l2fold, fold, transformed_l2fold, significant, status)</p>
<dl class="attribute">
<dt id="Expression.GeneExpressionResult.control_mean">
<tt class="descname">control_mean</tt><a class="headerlink" href="#Expression.GeneExpressionResult.control_mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 5</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.control_name">
<tt class="descname">control_name</tt><a class="headerlink" href="#Expression.GeneExpressionResult.control_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 4</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.control_std">
<tt class="descname">control_std</tt><a class="headerlink" href="#Expression.GeneExpressionResult.control_std" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 6</p>
</dd></dl>

<dl class="method">
<dt id="Expression.GeneExpressionResult.count">
<tt class="descname">count</tt><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span> &rarr; integer -- return number of occurrences of value<a class="headerlink" href="#Expression.GeneExpressionResult.count" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.fold">
<tt class="descname">fold</tt><a class="headerlink" href="#Expression.GeneExpressionResult.fold" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 10</p>
</dd></dl>

<dl class="method">
<dt id="Expression.GeneExpressionResult.index">
<tt class="descname">index</tt><span class="sig-paren">(</span><em>value</em><span class="optional">[</span>, <em>start</em><span class="optional">[</span>, <em>stop</em><span class="optional">]</span><span class="optional">]</span><span class="sig-paren">)</span> &rarr; integer -- return first index of value.<a class="headerlink" href="#Expression.GeneExpressionResult.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Raises ValueError if the value is not present.</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.l2fold">
<tt class="descname">l2fold</tt><a class="headerlink" href="#Expression.GeneExpressionResult.l2fold" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 9</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.pvalue">
<tt class="descname">pvalue</tt><a class="headerlink" href="#Expression.GeneExpressionResult.pvalue" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 7</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.qvalue">
<tt class="descname">qvalue</tt><a class="headerlink" href="#Expression.GeneExpressionResult.qvalue" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 8</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.significant">
<tt class="descname">significant</tt><a class="headerlink" href="#Expression.GeneExpressionResult.significant" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 12</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.status">
<tt class="descname">status</tt><a class="headerlink" href="#Expression.GeneExpressionResult.status" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 13</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.test_id">
<tt class="descname">test_id</tt><a class="headerlink" href="#Expression.GeneExpressionResult.test_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.transformed_l2fold">
<tt class="descname">transformed_l2fold</tt><a class="headerlink" href="#Expression.GeneExpressionResult.transformed_l2fold" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 11</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.treatment_mean">
<tt class="descname">treatment_mean</tt><a class="headerlink" href="#Expression.GeneExpressionResult.treatment_mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.treatment_name">
<tt class="descname">treatment_name</tt><a class="headerlink" href="#Expression.GeneExpressionResult.treatment_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.GeneExpressionResult.treatment_std">
<tt class="descname">treatment_std</tt><a class="headerlink" href="#Expression.GeneExpressionResult.treatment_std" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 3</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="Expression.writeExpressionResults">
<tt class="descclassname">Expression.</tt><tt class="descname">writeExpressionResults</tt><span class="sig-paren">(</span><em>outfile</em>, <em>result</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.writeExpressionResults" title="Permalink to this definition">¶</a></dt>
<dd><p>output expression results table.</p>
</dd></dl>

<dl class="class">
<dt id="Expression.WelchsTTest">
<em class="property">class </em><tt class="descclassname">Expression.</tt><tt class="descname">WelchsTTest</tt><a class="headerlink" href="#Expression.WelchsTTest" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>base class for computing expression differences.</p>
</dd></dl>

<dl class="class">
<dt id="Expression.SAMR">
<em class="property">class </em><tt class="descclassname">Expression.</tt><tt class="descname">SAMR</tt><a class="headerlink" href="#Expression.SAMR" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>SAM analysis of microarray data.</p>
<p>Use the Two-Class Unpaired Case Assuming Unequal Variances.</p>
<p>This uses the samr library.</p>
<p>Significant genes are either called at <em>fdr</em> or the
top <em>ngenes</em> are returned.</p>
<p><em>treatments</em> and <em>control</em> are arrays of
arrays of expression values.</p>
<p>See</p>
<p><a class="reference external" href="https://stat.ethz.ch/pipermail/bioconductor/2008-July/023251.html">https://stat.ethz.ch/pipermail/bioconductor/2008-July/023251.html</a></p>
<p>for an explanation of the differences between siggens SAM
and Excel SAM. This version is parameterised to reproduce Excel SAM
by setting:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">var</span><span class="o">.</span><span class="n">equal</span> <span class="o">=</span> <span class="n">TRUE</span>
<span class="n">med</span> <span class="o">=</span> <span class="n">TRUE</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">SAM requires log2 scaled expression levels.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="Expression.SAM">
<em class="property">class </em><tt class="descclassname">Expression.</tt><tt class="descname">SAM</tt><a class="headerlink" href="#Expression.SAM" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>SAM analysis of microarray data.</p>
<p>Use the Two-Class Unpaired Case Assuming Unequal Variances.</p>
<p>This uses the siggenes library. Note that there is also
an rsam package at:</p>
<p><a class="reference external" href="http://rss.acs.unt.edu/Rdoc/library/samr/html/samr.html">http://rss.acs.unt.edu/Rdoc/library/samr/html/samr.html</a></p>
<p>Significant genes are either called at <em>fdr</em> or the
top <em>ngenes</em> are returned.</p>
<p><em>treatments</em> and <em>control</em> are arrays of
arrays of expression values.</p>
<p>See
<a class="reference external" href="https://stat.ethz.ch/pipermail/bioconductor/2008-July/023251.html">https://stat.ethz.ch/pipermail/bioconductor/2008-July/023251.html</a></p>
<p>for an explanation of the differences between siggens SAM
and Excel SAM. To parameterize the FDR to excel sam, set the
flag <em>use_excel_sam</em>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">SAM requires log2 scaled expression levels.</p>
</div>
<p>I ran into trouble using this library. I was not able to
reproduce the same results from the original SAM study getting
differences in d and in the fdr.</p>
<p>fold change is treatment / control.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.loadTagData">
<tt class="descclassname">Expression.</tt><tt class="descname">loadTagData</tt><span class="sig-paren">(</span><em>tags_filename</em>, <em>design_filename</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.loadTagData" title="Permalink to this definition">¶</a></dt>
<dd><p>load tag data for deseq/edger analysis.</p>
<p><em>Infile</em> is a tab-separated file with counts.</p>
<p><em>design_file</em> is a tab-separated file with the
experimental design with four columns:</p>
<div class="highlight-python"><div class="highlight"><pre>track   include group   pair
CW-CD14-R1      0       CD14    1
CW-CD14-R2      0       CD14    1
CW-CD14-R3      1       CD14    1
CW-CD4-R1       1       CD4     1
FM-CD14-R1      1       CD14    2
FM-CD4-R2       0       CD4     2
FM-CD4-R3       0       CD4     2
FM-CD4-R4       0       CD4     2
</pre></div>
</div>
<dl class="docutils">
<dt>track</dt>
<dd>name of track - should correspond to column header in <em>infile</em></dd>
<dt>include</dt>
<dd>flag to indicate whether or not to include this data</dd>
<dt>group</dt>
<dd>group indicator - experimental group</dd>
<dt>pair</dt>
<dd>pair that sample belongs to (for paired tests)</dd>
</dl>
<p>This method creates various R objects:</p>
<p>countsTable : data frame with counts.
groups : vector with groups
pairs  : vector with pairs</p>
</dd></dl>

<dl class="function">
<dt id="Expression.filterTagData">
<tt class="descclassname">Expression.</tt><tt class="descname">filterTagData</tt><span class="sig-paren">(</span><em>filter_min_counts_per_row=1</em>, <em>filter_min_counts_per_sample=10</em>, <em>filter_percentile_rowsums=0</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.filterTagData" title="Permalink to this definition">¶</a></dt>
<dd><p>filter tag data.</p>
<ul>
<li><p class="first">remove rows with at least x number of counts</p>
</li>
<li><p class="first">remove samples with a maximum of <em>min_sample_counts</em></p>
</li>
<li><dl class="first docutils">
<dt>remove the lowest percentile of rows in the table, sorted</dt>
<dd><p class="first last">by total tags per row</p>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="Expression.groupTagData">
<tt class="descclassname">Expression.</tt><tt class="descname">groupTagData</tt><span class="sig-paren">(</span><em>ref_group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.groupTagData" title="Permalink to this definition">¶</a></dt>
<dd><p>compute groups and pairs from tag data table.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.plotCorrelationHeatmap">
<tt class="descclassname">Expression.</tt><tt class="descname">plotCorrelationHeatmap</tt><span class="sig-paren">(</span><em>method='correlation'</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.plotCorrelationHeatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>plot a heatmap of correlations derived from
countsTable.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.plotPairs">
<tt class="descclassname">Expression.</tt><tt class="descname">plotPairs</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Expression.plotPairs" title="Permalink to this definition">¶</a></dt>
<dd><p>requires counts table</p>
</dd></dl>

<dl class="function">
<dt id="Expression.plotPCA">
<tt class="descclassname">Expression.</tt><tt class="descname">plotPCA</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Expression.plotPCA" title="Permalink to this definition">¶</a></dt>
<dd><p>plot a PCA plot from countsTable.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.runEdgeR">
<tt class="descclassname">Expression.</tt><tt class="descname">runEdgeR</tt><span class="sig-paren">(</span><em>outfile</em>, <em>outfile_prefix='edger.'</em>, <em>fdr=0.1</em>, <em>prefix=''</em>, <em>dispersion=None</em>, <em>ref_group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.runEdgeR" title="Permalink to this definition">¶</a></dt>
<dd><p>run EdgeR on countsTable.</p>
<p>Results are stored in <em>outfile</em> and files prefixed by <em>outfile_prefix</em>.</p>
<p>The dispersion is usually measuered from replicates. If there are no 
replicates, you need to set the <em>dispersion</em> explicitely.</p>
<p>See page 13 of the EdgeR user guide:</p>
<div class="highlight-python"><div class="highlight"><pre>2. Simply pick a reasonable dispersion value, based on your experience with similar data,
and use that. Although subjective, this is still more defensible than assuming Poisson
variation. Typical values are dispersion=0.4 for human data, dispersion=0.1 for data
on genetically identical model organisms or dispersion=0.01 for technical replicates.
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="Expression.deseqPlotSizeFactors">
<tt class="descclassname">Expression.</tt><tt class="descname">deseqPlotSizeFactors</tt><span class="sig-paren">(</span><em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.deseqPlotSizeFactors" title="Permalink to this definition">¶</a></dt>
<dd><p>plot size factors - requires cds object.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.deseqOutputSizeFactors">
<tt class="descclassname">Expression.</tt><tt class="descname">deseqOutputSizeFactors</tt><span class="sig-paren">(</span><em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.deseqOutputSizeFactors" title="Permalink to this definition">¶</a></dt>
<dd><p>output size factors - requires cds object.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.deseqPlotCorrelationHeatmap">
<tt class="descclassname">Expression.</tt><tt class="descname">deseqPlotCorrelationHeatmap</tt><span class="sig-paren">(</span><em>outfile</em>, <em>vsd</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.deseqPlotCorrelationHeatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>plot a heatmap</p>
<p>Use variance stabilized data in object vsd.
Should be &#8216;blind&#8217;, as then the transform is
not informed by the experimental design.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.deseqPlotGeneHeatmap">
<tt class="descclassname">Expression.</tt><tt class="descname">deseqPlotGeneHeatmap</tt><span class="sig-paren">(</span><em>outfile</em>, <em>data</em>, <em>Rowv=False</em>, <em>Colv=False</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.deseqPlotGeneHeatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>plot a heatmap of all genes</p>
<p>Use variance stabilized data in object vsd.
Should be &#8216;blind&#8217;, as then the transform is
not informed by the experimental design.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.deseqPlotPCA">
<tt class="descclassname">Expression.</tt><tt class="descname">deseqPlotPCA</tt><span class="sig-paren">(</span><em>outfile</em>, <em>vsd</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.deseqPlotPCA" title="Permalink to this definition">¶</a></dt>
<dd><p>plot a PCA</p>
<p>Use variance stabilized data in object vsd.
Should be &#8216;blind&#8217;, as then the transform is
not informed by the experimental design.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.deseqPlotPairs">
<tt class="descclassname">Expression.</tt><tt class="descname">deseqPlotPairs</tt><span class="sig-paren">(</span><em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.deseqPlotPairs" title="Permalink to this definition">¶</a></dt>
<dd><p>requires counts table</p>
</dd></dl>

<dl class="function">
<dt id="Expression.deseqPlotPvaluesAgainstRowsums">
<tt class="descclassname">Expression.</tt><tt class="descname">deseqPlotPvaluesAgainstRowsums</tt><span class="sig-paren">(</span><em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.deseqPlotPvaluesAgainstRowsums" title="Permalink to this definition">¶</a></dt>
<dd><p>plot pvalues against row sum rank.</p>
<p>This plot is useful to see if quantile filtering could
be applied.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.deseqParseResults">
<tt class="descclassname">Expression.</tt><tt class="descname">deseqParseResults</tt><span class="sig-paren">(</span><em>control_name</em>, <em>treatment_name</em>, <em>fdr</em>, <em>vsd=False</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.deseqParseResults" title="Permalink to this definition">¶</a></dt>
<dd><p>parse deseq output.</p>
<p>retrieve deseq results from object &#8216;res&#8217; in R namespace.</p>
<p>The &#8216;res&#8217; object is a dataframe with the following columns (from the DESeq manual):</p>
<dl class="docutils">
<dt>id: The ID of the observable, taken from the row names of the</dt>
<dd>counts slots.</dd>
<dt>baseMean: The base mean (i.e., mean of the counts divided by the size</dt>
<dd>factors) for the counts for both conditions</dd>
<dt>baseMeanA: The base mean (i.e., mean of the counts divided by the size</dt>
<dd>factors) for the counts for condition A</dd>
</dl>
<p>baseMeanB: The base mean for condition B</p>
<p>foldChange: The ratio meanB/meanA</p>
<p>log2FoldChange: The log2 of the fold change</p>
<p>pval: The p value for rejecting the null hypothesis &#8216;meanA==meanB&#8217;</p>
<dl class="docutils">
<dt>padj: The adjusted p values (adjusted with &#8216;p.adjust( pval,</dt>
<dd>method=&#8221;BH&#8221;)&#8217;)</dd>
<dt>vsd_log2FoldChange: The log2 fold change after variance stabilization.</dt>
<dd>This data field is not part of DESeq proper, but has been added
in this module in the runDESeq() method.</dd>
</dl>
<p>Here, &#8216;conditionA&#8217; is &#8216;control&#8217; and &#8216;conditionB&#8217; is &#8216;treatment&#8217;
such that a foldChange of 2 means that treatment is twice
upregulated compared to control.</p>
<p>Returns a list of results.</p>
<p>If vsd is True, the log fold change will be computed from the variance
stabilized data.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.runDESeq">
<tt class="descclassname">Expression.</tt><tt class="descname">runDESeq</tt><span class="sig-paren">(</span><em>outfile</em>, <em>outfile_prefix='deseq.'</em>, <em>fdr=0.1</em>, <em>prefix=''</em>, <em>fit_type='parametric'</em>, <em>dispersion_method='pooled'</em>, <em>sharing_mode='maximum'</em>, <em>ref_group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.runDESeq" title="Permalink to this definition">¶</a></dt>
<dd><p>run DESeq on countsTable.</p>
<p>Results are stored in <em>outfile</em> and files prefixed by <em>outfile_prefix</em>.</p>
<p>The current analysis follows the analysis as outlined in version
1.14.0</p>
<p>DESeq ignores any pair information in the design matrix.</p>
<p>The output is treatment and control. Fold change values are
computed as treatment divided by control.</p>
</dd></dl>

<dl class="class">
<dt id="Expression.Design">
<em class="property">class </em><tt class="descclassname">Expression.</tt><tt class="descname">Design</tt><a class="headerlink" href="#Expression.Design" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">tuple</span></tt></p>
<p>Design(include, group, pair)</p>
<dl class="method">
<dt id="Expression.Design.count">
<tt class="descname">count</tt><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span> &rarr; integer -- return number of occurrences of value<a class="headerlink" href="#Expression.Design.count" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Expression.Design.group">
<tt class="descname">group</tt><a class="headerlink" href="#Expression.Design.group" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.Design.include">
<tt class="descname">include</tt><a class="headerlink" href="#Expression.Design.include" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="method">
<dt id="Expression.Design.index">
<tt class="descname">index</tt><span class="sig-paren">(</span><em>value</em><span class="optional">[</span>, <em>start</em><span class="optional">[</span>, <em>stop</em><span class="optional">]</span><span class="optional">]</span><span class="sig-paren">)</span> &rarr; integer -- return first index of value.<a class="headerlink" href="#Expression.Design.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Raises ValueError if the value is not present.</p>
</dd></dl>

<dl class="attribute">
<dt id="Expression.Design.pair">
<tt class="descname">pair</tt><a class="headerlink" href="#Expression.Design.pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="Expression.readDesignFile">
<tt class="descclassname">Expression.</tt><tt class="descname">readDesignFile</tt><span class="sig-paren">(</span><em>design_file</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.readDesignFile" title="Permalink to this definition">¶</a></dt>
<dd><p>reads a design file.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.plotTagStats">
<tt class="descclassname">Expression.</tt><tt class="descname">plotTagStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>design_file</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.plotTagStats" title="Permalink to this definition">¶</a></dt>
<dd><p>provide summary plots for tag data.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.plotDETagStats">
<tt class="descclassname">Expression.</tt><tt class="descname">plotDETagStats</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.plotDETagStats" title="Permalink to this definition">¶</a></dt>
<dd><p>provide summary plots for tag data.</p>
<p>Stratify boxplots and densities according to differential expression calls.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.parseCuffdiff">
<tt class="descclassname">Expression.</tt><tt class="descname">parseCuffdiff</tt><span class="sig-paren">(</span><em>infile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.parseCuffdiff" title="Permalink to this definition">¶</a></dt>
<dd><p>parse a cuffdiff .diff output file.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.loadCuffdiff">
<tt class="descclassname">Expression.</tt><tt class="descname">loadCuffdiff</tt><span class="sig-paren">(</span><em>infile</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.loadCuffdiff" title="Permalink to this definition">¶</a></dt>
<dd><p>load results from differential expression analysis and produce
summary plots.</p>
<p>Note: converts from ln(fold change) to log2 fold change.</p>
<p>The cuffdiff output is parsed.</p>
<p>Pairwise comparisons in which one gene is not expressed (fpkm &lt; fpkm_silent)
are set to status &#8216;NOCALL&#8217;. These transcripts might nevertheless be significant.</p>
<p>This requires the cummeRbund library to be present in R.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.runCuffdiff">
<tt class="descclassname">Expression.</tt><tt class="descname">runCuffdiff</tt><span class="sig-paren">(</span><em>bamfiles</em>, <em>design_file</em>, <em>geneset_file</em>, <em>outfile</em>, <em>cuffdiff_options=''</em>, <em>threads=4</em>, <em>fdr=0.1</em>, <em>mask_file=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.runCuffdiff" title="Permalink to this definition">¶</a></dt>
<dd><p>estimate differential expression using cuffdiff.</p>
<dl class="docutils">
<dt>infiles</dt>
<dd>bam files</dd>
<dt>geneset_file</dt>
<dd>geneset to use for the analysis</dd>
<dt>design_file</dt>
<dd>design file describing which differential expression to test</dd>
</dl>
<p>Replicates within each track are grouped.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.runMockAnalysis">
<tt class="descclassname">Expression.</tt><tt class="descname">runMockAnalysis</tt><span class="sig-paren">(</span><em>outfile</em>, <em>outfile_prefix</em>, <em>ref_group=None</em>, <em>pseudo_counts=0</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.runMockAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>run a mock analysis on a count table.</p>
<p>compute fold enrichment values, but do not normalize or
perform any test.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.outputTagSummary">
<tt class="descclassname">Expression.</tt><tt class="descname">outputTagSummary</tt><span class="sig-paren">(</span><em>filename_tags</em>, <em>outfile</em>, <em>output_filename_pattern</em>, <em>filename_design=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.outputTagSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>output summary values for a count table.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.dumpTagData">
<tt class="descclassname">Expression.</tt><tt class="descname">dumpTagData</tt><span class="sig-paren">(</span><em>filename_tags</em>, <em>filename_design</em>, <em>outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.dumpTagData" title="Permalink to this definition">¶</a></dt>
<dd><p>output filtered tag table.</p>
</dd></dl>

<dl class="function">
<dt id="Expression.loadTagDataPandas">
<tt class="descclassname">Expression.</tt><tt class="descname">loadTagDataPandas</tt><span class="sig-paren">(</span><em>tags_filename</em>, <em>design_filename</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.loadTagDataPandas" title="Permalink to this definition">¶</a></dt>
<dd><p>load tag data for deseq/edger analysis.</p>
<p><em>Infile</em> is a tab-separated file with counts.</p>
<p><em>design_file</em> is a tab-separated file with the
experimental design with four columns:</p>
<div class="highlight-python"><div class="highlight"><pre>track   include group   pair
CW-CD14-R1      0       CD14    1
CW-CD14-R2      0       CD14    1
CW-CD14-R3      1       CD14    1
CW-CD4-R1       1       CD4     1
FM-CD14-R1      1       CD14    2
FM-CD4-R2       0       CD4     2
FM-CD4-R3       0       CD4     2
FM-CD4-R4       0       CD4     2
</pre></div>
</div>
<dl class="docutils">
<dt>track</dt>
<dd>name of track - should correspond to column header in <em>infile</em></dd>
<dt>include</dt>
<dd>flag to indicate whether or not to include this data</dd>
<dt>group</dt>
<dd>group indicator - experimental group</dd>
<dt>pair</dt>
<dd>pair that sample belongs to (for paired tests)</dd>
</dl>
<p>This method creates various R objects:</p>
<p>countsTable : data frame with counts.
groups : vector with groups
pairs  : vector with pairs</p>
</dd></dl>

<dl class="function">
<dt id="Expression.filterTagDataPandas">
<tt class="descclassname">Expression.</tt><tt class="descname">filterTagDataPandas</tt><span class="sig-paren">(</span><em>counts_table</em>, <em>design_table</em>, <em>filter_min_counts_per_row=1</em>, <em>filter_min_counts_per_sample=10</em>, <em>filter_percentile_rowsums=0</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.filterTagDataPandas" title="Permalink to this definition">¶</a></dt>
<dd><p>filter tag data.</p>
<ul>
<li><p class="first">remove rows with at least x number of counts</p>
</li>
<li><p class="first">remove samples with a maximum of <em>min_sample_counts</em></p>
</li>
<li><dl class="first docutils">
<dt>remove the lowest percentile of rows in the table, sorted</dt>
<dd><p class="first last">by total tags per row</p>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="Expression.runTTest">
<tt class="descclassname">Expression.</tt><tt class="descname">runTTest</tt><span class="sig-paren">(</span><em>outfile</em>, <em>outfile_prefix</em>, <em>fdr</em>, <em>ref_group=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Expression.runTTest" title="Permalink to this definition">¶</a></dt>
<dd><p>apply a ttest on the data.</p>
<p>For the T-test it is best to use FPKM values as
this method does not perform any library normalization.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Expression.py - wrap various differential expression tools</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#documentation">Documentation</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="CSV2DB.html"
                        title="previous chapter">CSV2DB.py - utilities for uploading a table to database</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Fastq.html"
                        title="next chapter">Fastq.py - methods for dealing with fastq files</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/modules/Expression.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Fastq.html" title="Fastq.py - methods for dealing with fastq files"
             >next</a> |</li>
        <li class="right" >
          <a href="CSV2DB.html" title="CSV2DB.py - utilities for uploading a table to database"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" >Modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>