
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>GO.py - compute GO enrichment from gene lists &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Modules" href="../modules.html" />
    <link rel="next" title="GenomicIO.py - Subroutines for working on I/O of large genomic files" href="GenomicIO.html" />
    <link rel="prev" title="Bed.py - Tools for working with bed files" href="Bed.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="GenomicIO.html" title="GenomicIO.py - Subroutines for working on I/O of large genomic files"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Bed.html" title="Bed.py - Tools for working with bed files"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" accesskey="U">Modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-GO"></span><div class="section" id="go-py-compute-go-enrichment-from-gene-lists">
<h1>GO.py - compute GO enrichment from gene lists<a class="headerlink" href="#go-py-compute-go-enrichment-from-gene-lists" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="exception">
<dt id="GO.Error">
<em class="property">exception </em><tt class="descclassname">GO.</tt><tt class="descname">Error</tt><a class="headerlink" href="#GO.Error" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/2.7/library/exceptions.html#exceptions.Exception" title="(in Python v2.7)"><tt class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></tt></a></p>
<p>Base class for exceptions in this module.</p>
</dd></dl>

<dl class="exception">
<dt id="GO.InputError">
<em class="property">exception </em><tt class="descclassname">GO.</tt><tt class="descname">InputError</tt><span class="sig-paren">(</span><em>message</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.InputError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#GO.Error" title="GO.Error"><tt class="xref py py-class docutils literal"><span class="pre">GO.Error</span></tt></a></p>
<p>Exception raised for errors in the input.</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd>expression &#8211; input expression in which the error occurred
message &#8211; explanation of the error</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="GO.readOntology">
<tt class="descclassname">GO.</tt><tt class="descname">readOntology</tt><span class="sig-paren">(</span><em>infile</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.readOntology" title="Permalink to this definition">¶</a></dt>
<dd><p>read ontology in OBO format from infile.</p>
<p>returns a dictionary of Ontology entries.</p>
</dd></dl>

<dl class="class">
<dt id="GO.GOSample">
<em class="property">class </em><tt class="descclassname">GO.</tt><tt class="descname">GOSample</tt><span class="sig-paren">(</span><em>mmin</em>, <em>mmax</em>, <em>mmean</em>, <em>mstddev</em>, <em>mprobovers</em>, <em>mprobunders</em>, <em>counts</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.GOSample" title="Permalink to this definition">¶</a></dt>
<dd><p>store results from sampling.</p>
</dd></dl>

<dl class="class">
<dt id="GO.GOResults">
<em class="property">class </em><tt class="descclassname">GO.</tt><tt class="descname">GOResults</tt><a class="headerlink" href="#GO.GOResults" title="Permalink to this definition">¶</a></dt>
<dd><p>container for go results.</p>
</dd></dl>

<dl class="function">
<dt id="GO.FilterByGOIds">
<tt class="descclassname">GO.</tt><tt class="descname">FilterByGOIds</tt><span class="sig-paren">(</span><em>gene2go</em>, <em>go2info</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.FilterByGOIds" title="Permalink to this definition">¶</a></dt>
<dd><p>filter gene_id to go_id lookup by a list of go_ids</p>
<p>returns a new gene2go mapping.</p>
<p>used to restrict GO terms to GO_slim and remove alternates</p>
<p>gene2go                             # starting set, map of genes to go terms
go2info                             # alt ids are repeats of superceding ids</p>
</dd></dl>

<dl class="function">
<dt id="GO.MapGO2Slims">
<tt class="descclassname">GO.</tt><tt class="descname">MapGO2Slims</tt><span class="sig-paren">(</span><em>gene2go</em>, <em>go2slim</em>, <em>ontology=None</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.MapGO2Slims" title="Permalink to this definition">¶</a></dt>
<dd><p>filter gene2go lookup by a list of go_ids in go2slim.</p>
<p>gene2go: map of genes to go terms
go2slim: map of go categories to goslim go categories</p>
<p>If ontology is given, missing descriptions of go entries
are added from the ontology.</p>
<p>returns a new gene2go mapping.</p>
</dd></dl>

<dl class="function">
<dt id="GO.GetGOSlims">
<tt class="descclassname">GO.</tt><tt class="descname">GetGOSlims</tt><span class="sig-paren">(</span><em>infile</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.GetGOSlims" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a map of go identifiers to slim categories</p>
<p>Input is the output of Chris Mungal&#8217;s map2slim.pl.</p>
</dd></dl>

<dl class="function">
<dt id="GO.GetGOFrequencies">
<tt class="descclassname">GO.</tt><tt class="descname">GetGOFrequencies</tt><span class="sig-paren">(</span><em>gene2go</em>, <em>genes</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.GetGOFrequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of each go category in gene list.</p>
<p>return a tuple containing:
* the total number of GO categories found.
* dictionary of counts per GO category
* dictionary of genes found with GO categories</p>
</dd></dl>

<dl class="function">
<dt id="GO.AnalyseGO">
<tt class="descclassname">GO.</tt><tt class="descname">AnalyseGO</tt><span class="sig-paren">(</span><em>gene2go</em>, <em>genes</em>, <em>genes_background=()</em>, <em>do_probabilities=True</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.AnalyseGO" title="Permalink to this definition">¶</a></dt>
<dd><p>analyse go ids.</p>
<p>goids: list of goids to analyse
genes: sample set of genes
genes_background: background set of genes (default: all)</p>
</dd></dl>

<dl class="function">
<dt id="GO.GetGOStatement">
<tt class="descclassname">GO.</tt><tt class="descname">GetGOStatement</tt><span class="sig-paren">(</span><em>go_type</em>, <em>database</em>, <em>species</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.GetGOStatement" title="Permalink to this definition">¶</a></dt>
<dd><p>build statement to get GO assignments for genes from ENSEMBL.</p>
</dd></dl>

<dl class="function">
<dt id="GO.ReadGene2GOFromDatabase">
<tt class="descclassname">GO.</tt><tt class="descname">ReadGene2GOFromDatabase</tt><span class="sig-paren">(</span><em>dbhandle</em>, <em>go_type</em>, <em>database</em>, <em>species</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.ReadGene2GOFromDatabase" title="Permalink to this definition">¶</a></dt>
<dd><p>read go assignments from ensembl database.</p>
<p>returns a dictionary of lists.
(one to many mapping of genes to GO categories)
and a dictionary of go-term to go information</p>
<p>Note: assumes that external_db_id for GO is 1000</p>
</dd></dl>

<dl class="function">
<dt id="GO.DumpGOFromDatabase">
<tt class="descclassname">GO.</tt><tt class="descname">DumpGOFromDatabase</tt><span class="sig-paren">(</span><em>outfile</em>, <em>dbhandle</em>, <em>options</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.DumpGOFromDatabase" title="Permalink to this definition">¶</a></dt>
<dd><p>read go assignments from database.</p>
<p>and dump them into a flatfile.
(one to many mapping of genes to GO categories)
and a dictionary of go-term to go information</p>
</dd></dl>

<dl class="function">
<dt id="GO.ReadGene2GOFromFile">
<tt class="descclassname">GO.</tt><tt class="descname">ReadGene2GOFromFile</tt><span class="sig-paren">(</span><em>infile</em>, <em>synonyms={}</em>, <em>obsolete={}</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.ReadGene2GOFromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>reads GO mappings for all go_types from a
file.</p>
<p>If synonyms is given, goids in synynoms will be translated.
Terms in <em>obsolete</em> will be discarded.</p>
<p>returns two maps: gene2go maps genes to go categories
and go2info maps go categories to information.</p>
</dd></dl>

<dl class="function">
<dt id="GO.CountGO">
<tt class="descclassname">GO.</tt><tt class="descname">CountGO</tt><span class="sig-paren">(</span><em>gene2go</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.CountGO" title="Permalink to this definition">¶</a></dt>
<dd><p>count number of genes and go categories in mapping.</p>
</dd></dl>

<dl class="function">
<dt id="GO.removeCategories">
<tt class="descclassname">GO.</tt><tt class="descname">removeCategories</tt><span class="sig-paren">(</span><em>gene2go</em>, <em>categories</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.removeCategories" title="Permalink to this definition">¶</a></dt>
<dd><p>remove all genes that map to <em>categories</em>.</p>
</dd></dl>

<dl class="function">
<dt id="GO.countGOs">
<tt class="descclassname">GO.</tt><tt class="descname">countGOs</tt><span class="sig-paren">(</span><em>gene2gos</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.countGOs" title="Permalink to this definition">¶</a></dt>
<dd><p>return map of number of genes and go categories in mapping.</p>
</dd></dl>

<dl class="function">
<dt id="GO.ReadGeneLists">
<tt class="descclassname">GO.</tt><tt class="descname">ReadGeneLists</tt><span class="sig-paren">(</span><em>filename_genes</em>, <em>gene_pattern=None</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.ReadGeneLists" title="Permalink to this definition">¶</a></dt>
<dd><p>read gene lists from filename in matrix.</p>
<p>returns a tuple (list of all genes, dictionary of gene lists)</p>
</dd></dl>

<dl class="function">
<dt id="GO.buildGO2Genes">
<tt class="descclassname">GO.</tt><tt class="descname">buildGO2Genes</tt><span class="sig-paren">(</span><em>gene2gos</em>, <em>ancestors=None</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.buildGO2Genes" title="Permalink to this definition">¶</a></dt>
<dd><p>invert the dictionary genes2go.</p>
<p>If ancestors is given, add missing ancestral information.</p>
</dd></dl>

<dl class="function">
<dt id="GO.GetCode">
<tt class="descclassname">GO.</tt><tt class="descname">GetCode</tt><span class="sig-paren">(</span><em>v</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.GetCode" title="Permalink to this definition">¶</a></dt>
<dd><p>return a code for over/underrepresentation.</p>
</dd></dl>

<dl class="function">
<dt id="GO.convertGo2Goslim">
<tt class="descclassname">GO.</tt><tt class="descname">convertGo2Goslim</tt><span class="sig-paren">(</span><em>options</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.convertGo2Goslim" title="Permalink to this definition">¶</a></dt>
<dd><p>read gene list with GO assignments and convert to GO slim categories.</p>
</dd></dl>

<dl class="function">
<dt id="GO.outputResults">
<tt class="descclassname">GO.</tt><tt class="descname">outputResults</tt><span class="sig-paren">(</span><em>outfile</em>, <em>pairs</em>, <em>go2info</em>, <em>options</em>, <em>fdrs=None</em>, <em>samples=None</em>, <em>gene2go=None</em>, <em>foreground=None</em>, <em>gene2name=None</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.outputResults" title="Permalink to this definition">¶</a></dt>
<dd><p>output GO results to outfile.</p>
<p>If foreground is given, output a list of gene identifiers in the foreground.
If gene2name is given, output a columns with gene names (instead of identifiers)</p>
</dd></dl>

<dl class="function">
<dt id="GO.getFileName">
<tt class="descclassname">GO.</tt><tt class="descname">getFileName</tt><span class="sig-paren">(</span><em>options</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.getFileName" title="Permalink to this definition">¶</a></dt>
<dd><p>return a filename</p>
<p>Placeholders in filename are string-substituted with the 
dictionary in kwargs.</p>
</dd></dl>

<dl class="function">
<dt id="GO.buildMatrix">
<tt class="descclassname">GO.</tt><tt class="descname">buildMatrix</tt><span class="sig-paren">(</span><em>results</em>, <em>valuef</em>, <em>dtype=&lt;type 'float'&gt;</em>, <em>default=0</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.buildMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>build a matrix from a field in <em>results</em></p>
<p>The value stored in the matrix is accessed via <em>valuef</em>.</p>
</dd></dl>

<dl class="function">
<dt id="GO.selectSignificantResults">
<tt class="descclassname">GO.</tt><tt class="descname">selectSignificantResults</tt><span class="sig-paren">(</span><em>pairs</em>, <em>fdrs</em>, <em>options</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.selectSignificantResults" title="Permalink to this definition">¶</a></dt>
<dd><p>select a set of significant results.</p>
</dd></dl>

<dl class="function">
<dt id="GO.outputMultipleGeneListResults">
<tt class="descclassname">GO.</tt><tt class="descname">outputMultipleGeneListResults</tt><span class="sig-paren">(</span><em>results</em>, <em>all_genelists_with_results</em>, <em>test_ontology</em>, <em>go2info</em>, <em>options</em>, <em>section</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.outputMultipleGeneListResults" title="Permalink to this definition">¶</a></dt>
<dd><p>select a set of significant results.</p>
</dd></dl>

<dl class="function">
<dt id="GO.pairwiseGOEnrichment">
<tt class="descclassname">GO.</tt><tt class="descname">pairwiseGOEnrichment</tt><span class="sig-paren">(</span><em>results_per_genelist</em>, <em>labels</em>, <em>test_ontology</em>, <em>go2info</em>, <em>options</em><span class="sig-paren">)</span><a class="headerlink" href="#GO.pairwiseGOEnrichment" title="Permalink to this definition">¶</a></dt>
<dd><p>compute pairwise enrichment between sets.</p>
<p>The purpose of this method is to find if there are categories that are differently enriched
in a pair of gene lists.</p>
<p>The appropriate test here is the Chi-Squared test.</p>
<p>The assumption is that the background set is the same in all gene lists.</p>
<p>The workflow is thus:</p>
<div class="highlight-python"><div class="highlight"><pre>for each combination of two gene lists:
    for each GO category:
        get counts in foreground, total counts of foreground
        compute chi-square enrichment output
        save P-value
    apply fdr - output significant differences.
</pre></div>
</div>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">GO.py - compute GO enrichment from gene lists</a><ul>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Bed.html"
                        title="previous chapter">Bed.py - Tools for working with bed files</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="GenomicIO.html"
                        title="next chapter">GenomicIO.py - Subroutines for working on I/O of large genomic files</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/modules/GO.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="GenomicIO.html" title="GenomicIO.py - Subroutines for working on I/O of large genomic files"
             >next</a> |</li>
        <li class="right" >
          <a href="Bed.html" title="Bed.py - Tools for working with bed files"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" >Modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>