
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Predictor.py - running gene predictions with exonerate &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Modules" href="../modules.html" />
    <link rel="next" title="BlastAlignments.py - tools for working with alignments" href="BlastAlignments.html" />
    <link rel="prev" title="PredictionParser.py - Parser for exonerate/genewise output" href="PredictionParser.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="BlastAlignments.html" title="BlastAlignments.py - tools for working with alignments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="PredictionParser.html" title="PredictionParser.py - Parser for exonerate/genewise output"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" accesskey="U">Modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-Predictor"></span><div class="section" id="predictor-py-running-gene-predictions-with-exonerate">
<h1>Predictor.py - running gene predictions with exonerate<a class="headerlink" href="#predictor-py-running-gene-predictions-with-exonerate" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body">Andreas Heger</td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a></h2>
<dl class="exception">
<dt id="Predictor.Error">
<em class="property">exception </em><tt class="descclassname">Predictor.</tt><tt class="descname">Error</tt><a class="headerlink" href="#Predictor.Error" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/2.7/library/exceptions.html#exceptions.Exception" title="(in Python v2.7)"><tt class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></tt></a></p>
<p>Base class for exceptions in this module.</p>
</dd></dl>

<dl class="exception">
<dt id="Predictor.ParsingError">
<em class="property">exception </em><tt class="descclassname">Predictor.</tt><tt class="descname">ParsingError</tt><span class="sig-paren">(</span><em>message</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.ParsingError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Predictor.Error" title="Predictor.Error"><tt class="xref py py-class docutils literal"><span class="pre">Predictor.Error</span></tt></a></p>
<p>Exception raised for errors in the input.</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd>expression &#8211; input expression in which the error occurred
message &#8211; explanation of the error</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="Predictor.PredictorGenewise">
<em class="property">class </em><tt class="descclassname">Predictor.</tt><tt class="descname">PredictorGenewise</tt><span class="sig-paren">(</span><em>parser=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictorGenewise" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">Predictor.Predictor</span></tt></p>
<p>run genewise as predictor.</p>
<p>genewise only runs a single peptide sequence against a single
DNA sequences. genewisedb runs collections of peptide sequences
against collections of DNA sequences. Note that this might take a while
so it is prudent to spread it over the cluster.</p>
<dl class="method">
<dt id="Predictor.PredictorGenewise.getParser">
<tt class="descname">getParser</tt><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictorGenewise.getParser" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the parser object.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictorGenewise.predictFromRange">
<tt class="descname">predictFromRange</tt><span class="sig-paren">(</span><em>peptide_sequence</em>, <em>contig</em>, <em>strand</em>, <em>start</em>, <em>end</em>, <em>fasta</em>, <em>options=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictorGenewise.predictFromRange" title="Permalink to this definition">¶</a></dt>
<dd><p>predict from a peptide sequence within a region.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictorGenewise.predictFromSequences">
<tt class="descname">predictFromSequences</tt><span class="sig-paren">(</span><em>peptide_sequence</em>, <em>genomic_sequence</em>, <em>options=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictorGenewise.predictFromSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>predict from two explicitely given sequences.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictorGenewise.setCluster">
<tt class="descname">setCluster</tt><span class="sig-paren">(</span><em>cluster</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictorGenewise.setCluster" title="Permalink to this definition">¶</a></dt>
<dd><p>set cluster object.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictorGenewise.setDryRun">
<tt class="descname">setDryRun</tt><span class="sig-paren">(</span><em>flag=True</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictorGenewise.setDryRun" title="Permalink to this definition">¶</a></dt>
<dd><p>setup dry run.</p>
<p>Gene prediction is not run, but files are set up as if it were.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictorGenewise.setLogLevel">
<tt class="descname">setLogLevel</tt><span class="sig-paren">(</span><em>level</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictorGenewise.setLogLevel" title="Permalink to this definition">¶</a></dt>
<dd><p>set loglevel.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictorGenewise.setOutputOptions">
<tt class="descname">setOutputOptions</tt><span class="sig-paren">(</span><em>options</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictorGenewise.setOutputOptions" title="Permalink to this definition">¶</a></dt>
<dd><p>set output options.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Predictor.PredictionParser">
<em class="property">class </em><tt class="descclassname">Predictor.</tt><tt class="descname">PredictionParser</tt><span class="sig-paren">(</span><em>peptides=None</em>, <em>peptide=None</em>, <em>genomes=None</em>, <em>genome=None</em>, <em>filename_genome=None</em>, <em>filename_peptides=None</em>, <em>indexed_genome=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParser" title="Permalink to this definition">¶</a></dt>
<dd><p>base class of prediction parsers.</p>
<p>pass sequence data.</p>
<dl class="method">
<dt id="Predictor.PredictionParser.parse">
<tt class="descname">parse</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParser.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>parse lines.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParser.getPeptideSequence">
<tt class="descname">getPeptideSequence</tt><span class="sig-paren">(</span><em>query_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParser.getPeptideSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns peptide sequence for query_token.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParser.addGenomicSequence">
<tt class="descname">addGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sequence</em>, <em>strand='+'</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParser.addGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>add a sequence so that the parser nkows about it.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParser.deleteGenomicSequence">
<tt class="descname">deleteGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParser.deleteGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>remove a genomic sequence from parser.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParser.getGenomicSequence">
<tt class="descname">getGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sbjct_strand=None</em>, <em>sbjct_from=None</em>, <em>sbjct_to=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParser.getGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParser.getGenomicSequenceLength">
<tt class="descname">getGenomicSequenceLength</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParser.getGenomicSequenceLength" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Predictor.PredictionParserGenewise">
<em class="property">class </em><tt class="descclassname">Predictor.</tt><tt class="descname">PredictionParserGenewise</tt><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Predictor.PredictionParser" title="Predictor.PredictionParser"><tt class="xref py py-class docutils literal"><span class="pre">Predictor.PredictionParser</span></tt></a></p>
<p>Parse output from genewise.</p>
<dl class="method">
<dt id="Predictor.PredictionParserGenewise.parse">
<tt class="descname">parse</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>call the various parsers for a chunk of data.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.parseGFF">
<tt class="descname">parseGFF</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.parseGFF" title="Permalink to this definition">¶</a></dt>
<dd><p>extract strands from the gff section.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.parseSummary">
<tt class="descname">parseSummary</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.parseSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>parse summary entry of genewise.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.parseTranslation">
<tt class="descname">parseTranslation</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.parseTranslation" title="Permalink to this definition">¶</a></dt>
<dd><p>read a translation entry from genewise.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.parsePeptide">
<tt class="descname">parsePeptide</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.parsePeptide" title="Permalink to this definition">¶</a></dt>
<dd><p>read a peptide entry from genewise.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.parseAlb">
<tt class="descname">parseAlb</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.parseAlb" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse genewise logical block alignment and return aligment.</p>
<dl class="docutils">
<dt>( alignment of peptide to genome,</dt>
<dd>alignment of peptide to translation )</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.addGenomicSequence">
<tt class="descname">addGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sequence</em>, <em>strand='+'</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.addGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>add a sequence so that the parser nkows about it.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.deleteGenomicSequence">
<tt class="descname">deleteGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.deleteGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>remove a genomic sequence from parser.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.getGenomicSequence">
<tt class="descname">getGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sbjct_strand=None</em>, <em>sbjct_from=None</em>, <em>sbjct_to=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.getGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.getGenomicSequenceLength">
<tt class="descname">getGenomicSequenceLength</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.getGenomicSequenceLength" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewise.getPeptideSequence">
<tt class="descname">getPeptideSequence</tt><span class="sig-paren">(</span><em>query_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewise.getPeptideSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns peptide sequence for query_token.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Predictor.PredictionParserGenewiseHMM">
<em class="property">class </em><tt class="descclassname">Predictor.</tt><tt class="descname">PredictionParserGenewiseHMM</tt><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Predictor.PredictionParserGenewise" title="Predictor.PredictionParserGenewise"><tt class="xref py py-class docutils literal"><span class="pre">Predictor.PredictionParserGenewise</span></tt></a></p>
<p>Parse output from genewise.</p>
<p>This subclass reads the profile from the file to extract
the length of a query.</p>
<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.extractQueryLength">
<tt class="descname">extractQueryLength</tt><span class="sig-paren">(</span><em>filename_peptides</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.extractQueryLength" title="Permalink to this definition">¶</a></dt>
<dd><p>extract query length from HMM.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.addGenomicSequence">
<tt class="descname">addGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sequence</em>, <em>strand='+'</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.addGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>add a sequence so that the parser nkows about it.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.deleteGenomicSequence">
<tt class="descname">deleteGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.deleteGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>remove a genomic sequence from parser.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.getGenomicSequence">
<tt class="descname">getGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sbjct_strand=None</em>, <em>sbjct_from=None</em>, <em>sbjct_to=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.getGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.getGenomicSequenceLength">
<tt class="descname">getGenomicSequenceLength</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.getGenomicSequenceLength" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.getPeptideSequence">
<tt class="descname">getPeptideSequence</tt><span class="sig-paren">(</span><em>query_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.getPeptideSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns peptide sequence for query_token.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.parse">
<tt class="descname">parse</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>call the various parsers for a chunk of data.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.parseAlb">
<tt class="descname">parseAlb</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.parseAlb" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse genewise logical block alignment and return aligment.</p>
<dl class="docutils">
<dt>( alignment of peptide to genome,</dt>
<dd>alignment of peptide to translation )</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.parseGFF">
<tt class="descname">parseGFF</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.parseGFF" title="Permalink to this definition">¶</a></dt>
<dd><p>extract strands from the gff section.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.parsePeptide">
<tt class="descname">parsePeptide</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.parsePeptide" title="Permalink to this definition">¶</a></dt>
<dd><p>read a peptide entry from genewise.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.parseSummary">
<tt class="descname">parseSummary</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.parseSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>parse summary entry of genewise.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserGenewiseHMM.parseTranslation">
<tt class="descname">parseTranslation</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserGenewiseHMM.parseTranslation" title="Permalink to this definition">¶</a></dt>
<dd><p>read a translation entry from genewise.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Predictor.PredictionParserExonerate">
<em class="property">class </em><tt class="descclassname">Predictor.</tt><tt class="descname">PredictionParserExonerate</tt><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserExonerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Predictor.PredictionParser" title="Predictor.PredictionParser"><tt class="xref py py-class docutils literal"><span class="pre">Predictor.PredictionParser</span></tt></a></p>
<p>Parse output from exonerate.</p>
<dl class="method">
<dt id="Predictor.PredictionParserExonerate.addGenomicSequence">
<tt class="descname">addGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sequence</em>, <em>strand='+'</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserExonerate.addGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>add a sequence so that the parser nkows about it.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserExonerate.deleteGenomicSequence">
<tt class="descname">deleteGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserExonerate.deleteGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>remove a genomic sequence from parser.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserExonerate.getGenomicSequence">
<tt class="descname">getGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sbjct_strand=None</em>, <em>sbjct_from=None</em>, <em>sbjct_to=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserExonerate.getGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserExonerate.getGenomicSequenceLength">
<tt class="descname">getGenomicSequenceLength</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserExonerate.getGenomicSequenceLength" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserExonerate.getPeptideSequence">
<tt class="descname">getPeptideSequence</tt><span class="sig-paren">(</span><em>query_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserExonerate.getPeptideSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns peptide sequence for query_token.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Predictor.PredictionParserEcho">
<em class="property">class </em><tt class="descclassname">Predictor.</tt><tt class="descname">PredictionParserEcho</tt><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserEcho" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Predictor.PredictionParser" title="Predictor.PredictionParser"><tt class="xref py py-class docutils literal"><span class="pre">Predictor.PredictionParser</span></tt></a></p>
<p>simply output predictions - do not parse.</p>
<dl class="method">
<dt id="Predictor.PredictionParserEcho.parse">
<tt class="descname">parse</tt><span class="sig-paren">(</span><em>lines</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserEcho.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>call the various parsers for a chunk of data.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserEcho.addGenomicSequence">
<tt class="descname">addGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sequence</em>, <em>strand='+'</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserEcho.addGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>add a sequence so that the parser nkows about it.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserEcho.deleteGenomicSequence">
<tt class="descname">deleteGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserEcho.deleteGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>remove a genomic sequence from parser.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserEcho.getGenomicSequence">
<tt class="descname">getGenomicSequence</tt><span class="sig-paren">(</span><em>sbjct_token</em>, <em>sbjct_strand=None</em>, <em>sbjct_from=None</em>, <em>sbjct_to=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserEcho.getGenomicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserEcho.getGenomicSequenceLength">
<tt class="descname">getGenomicSequenceLength</tt><span class="sig-paren">(</span><em>sbjct_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserEcho.getGenomicSequenceLength" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a genomic sequence for given coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Predictor.PredictionParserEcho.getPeptideSequence">
<tt class="descname">getPeptideSequence</tt><span class="sig-paren">(</span><em>query_token</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.PredictionParserEcho.getPeptideSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>returns peptide sequence for query_token.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Predictor.TandemPredictor">
<em class="property">class </em><tt class="descclassname">Predictor.</tt><tt class="descname">TandemPredictor</tt><span class="sig-paren">(</span><em>predictor</em>, <em>upstream_queries=</em>, <span class="optional">[</span><span class="optional">]</span><em>downstream_queries=</em><span class="optional">[</span><span class="optional">]</span><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.TandemPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>run a tandem prediction</p>
<p>If a match is found, look for further related matches upstream or
downstream.</p>
<p>This is a hack that works for one additional query downstream or
upstream. To be fully workable, sort out reverse strand predictions
and overlapping predictions. Also, make sure that if several results
are obtained, they are not spanning each other.</p>
<p>run predictor on queries.</p>
<dl class="method">
<dt id="Predictor.TandemPredictor.runTandem">
<tt class="descname">runTandem</tt><span class="sig-paren">(</span><em>filename_peptides</em>, <em>filename_genome</em>, <em>options</em>, <em>sbjct_token</em>, <em>sbjct_strand</em>, <em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.TandemPredictor.runTandem" title="Permalink to this definition">¶</a></dt>
<dd><p>run a tandem prediction with region given by start, end.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Predictor.RecursivePredictor">
<em class="property">class </em><tt class="descclassname">Predictor.</tt><tt class="descname">RecursivePredictor</tt><span class="sig-paren">(</span><em>predictor</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.RecursivePredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>run recursive prediction.</p>
<p>Note: this will always run the full peptide set against the matched fragments,
so only use this if you have only a single input sequence/hmm and the matches
returned are not repetetive.</p>
</dd></dl>

<dl class="function">
<dt id="Predictor.getPredictor">
<tt class="descclassname">Predictor.</tt><tt class="descname">getPredictor</tt><span class="sig-paren">(</span><em>options</em>, <em>parser_options</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.getPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>setup predictor.</p>
</dd></dl>

<dl class="function">
<dt id="Predictor.runLocally">
<tt class="descclassname">Predictor.</tt><tt class="descname">runLocally</tt><span class="sig-paren">(</span><em>filename_peptides</em>, <em>filename_genome</em>, <em>options</em>, <em>parser_options</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.runLocally" title="Permalink to this definition">¶</a></dt>
<dd><p>run jobs locally.</p>
</dd></dl>

<dl class="function">
<dt id="Predictor.runOnCluster">
<tt class="descclassname">Predictor.</tt><tt class="descname">runOnCluster</tt><span class="sig-paren">(</span><em>filename_peptides</em>, <em>filename_genome</em>, <em>options</em>, <em>parser_options</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.runOnCluster" title="Permalink to this definition">¶</a></dt>
<dd><p>run jobs on cluster.</p>
<p>Warning: does not work perfectly - there are some problems
with the parsers which surface when called from inside a makefile,
but (strangely) not when called from the shell directly.</p>
</dd></dl>

<dl class="function">
<dt id="Predictor.predictFromGFF">
<tt class="descclassname">Predictor.</tt><tt class="descname">predictFromGFF</tt><span class="sig-paren">(</span><em>options</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.predictFromGFF" title="Permalink to this definition">¶</a></dt>
<dd><p>run predictions from a gff file that associates queries with genomic regions.</p>
</dd></dl>

<dl class="function">
<dt id="Predictor.predictFromPairs">
<tt class="descclassname">Predictor.</tt><tt class="descname">predictFromPairs</tt><span class="sig-paren">(</span><em>options</em><span class="sig-paren">)</span><a class="headerlink" href="#Predictor.predictFromPairs" title="Permalink to this definition">¶</a></dt>
<dd><p>run predictions from two fasta files, one with peptide sequences and the
other with genomic sequences.</p>
<p>Sequences are associated via the identifier.</p>
<p>The cds file should be indexd.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Predictor.py - running gene predictions with exonerate</a><ul>
<li><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="PredictionParser.html"
                        title="previous chapter">PredictionParser.py - Parser for exonerate/genewise output</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="BlastAlignments.html"
                        title="next chapter">BlastAlignments.py - tools for working with alignments</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/modules/Predictor.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="BlastAlignments.html" title="BlastAlignments.py - tools for working with alignments"
             >next</a> |</li>
        <li class="right" >
          <a href="PredictionParser.html" title="PredictionParser.py - Parser for exonerate/genewise output"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" >Modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>