
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>TreeTools.py - tools for dealing with domain trees. &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Modules" href="../modules.html" />
    <link rel="next" title="RateEstimation.py - utilities for computing rate estimates for codon models." href="RateEstimation.html" />
    <link rel="prev" title="Tree.py - A derivation of the tree class from Biopython" href="Tree.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="RateEstimation.html" title="RateEstimation.py - utilities for computing rate estimates for codon models."
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Tree.html" title="Tree.py - A derivation of the tree class from Biopython"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" accesskey="U">Modules</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-TreeTools"></span><div class="section" id="treetools-py-tools-for-dealing-with-domain-trees">
<h1>TreeTools.py - tools for dealing with domain trees.<a class="headerlink" href="#treetools-py-tools-for-dealing-with-domain-trees" title="Permalink to this headline">¶</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Author:</th><td class="field-body"></td>
</tr>
<tr class="field-even field"><th class="field-name">Release:</th><td class="field-body">$Id$</td>
</tr>
<tr class="field-odd field"><th class="field-name">Date:</th><td class="field-body">April 02, 2014</td>
</tr>
<tr class="field-even field"><th class="field-name">Tags:</th><td class="field-body">Python</td>
</tr>
</tbody>
</table>
<p>A tree is a list of nodes:</p>
<p>node 0 is the root</p>
<p>A node is a list consisting of:</p>
<p>[level, parent, left_child, right_child, [ranges]]</p>
<p>The best way of course would have been to do this object
oriented, but I thought that might be overkill.</p>
<dl class="function">
<dt id="TreeTools.SetChildren">
<tt class="descclassname">TreeTools.</tt><tt class="descname">SetChildren</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.SetChildren" title="Permalink to this definition">¶</a></dt>
<dd><p>sets children correctly.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.CollapseTree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">CollapseTree</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.CollapseTree" title="Permalink to this definition">¶</a></dt>
<dd><p>remove all empty levels in the tree and renumber nodes
so that they are continuous.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.RemoveEmptyNodes">
<tt class="descclassname">TreeTools.</tt><tt class="descname">RemoveEmptyNodes</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.RemoveEmptyNodes" title="Permalink to this definition">¶</a></dt>
<dd><p>remove all empty node from the tree and
renumber nodes, so that they are continuous.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.PrintTree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">PrintTree</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.PrintTree" title="Permalink to this definition">¶</a></dt>
<dd><p>print tree.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.RemoveSmallDomains">
<tt class="descclassname">TreeTools.</tt><tt class="descname">RemoveSmallDomains</tt><span class="sig-paren">(</span><em>tree</em>, <em>min_segment=10</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.RemoveSmallDomains" title="Permalink to this definition">¶</a></dt>
<dd><p>remove small ranges from the tree.</p>
<p>If a small range is encountered, it is attributed to
the other child.</p>
<p>Note: this breaks down, if both children are below
the minimum segment size!</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.TruncateTree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">TruncateTree</tt><span class="sig-paren">(</span><em>tree</em>, <em>min_segment=10</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.TruncateTree" title="Permalink to this definition">¶</a></dt>
<dd><p>truncate tree, if a segment is less than a certain size.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Newick2Nexus">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Newick2Nexus</tt><span class="sig-paren">(</span><em>infile</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Newick2Nexus" title="Permalink to this definition">¶</a></dt>
<dd><p>convert newick formatted tree(s) into a nexus object.</p>
<p>Multiple trees are separated by a semicolon. Tree names can
be given by fasta-style separators, i.e., lines starting with
&#8216;&gt;&#8217;.</p>
<p>If the token [&amp;&amp;NHX is found in the tree, it is assumed to be
output from njtree and support values are added. Support values are
added in the format taxon:support:branchlength</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Nexus2Newick">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Nexus2Newick</tt><span class="sig-paren">(</span><em>nexus</em>, <em>with_branchlengths=True</em>, <em>with_names=False</em>, <em>write_all_taxa=False</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Nexus2Newick" title="Permalink to this definition">¶</a></dt>
<dd><p>convert nexus tree format to newick format.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Tree2Newick">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Tree2Newick</tt><span class="sig-paren">(</span><em>tree</em>, <em>with_branch_lengths=True</em>, <em>write_all_taxa=False</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Tree2Newick" title="Permalink to this definition">¶</a></dt>
<dd><p>convert tree to newick format.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Newick2Tree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Newick2Tree</tt><span class="sig-paren">(</span><em>txt</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Newick2Tree" title="Permalink to this definition">¶</a></dt>
<dd><p>convert tree to newick format.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.WriteNexus">
<tt class="descclassname">TreeTools.</tt><tt class="descname">WriteNexus</tt><span class="sig-paren">(</span><em>nexus</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.WriteNexus" title="Permalink to this definition">¶</a></dt>
<dd><p>write nexus file format.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetTaxa">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetTaxa</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetTaxa" title="Permalink to this definition">¶</a></dt>
<dd><p>retrieve taxa in a tree.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetTaxonomicNames">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetTaxonomicNames</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetTaxonomicNames" title="Permalink to this definition">¶</a></dt>
<dd><p>get list of taxa.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Branchlength2Support">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Branchlength2Support</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Branchlength2Support" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy values stored in data.branchlength to data.support, and do not set branchlength to 0.0</p>
<p>This is necessary when support has been stored as branchlength (e.g. paup), and has thus
been read in as branchlength.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Species2Genes">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Species2Genes</tt><span class="sig-paren">(</span><em>nexus</em>, <em>map_species2genes</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Species2Genes" title="Permalink to this definition">¶</a></dt>
<dd><p>convert a species tree to a gene tree.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Genes2Species">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Genes2Species</tt><span class="sig-paren">(</span><em>nexus</em>, <em>map_gene2species</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Genes2Species" title="Permalink to this definition">¶</a></dt>
<dd><p>convert a gene tree into a species tree.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.BuildMapSpecies2Genes">
<tt class="descclassname">TreeTools.</tt><tt class="descname">BuildMapSpecies2Genes</tt><span class="sig-paren">(</span><em>genes</em>, <em>pattern_species='^([^|]+)[|]'</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.BuildMapSpecies2Genes" title="Permalink to this definition">¶</a></dt>
<dd><p>read genes from infile and build a map of species to genes.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetMonophyleticPairs">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetMonophyleticPairs</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetMonophyleticPairs" title="Permalink to this definition">¶</a></dt>
<dd><p>build list of monophyletic pairs in tree.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetTaxaForSpecies">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetTaxaForSpecies</tt><span class="sig-paren">(</span><em>tree</em>, <em>species</em>, <em>pattern_species='^([^|]+)[|]'</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetTaxaForSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>get all taxa of a given species.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.IsMonophyleticForSpecies">
<tt class="descclassname">TreeTools.</tt><tt class="descname">IsMonophyleticForSpecies</tt><span class="sig-paren">(</span><em>tree</em>, <em>species</em>, <em>pattern_species='^([^|]+)[|]'</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.IsMonophyleticForSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>check if a tree is monophyletic for a species.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.IsMonophyleticForTaxa">
<tt class="descclassname">TreeTools.</tt><tt class="descname">IsMonophyleticForTaxa</tt><span class="sig-paren">(</span><em>tree</em>, <em>taxa</em>, <em>support=None</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.IsMonophyleticForTaxa" title="Permalink to this definition">¶</a></dt>
<dd><p>check if a tree is monophyletic for a list of taxa.</p>
<p>If support is given, minimum support is checked.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetLeaves">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetLeaves</tt><span class="sig-paren">(</span><em>tree</em>, <em>node</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetLeaves" title="Permalink to this definition">¶</a></dt>
<dd><p>get leaves in tree below node.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.IsSingleSpecies">
<tt class="descclassname">TreeTools.</tt><tt class="descname">IsSingleSpecies</tt><span class="sig-paren">(</span><em>tree</em>, <em>node</em>, <em>pattern_species='^([^|]+)[|]'</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.IsSingleSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>check if list of taxa below node contains the same species.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Transcript2GeneTree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Transcript2GeneTree</tt><span class="sig-paren">(</span><em>tree</em>, <em>map_transcript2gene</em>, <em>map_gene2transcripts</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Transcript2GeneTree" title="Permalink to this definition">¶</a></dt>
<dd><p>convert a transcript tree into a gene tree.</p>
<p>supply a map for mapping transcripts to genes.</p>
<p>The procedure for converting a transcript tree into a gene tree:</p>
<p>If there are two genes, and they are monophyletic, no matter how many
transcripts, the order is as follows:</p>
<blockquote>
<div><p>1 Merge all nodes into two, one for each gene.</p>
<dl class="docutils">
<dt>2 The distance between the genes is the minimum distance observed between</dt>
<dd>two transcripts from different genes. Half of this will be set as the
branch length from the gene leaves.</dd>
</dl>
</div></blockquote>
<p>If this is not possible for a set of genes, the procedure will fail and not
return a gene tree.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.MapTaxa">
<tt class="descclassname">TreeTools.</tt><tt class="descname">MapTaxa</tt><span class="sig-paren">(</span><em>tree</em>, <em>mapping</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.MapTaxa" title="Permalink to this definition">¶</a></dt>
<dd><p>map taxon names in all trees.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetCommonAncestor">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetCommonAncestor</tt><span class="sig-paren">(</span><em>tree</em>, <em>taxa</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetCommonAncestor" title="Permalink to this definition">¶</a></dt>
<dd><p>retrieve common ancestor for a list of taxa.</p>
<p>Reroot tree. Check if it is monopyletic. If it is, return root,
otherwise, return -1.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.TreeDFS">
<tt class="descclassname">TreeTools.</tt><tt class="descname">TreeDFS</tt><span class="sig-paren">(</span><em>tree</em>, <em>node_id</em>, <em>pre_function=&lt;function Nop at 0xfbbf938&gt;</em>, <em>descend_condition=&lt;function Nop at 0xfbbf938&gt;</em>, <em>post_function=&lt;function Nop at 0xfbbf938&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.TreeDFS" title="Permalink to this definition">¶</a></dt>
<dd><p>BFS tree tree traversal starting at node_id.</p>
<p>Apply functions pre_function at first and
post_function at last visit of a node.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetMaxIndex">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetMaxIndex</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetMaxIndex" title="Permalink to this definition">¶</a></dt>
<dd><p>get maximum node number.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetBranchLengths">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetBranchLengths</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetBranchLengths" title="Permalink to this definition">¶</a></dt>
<dd><p>return an array with minimum and maximum branch length.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Reroot">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Reroot</tt><span class="sig-paren">(</span><em>tree</em>, <em>taxa</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Reroot" title="Permalink to this definition">¶</a></dt>
<dd><p>reroot tree with taxa - the list of
taxa does not need to be monophyletic.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetSubsets">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetSubsets</tt><span class="sig-paren">(</span><em>tree</em>, <em>node=None</em>, <em>with_decoration=True</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetSubsets" title="Permalink to this definition">¶</a></dt>
<dd><p>return subsets below a certain node including
their height (distance from leaves) and branchlength</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.CountBranchPoints">
<tt class="descclassname">TreeTools.</tt><tt class="descname">CountBranchPoints</tt><span class="sig-paren">(</span><em>tree</em>, <em>taxa</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.CountBranchPoints" title="Permalink to this definition">¶</a></dt>
<dd><p>count the number branch points together with their
distances for a given list of taxa.</p>
<p>return a list of branch points</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.IsCompatible">
<tt class="descclassname">TreeTools.</tt><tt class="descname">IsCompatible</tt><span class="sig-paren">(</span><em>tree1</em>, <em>tree2</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.IsCompatible" title="Permalink to this definition">¶</a></dt>
<dd><p>check if two trees are compatible.</p>
<p>note: this will delete support information.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Tree2Graph">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Tree2Graph</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Tree2Graph" title="Permalink to this definition">¶</a></dt>
<dd><p>return tree as a list of edges in a graph.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Graph2Tree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Graph2Tree</tt><span class="sig-paren">(</span><em>links</em>, <em>label_ancestral_nodes=False</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Graph2Tree" title="Permalink to this definition">¶</a></dt>
<dd><p>build tree from list of nodes.</p>
<p>Assumption is that links always point from parent to child.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetAllNodes">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetAllNodes</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetAllNodes" title="Permalink to this definition">¶</a></dt>
<dd><p>return all nodes in the tree.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetDistancesBetweenTaxa">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetDistancesBetweenTaxa</tt><span class="sig-paren">(</span><em>tree</em>, <em>taxa1</em>, <em>taxa2</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetDistancesBetweenTaxa" title="Permalink to this definition">¶</a></dt>
<dd><p>get average branchlength between taxa1 and taxa2.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.PruneTerminal">
<tt class="descclassname">TreeTools.</tt><tt class="descname">PruneTerminal</tt><span class="sig-paren">(</span><em>tree</em>, <em>taxon</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.PruneTerminal" title="Permalink to this definition">¶</a></dt>
<dd><p>Prunes a terminal taxon from the tree.</p>
<p>id_of_previous_node = prune(tree,taxon)
If taxon is from a bifurcation, the connecting node will be collapsed
and its branchlength added to remaining terminal node. This might be no
longer a meaningful value.</p>
<p>direct copy of Nexus.Trees.py - don&#8217;t know why have a separate method,
maybe there was a bug in Nexus.Trees.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetSubtree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetSubtree</tt><span class="sig-paren">(</span><em>tree</em>, <em>node_id</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetSubtree" title="Permalink to this definition">¶</a></dt>
<dd><p>return a copy of tree from node_id downwards.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.Unroot">
<tt class="descclassname">TreeTools.</tt><tt class="descname">Unroot</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.Unroot" title="Permalink to this definition">¶</a></dt>
<dd><p>unroot tree.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetSize">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetSize</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetSize" title="Permalink to this definition">¶</a></dt>
<dd><p>return the length of the tree. This is the maximum node_id + 1.</p>
<p>This quantity is useful for tree traversal while updating
a container.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.PruneTree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">PruneTree</tt><span class="sig-paren">(</span><em>tree</em>, <em>taxa</em>, <em>keep_distance_to_root=False</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.PruneTree" title="Permalink to this definition">¶</a></dt>
<dd><p>prune tree: keep only those taxa in list.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetNodeMap">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetNodeMap</tt><span class="sig-paren">(</span><em>tree1</em>, <em>tree2</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetNodeMap" title="Permalink to this definition">¶</a></dt>
<dd><p>map nodes between tree1 and tree2.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.ReconciliateByRio">
<tt class="descclassname">TreeTools.</tt><tt class="descname">ReconciliateByRio</tt><span class="sig-paren">(</span><em>gene_tree</em>, <em>species_tree</em>, <em>extract_species</em>, <em>extract_gene=None</em>, <em>outgroup_species=None</em>, <em>min_branch_length=0.0</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.ReconciliateByRio" title="Permalink to this definition">¶</a></dt>
<dd><p>Gene tree G and species tree S</p>
<p>If outgroup_species is given: trees will be cut of
as soon as one of the outgroup species is part of a subtree.
The corresponding node type will be out-paralog. Out-paralog
relationship is cast upwards.</p>
<p>Input trees are rooted and binary.</p>
<p>Output: gene tree with duplication/speciation assigned to each node.</p>
<p>Initialization:</p>
<blockquote>
<div><p>Number nodes in S in pre-order traversal (root = 1), such
that child nodes are always larger than parent nodes.</p>
<p>For each external node g of G, set M(g) to the number of the
external node in S with the matching species name.</p>
</div></blockquote>
<p>Recursion:</p>
<blockquote>
<div><p>Visit each internal node g of G in post-order traversal, (i.e.
from leaves to root):</p>
<div class="highlight-python"><div class="highlight"><pre>set a = M(g1) # g1 = first child of current node g
set b = M(g2) # g2 = second child of current node g

while a != b:
    if a &gt; b:
          set a = parent of node a in species tree
    else:
          set b = parent of node b in species tree
set M(g) = a

if M(g) == M(g1) or M(g) == M(g2):
    g is duplication
else:
    g is speciation
</pre></div>
</div>
</div></blockquote>
<p>The algorithm returns an array for each node with its type.</p>
<p>If extract_gene is given, the algorithm will label transcription nodes
for alternative transcripts (duplications involving the same gene).</p>
<p>The algorithm has been extended to accomodate the following test cases:</p>
<dl class="docutils">
<dt>Alternative transcripts</dt>
<dd><p class="first">Alternative transcripts that span genes from other species are permitted,
if at most one gene of the other species is involved.</p>
<p class="last">To avoid over-counting of speciation events, the one subtree with the
least species is masked.</p>
</dd>
</dl>
<p>If the branch length of a node in the gene tree is shorter than min_branch_length,
the resultant node is masked, because the topology might be dodgy.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.CountDuplications">
<tt class="descclassname">TreeTools.</tt><tt class="descname">CountDuplications</tt><span class="sig-paren">(</span><em>gene_tree</em>, <em>species_tree</em>, <em>node_types</em>, <em>extract_species</em>, <em>extract_gene=None</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.CountDuplications" title="Permalink to this definition">¶</a></dt>
<dd><p>count duplications.</p>
<p>given are gene and species tree and node types (duplication/speciation)</p>
<p>extract_species gives the species for an OTU in the gene tree</p>
<p>Extract_gene gives the gene for an OTU in the gene tree. If not given,
all transcripts are counted as unique.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetParentNodeWhereTrue">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetParentNodeWhereTrue</tt><span class="sig-paren">(</span><em>node_id</em>, <em>tree</em>, <em>stop_function</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetParentNodeWhereTrue" title="Permalink to this definition">¶</a></dt>
<dd><p>walk up in gene tree and stop where stop_function is true.</p>
<p>The walk finishes at the root.</p>
<p>returns tuple of node and distance.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetChildNodesWhereTrue">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetChildNodesWhereTrue</tt><span class="sig-paren">(</span><em>node_id</em>, <em>tree</em>, <em>stop_function</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetChildNodesWhereTrue" title="Permalink to this definition">¶</a></dt>
<dd><p>walk down in tree and stop where stop_function is true</p>
<p>The walk finishes at the leaves.</p>
<p>returns a list of tuples of nodes and distance.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.GetDistanceToRoot">
<tt class="descclassname">TreeTools.</tt><tt class="descname">GetDistanceToRoot</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.GetDistanceToRoot" title="Permalink to this definition">¶</a></dt>
<dd><p>return list with distance to root for each node.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.traverseGraph">
<tt class="descclassname">TreeTools.</tt><tt class="descname">traverseGraph</tt><span class="sig-paren">(</span><em>graph</em>, <em>start</em>, <em>block=</em><span class="optional">[</span><span class="optional">]</span><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.traverseGraph" title="Permalink to this definition">¶</a></dt>
<dd><p>traverse graph, go not passed nodes in block.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.convertTree2Graph">
<tt class="descclassname">TreeTools.</tt><tt class="descname">convertTree2Graph</tt><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.convertTree2Graph" title="Permalink to this definition">¶</a></dt>
<dd><p>convert tree to a graph.</p>
</dd></dl>

<dl class="function">
<dt id="TreeTools.calculatePatternsFromTree">
<tt class="descclassname">TreeTools.</tt><tt class="descname">calculatePatternsFromTree</tt><span class="sig-paren">(</span><em>tree</em>, <em>sort_order</em><span class="sig-paren">)</span><a class="headerlink" href="#TreeTools.calculatePatternsFromTree" title="Permalink to this definition">¶</a></dt>
<dd><p>calculate patterns from a tree.</p>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h4>Previous topic</h4>
  <p class="topless"><a href="Tree.html"
                        title="previous chapter">Tree.py - A derivation of the tree class from Biopython</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="RateEstimation.html"
                        title="next chapter">RateEstimation.py - utilities for computing rate estimates for codon models.</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/modules/TreeTools.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="RateEstimation.html" title="RateEstimation.py - utilities for computing rate estimates for codon models."
             >next</a> |</li>
        <li class="right" >
          <a href="Tree.html" title="Tree.py - A derivation of the tree class from Biopython"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../modules.html" >Modules</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>