
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>GPipe - Gene prediction pipeline &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Lecgacy pipelines" href="../MakePipelines.html" />
    <link rel="next" title="454 Transcript mapping pipeline" href="MapTranscripts454.html" />
    <link rel="prev" title="Transcript comparison pipeline" href="CompareTranscripts.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="MapTranscripts454.html" title="454 Transcript mapping pipeline"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="CompareTranscripts.html" title="Transcript comparison pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../MakePipelines.html" accesskey="U">Lecgacy pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="gpipe-gene-prediction-pipeline">
<h1>GPipe - Gene prediction pipeline<a class="headerlink" href="#gpipe-gene-prediction-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This document describes the pipeline of the Chris Ponting group
for predicting genes by homology. The input is a set of known transcripts
from a reference genome and the masked genomic sequence of a
target genome.</p>
<p>The pipeline predicts genes in a two-step procedure:</p>
<ol class="arabic simple">
<li>Regions of similarity between the known transcripts and the reference genome
are identified using a quick heuristic search.</li>
<li>The regions of similarity of step 1 are submitted to a sensitive, but slower gene
prediction program.</li>
</ol>
<p>The pipeline contains many options to mask sequences, analyse and quality
control the predictions and store the results in a relational database.
This manual describes how to setup up the pipeline, run it on our cluster,
and analyse the results.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The pipeline is tailored to the computational setup within the Ponting group.
Porting it elsewhere will require a significant amount of software installation
and configuration.</p>
</div>
</div>
<div class="section" id="setting-up">
<h2>Setting up<a class="headerlink" href="#setting-up" title="Permalink to this headline">¶</a></h2>
<p>The pipeline consists of a set of scripts and a makefile,
that glues together the various scripts. This section tells
you what programs are needed to be installed (<a class="reference internal" href="#requirements">Requirements</a>),
how to install the software from the pipeline (<a class="reference internal" href="#installation">Installation</a>).</p>
<p>Next, the input files need to prepared (<a class="reference internal" href="#input-files">Input files</a>).</p>
<p>Before starting the pipeline, you need to configure your environment
and the pipeline (<a class="reference internal" href="#configuration">Configuration</a>).</p>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>Gpipe requires following software to be installed:</p>
<blockquote>
<div><dl class="docutils">
<dt>postgres</dt>
<dd>Gpipe currently requires a postgres installation.</dd>
<dt>seq</dt>
<dd>Seg is a program to mask low complexity regions in protein sequences.</dd>
<dt>exonerate</dt>
<dd>Exonerate is a program to align a peptide sequence to a genomic sequence
(other alignment modes are possible). It offers heuristic modes, that allow for fast
scanning of large chunks of genomic DNA, and exhaustive modes, that do a full dynamic
programming mode. It is available at <a class="reference external" href="http://www.ebi.ac.uk/~guy/exonerate">http://www.ebi.ac.uk/~guy/exonerate</a>.</dd>
<dt>python</dt>
<dd>Most scripts require python...</dd>
<dt>perl</dt>
<dd>...unless they are written in perl</dd>
<dt>alignlib</dt>
<dd>a library for sequence alignments and its python interface. See
<a class="reference external" href="http://sourceforge.net/projects/alignlib">http://sourceforge.net/projects/alignlib</a>.</dd>
<dt>SGE</dt>
<dd>Sun Grid Engine. Other schedulers might work.</dd>
<dt>seq_pairs_kaks</dt>
<dd>Leo&#8217;s wrapper around PAML (optional, only required for step12).</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h3>
<p>Untar and unpack the source code tarball. The directory in which it ends up
is the <a class="reference internal" href="#term-source-directory"><em class="xref std std-term">source directory</em></a>.</p>
<p>Gpipe works in a <a class="reference internal" href="#term-working-directory"><em class="xref std std-term">working directory</em></a>. To set up the pipeline with the current
directory as the <a class="reference internal" href="#term-working-directory"><em class="xref std std-term">working directory</em></a>, run:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">python</span> <span class="o">&lt;</span><span class="n">src</span><span class="o">&gt;</span><span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">method</span><span class="o">=</span><span class="n">gpipe</span> <span class="o">--</span><span class="n">project</span><span class="o">=</span><span class="n">project_name</span> <span class="o">&gt;</span> <span class="n">setup</span><span class="o">.</span><span class="n">log</span>
</pre></div>
</div>
<p><tt class="file docutils literal"><span class="pre">src</span></tt> is the location of the <a class="reference internal" href="#term-source-directory"><em class="xref std std-term">source directory</em></a>.</p>
<p>This will create a makefile in the current directory and give the project the
name <tt class="docutils literal"><span class="pre">project_name</span></tt>. The latter will be used as the name of the database schema
in postgres in which data will be stored.</p>
<p>Note that gpipe assumes that you use <tt class="docutils literal"><span class="pre">bash</span></tt> as you shell. In particular, it requires
that two functions are in your environment:</p>
<div class="highlight-python"><div class="highlight"><pre>#-----------------------------------
# helper functions for detecting errors in pipes
#-----------------------------------
detect_pipe_error_helper()
{
 while [ &quot;$#&quot; != 0 ] ; do
     # there was an error in at least one program of the pipe
     if [ &quot;$1&quot; != 0 ] ; then return 1 ; fi
     shift 1
 done
 return 0
}

detect_pipe_error()
{
 detect_pipe_error_helper &quot;${PIPESTATUS[@]}&quot;
 return $?
}
</pre></div>
</div>
<p>Once the code is in place, add the input files to the working and make sure
that all the other requirements are fulfilled.</p>
</div>
<div class="section" id="input-files">
<h3>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h3>
<p>Gpipe requires 5 files to run. These input files contain the reference gene set to predict with and
the genome sequence to predict in. Sample data is available at
<a class="reference external" href="http://genserv.anat.ox.ac.uk/downloads/software/gpipe/sampledata/gpipe_sample_data.tar">http://genserv.anat.ox.ac.uk/downloads/software/gpipe/sampledata/gpipe_sample_data.tar</a>.
To drop the sample data into your <a class="reference internal" href="#term-working-directory"><em class="xref std std-term">working directory</em></a>, type:</p>
<div class="highlight-python"><div class="highlight"><pre>wget ttp://genserv.anat.ox.ac.uk/downloads/software/gpipe/sampledata/gpipe_sample_data.tar
tar -xf gpipe_sample_data.tar
gunzip *
</pre></div>
</div>
<p>The filenames and their contents are:</p>
<dl class="docutils">
<dt><tt class="file docutils literal"><span class="pre">peptides.fasta</span></tt></dt>
<dd><p class="first">A fasta-formatted file with peptide sequences. Each sequence is on a single line.
The identifier of a sequence is taken from the description line
with the pattern <tt class="docutils literal"><span class="pre">&gt;(\S+)</span></tt> (characters between &gt; and first white-space).
For example:</p>
<div class="last highlight-python"><div class="highlight"><pre>&gt;CG11023-RA
MGERDQPQSSERISIFNPPVYTQHQVRNEAPYIPTTFDLLSDDEESSQRVANAGPSFRPL...
&gt;CG2671-RA
MLKFIRGKGQQPSADRHRLQKDLFAYRKTAQHGFPHKPSALAYDPVLKLMAIGTQTGALK...
...
</pre></div>
</div>
</dd>
<dt><tt class="file docutils literal"><span class="pre">genome.fasta</span></tt> and <tt class="file docutils literal"><span class="pre">genome.idx</span></tt></dt>
<dd><p class="first">A fasta-formatted file with the genomic sequence together with its index. This file can be created
from a collection of individual fasta formatted files (for example, per chromosome files) using the command:</p>
<div class="last highlight-python"><div class="highlight"><pre>python &lt;src&gt;index_fasta.py genome &lt;somedir&gt;my_dna_sequences*.fa.gz &gt; genome.log
</pre></div>
</div>
</dd>
<dt><tt class="file docutils literal"><span class="pre">reference.exons</span></tt></dt>
<dd><p class="first">A table with gene models from the reference gene set.
This is a tab-formatted table with the following columns:</p>
<dl class="docutils">
<dt>transcript name</dt>
<dd>name of the transcript consistent with <tt class="file docutils literal"><span class="pre">peptides.fasta</span></tt></dd>
<dt>contig name</dt>
<dd>name of the DNA segment the transcript is located on</dd>
<dt>strand</dt>
<dd>strand</dd>
<dt>phase</dt>
<dd>phase of that particular exon</dd>
<dt>exon-id</dt>
<dd>numerical number of exon</dd>
<dt>peptide_start</dt>
<dd>start of exon in transcript sequence</dd>
<dt>peptide_end</dt>
<dd>end of exon in transcript sequence</dd>
<dt>genome_start</dt>
<dd>start of exon on contig</dd>
<dt>genome_end</dt>
<dd>end of exon on contig</dd>
</dl>
<p>Coordinates are 0-based, half-open intervals. Genomic coordinates are forward/reverse strand coordinates.</p>
<p>For example:</p>
<div class="last highlight-python"><div class="highlight"><pre>CG10000-RA chr3R   -1      0       1       0       126     24577165        24577291
CG10000-RA chr3R   -1      0       2       126     287     24576946        24577107
CG10000-RA chr3R   -1      1       3       287     466     24576706        24576885
CG10000-RA chr3R   -1      2       4       466     930     24576187        24576651
CG10000-RA chr3R   -1      0       5       930     1100    24575892        24576062
CG10000-RA chr3R   -1      1       6       1100    1280    24575573        24575753
CG10000-RA chr3R   -1      1       7       1280    1677    24574936        24575330
CG10001-RA chr3R   -1      0       1       0       540     24569207        24569747
CG10001-RA chr3R   -1      0       2       540     819     24566427        24566706
CG10001-RA chr3R   -1      0       3       819     978     24566193        24566352
</pre></div>
</div>
</dd>
<dt><tt class="file docutils literal"><span class="pre">map_rep2mem</span></tt></dt>
<dd><p class="first">A table linking genes to transcripts. This tab-formatted table contains the following columns</p>
<dl class="last docutils">
<dt>rep</dt>
<dd>A gene identifier</dd>
<dt>mem</dt>
<dd>A transcript identifier</dd>
<dt>size</dt>
<dd>Transcript size</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h3>
<p>To configure the pipeline, options can be set in the <tt class="file docutils literal"><span class="pre">Makefile</span></tt> in the
<a class="reference internal" href="#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
<p>Options that might need to be changed:</p>
<blockquote>
<div><dl class="docutils">
<dt>PARAM_PSQL_DATABASE</dt>
<dd>The psql database</dd>
<dt>PARAM_PSQL_HOST</dt>
<dd>The psql host</dd>
<dt>PARAM_PSQL_USER</dt>
<dd>The psql username</dd>
</dl>
</div></blockquote>
</div>
</div>
<div class="section" id="running-the-pipeline">
<h2>Running the pipeline<a class="headerlink" href="#running-the-pipeline" title="Permalink to this headline">¶</a></h2>
<p>The pipeline uses makefiles to control script logic. Before executing any make commands,
run:</p>
<div class="highlight-python"><div class="highlight"><pre>source setup.csh
</pre></div>
</div>
<p>to update your paths and other environment variables.</p>
<p>Before first running the pipeline, some maintenance work need to
be performed like creating the database schema and the tables.
To prepare the pipeline, run:</p>
<div class="highlight-python"><div class="highlight"><pre>make prepare
</pre></div>
</div>
<p>This needs to be done only once.</p>
<p>To run the pipeline, type:</p>
<div class="highlight-python"><div class="highlight"><pre>make all
</pre></div>
</div>
<p>Gpipe writes status messages to the file <tt class="file docutils literal"><span class="pre">log</span></tt>
in the <a class="reference internal" href="#term-working-directory"><em class="xref std std-term">working directory</em></a>.</p>
</div>
<div class="section" id="results">
<h2>Results<a class="headerlink" href="#results" title="Permalink to this headline">¶</a></h2>
<p>Results of the gpipe run are stored in the psql database.</p>
<p>The view <tt class="docutils literal"><span class="pre">overview</span></tt> aggregates most results into a single
table for easy access.</p>
</div>
<div class="section" id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Permalink to this headline">¶</a></h2>
<p>If something goes wrong, the first step is to look at
the command line that caused the problem. To see the command
executed, run:</p>
<div class="highlight-python"><div class="highlight"><pre>make -n &lt;target&gt;
</pre></div>
</div>
<p>We use Sun Grid Engine as job queueing system and assume that for all nodes
the code and data can be reached via the same mount point. All jobs that are run on
the cluster are prefixed by the MAKE variable $(CMD_REMOTE_SUBMIT). You can set this
variable to the empty string to run everything locally or on a mosix cluster:</p>
<blockquote>
<div>make all CMD_REMOTE_SUBMIT=</div></blockquote>
</div>
<div class="section" id="steps">
<h2>Steps<a class="headerlink" href="#steps" title="Permalink to this headline">¶</a></h2>
<p>The pipeline proceeds in 12 steps, which are:</p>
<dl class="docutils">
<dt>Step1</dt>
<dd>Masking of protein sequences</dd>
<dt>Step2</dt>
<dd>Selecting representative transcipts to search with</dd>
<dt>Step3</dt>
<dd>Running exonerate</dd>
<dt>Step4</dt>
<dd>Running TBLASTN (disabled)</dd>
<dt>Step5</dt>
<dd>Collating putative gene-containing regions</dd>
<dt>Step6</dt>
<dd>Predicting genes for representative transcripts.</dd>
<dt>Step7</dt>
<dd>Predicting genes for redundant (alternative) transcripts</dd>
<dt>Step8</dt>
<dd>Predicting genes for member sequences</dd>
<dt>Step9</dt>
<dd>Analysing the predictions</dd>
<dt>Step10</dt>
<dd>Quality control of predictions</dd>
<dt>Step11</dt>
<dd>Removing redundant/erroneous predictions</dd>
<dt>Step12</dt>
<dd>Filter by ks (optional)</dd>
</dl>
</div>
<div class="section" id="glossary">
<h2>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h2>
<dl class="glossary docutils">
<dt id="term-working-directory">working directory</dt>
<dd>The working directory. Location of the data files and results. All commands in this tutorial are
executed in the working directory.</dd>
<dt id="term-source-directory">source directory</dt>
<dd>The location of the source code. The place where the script <tt class="file docutils literal"><span class="pre">setup.py</span></tt> resides.</dd>
</dl>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">GPipe - Gene prediction pipeline</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#setting-up">Setting up</a><ul>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#input-files">Input files</a></li>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
</ul>
</li>
<li><a class="reference internal" href="#running-the-pipeline">Running the pipeline</a></li>
<li><a class="reference internal" href="#results">Results</a></li>
<li><a class="reference internal" href="#troubleshooting">Troubleshooting</a></li>
<li><a class="reference internal" href="#steps">Steps</a></li>
<li><a class="reference internal" href="#glossary">Glossary</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="CompareTranscripts.html"
                        title="previous chapter">Transcript comparison pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="MapTranscripts454.html"
                        title="next chapter">454 Transcript mapping pipeline</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/make_pipelines/Gpipe.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="MapTranscripts454.html" title="454 Transcript mapping pipeline"
             >next</a> |</li>
        <li class="right" >
          <a href="CompareTranscripts.html" title="Transcript comparison pipeline"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../MakePipelines.html" >Lecgacy pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>