
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>OPTIC &mdash; CGAT  documentation</title>
    
    <link rel="stylesheet" href="../_static/cgat.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/sorttable.js"></script>
    <script type="text/javascript" src="../_static/jquery.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.dataTables.min.js"></script>
    <link rel="top" title="CGAT  documentation" href="../index.html" />
    <link rel="up" title="Lecgacy pipelines" href="../MakePipelines.html" />
    <link rel="next" title="Contributing to CGAT code" href="../developing.html" />
    <link rel="prev" title="454 Transcript mapping pipeline" href="MapTranscripts454.html" />
   <script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
        $('.scrollable').dataTable({
              "sScrollY":"200px",
               "sScrollX":"100%",
               "bAutoWidth":false,
               "bPaginate":false });
     });
    </script>
     

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../developing.html" title="Contributing to CGAT code"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="MapTranscripts454.html" title="454 Transcript mapping pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../MakePipelines.html" accesskey="U">Lecgacy pipelines</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="optic">
<h1>OPTIC<a class="headerlink" href="#optic" title="Permalink to this headline">¶</a></h1>
<div class="section" id="purpose">
<h2>Purpose<a class="headerlink" href="#purpose" title="Permalink to this headline">¶</a></h2>
<p>The optic pipeline performs orthology assignment in
a group of species.</p>
</div>
<div class="section" id="setting-up">
<h2>Setting up<a class="headerlink" href="#setting-up" title="Permalink to this headline">¶</a></h2>
<div class="section" id="download-cgat-code">
<h3>Download CGAT code<a class="headerlink" href="#download-cgat-code" title="Permalink to this headline">¶</a></h3>
<p>CGAT code can be obtained by checking the latest version out from
mercurial:</p>
<p>hg clone <a class="reference external" href="http://www.cgat.org/hg/cgat">http://www.cgat.org/hg/cgat</a></p>
<p>In the following, the location of the checked out code will be
referred to as &lt;src&gt;.</p>
</div>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>Optic requires the following tools to be installed.</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="22%" />
<col width="55%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><em>Program</em></td>
<td><em>Version</em></td>
<td><em>Purpose</em></td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://www.postgresql.org/">postgres</a></td>
<td>&nbsp;</td>
<td>database</td>
</tr>
<tr class="row-odd"><td><a class="reference external" href="http://www.drive5.com/muscle/">muscle</a></td>
<td>&gt;=3.81.3</td>
<td>Multiple alignment</td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://www.cgat.org/~andreas/leotools-0.1.x86_64.linux.tar.gz">phyop</a></td>
<td>&nbsp;</td>
<td>read mapping</td>
</tr>
<tr class="row-odd"><td><a class="reference external" href="http://treesoft.sourceforge.net/treebest.shtml">treebest</a></td>
<td>&gt;=0.1</td>
<td>bam/sam files</td>
</tr>
<tr class="row-even"><td><a class="reference external" href="http://abacus.gene.ucl.ac.uk/software/paml.html">paml</a></td>
<td>&gt;=4.4c</td>
<td>evolutionary rate estimation</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genomes">
<h3>Genomes<a class="headerlink" href="#genomes" title="Permalink to this headline">¶</a></h3>
<p>The first step is to build the files with the genomic
sequences. Download fasta files with genomic secquences
and index them using the &lt;src&gt;/IndexedFasta.py tools.</p>
<p>Store the genomes in a separate directory (referred to
later as &lt;genome&gt;).</p>
</div>
<div class="section" id="genesets">
<h3>Genesets<a class="headerlink" href="#genesets" title="Permalink to this headline">¶</a></h3>
<p>The second step is to upload the genesets into the database. The
pipeline expects protein coding transcripts from ENSEMBL and requires
two files that can be downloaded from the <a class="reference external" href="http://www.ensembl.org/info/data/ftp/index.html">ENSEMBL ftp site</a>:</p>
<ul>
<li><p class="first">File with peptide sequences, usually called <tt class="docutils literal"><span class="pre">species</span></tt>.pep.all.fa.gz</p>
</li>
<li><dl class="first docutils">
<dt>File with exon coordinates in gtf format, usually called</dt>
<dd><p class="first last"><tt class="docutils literal"><span class="pre">species</span></tt>.gtf.gz</p>
</dd>
</dl>
</li>
</ul>
<p>The following example shows how to upload the human gene set. We are
going to be using ensembl version 62 on hg19:</p>
<div class="highlight-python"><div class="highlight"><pre>mkdir -p genesets/hs62

cd genesets/hs62

ln -s &lt;genomes&gt;hg19.fasta genome.fasta

ln -s &lt;genomes&gt;hg19.idx genome.idx

ln -s &lt;mirror&gt;Homo_sapiens.GRCh37.62.gtf.gz reference.gtf.gz

ln -s &lt;mirror&gt;Homo_sapiens.GRCh37.62.pep.all.fa.gz reference.pep.fa.gz
</pre></div>
</div>
<p>Create the makefile:</p>
<div class="highlight-python"><div class="highlight"><pre>python &lt;src&gt;setup.py -m ensembl -p cgat_hs62
</pre></div>
</div>
<p>Create database tables:</p>
<div class="highlight-python"><div class="highlight"><pre>make prepare
</pre></div>
</div>
<p>Upload data:</p>
<div class="highlight-python"><div class="highlight"><pre>make all
</pre></div>
</div>
<p>To verify all was ok, look at the file <tt class="file docutils literal"><span class="pre">predictions.check</span></tt>. This file
compares the ENSEMBL supplied peptide sequences with those that have
been uploaded into the database.</p>
<p>The data is now stored in the database schema cgat_hs62.</p>
<p>You need to do this for all species that you want to run OPTIC on.</p>
<p>Make sure that the naming is consistent with the genomes. Thus,
<tt class="docutils literal"><span class="pre">hg19</span></tt> should both refer to the gene set for human, but also
to the genomic sequence files (<tt class="file docutils literal"><span class="pre">hg19.fasta</span></tt>) in the
&lt;genomes&gt; directory.</p>
</div>
</div>
<div class="section" id="running-optic">
<h2>Running OPTIC<a class="headerlink" href="#running-optic" title="Permalink to this headline">¶</a></h2>
<p>The optic pipeline works from several directories.</p>
<p>Create a working directory:</p>
<div class="highlight-python"><div class="highlight"><pre>mkdir optic
</pre></div>
</div>
<p>Create a makefile:</p>
<div class="highlight-python"><div class="highlight"><pre>python &lt;src&gt;setup.py -m optic -p optic
cd optic
</pre></div>
</div>
<p>Enter the <tt class="file docutils literal"><span class="pre">data</span></tt> directory and create the following files:</p>
<dl class="docutils">
<dt>Makefile.inc</dt>
<dd><p class="first">common makefile file options. For example:</p>
<div class="last highlight-python"><div class="highlight"><pre>## Global configuration options for OPTIC
PARAM_PROJECT_NAME=cgat_proj007

DIR_TMP=/tmp/

PARAM_DIR_DATA=&lt;optic&gt;data/

PARAM_SRC_SCHEMAS=cgat_hs62 cgat_mm62 cgat_gg62 cgat_ac62 cgat_xt62
cgat_dr62 cgat_oa65

PARAM_SPECIES_TREE=((((((cgat_hs62,cgat_mm62),cgat_oa65),cgat_gg62),cgat_ac62),cgat_xt62),cgat_dr62);

PARAM_ANALYSIS_DUPLICATIONS_OUTGROUPS=cgat_dr62

## CGAT cluster params

DIR_SCRIPTS=&lt;src&gt;/

PARAM_QUEUE=all.q
PARAM_QUEUE_LOCAL=all.q
PARAM_QUEUE_SERVER=all.q
</pre></div>
</div>
</dd>
<dt>schema2sp</dt>
<dd><p class="first">tsv file mapping species in database schema to swissprot name (required
for treebest):</p>
<div class="last highlight-python"><div class="highlight"><pre># map of species names to swiss prot taxonomic names
# used for njtree
schema  sp
cgat_hs62       HUMAN
cgat_mm62       MOUSE
cgat_xt62       XENTR
cgat_dr62       DANRE
cgat_ac62       ANOCA
cgat_gg62       CHICK
cgat_oa65       ORNAN
</pre></div>
</div>
</dd>
<dt>species_tree</dt>
<dd><p class="first">the phylogeny of the species in newick format, for example:</p>
<div class="last highlight-python"><div class="highlight"><pre><span class="p">((((((</span><span class="n">cgat_hs62</span><span class="p">,</span><span class="n">cgat_mm62</span><span class="p">),</span><span class="n">cgat_oa65</span><span class="p">),</span><span class="n">cgat_gg62</span><span class="p">),</span><span class="n">cgat_ac62</span><span class="p">),</span><span class="n">cgat_xt62</span><span class="p">),</span><span class="n">cgat_dr62</span><span class="p">);</span>
</pre></div>
</div>
</dd>
<dt>species_tree_permutations</dt>
<dd>the phylogeny of the species and permutations of it. Usually
just a copy of species_tree</dd>
</dl>
<p>files needed for web server:</p>
<dl class="docutils">
<dt>schema2colour</dt>
<dd><p class="first">map species name to colour:</p>
<div class="last highlight-python"><div class="highlight"><pre>cgat_hs62       255,204,0
cgat_dr62       255,102,204
cgat_xt62       204,204,0
cgat_mm62       204,102,255
cgat_ac62       102,255,255
cgat_gg62       125,125,255
cgat_oa65       255,255,102
</pre></div>
</div>
</dd>
<dt>schema2name</dt>
<dd><p class="first">map species name to real name:</p>
<div class="last highlight-python"><div class="highlight"><pre>schema  name
cgat_hs62       H. sapiens
cgat_mm62       M. musculus
cgat_ac62       A. carolinensis
cgat_xt62       X. tropicalis
cgat_dr62       D. rerio
cgat_gg62       G. gallus
cgat_oa65       O. anatinus
</pre></div>
</div>
</dd>
<dt>schema2url</dt>
<dd><p class="first">map species name to ENSEMBL URL:</p>
<div class="last highlight-python"><div class="highlight"><pre>cgat_hs62       displayGene?schema=%(species)s&amp;gene_id=%(gene)s
cgat_mm62       displayGene?schema=%(species)s&amp;gene_id=%(gene)s
cgat_ac62       displayGene?schema=%(species)s&amp;gene_id=%(gene)s
cgat_xt62       displayGene?schema=%(species)s&amp;gene_id=%(gene)s
cgat_dr62       displayGene?schema=%(species)s&amp;gene_id=%(gene)s
cgat_gg62       displayGene?schema=%(species)s&amp;gene_id=%(gene)s
cgat_oa65       displayGene?schema=%(species)s&amp;gene_id=%(gene)s
</pre></div>
</div>
</dd>
</dl>
<div class="section" id="preparing-data">
<h3>Preparing data<a class="headerlink" href="#preparing-data" title="Permalink to this headline">¶</a></h3>
<p>Create export data files. This will create fasta file and exon
boundary files for all species:</p>
<div class="highlight-python"><div class="highlight"><pre>make -C export export_clustering
</pre></div>
</div>
</div>
<div class="section" id="phyop">
<h3>Phyop<a class="headerlink" href="#phyop" title="Permalink to this headline">¶</a></h3>
<p>Setup pairwise phyop runs and run them:</p>
<div class="highlight-python"><div class="highlight"><pre>make -C orthology_pairwise prepare

nice -19 nohup make -C orthology_pairwise all
</pre></div>
</div>
<p>Wait a while...</p>
</div>
<div class="section" id="clustering">
<h3>Clustering<a class="headerlink" href="#clustering" title="Permalink to this headline">¶</a></h3>
<p>The clustering step combines pairwise orthology assignments into
clusters of potential orthologs:</p>
<div class="highlight-python"><div class="highlight"><pre>make -C orthology_multiple prepare
make -C orthology_mulitple all
</pre></div>
</div>
</div>
<div class="section" id="multiple-alignment">
<h3>Multiple alignment<a class="headerlink" href="#multiple-alignment" title="Permalink to this headline">¶</a></h3>
<p>Next, multiple alignments are built for each cluster:</p>
<div class="highlight-python"><div class="highlight"><pre>make -C malis prepare
make -C malis all
make -C malis summary.dir
make -C malis summary
</pre></div>
</div>
</div>
<div class="section" id="orthologous-groups">
<h3>Orthologous groups<a class="headerlink" href="#orthologous-groups" title="Permalink to this headline">¶</a></h3>
<p>Based on the multiple alignments, trees are built within each cluster
and the trees are split into orthologous groups:</p>
<div class="highlight-python"><div class="highlight"><pre>make -C paralogy_trees prepare
make -C paralogy_trees all
make -C paralogy_trees analysis
make -C paralogy_trees summary
</pre></div>
</div>
</div>
</div>
<div class="section" id="configuration">
<h2>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h2>
<p>Edit the <tt class="file docutils literal"><span class="pre">Makefile</span></tt> to configure the pipeline.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/cgat_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">OPTIC</a><ul>
<li><a class="reference internal" href="#purpose">Purpose</a></li>
<li><a class="reference internal" href="#setting-up">Setting up</a><ul>
<li><a class="reference internal" href="#download-cgat-code">Download CGAT code</a></li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
<li><a class="reference internal" href="#genomes">Genomes</a></li>
<li><a class="reference internal" href="#genesets">Genesets</a></li>
</ul>
</li>
<li><a class="reference internal" href="#running-optic">Running OPTIC</a><ul>
<li><a class="reference internal" href="#preparing-data">Preparing data</a></li>
<li><a class="reference internal" href="#phyop">Phyop</a></li>
<li><a class="reference internal" href="#clustering">Clustering</a></li>
<li><a class="reference internal" href="#multiple-alignment">Multiple alignment</a></li>
<li><a class="reference internal" href="#orthologous-groups">Orthologous groups</a></li>
</ul>
</li>
<li><a class="reference internal" href="#configuration">Configuration</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="MapTranscripts454.html"
                        title="previous chapter">454 Transcript mapping pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../developing.html"
                        title="next chapter">Contributing to CGAT code</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/make_pipelines/Optic.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../developing.html" title="Contributing to CGAT code"
             >next</a> |</li>
        <li class="right" >
          <a href="MapTranscripts454.html" title="454 Transcript mapping pipeline"
             >previous</a> |</li>
        <li><a href="../index.html">CGAT  documentation</a> &raquo;</li>
          <li><a href="../Pipelines.html" >CGAT Pipelines</a> &raquo;</li>
          <li><a href="../MakePipelines.html" >Lecgacy pipelines</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2011, Andreas Heger.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>